<h1 align="center">
<img src="https://raw.githubusercontent.com/Starlitnightly/omicverse/master/README.assets/logo.png" width="400">
</h1><br>

[![pypi-badge](https://img.shields.io/pypi/v/omicverse)](https://pypi.org/project/omicverse) [![Documentation Status](https://readthedocs.org/projects/omicverse/badge/?version=latest)](https://omicverse.readthedocs.io/en/latest/?badge=latest) [![pypiDownloads](https://static.pepy.tech/badge/omicverse)](https://pepy.tech/project/omicverse) [![condaDownloads](https://img.shields.io/conda/dn/conda-forge/omicverse?logo=Anaconda)](https://anaconda.org/conda-forge/omicverse) [![License:GPL](https://img.shields.io/badge/license-GNU-blue)](https://img.shields.io/apm/l/vim-mode) [![scverse](https://img.shields.io/badge/scverse-ecosystem-blue.svg?labelColor=yellow)](https://scverse.org/) [![Pytest](https://github.com/Starlitnightly/omicverse/workflows/py38|py39/badge.svg)](https://github.com/Starlitnightly/omicverse/) 

**`OmicVerse`** is the fundamental package for multi omics included **bulk ,single cell and spatial RNA-seq** analysis with Python. For more information, please read our paper: [OmicVerse: A single pipeline for exploring the entire transcriptome universe](https://www.biorxiv.org/content/10.1101/2023.06.06.543913v2)

> [!IMPORTANT]
>
> **Star Us**, You will receive all release notifications from GitHub without any delay \~ â­ï¸
>
> If you like **OmicVerse** and want to support our mission, please consider making a [ğŸ’—donation](https://afdian.net/a/starlitnightly) to support our efforts.

<details>
  <summary><kbd>Star History</kbd></summary>
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://api.star-history.com/svg?repos=Starlitnightly%2Fomicverse&theme=dark&type=Date">
    <img width="100%" src="https://api.star-history.com/svg?repos=Starlitnightly%2Fomicverse&type=Date">
  </picture>
</details>





## `1` [Introduction][docs-feat-provider]

The original name of the omicverse was [Pyomic](https://pypi.org/project/Pyomic/), but we wanted to address a whole universe of transcriptomics, so we changed the name to **`OmicVerse`**, it aimed to solve all task in RNA-seq.

> [!NOTE]
> **BulkTrajBlend** algorithm in OmicVerse that combines Beta-Variational AutoEncoder for deconvolution and graph neural networks for overlapping community discovery to effectively interpolate and restore the continuity of **"omission"** cells in the original scRNA-seq data.

![omicverse-light](omicverse_guide/docs/img/omicverse.png#gh-light-mode-only)
![omicverse-dark](omicverse_guide/docs/img/omicverse_dark.png#gh-dark-mode-only)


## `2` [Directory structure](#)

````shell
.
â”œâ”€â”€ omicverse                  # Main Python package
â”œâ”€â”€ omicverse_guide            # Documentation files
â”œâ”€â”€ sample                     # Some test data
â”œâ”€â”€ LICENSE
â””â”€â”€ README.md
````

## `3` [Getting Started ](#)

OmicVerse can be installed via conda or pypi and you need to install `pytorch` at first. Please refer to the [installation tutorial](https://starlitnightly.github.io/omicverse/Installation_guild/) for more detailed installation steps and adaptations for different platforms (`Windows`, `Linux` or `Mac OS`).

You can use `conda install omicverse -c conda-forge` or `pip install -U omicverse` for installation.

Please checkout the documentations and tutorials at [omicverse page](https://starlitnightly.github.io/omicverse/) or [omicverse.readthedocs.io](https://omicverse.readthedocs.io/en/latest/index.html).

## `4` [Data Framework and Reference](#)

The omicverse is implemented as an infrastructure based on the following four data structures.

<div align="center">
<table>
  <tr>
    <td> <a href="https://github.com/pandas-dev/pandas">pandas</a></td>
    <td> <a href="https://github.com/scverse/anndata">anndata</a></td>
    <td> <a href="https://github.com/numpy/numpy">numpy</a></td>
    <td> <a href="https://github.com/scverse/mudata">mudata</a></td>
  </tr>

</table>
</div>

---

The table contains the tools have been published 

<div align="center">
<table>

  <tr>
    <td align="center">Scanpy<br><a href="https://github.com/scverse/scanpy">ğŸ“¦</a> <a href="https://link.springer.com/article/10.1186/s13059-017-1382-0">ğŸ“–</a></td>
    <td align="center">dynamicTreeCut<br><a href="https://github.com/kylessmith/dynamicTreeCut">ğŸ“¦</a> <a href="https://academic.oup.com/bioinformatics/article/24/5/719/200751">ğŸ“–</a></td>
    <td align="center">scDrug<br><a href="https://github.com/ailabstw/scDrug">ğŸ“¦</a> <a href="https://www.sciencedirect.com/science/article/pii/S2001037022005505">ğŸ“–</a></td>
    <td align="center">MOFA<br><a href="https://github.com/bioFAM/mofapy2">ğŸ“¦</a> <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1">ğŸ“–</a></td>
    <td align="center">COSG<br><a href="https://github.com/genecell/COSG">ğŸ“¦</a> <a href="https://academic.oup.com/bib/advance-article-abstract/doi/10.1093/bib/bbab579/6511197?redirectedFrom=fulltext">ğŸ“–</a></td>
    <td align="center">CellphoneDB<br><a href="https://github.com/ventolab/CellphoneDB">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41586-018-0698-6">ğŸ“–</a></td>
    </tr>

  <tr>
    <td align="center">AUCell<br><a href="https://github.com/aertslab/AUCell">ğŸ“¦</a> <a href="https://bioconductor.org/packages/AUCell">ğŸ“–</a></td>
    <td align="center">Bulk2Space<br><a href="https://github.com/ZJUFanLab/bulk2space">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-022-34271-z">ğŸ“–</a></td>
    <td align="center">SCSA<br><a href="https://github.com/bioinfo-ibms-pumc/SCSA">ğŸ“¦</a> <a href="https://doi.org/10.3389/fgene.2020.00490">ğŸ“–</a></td>
    <td align="center">WGCNA<br><a href="http://www.genetics.ucla.edu/labs/horvath/CoexpressionNetwork/Rpackages/WGCNA">ğŸ“¦</a> <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559">ğŸ“–</a></td>
    <td align="center">VIA<br><a href="https://github.com/ShobiStassen/VIA">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-021-25773-3">ğŸ“–</a></td>
    <td align="center">pyDEseq2<br><a href="https://github.com/owkin/PyDESeq2">ğŸ“¦</a> <a href="https://www.biorxiv.org/content/10.1101/2022.12.14.520412v1">ğŸ“–</a></td>
</tr>

  <tr>
    <td align="center">NOCD<br><a href="https://github.com/shchur/overlapping-community-detection">ğŸ“¦</a> <a href="https://arxiv.org/abs/1909.12201">ğŸ“–</a></td>
    <td align="center">SIMBA<br><a href="https://github.com/pinellolab/simba">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-023-01899-8">ğŸ“–</a></td>
    <td align="center">GLUE<br><a href="https://github.com/gao-lab/GLUE">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41587-022-01284-4">ğŸ“–</a></td>
    <td align="center">MetaTiME<br><a href="https://github.com/yi-zhang/MetaTiME">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-023-38333-8">ğŸ“–</a></td>
    <td align="center">TOSICA<br><a href="https://github.com/JackieHanLab/TOSICA">ğŸ“¦</a> <a href="https://doi.org/10.1038/s41467-023-35923-4">ğŸ“–</a></td>
    <td align="center">Harmony<br><a href="https://github.com/slowkow/harmonypy/">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-019-0619-0">ğŸ“–</a></td>
  </tr>

  <tr>
    <td align="center">Scanorama<br><a href="https://github.com/brianhie/scanorama">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41587-019-0113-3">ğŸ“–</a></td>
    <td align="center">Combat<br><a href="https://github.com/epigenelabs/pyComBat/">ğŸ“¦</a> <a href="https://doi.org/10.1101/2020.03.17.995431">ğŸ“–</a></td>
    <td align="center">TAPE<br><a href="https://github.com/poseidonchan/TAPE">ğŸ“¦</a> <a href="https://doi.org/10.1038/s41467-022-34550-9">ğŸ“–</a></td>
    <td align="center">SEACells<br><a href="https://github.com/dpeerlab/SEACells">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41587-023-01716-9">ğŸ“–</a></td>
    <td align="center">Palantir<br><a href="https://github.com/dpeerlab/Palantir">ğŸ“¦</a> <a href="https://doi.org/10.1038/s41587-019-0068-49">ğŸ“–</a></td>
    <td align="center">STAGATE<br><a href="https://github.com/QIFEIDKN/STAGATE_pyG">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-022-29439-6">ğŸ“–</a></td>
  </tr>

  <tr>
    <td align="center">scVI<br><a href="https://github.com/scverse/scvi-tools">ğŸ“¦</a> <a href="https://doi.org/10.1038/s41587-021-01206-w">ğŸ“–</a></td>
    <td align="center">MIRA<br><a href="https://github.com/cistrome/MIRA">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-022-01595-z">ğŸ“–</a></td>
    <td align="center">Tangram<br><a href="https://github.com/broadinstitute/Tangram/">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-021-01264-7">ğŸ“–</a></td>
    <td align="center">STAligner<br><a href="https://github.com/zhoux85/STAligner">ğŸ“¦</a> <a href="https://doi.org/10.1038/s43588-023-00528-w">ğŸ“–</a></td>
    <td align="center">CEFCON<br><a href="https://github.com/WPZgithub/CEFCON">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-023-44103-3">ğŸ“–</a></td>
    <td align="center">PyComplexHeatmap<br><a href="https://github.com/DingWB/PyComplexHeatmap">ğŸ“¦</a> <a href="https://doi.org/10.1002/imt2.115">ğŸ“–</a></td>
      </tr>

  <tr>
    <td align="center">STT<br><a href="https://github.com/cliffzhou92/STT/">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-024-02266-x#Sec2">ğŸ“–</a></td>
    <td align="center">SLAT<br><a href="https://github.com/gao-lab/SLAT">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-023-43105-5">ğŸ“–</a></td>
    <td align="center">GPTCelltype<br><a href="https://github.com/Winnie09/GPTCelltype">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-024-02235-4">ğŸ“–</a></td>
    <td align="center">PROST<br><a href="https://github.com/Tang-Lab-super/PROST">ğŸ“¦</a> <a href="https://doi.org/10.1038/s41467-024-44835-w">ğŸ“–</a></td>
    <td align="center">CytoTrace2<br><a href="https://github.com/digitalcytometry/cytotrace2">ğŸ“¦</a> <a href="https://doi.org/10.1101/2024.03.19.585637">ğŸ“–</a></td>
    <td align="center">GraphST<br><a href="https://github.com/JinmiaoChenLab/GraphST">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-023-36796-3#citeas">ğŸ“–</a></td>
  </tr>

  <tr>
    <td align="center">COMPOSITE<br><a href="https://github.com/CHPGenetics/COMPOSITE/">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41467-024-49448-x#Abs1">ğŸ“–</a></td>
    <td align="center">mellon<br><a href="https://github.com/settylab/mellon">ğŸ“¦</a> <a href="https://www.nature.com/articles/s41592-024-02302-w">ğŸ“–</a></td>
  </tr>
</table>
</div>

---

**Included Package not published or preprint**

- [1] [Cellula](https://github.com/andrecossa5/Cellula/) is to provide a toolkit for the exploration of scRNA-seq. These tools perform common single-cell analysis tasks
- [2] [pegasus](https://github.com/lilab-bcb/pegasus/) is a tool for analyzing transcriptomes of millions of single cells. It is a command line tool, a python package and a base for Cloud-based analysis workflows.
- [3] [cNMF](https://github.com/dylkot/cNMF) is an analysis pipeline for inferring gene expression programs from single-cell RNA-Seq (scRNA-Seq) data.

## `5` [Contact](#)

- Zehua Zeng ([starlitnightly@163.com](mailto:starlitnightly@163.com) or [zehuazeng@xs.ustb.edu.cn](mailto:zehuazeng@xs.ustb.edu.cn))
- Lei Hu ([hulei@westlake.edu.cn](mailto:hulei@westlake.edu.cn))

## `6` [Developer Guild and Contributing](#)

If you would like to contribute to omicverse, please refer to our [developer documentation](https://omicverse.readthedocs.io/en/latest/Developer_guild/).

<table align="center">
    <tr>
      <th colspan="2">
        <br><img src="https://contrib.rocks/image?repo=Starlitnightly/omicverse"><br><br>
      </th>
    </tr>
</table>


> [!IMPORTANT]  
> We would like to thank the following WeChat Official Accounts for promoting Omicverse.
> <p align="center"> <a href="https://mp.weixin.qq.com/s/egAnRfr3etccU_RsN-zIlg" target="_blank" rel="noreferrer"> <img src="README.assets/image-20230701163953794.png" alt="linux" width="50" height="50"/> </a> <a href="https://zhuanlan.zhihu.com/c_1257815636945915904?page=3" target="_blank" rel="noreferrer"> <img src="README.assets/WechatIMG688.png" alt="linux" width="50" height="50"/> </a> </p>





## `7` [Other](#)

If you would like to sponsor the development of our project, you can go to the afdian website (https://afdian.net/a/starlitnightly) and sponsor us.


Copyright Â© 2024 [112 Lab](https://112lab.asia/). <br />
This project is [GPL3.0](./LICENSE) licensed.

<!-- LINK GROUP -->
[docs-feat-provider]: https://starlitnightly.github.io/omicverse/
