
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api scgsea - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.single.geneset_aucell">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api scgsea
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.geneset_aucell">
    geneset_aucell()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.pathway_aucell">
    pathway_aucell()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.pathway_aucell_enrichment">
    pathway_aucell_enrichment()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.pathway_enrichment">
    pathway_enrichment()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.pathway_enrichment_plot">
    pathway_enrichment_plot()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api scgsea</h1>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.geneset_aucell">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">geneset_aucell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">geneset_name</span><span class="p">,</span> <span class="n">geneset</span><span class="p">,</span> <span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.geneset_aucell" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Calculate the AUC-ell score for a given gene set.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData object
Annotated data matrix containing gene expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>geneset_name</code></td>
<td>
</td>
<td><p>str
Name of the gene set.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>geneset</code></td>
<td>
</td>
<td><p>list of str
List of gene symbols for the gene set.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>AUC_threshold</code></td>
<td>
</td>
<td><p>float, optional
AUC threshold used to determine significant interactions (default is 0.01).</p></td>
<td>
<code>0.01</code>
</td>
</tr>
<tr>
<td><code>seed</code></td>
<td>
</td>
<td><p>int, optional
Seed used to initialize the random number generator (default is 42).</p></td>
<td>
<code>42</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
</td>
<td><p>None
Adds a column to the 'obs' attribute of the adata object containing the AUC-ell score for the gene set.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scgsea.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="37 "></span><span class="k">def</span> <span class="nf">geneset_aucell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">geneset_name</span><span class="p">,</span><span class="n">geneset</span><span class="p">,</span><span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span><span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">):</span>
<span class="linenos" data-linenos="38 "></span>
<span class="linenos" data-linenos="39 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="40 "></span><span class="sd">    Calculate the AUC-ell score for a given gene set.</span>
<span class="linenos" data-linenos="41 "></span>
<span class="linenos" data-linenos="42 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="43 "></span><span class="sd">        adata: AnnData object</span>
<span class="linenos" data-linenos="44 "></span><span class="sd">            Annotated data matrix containing gene expression data.</span>
<span class="linenos" data-linenos="45 "></span><span class="sd">        geneset_name: str</span>
<span class="linenos" data-linenos="46 "></span><span class="sd">            Name of the gene set.</span>
<span class="linenos" data-linenos="47 "></span><span class="sd">        geneset: list of str</span>
<span class="linenos" data-linenos="48 "></span><span class="sd">            List of gene symbols for the gene set.</span>
<span class="linenos" data-linenos="49 "></span><span class="sd">        AUC_threshold: float, optional</span>
<span class="linenos" data-linenos="50 "></span><span class="sd">            AUC threshold used to determine significant interactions (default is 0.01).</span>
<span class="linenos" data-linenos="51 "></span><span class="sd">        seed: int, optional</span>
<span class="linenos" data-linenos="52 "></span><span class="sd">            Seed used to initialize the random number generator (default is 42).</span>
<span class="linenos" data-linenos="53 "></span>
<span class="linenos" data-linenos="54 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="55 "></span><span class="sd">        None</span>
<span class="linenos" data-linenos="56 "></span><span class="sd">           Adds a column to the 'obs' attribute of the adata object containing the AUC-ell score for the gene set.</span>
<span class="linenos" data-linenos="57 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="58 "></span>
<span class="linenos" data-linenos="59 "></span>    <span class="n">check_ctxcore</span><span class="p">()</span>
<span class="linenos" data-linenos="60 "></span>    <span class="k">global</span> <span class="n">ctxcore_install</span>
<span class="linenos" data-linenos="61 "></span>    <span class="k">if</span> <span class="n">ctxcore_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="62 "></span>        <span class="k">global</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="63 "></span>        <span class="k">global</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="64 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.recovery</span> <span class="kn">import</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="65 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.genesig</span> <span class="kn">import</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="66 "></span>
<span class="linenos" data-linenos="67 "></span>
<span class="linenos" data-linenos="68 "></span>    <span class="n">matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
<span class="linenos" data-linenos="69 "></span>    <span class="n">percentiles</span> <span class="o">=</span> <span class="n">derive_auc_threshold</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
<span class="linenos" data-linenos="70 "></span>    <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">percentiles</span><span class="p">[</span><span class="n">AUC_threshold</span><span class="p">]</span>
<span class="linenos" data-linenos="71 "></span>
<span class="linenos" data-linenos="72 "></span>    <span class="n">df_rnk</span><span class="o">=</span><span class="n">create_rankings</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">seed</span><span class="p">)</span>
<span class="linenos" data-linenos="73 "></span>    <span class="n">rnk</span><span class="o">=</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">df_rnk</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">geneset</span><span class="p">)]</span>
<span class="linenos" data-linenos="74 "></span>
<span class="linenos" data-linenos="75 "></span>    <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">empty</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">geneset</span><span class="p">)))</span> <span class="o">&lt;</span> <span class="mf">0.80</span><span class="p">:</span>
<span class="linenos" data-linenos="76 "></span>        <span class="nb">print</span><span class="p">(</span>
<span class="linenos" data-linenos="77 "></span>            <span class="sa">f</span><span class="s2">"Less than 80% of the genes in </span><span class="si">{</span><span class="n">geneset_name</span><span class="si">}</span><span class="s2"> are present in the "</span>
<span class="linenos" data-linenos="78 "></span>            <span class="s2">"expression matrix."</span>
<span class="linenos" data-linenos="79 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="80 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'</span><span class="si">{}</span><span class="s1">_aucell'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">geneset_name</span><span class="p">)]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="linenos" data-linenos="81 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="82 "></span>        <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">geneset</span><span class="p">])</span>
<span class="linenos" data-linenos="83 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'</span><span class="si">{}</span><span class="s1">_aucell'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">geneset_name</span><span class="p">)]</span><span class="o">=</span><span class="n">aucs</span><span class="p">(</span><span class="n">rnk</span><span class="p">,</span>
<span class="linenos" data-linenos="84 "></span>            <span class="nb">len</span><span class="p">(</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span><span class="n">weights</span><span class="p">,</span><span class="n">auc_threshold</span>
<span class="linenos" data-linenos="85 "></span>            <span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.pathway_aucell">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">pathway_aucell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathway_names</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">,</span> <span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.pathway_aucell" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Calculates the area under the curve (AUC) for a set of pathways in an AnnData object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData object
AnnData object containing the data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pathway_names</code></td>
<td>
</td>
<td><p>list of str
Names of the pathways to analyze.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pathways_dict</code></td>
<td>
</td>
<td><p>dict
Dictionary containing the gene sets for each pathway.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>AUC_threshold</code></td>
<td>
</td>
<td><p>float, optional (default: 0.01)
AUC threshold to use for determining significant gene-pathway associations.</p></td>
<td>
<code>0.01</code>
</td>
</tr>
<tr>
<td><code>seed</code></td>
<td>
</td>
<td><p>int, optional (default: 42)
Random seed for reproducibility.</p></td>
<td>
<code>42</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
</td>
<td><p>None
The function modifies the <code>adata.obs</code> attribute of the input AnnData object.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scgsea.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 87 "></span><span class="k">def</span> <span class="nf">pathway_aucell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">pathway_names</span><span class="p">,</span><span class="n">pathways_dict</span><span class="p">,</span><span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span><span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">):</span>
<span class="linenos" data-linenos=" 88 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos=" 89 "></span><span class="sd">    Calculates the area under the curve (AUC) for a set of pathways in an AnnData object.</span>
<span class="linenos" data-linenos=" 90 "></span>
<span class="linenos" data-linenos=" 91 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos=" 92 "></span><span class="sd">        adata: AnnData object</span>
<span class="linenos" data-linenos=" 93 "></span><span class="sd">            AnnData object containing the data.</span>
<span class="linenos" data-linenos=" 94 "></span><span class="sd">        pathway_names: list of str</span>
<span class="linenos" data-linenos=" 95 "></span><span class="sd">            Names of the pathways to analyze.</span>
<span class="linenos" data-linenos=" 96 "></span><span class="sd">        pathways_dict: dict</span>
<span class="linenos" data-linenos=" 97 "></span><span class="sd">            Dictionary containing the gene sets for each pathway.</span>
<span class="linenos" data-linenos=" 98 "></span><span class="sd">        AUC_threshold: float, optional (default: 0.01)</span>
<span class="linenos" data-linenos=" 99 "></span><span class="sd">            AUC threshold to use for determining significant gene-pathway associations.</span>
<span class="linenos" data-linenos="100 "></span><span class="sd">        seed: int, optional (default: 42)</span>
<span class="linenos" data-linenos="101 "></span><span class="sd">            Random seed for reproducibility.</span>
<span class="linenos" data-linenos="102 "></span>
<span class="linenos" data-linenos="103 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">        None</span>
<span class="linenos" data-linenos="105 "></span><span class="sd">            The function modifies the `adata.obs` attribute of the input AnnData object.</span>
<span class="linenos" data-linenos="106 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="107 "></span>
<span class="linenos" data-linenos="108 "></span>
<span class="linenos" data-linenos="109 "></span>    <span class="n">check_ctxcore</span><span class="p">()</span>
<span class="linenos" data-linenos="110 "></span>    <span class="k">global</span> <span class="n">ctxcore_install</span>
<span class="linenos" data-linenos="111 "></span>    <span class="k">if</span> <span class="n">ctxcore_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="112 "></span>        <span class="k">global</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="113 "></span>        <span class="k">global</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="114 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.recovery</span> <span class="kn">import</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="115 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.genesig</span> <span class="kn">import</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="116 "></span>
<span class="linenos" data-linenos="117 "></span>    <span class="n">matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
<span class="linenos" data-linenos="118 "></span>    <span class="n">percentiles</span> <span class="o">=</span> <span class="n">derive_auc_threshold</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
<span class="linenos" data-linenos="119 "></span>    <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">percentiles</span><span class="p">[</span><span class="n">AUC_threshold</span><span class="p">]</span>
<span class="linenos" data-linenos="120 "></span>    <span class="n">df_rnk</span><span class="o">=</span><span class="n">create_rankings</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">seed</span><span class="p">)</span>
<span class="linenos" data-linenos="121 "></span>
<span class="linenos" data-linenos="122 "></span>    <span class="k">for</span> <span class="n">pathway_name</span> <span class="ow">in</span> <span class="n">pathway_names</span><span class="p">:</span>
<span class="linenos" data-linenos="123 "></span>        <span class="n">pathway_genes</span><span class="o">=</span><span class="n">pathways_dict</span><span class="p">[</span><span class="n">pathway_name</span><span class="p">]</span>
<span class="linenos" data-linenos="124 "></span>        <span class="n">rnk</span><span class="o">=</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">df_rnk</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">pathway_genes</span><span class="p">)]</span>
<span class="linenos" data-linenos="125 "></span>        <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">empty</span> <span class="ow">or</span> <span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pathway_genes</span><span class="p">)))</span> <span class="o">&lt;</span> <span class="mf">0.80</span><span class="p">:</span>
<span class="linenos" data-linenos="126 "></span>            <span class="nb">print</span><span class="p">(</span>
<span class="linenos" data-linenos="127 "></span>                <span class="sa">f</span><span class="s2">"Less than 80% of the genes in </span><span class="si">{</span><span class="n">pathway_name</span><span class="si">}</span><span class="s2"> are present in the "</span>
<span class="linenos" data-linenos="128 "></span>                <span class="s2">"expression matrix."</span>
<span class="linenos" data-linenos="129 "></span>            <span class="p">)</span>
<span class="linenos" data-linenos="130 "></span>            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'</span><span class="si">{}</span><span class="s1">_aucell'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pathway_name</span><span class="p">)]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="linenos" data-linenos="131 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="132 "></span>            <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pathway_genes</span><span class="p">])</span>
<span class="linenos" data-linenos="133 "></span>            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'</span><span class="si">{}</span><span class="s1">_aucell'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">pathway_name</span><span class="p">)]</span><span class="o">=</span><span class="n">aucs</span><span class="p">(</span><span class="n">rnk</span><span class="p">,</span>
<span class="linenos" data-linenos="134 "></span>                <span class="nb">len</span><span class="p">(</span><span class="n">df_rnk</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span><span class="n">weights</span><span class="p">,</span><span class="n">auc_threshold</span>
<span class="linenos" data-linenos="135 "></span>                <span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.pathway_aucell_enrichment">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">pathway_aucell_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">,</span> <span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.pathway_aucell_enrichment" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Enriches cell annotations with pathway activity scores using the AUC-ell method.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>AnnData object
AnnData object containing the expression matrix.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pathways_dict</code></td>
<td>
</td>
<td><p>dict
A dictionary where keys are pathway names and values are lists of genes associated with each pathway.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>AUC_threshold</code></td>
<td>
</td>
<td><p>float, optional
The threshold for calculating the area under the curve (AUC) values using the AUC-ell method. The default is 0.01.</p></td>
<td>
<code>0.01</code>
</td>
</tr>
<tr>
<td><code>seed</code></td>
<td>
</td>
<td><p>int, optional
The seed to use for the random number generator. The default is 42.</p></td>
<td>
<code>42</code>
</td>
</tr>
<tr>
<td><code>num_workers</code></td>
<td>
</td>
<td><p>int, optional
The number of workers to use for parallel processing. The default is 1.</p></td>
<td>
<code>1</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata_aucs</code></td> <td>
</td>
<td><p>AnnData object
AnnData object containing the pathway activity scores for each cell in the input AnnData object.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scgsea.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="137 "></span><span class="k">def</span> <span class="nf">pathway_aucell_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">pathways_dict</span><span class="p">,</span><span class="n">AUC_threshold</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span><span class="n">seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span><span class="n">num_workers</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="linenos" data-linenos="138 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="139 "></span><span class="sd">    Enriches cell annotations with pathway activity scores using the AUC-ell method.</span>
<span class="linenos" data-linenos="140 "></span>
<span class="linenos" data-linenos="141 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="142 "></span><span class="sd">        adata: AnnData object</span>
<span class="linenos" data-linenos="143 "></span><span class="sd">            AnnData object containing the expression matrix.</span>
<span class="linenos" data-linenos="144 "></span><span class="sd">        pathways_dict: dict</span>
<span class="linenos" data-linenos="145 "></span><span class="sd">            A dictionary where keys are pathway names and values are lists of genes associated with each pathway.</span>
<span class="linenos" data-linenos="146 "></span><span class="sd">        AUC_threshold: float, optional</span>
<span class="linenos" data-linenos="147 "></span><span class="sd">            The threshold for calculating the area under the curve (AUC) values using the AUC-ell method. The default is 0.01.</span>
<span class="linenos" data-linenos="148 "></span><span class="sd">        seed: int, optional</span>
<span class="linenos" data-linenos="149 "></span><span class="sd">            The seed to use for the random number generator. The default is 42.</span>
<span class="linenos" data-linenos="150 "></span><span class="sd">        num_workers: int, optional</span>
<span class="linenos" data-linenos="151 "></span><span class="sd">            The number of workers to use for parallel processing. The default is 1.</span>
<span class="linenos" data-linenos="152 "></span>
<span class="linenos" data-linenos="153 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="154 "></span><span class="sd">        adata_aucs: AnnData object</span>
<span class="linenos" data-linenos="155 "></span><span class="sd">            AnnData object containing the pathway activity scores for each cell in the input AnnData object.</span>
<span class="linenos" data-linenos="156 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="157 "></span>
<span class="linenos" data-linenos="158 "></span>    <span class="n">check_ctxcore</span><span class="p">()</span>
<span class="linenos" data-linenos="159 "></span>    <span class="k">global</span> <span class="n">ctxcore_install</span>
<span class="linenos" data-linenos="160 "></span>    <span class="k">if</span> <span class="n">ctxcore_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="161 "></span>        <span class="k">global</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="162 "></span>        <span class="k">global</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="163 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.recovery</span> <span class="kn">import</span> <span class="n">aucs</span>
<span class="linenos" data-linenos="164 "></span>        <span class="kn">from</span> <span class="nn">ctxcore.genesig</span> <span class="kn">import</span> <span class="n">GeneSignature</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span>    <span class="n">test_gmt</span><span class="o">=</span><span class="p">[]</span>
<span class="linenos" data-linenos="167 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pathways_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="linenos" data-linenos="168 "></span>        <span class="n">test_gmt</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">GeneSignature</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">i</span><span class="p">,</span><span class="n">gene2weight</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">pathways_dict</span><span class="p">[</span><span class="n">i</span><span class="p">],[</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pathways_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]]))))</span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span>    <span class="n">matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
<span class="linenos" data-linenos="171 "></span>    <span class="n">percentiles</span> <span class="o">=</span> <span class="n">derive_auc_threshold</span><span class="p">(</span><span class="n">matrix</span><span class="p">)</span>
<span class="linenos" data-linenos="172 "></span>    <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">percentiles</span><span class="p">[</span><span class="n">AUC_threshold</span><span class="p">]</span>
<span class="linenos" data-linenos="173 "></span>
<span class="linenos" data-linenos="174 "></span>    <span class="n">aucs_mtx</span> <span class="o">=</span> <span class="n">aucell</span><span class="p">(</span><span class="n">matrix</span><span class="p">,</span> <span class="n">signatures</span><span class="o">=</span><span class="n">test_gmt</span><span class="p">,</span> <span class="n">auc_threshold</span><span class="o">=</span><span class="n">auc_threshold</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">)</span>
<span class="linenos" data-linenos="175 "></span>    <span class="n">adata_aucs</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">aucs_mtx</span><span class="p">)</span>
<span class="linenos" data-linenos="176 "></span>    <span class="k">return</span> <span class="n">adata_aucs</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.pathway_enrichment">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">pathway_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">'Human'</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="s1">'louvain'</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">logfc_threshold</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">pvalue_type</span><span class="o">=</span><span class="s1">'adjust'</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.pathway_enrichment" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Perform pathway enrichment analysis on gene expression data.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>anndata.AnnData
Annotated data matrix containing gene expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pathways_dict</code></td>
<td>
</td>
<td><p>dict
A dictionary of gene sets with their gene members.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>organism</code></td>
<td>
</td>
<td><p>str, optional (default: 'Human')
The organism to be used for the enrichment analysis. Can be either 'Human' or 'Mouse'.</p></td>
<td>
<code>'Human'</code>
</td>
</tr>
<tr>
<td><code>group_by</code></td>
<td>
</td>
<td><p>str, optional (default: 'louvain')
The group label of the cells in adata.obs to perform the enrichment analysis on.</p></td>
<td>
<code>'louvain'</code>
</td>
</tr>
<tr>
<td><code>cutoff</code></td>
<td>
</td>
<td><p>float, optional (default: 0.05)
The adjusted p-value cutoff used to filter enriched pathways.</p></td>
<td>
<code>0.05</code>
</td>
</tr>
<tr>
<td><code>logfc_threshold</code></td>
<td>
</td>
<td><p>float, optional (default: 2)
The log2 fold change cutoff used to define differentially expressed genes.</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>pvalue_type</code></td>
<td>
</td>
<td><p>str, optional (default: 'adjust')
The type of p-value used for filtering enriched pathways. Can be either 'adjust' or 'raw'.</p></td>
<td>
<code>'adjust'</code>
</td>
</tr>
<tr>
<td><code>plot</code></td>
<td>
</td>
<td><p>bool, optional (default: True)
If True, generate a bar plot for each cluster of enriched pathways.</p></td>
<td>
<code>True</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>enrich_res</code></td> <td>
</td>
<td><p>pandas.DataFrame
A pandas dataframe containing the enriched pathways information including term name, 
p-value, adjusted p-value, overlap genes, odds ratio, log2 fold change, and cluster name.</p></td>
</tr>
</tbody>
</table>
<p><strong>Examples:</strong></p>
<div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Perform pathway enrichment analysis on adata using pathways_dict</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">pathway_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">)</span>
</code></pre></div>
<details class="reference">
<summary>Reference</summary>
<p>The code for pathway_enrichment() function was adapted from scanpy workflows: 
https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html#Gene-set-enrichment-analysis.</p>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scgsea.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="179 "></span><span class="k">def</span> <span class="nf">pathway_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">,</span><span class="n">organism</span><span class="o">=</span><span class="s1">'Human'</span><span class="p">,</span><span class="n">group_by</span><span class="o">=</span><span class="s1">'louvain'</span><span class="p">,</span> 
<span class="linenos" data-linenos="180 "></span>                       <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">logfc_threshold</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">pvalue_type</span><span class="o">=</span><span class="s1">'adjust'</span><span class="p">,</span><span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="linenos" data-linenos="181 "></span>
<span class="linenos" data-linenos="182 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="183 "></span><span class="sd">    Perform pathway enrichment analysis on gene expression data.</span>
<span class="linenos" data-linenos="184 "></span>
<span class="linenos" data-linenos="185 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="186 "></span><span class="sd">        adata: anndata.AnnData</span>
<span class="linenos" data-linenos="187 "></span><span class="sd">            Annotated data matrix containing gene expression data.</span>
<span class="linenos" data-linenos="188 "></span><span class="sd">        pathways_dict: dict</span>
<span class="linenos" data-linenos="189 "></span><span class="sd">            A dictionary of gene sets with their gene members.</span>
<span class="linenos" data-linenos="190 "></span><span class="sd">        organism: str, optional (default: 'Human')</span>
<span class="linenos" data-linenos="191 "></span><span class="sd">            The organism to be used for the enrichment analysis. Can be either 'Human' or 'Mouse'.</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        group_by: str, optional (default: 'louvain')</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">            The group label of the cells in adata.obs to perform the enrichment analysis on.</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        cutoff: float, optional (default: 0.05)</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">            The adjusted p-value cutoff used to filter enriched pathways.</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        logfc_threshold: float, optional (default: 2)</span>
<span class="linenos" data-linenos="197 "></span><span class="sd">            The log2 fold change cutoff used to define differentially expressed genes.</span>
<span class="linenos" data-linenos="198 "></span><span class="sd">        pvalue_type: str, optional (default: 'adjust')</span>
<span class="linenos" data-linenos="199 "></span><span class="sd">            The type of p-value used for filtering enriched pathways. Can be either 'adjust' or 'raw'.</span>
<span class="linenos" data-linenos="200 "></span><span class="sd">        plot: bool, optional (default: True)</span>
<span class="linenos" data-linenos="201 "></span><span class="sd">            If True, generate a bar plot for each cluster of enriched pathways.</span>
<span class="linenos" data-linenos="202 "></span>
<span class="linenos" data-linenos="203 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="204 "></span><span class="sd">        enrich_res: pandas.DataFrame</span>
<span class="linenos" data-linenos="205 "></span><span class="sd">            A pandas dataframe containing the enriched pathways information including term name, </span>
<span class="linenos" data-linenos="206 "></span><span class="sd">            p-value, adjusted p-value, overlap genes, odds ratio, log2 fold change, and cluster name.</span>
<span class="linenos" data-linenos="207 "></span>
<span class="linenos" data-linenos="208 "></span><span class="sd">    Examples:</span>
<span class="linenos" data-linenos="209 "></span><span class="sd">        &gt;&gt;&gt; # Perform pathway enrichment analysis on adata using pathways_dict</span>
<span class="linenos" data-linenos="210 "></span><span class="sd">        &gt;&gt;&gt; res = pathway_enrichment(adata, pathways_dict)</span>
<span class="linenos" data-linenos="211 "></span>
<span class="linenos" data-linenos="212 "></span><span class="sd">    Reference:</span>
<span class="linenos" data-linenos="213 "></span><span class="sd">        The code for pathway_enrichment() function was adapted from scanpy workflows: </span>
<span class="linenos" data-linenos="214 "></span><span class="sd">        https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html#Gene-set-enrichment-analysis.</span>
<span class="linenos" data-linenos="215 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="216 "></span>
<span class="linenos" data-linenos="217 "></span>    <span class="c1">#import gseapy as gp</span>
<span class="linenos" data-linenos="218 "></span>    <span class="kn">from</span> <span class="nn">..externel.gseapy</span> <span class="kn">import</span> <span class="n">enrichr</span>
<span class="linenos" data-linenos="219 "></span>    <span class="n">df_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="220 "></span>    <span class="n">cluster_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="221 "></span>    <span class="n">celltypes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">group_by</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="linenos" data-linenos="222 "></span>
<span class="linenos" data-linenos="223 "></span>    <span class="k">for</span> <span class="n">celltype</span> <span class="ow">in</span> <span class="n">celltypes</span><span class="p">:</span>
<span class="linenos" data-linenos="224 "></span>        <span class="n">degs</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">celltype</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s1">'rank_genes_groups'</span><span class="p">,</span> <span class="n">log2fc_min</span><span class="o">=</span><span class="n">logfc_threshold</span><span class="p">,</span> 
<span class="linenos" data-linenos="225 "></span>                                    <span class="n">pval_cutoff</span><span class="o">=</span><span class="n">cutoff</span><span class="p">)[</span><span class="s1">'names'</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="linenos" data-linenos="226 "></span>        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">degs</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
<span class="linenos" data-linenos="227 "></span>            <span class="n">degs</span> <span class="o">=</span> <span class="p">[</span><span class="n">degs</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>
<span class="linenos" data-linenos="228 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="229 "></span>            <span class="n">degs</span> <span class="o">=</span> <span class="n">degs</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="230 "></span>
<span class="linenos" data-linenos="231 "></span>        <span class="k">if</span> <span class="ow">not</span> <span class="n">degs</span><span class="p">:</span>
<span class="linenos" data-linenos="232 "></span>            <span class="k">continue</span>
<span class="linenos" data-linenos="233 "></span>        <span class="k">if</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'Mouse'</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'mouse'</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'mm'</span><span class="p">):</span>
<span class="linenos" data-linenos="234 "></span>            <span class="n">background</span><span class="o">=</span><span class="s1">'mmusculus_gene_ensembl'</span>
<span class="linenos" data-linenos="235 "></span>        <span class="k">elif</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'Human'</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'human'</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">organism</span> <span class="o">==</span> <span class="s1">'hs'</span><span class="p">):</span>
<span class="linenos" data-linenos="236 "></span>            <span class="n">background</span><span class="o">=</span><span class="s1">'hsapiens_gene_ensembl'</span>
<span class="linenos" data-linenos="237 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="238 "></span>            <span class="n">background</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="239 "></span>        <span class="n">enr</span> <span class="o">=</span> <span class="n">enrichr</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">degs</span><span class="p">,</span>
<span class="linenos" data-linenos="240 "></span>                <span class="n">description</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos="241 "></span>                <span class="n">gene_sets</span><span class="o">=</span><span class="n">pathways_dict</span><span class="p">,</span>
<span class="linenos" data-linenos="242 "></span>                <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span>
<span class="linenos" data-linenos="243 "></span>                <span class="n">background</span><span class="o">=</span><span class="n">background</span><span class="p">,</span>
<span class="linenos" data-linenos="244 "></span>                <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="linenos" data-linenos="245 "></span>                <span class="p">)</span>
<span class="linenos" data-linenos="246 "></span>        <span class="k">if</span> <span class="p">(</span><span class="n">enr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">enr</span><span class="p">,</span> <span class="s1">'res2d'</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">enr</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
<span class="linenos" data-linenos="247 "></span>            <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">enr</span><span class="o">.</span><span class="n">res2d</span><span class="p">)</span>
<span class="linenos" data-linenos="248 "></span>            <span class="n">cluster_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">celltype</span><span class="p">)</span>
<span class="linenos" data-linenos="249 "></span>
<span class="linenos" data-linenos="250 "></span>    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'Cluster'</span><span class="p">,</span> <span class="s1">'Gene_set'</span><span class="p">,</span> <span class="s1">'Term'</span><span class="p">,</span> <span class="s1">'Overlap'</span><span class="p">,</span> <span class="s1">'Odds Ratio'</span><span class="p">,</span><span class="s1">'P-value'</span><span class="p">,</span> <span class="s1">'Adjusted P-value'</span><span class="p">,</span> <span class="s1">'Genes'</span><span class="p">]</span>
<span class="linenos" data-linenos="251 "></span>
<span class="linenos" data-linenos="252 "></span>    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">columns</span><span class="p">)</span>
<span class="linenos" data-linenos="253 "></span>    <span class="k">for</span> <span class="n">cluster_ind</span><span class="p">,</span> <span class="n">df_</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">cluster_list</span><span class="p">,</span> <span class="n">df_list</span><span class="p">):</span>
<span class="linenos" data-linenos="254 "></span>        <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="p">[</span><span class="n">df_</span><span class="p">[</span><span class="s1">'Adjusted P-value'</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">cutoff</span><span class="p">]</span>
<span class="linenos" data-linenos="255 "></span>        <span class="n">df_</span> <span class="o">=</span> <span class="n">df_</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">Cluster</span> <span class="o">=</span> <span class="n">cluster_ind</span><span class="p">)</span>
<span class="linenos" data-linenos="256 "></span>        <span class="k">if</span> <span class="p">(</span><span class="n">df_</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
<span class="linenos" data-linenos="257 "></span>            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="p">,</span> <span class="n">df_</span><span class="p">[</span><span class="n">columns</span><span class="p">]],</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="258 "></span>            <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df_</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s1">'Term'</span><span class="p">,</span> <span class="s1">'Adjusted P-value'</span><span class="p">]][:</span><span class="nb">min</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="n">df_</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
<span class="linenos" data-linenos="259 "></span>            <span class="n">df_tmp</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'('</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df_tmp</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">]]</span>
<span class="linenos" data-linenos="260 "></span>            <span class="n">df_tmp</span><span class="p">[</span><span class="s1">'-log_adj_p'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">df_tmp</span><span class="p">[</span><span class="s1">'Adjusted P-value'</span><span class="p">])</span>
<span class="linenos" data-linenos="261 "></span>            <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">'-log_adj_p'</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="262 "></span>            <span class="k">if</span> <span class="n">plot</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="263 "></span>                <span class="n">ax</span> <span class="o">=</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="s1">'-log_adj_p'</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">'Term'</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">grid</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="linenos" data-linenos="264 "></span>                <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">'Cluster </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cluster_ind</span><span class="p">))</span>
<span class="linenos" data-linenos="265 "></span>                <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">''</span><span class="p">)</span>
<span class="linenos" data-linenos="266 "></span>                <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'-log(Adjusted P-value)'</span><span class="p">)</span>
<span class="linenos" data-linenos="267 "></span>            <span class="c1">#pp.savefig(ax.figure, bbox_inches='tight')</span>
<span class="linenos" data-linenos="268 "></span>            <span class="c1">#plt.close()</span>
<span class="linenos" data-linenos="269 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="270 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'No pathway with an adjusted P-value less than the cutoff (=</span><span class="si">{}</span><span class="s1">) for cluster </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cutoff</span><span class="p">,</span> <span class="n">cluster_ind</span><span class="p">))</span>
<span class="linenos" data-linenos="271 "></span>
<span class="linenos" data-linenos="272 "></span>    <span class="k">if</span> <span class="n">pvalue_type</span><span class="o">==</span><span class="s1">'adjust'</span><span class="p">:</span>
<span class="linenos" data-linenos="273 "></span>        <span class="n">enrich_res</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">'Adjusted P-value'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">cutoff</span><span class="p">]</span>
<span class="linenos" data-linenos="274 "></span>        <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Adjusted P-value'</span><span class="p">])</span>
<span class="linenos" data-linenos="275 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="276 "></span>        <span class="n">enrich_res</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">'P-value'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">cutoff</span><span class="p">]</span>
<span class="linenos" data-linenos="277 "></span>        <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'P-value'</span><span class="p">])</span>
<span class="linenos" data-linenos="278 "></span>    <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'logc'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Odds Ratio'</span><span class="p">])</span>
<span class="linenos" data-linenos="279 "></span>    <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'num'</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'/'</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Overlap'</span><span class="p">]]</span>
<span class="linenos" data-linenos="280 "></span>    <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'fraction'</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'/'</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span><span class="o">/</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'/'</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Overlap'</span><span class="p">]]</span>
<span class="linenos" data-linenos="281 "></span>
<span class="linenos" data-linenos="282 "></span>    <span class="k">return</span> <span class="n">enrich_res</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.pathway_enrichment_plot">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">pathway_enrichment_plot</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">,</span> <span class="n">term_num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">return_table</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">plot_title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="o">**</span><span class="n">kwds</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.pathway_enrichment_plot" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Visualize the pathway enrichment analysis results as a heatmap.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>enrich_res</code></td>
<td>
</td>
<td><p>pandas DataFrame
The output from the pathway_enrichment() function.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>term_num</code></td>
<td>
</td>
<td><p>int, optional
The number of enriched terms to display for each cluster. Default is 5.</p></td>
<td>
<code>5</code>
</td>
</tr>
<tr>
<td><code>return_table</code></td>
<td>
</td>
<td><p>bool, optional
Whether to return the heatmap table as a DataFrame. Default is False.</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>tuple, optional
The size of the plot. Default is (3,10).</p></td>
<td>
<code>(3, 10)</code>
</td>
</tr>
<tr>
<td><code>plot_title</code></td>
<td>
</td>
<td><p>str, optional
The title of the plot. Default is an empty string.</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>**kwds</code></td>
<td>
</td>
<td><p>optional
Other keyword arguments to pass to the seaborn heatmap function.</p></td>
<td>
<code>{}</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ax</code></td> <td>
</td>
<td><p>matplotlib Axes object
The heatmap plot.</p></td>
</tr>
</tbody>
</table>
<p><strong>Examples:</strong></p>
<div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">pathway_enrichment</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pathways_dict</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pathway_enrichment_plot</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="n">term_num</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">return_table</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">12</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Blues'</span><span class="p">)</span>
</code></pre></div>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scgsea.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="284 "></span><span class="k">def</span> <span class="nf">pathway_enrichment_plot</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">,</span><span class="n">term_num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">return_table</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">plot_title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="o">**</span><span class="n">kwds</span><span class="p">):</span>
<span class="linenos" data-linenos="285 "></span>
<span class="linenos" data-linenos="286 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="287 "></span><span class="sd">    Visualize the pathway enrichment analysis results as a heatmap.</span>
<span class="linenos" data-linenos="288 "></span>
<span class="linenos" data-linenos="289 "></span><span class="sd">    Parameters:</span>
<span class="linenos" data-linenos="290 "></span><span class="sd">        enrich_res: pandas DataFrame</span>
<span class="linenos" data-linenos="291 "></span><span class="sd">            The output from the pathway_enrichment() function.</span>
<span class="linenos" data-linenos="292 "></span><span class="sd">        term_num: int, optional</span>
<span class="linenos" data-linenos="293 "></span><span class="sd">            The number of enriched terms to display for each cluster. Default is 5.</span>
<span class="linenos" data-linenos="294 "></span><span class="sd">        return_table: bool, optional</span>
<span class="linenos" data-linenos="295 "></span><span class="sd">            Whether to return the heatmap table as a DataFrame. Default is False.</span>
<span class="linenos" data-linenos="296 "></span><span class="sd">        figsize: tuple, optional</span>
<span class="linenos" data-linenos="297 "></span><span class="sd">            The size of the plot. Default is (3,10).</span>
<span class="linenos" data-linenos="298 "></span><span class="sd">        plot_title: str, optional</span>
<span class="linenos" data-linenos="299 "></span><span class="sd">            The title of the plot. Default is an empty string.</span>
<span class="linenos" data-linenos="300 "></span><span class="sd">        **kwds: optional</span>
<span class="linenos" data-linenos="301 "></span><span class="sd">            Other keyword arguments to pass to the seaborn heatmap function.</span>
<span class="linenos" data-linenos="302 "></span>
<span class="linenos" data-linenos="303 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="304 "></span><span class="sd">        ax: matplotlib Axes object</span>
<span class="linenos" data-linenos="305 "></span><span class="sd">            The heatmap plot.</span>
<span class="linenos" data-linenos="306 "></span>
<span class="linenos" data-linenos="307 "></span><span class="sd">    Examples:</span>
<span class="linenos" data-linenos="308 "></span><span class="sd">        &gt;&gt;&gt; res = pathway_enrichment(adata, pathways_dict)</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">        &gt;&gt;&gt; pathway_enrichment_plot(res, term_num=10, return_table=True, figsize=(6,12), cmap='Blues')</span>
<span class="linenos" data-linenos="310 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="311 "></span>
<span class="linenos" data-linenos="312 "></span>    <span class="n">celltypes</span><span class="o">=</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Cluster'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="linenos" data-linenos="313 "></span>    <span class="n">plot_heatmap</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">celltypes</span><span class="p">)</span>
<span class="linenos" data-linenos="314 "></span>    <span class="k">for</span> <span class="n">celltype</span> <span class="ow">in</span> <span class="n">celltypes</span><span class="p">:</span>
<span class="linenos" data-linenos="315 "></span>        <span class="n">res_test</span><span class="o">=</span><span class="n">enrich_res</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Cluster'</span><span class="p">]</span><span class="o">==</span><span class="n">celltype</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">term_num</span><span class="p">]</span>
<span class="linenos" data-linenos="316 "></span>        <span class="n">plot_heatmap_test</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">res_test</span><span class="p">[[</span><span class="s1">'Term'</span><span class="p">,</span><span class="s1">'logp'</span><span class="p">]])</span>
<span class="linenos" data-linenos="317 "></span>        <span class="n">plot_heatmap_test</span><span class="o">=</span><span class="n">plot_heatmap_test</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">plot_heatmap_test</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">])</span>
<span class="linenos" data-linenos="318 "></span>        <span class="k">del</span> <span class="n">plot_heatmap_test</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">]</span>
<span class="linenos" data-linenos="319 "></span>        <span class="c1">#plot_heatmap[plot_heatmap_test.index]=0</span>
<span class="linenos" data-linenos="320 "></span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">plot_heatmap_test</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
<span class="linenos" data-linenos="321 "></span>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">enrich_res</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Cluster'</span><span class="p">]</span><span class="o">==</span><span class="n">celltype</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">]</span><span class="o">==</span><span class="n">i</span><span class="p">),</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos="322 "></span>                <span class="n">plot_heatmap</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">celltype</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">enrich_res</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Cluster'</span><span class="p">]</span><span class="o">==</span><span class="n">celltype</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">enrich_res</span><span class="p">[</span><span class="s1">'Term'</span><span class="p">]</span><span class="o">==</span><span class="n">i</span><span class="p">),</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="323 "></span>            <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="324 "></span>                <span class="n">plot_heatmap</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">celltype</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="325 "></span>    <span class="n">plot_heatmap</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="326 "></span>    <span class="k">if</span> <span class="n">return_table</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="327 "></span>        <span class="k">return</span> <span class="n">plot_heatmap</span>
<span class="linenos" data-linenos="328 "></span>
<span class="linenos" data-linenos="329 "></span>    <span class="n">fig</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
<span class="linenos" data-linenos="330 "></span>    <span class="n">axr</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">plot_heatmap</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span><span class="o">**</span><span class="n">kwds</span><span class="p">)</span>
<span class="linenos" data-linenos="331 "></span>    <span class="n">axr</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">right</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">labeltop</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="332 "></span>    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">plot_title</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="333 "></span>    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">(),</span> <span class="n">rotation</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="linenos" data-linenos="334 "></span>    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span> <span class="o">=</span> <span class="mi">90</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="linenos" data-linenos="335 "></span>
<span class="linenos" data-linenos="336 "></span>
<span class="linenos" data-linenos="337 "></span>    <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2024, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>