
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api via - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.single.pyVIA">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api via
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.pyVIA">
    pyVIA
  </a>
<nav aria-label="pyVIA" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.__init__">
    __init__()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.run">
    run()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.get_piechart_dict">
    get_piechart_dict()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.get_pseudotime">
    get_pseudotime()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_piechart_graph">
    plot_piechart_graph()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_stream">
    plot_stream()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_trajectory_gams">
    plot_trajectory_gams()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_lineage_probability">
    plot_lineage_probability()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_gene_trend">
    plot_gene_trend()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_gene_trend_heatmap">
    plot_gene_trend_heatmap()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._via.pyVIA.plot_clustergraph">
    plot_clustergraph()
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api via</h1>
<div class="doc doc-object doc-class">
<h2 class="doc doc-heading" id="omicverse.single.pyVIA">
<code>omicverse.single.pyVIA</code>
<a class="headerlink" href="#omicverse.single.pyVIA" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p class="doc doc-class-bases">
        Bases: <code>object</code></p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 21 "></span><span class="k">class</span> <span class="nc">pyVIA</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="linenos" data-linenos=" 22 "></span>
<span class="linenos" data-linenos=" 23 "></span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">adata_key</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'X_pca'</span><span class="p">,</span><span class="n">adata_ncomps</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'X_umap'</span><span class="p">,</span>
<span class="linenos" data-linenos=" 24 "></span>                 <span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">dist_std_local</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">jac_std_global</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 25 "></span>                 <span class="n">keep_all_local_dist</span><span class="o">=</span><span class="s1">'auto'</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">resolution_parameter</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">partition_type</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s2">"ModularityVP"</span><span class="p">,</span> <span class="n">small_pop</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="linenos" data-linenos=" 26 "></span>                 <span class="n">jac_weighted_edges</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">knn</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_iter_leiden</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
<span class="linenos" data-linenos=" 27 "></span>                 <span class="n">num_threads</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">'l2'</span><span class="p">,</span> <span class="n">time_smallpop</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
<span class="linenos" data-linenos=" 28 "></span>                 <span class="n">super_cluster_labels</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>                 <span class="n">super_node_degree_list</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">super_terminal_cells</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">x_lazy</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">alpha_teleport</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.99</span><span class="p">,</span>
<span class="linenos" data-linenos=" 29 "></span>                 <span class="n">root_user</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">super_terminal_clusters</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span>
<span class="linenos" data-linenos=" 30 "></span>                 <span class="n">is_coarse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">csr_full_graph</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">csr_array_locally_pruned</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">ig_full_graph</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos=" 31 "></span>                 <span class="n">full_neighbor_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">full_distance_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>  <span class="n">df_annot</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 32 "></span>                 <span class="n">preserve_disconnected_after_pruning</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos=" 33 "></span>                 <span class="n">secondary_annotations</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pseudotime_threshold_TS</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cluster_graph_pruning_std</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span>
<span class="linenos" data-linenos=" 34 "></span>                 <span class="n">visual_cluster_graph_pruning</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">neighboring_terminal_states_threshold</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">num_mcmc_simulations</span><span class="o">=</span><span class="mi">1300</span><span class="p">,</span>
<span class="linenos" data-linenos=" 35 "></span>                 <span class="n">piegraph_arrow_head_width</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="linenos" data-linenos=" 36 "></span>                 <span class="n">piegraph_edgeweight_scalingfactor</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">max_visual_outgoing_edges</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">via_coarse</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velocity_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 37 "></span>                 <span class="n">gene_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velo_weight</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">edgebundle_pruning</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">A_velo</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">CSM</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">edgebundle_pruning_twice</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pca_loadings</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">time_series</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos=" 38 "></span>                 <span class="n">time_series_labels</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">knn_sequential</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">knn_sequential_reverse</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">t_diff_step</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">single_cell_transition_matrix</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 39 "></span>                 <span class="n">embedding_type</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'via-mds'</span><span class="p">,</span><span class="n">do_compute_embedding</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">color_dict</span><span class="p">:</span><span class="nb">dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">user_defined_terminal_cell</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span> <span class="n">user_defined_terminal_group</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span>
<span class="linenos" data-linenos=" 40 "></span>                 <span class="n">do_gaussian_kernel_edgeweights</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">RW2_mode</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">working_dir_fp</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span><span class="s1">'/home/shobi/Trajectory/Datasets/'</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos=" 41 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos=" 42 "></span><span class="sd">        Initialize a pyVIA object.</span>
<span class="linenos" data-linenos=" 43 "></span>
<span class="linenos" data-linenos=" 44 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos=" 45 "></span><span class="sd">            adata: An AnnData object containing the scRNA-seq.</span>
<span class="linenos" data-linenos=" 46 "></span><span class="sd">            adata_key: the key of the AnnData in obsm to perform VIA on. default: 'X_pca'</span>
<span class="linenos" data-linenos=" 47 "></span><span class="sd">            adata_ncomps: the number of components to use from the AnnData in obsm to perform VIA on. default: 80</span>
<span class="linenos" data-linenos=" 48 "></span><span class="sd">            basis: the key of the AnnData in obsm to use as the basis for the embedding. default: 'X_umap'</span>
<span class="linenos" data-linenos=" 49 "></span><span class="sd">            clusters: the clusters to use for the VIA analysis. default: ''</span>
<span class="linenos" data-linenos=" 50 "></span><span class="sd">            dist_std_local: local level of pruning for PARC graph clustering stage. Range (0.1,3) higher numbers mean more edge retention</span>
<span class="linenos" data-linenos=" 51 "></span><span class="sd">            jac_std_global: (optional, default = 0.15, can also set as 'median') global level graph pruning for PARC clustering stage. Number of standard deviations below the network’s mean-jaccard-weighted edges. 0.1-1 provide reasonable pruning.higher value means less pruning (more edges retained). e.g. a value of 0.15 means all edges that are above mean(edgeweight)-0.15*std(edge-weights) are retained. We find both 0.15 and ‘median’ to yield good results/starting point and resulting in pruning away ~ 50-60% edges</span>
<span class="linenos" data-linenos=" 52 "></span><span class="sd">            labels: default is None. and PARC clusters are used for the viagraph. alternatively provide a list of clustermemberships that are integer values (not strings) to construct the viagraph using another clustering method or available annotations</span>
<span class="linenos" data-linenos=" 53 "></span><span class="sd">            keep_all_local_dist: default value of 'auto' means that for smaller datasets local-pruning is done prior to clustering, but for large datasets local pruning is set to False for speed.</span>
<span class="linenos" data-linenos=" 54 "></span>
<span class="linenos" data-linenos=" 55 "></span><span class="sd">            too_big_factor: (optional, default=0.4). Forces clusters &gt; 0.4*n_cells to be re-clustered</span>
<span class="linenos" data-linenos=" 56 "></span><span class="sd">            resolution_parameter: (float) the resolution parameter for the Louvain algorithm.</span>
<span class="linenos" data-linenos=" 57 "></span><span class="sd">            partition_type: (str, default: "ModularityVP") the partitioning algorithm to use.</span>
<span class="linenos" data-linenos=" 58 "></span><span class="sd">            small_pop: (int, default: 10) the number of cells to be considered in a small population.</span>
<span class="linenos" data-linenos=" 59 "></span><span class="sd">            jac_weighted_edges: (bool, default: True) whether to use weighted edges in the PARC clustering step.</span>
<span class="linenos" data-linenos=" 60 "></span><span class="sd">            knn: (int, optional, default: 30) the number of K-Nearest Neighbors for HNSWlib KNN graph. Larger knn means more graph connectivity. Lower knn means more loosely connected clusters/cells.</span>
<span class="linenos" data-linenos=" 61 "></span><span class="sd">            n_iter_leiden: (int) the number of iterations for the Leiden algorithm.</span>
<span class="linenos" data-linenos=" 62 "></span><span class="sd">            random_seed: (int) the random seed to pass to the clustering algorithm.</span>
<span class="linenos" data-linenos=" 63 "></span><span class="sd">            num_threads: (int) the number of threads to use for the clustering algorithm.</span>
<span class="linenos" data-linenos=" 64 "></span><span class="sd">            distance: (str, default: 'l2') the distance metric to use for graph construction and similarity. Options are 'l2', 'ip', and 'cosine'.</span>
<span class="linenos" data-linenos=" 65 "></span><span class="sd">            visual_cluster_graph_pruning: (float, default: 0.15) the pruning level for the cluster graph. This only comes into play if the user deliberately chooses not to use the default edge-bundling method of visualizating edges (draw_piechart_graph()) and instead calls draw_piechart_graph_nobundle(). It controls the number of edges plotted for visual effect. This does not impact computation of terminal states, pseudotime, or lineage likelihoods.</span>
<span class="linenos" data-linenos=" 66 "></span><span class="sd">            cluster_graph_pruning_std: (float, default: 0.15) the pruning level of the cluster graph. Often set to the same value as the PARC clustering level of jac_std_global. Reasonable range is [0.1, 1]. To retain more connectivity in the clustergraph underlying the trajectory computations, increase the value.</span>
<span class="linenos" data-linenos=" 67 "></span><span class="sd">            time_smallpop: (max time to be allowed handling singletons) the maximum time allowed to handle singletons.</span>
<span class="linenos" data-linenos=" 68 "></span><span class="sd">            super_cluster_labels:</span>
<span class="linenos" data-linenos=" 69 "></span><span class="sd">            super_node_degree_list:</span>
<span class="linenos" data-linenos=" 70 "></span><span class="sd">            super_terminal_cells:</span>
<span class="linenos" data-linenos=" 71 "></span><span class="sd">            x_lazy: (float, default: 0.95) 1-x = probability of staying in the same node (lazy). Values between 0.9-0.99 are reasonable.</span>
<span class="linenos" data-linenos=" 72 "></span><span class="sd">            alpha_teleport: (float, default: 0.99) 1-alpha is probability of jumping. Values between 0.95-0.99 are reasonable unless prior knowledge of teleportation.</span>
<span class="linenos" data-linenos=" 73 "></span><span class="sd">            root_user: (list, None) the root user list. Can be a list of strings, a list of int, or None. The default is None. When the root_user is set as None and an RNA velocity matrix is available, a root will be automatically computed. If the root_user is None and no velocity matrix is provided, then an arbitrary root is selected. If the root_user is ['celltype_earlystage'] where the str corresponds to an item in true_label, then a suitable starting point will be selected corresponding to this group.</span>
<span class="linenos" data-linenos=" 74 "></span><span class="sd">            preserve_disconnected: bool (default = True) If you believe there may be disconnected trajectories then set this to False</span>
<span class="linenos" data-linenos=" 75 "></span><span class="sd">            dataset: str Can be set to 'group' or '' (default). this refers to the type of root label (group level root or single cell index) you are going to provide. if your true_label has a sensible group of cells for a root then you can set dataset to 'group' and make the root parameter ['labelname_root_cell_type'] if your root corresponds to one particular cell then set dataset = '' (default)</span>
<span class="linenos" data-linenos=" 76 "></span><span class="sd">            embedding: ndarray (optional, default = None) embedding (e.g. precomputed tsne, umap, phate, via-umap) for plotting data. Size n_cells x 2 If an embedding is provided when running VIA, then a scatterplot colored by pseudotime, highlighting terminal fates</span>
<span class="linenos" data-linenos=" 77 "></span><span class="sd">            velo_weight: float (optional, default = 0.5) #float between [0,1]. the weight assigned to directionality and connectivity derived from scRNA-velocity</span>
<span class="linenos" data-linenos=" 78 "></span><span class="sd">            neighboring_terminal_states_threshold:int (default = 3). Candidates for terminal states that are neighbors of each other may be removed from the list if they have this number of more of terminal states as neighbors</span>
<span class="linenos" data-linenos=" 79 "></span><span class="sd">            knn_sequential:int (default =10) number of knn in the adjacent time-point for time-series data (t_i and t_i+1)</span>
<span class="linenos" data-linenos=" 80 "></span><span class="sd">            knn_sequential_reverse: int (default = 0) number of knn enforced from current to previous time point</span>
<span class="linenos" data-linenos=" 81 "></span><span class="sd">            t_diff_step: int (default =1) Number of permitted temporal intervals between connected nodes. If time data is labeled as [0,25,50,75,100,..] then t_diff_step=1 corresponds to '25' and only edges within t_diff_steps are retained</span>
<span class="linenos" data-linenos=" 82 "></span><span class="sd">            is_coarse:bool (default = True) If running VIA in two iterations where you wish to link the second fine-grained iteration with the initial iteration, then you set to False</span>
<span class="linenos" data-linenos=" 83 "></span><span class="sd">            via_coarse: VIA (default = None) If instantiating a second iteration of VIA that needs to be linked to a previous iteration (e.g. via0), then set via_coarse to the previous via0 object</span>
<span class="linenos" data-linenos=" 84 "></span><span class="sd">            df_annot: DataFrame (default None) used for the Mouse Organ data</span>
<span class="linenos" data-linenos=" 85 "></span><span class="sd">            preserve_disconnected_after_pruning:bool (default = True) If you believe there are disconnected trajectories then set this to False and test your hypothesis</span>
<span class="linenos" data-linenos=" 86 "></span><span class="sd">            A_velo: ndarray Cluster Graph Transition matrix based on rna velocity [n_clus x n_clus]</span>
<span class="linenos" data-linenos=" 87 "></span><span class="sd">            velocity_matrix: matrix (default None) matrix of size [n_samples x n_genes]. this is the velocity matrix computed by scVelo (or similar package) and stored in adata.layers['velocity']. The genes used for computing velocity should correspond to those useing in gene_matrix Requires gene_matrix to be provided too.</span>
<span class="linenos" data-linenos=" 88 "></span><span class="sd">            gene_matrix: matrix (default None) Only used if Velocity_matrix is available. matrix of size [n_samples x n_genes]. We recommend using a subset like HVGs rather than full set of genes. (need to densify input if taking from adata = adata.X.todense())</span>
<span class="linenos" data-linenos=" 89 "></span><span class="sd">            time_series: if the data has time-series labels then set to True</span>
<span class="linenos" data-linenos=" 90 "></span><span class="sd">            time_series_labels:list (default None) list of integer values of temporal annotations corresponding to e.g. hours (post fert), days, or sequential ordering</span>
<span class="linenos" data-linenos=" 91 "></span><span class="sd">            pca_loadings: array (default None) the loadings of the pcs used to project the cells (to projected euclidean location based on velocity). n_cells x n_pcs</span>
<span class="linenos" data-linenos=" 92 "></span><span class="sd">            secondary_annotations: None (default None)</span>
<span class="linenos" data-linenos=" 93 "></span><span class="sd">            edgebundle_pruning:float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</span>
<span class="linenos" data-linenos=" 94 "></span><span class="sd">            edgebundle_pruning_twice:bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</span>
<span class="linenos" data-linenos=" 95 "></span><span class="sd">            piegraph_arrow_head_width: float (default = 0.1) size of arrow heads in via cluster graph</span>
<span class="linenos" data-linenos=" 96 "></span><span class="sd">            piegraph_edgeweight_scalingfactor: (defaulf = 1.5) scaling factor for edge thickness in via cluster graph</span>
<span class="linenos" data-linenos=" 97 "></span><span class="sd">            max_visual_outgoing_edges: int (default =2) Rarely comes into play. Only used if the user later chooses to plot the via-graph without edgebunding using draw_piechart_graph_nobundle() Only allows max_visual_outgoing_edges to come out of any given node.</span>
<span class="linenos" data-linenos=" 98 "></span><span class="sd">            edgebundle_pruning:float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</span>
<span class="linenos" data-linenos=" 99 "></span><span class="sd">            edgebundle_pruning_twice:bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</span>
<span class="linenos" data-linenos="100 "></span><span class="sd">            pseudotime_threshold_TS: int (default = 30) corresponds to the criteria for a state to be considered a candidate terminal cell fate to be 30% or later of the computed pseudotime range</span>
<span class="linenos" data-linenos="101 "></span><span class="sd">            num_mcmc_simulations:int (default = 1300) number of random walk simulations conducted</span>
<span class="linenos" data-linenos="102 "></span><span class="sd">            embedding_type: str (default = 'via-mds', other options are 'via-umap' and 'via-force'</span>
<span class="linenos" data-linenos="103 "></span><span class="sd">            do_compute_embedding: bool (default = False) If you want an embedding (n_samples x2) to be computed on the basis of the via sc graph then set this to True</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">            do_gaussian_kernel_edgeweights: bool (default = False) Type of edgeweighting on the graph edges</span>
<span class="linenos" data-linenos="105 "></span>
<span class="linenos" data-linenos="106 "></span>
<span class="linenos" data-linenos="107 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="108 "></span>
<span class="linenos" data-linenos="109 "></span>
<span class="linenos" data-linenos="110 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span>
<span class="linenos" data-linenos="111 "></span>        <span class="c1">#self.adata_key = adata_key</span>
<span class="linenos" data-linenos="112 "></span>        <span class="n">data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">adata_key</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">adata_ncomps</span><span class="p">]</span>
<span class="linenos" data-linenos="113 "></span>        <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="114 "></span>        <span class="n">true_label</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span>
<span class="linenos" data-linenos="115 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="116 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">basis</span><span class="o">=</span><span class="n">basis</span>
<span class="linenos" data-linenos="117 "></span>
<span class="linenos" data-linenos="118 "></span>        <span class="k">if</span> <span class="n">root_user</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="119 "></span>             <span class="n">dataset</span><span class="o">=</span><span class="s1">'group'</span>
<span class="linenos" data-linenos="120 "></span>
<span class="linenos" data-linenos="121 "></span>
<span class="linenos" data-linenos="122 "></span>
<span class="linenos" data-linenos="123 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="n">VIA</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span><span class="n">true_label</span><span class="o">=</span><span class="n">true_label</span><span class="p">,</span>
<span class="linenos" data-linenos="124 "></span>                 <span class="n">dist_std_local</span><span class="o">=</span><span class="n">dist_std_local</span><span class="p">,</span><span class="n">jac_std_global</span><span class="o">=</span><span class="n">jac_std_global</span><span class="p">,</span><span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
<span class="linenos" data-linenos="125 "></span>                 <span class="n">keep_all_local_dist</span><span class="o">=</span><span class="n">keep_all_local_dist</span><span class="p">,</span><span class="n">too_big_factor</span><span class="o">=</span><span class="n">too_big_factor</span><span class="p">,</span><span class="n">resolution_parameter</span><span class="o">=</span><span class="n">resolution_parameter</span><span class="p">,</span><span class="n">partition_type</span><span class="o">=</span><span class="n">partition_type</span><span class="p">,</span><span class="n">small_pop</span><span class="o">=</span><span class="n">small_pop</span><span class="p">,</span>
<span class="linenos" data-linenos="126 "></span>                 <span class="n">jac_weighted_edges</span><span class="o">=</span><span class="n">jac_weighted_edges</span><span class="p">,</span><span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span><span class="n">n_iter_leiden</span><span class="o">=</span><span class="n">n_iter_leiden</span><span class="p">,</span><span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="linenos" data-linenos="127 "></span>                 <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span><span class="n">distance</span><span class="o">=</span><span class="n">distance</span><span class="p">,</span><span class="n">time_smallpop</span><span class="o">=</span><span class="n">time_smallpop</span><span class="p">,</span>
<span class="linenos" data-linenos="128 "></span>                 <span class="n">super_cluster_labels</span><span class="o">=</span><span class="n">super_cluster_labels</span><span class="p">,</span><span class="n">super_node_degree_list</span><span class="o">=</span><span class="n">super_node_degree_list</span><span class="p">,</span><span class="n">super_terminal_cells</span><span class="o">=</span><span class="n">super_terminal_cells</span><span class="p">,</span><span class="n">x_lazy</span><span class="o">=</span><span class="n">x_lazy</span><span class="p">,</span><span class="n">alpha_teleport</span><span class="o">=</span><span class="n">alpha_teleport</span><span class="p">,</span>
<span class="linenos" data-linenos="129 "></span>                 <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span><span class="n">preserve_disconnected</span><span class="o">=</span><span class="n">preserve_disconnected</span><span class="p">,</span><span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span><span class="n">super_terminal_clusters</span><span class="o">=</span><span class="n">super_terminal_clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="130 "></span>                 <span class="n">is_coarse</span><span class="o">=</span><span class="n">is_coarse</span><span class="p">,</span><span class="n">csr_full_graph</span><span class="o">=</span><span class="n">csr_full_graph</span><span class="p">,</span><span class="n">csr_array_locally_pruned</span><span class="o">=</span><span class="n">csr_array_locally_pruned</span><span class="p">,</span><span class="n">ig_full_graph</span><span class="o">=</span><span class="n">ig_full_graph</span><span class="p">,</span>
<span class="linenos" data-linenos="131 "></span>                 <span class="n">full_neighbor_array</span><span class="o">=</span><span class="n">full_neighbor_array</span><span class="p">,</span><span class="n">full_distance_array</span><span class="o">=</span><span class="n">full_distance_array</span><span class="p">,</span><span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">df_annot</span><span class="o">=</span><span class="n">df_annot</span><span class="p">,</span>
<span class="linenos" data-linenos="132 "></span>                 <span class="n">preserve_disconnected_after_pruning</span><span class="o">=</span><span class="n">preserve_disconnected_after_pruning</span><span class="p">,</span>
<span class="linenos" data-linenos="133 "></span>                 <span class="n">secondary_annotations</span><span class="o">=</span><span class="n">secondary_annotations</span><span class="p">,</span><span class="n">pseudotime_threshold_TS</span><span class="o">=</span><span class="n">pseudotime_threshold_TS</span><span class="p">,</span><span class="n">cluster_graph_pruning_std</span><span class="o">=</span><span class="n">cluster_graph_pruning_std</span><span class="p">,</span>
<span class="linenos" data-linenos="134 "></span>                 <span class="n">visual_cluster_graph_pruning</span><span class="o">=</span><span class="n">visual_cluster_graph_pruning</span><span class="p">,</span><span class="n">neighboring_terminal_states_threshold</span><span class="o">=</span><span class="n">neighboring_terminal_states_threshold</span><span class="p">,</span><span class="n">num_mcmc_simulations</span><span class="o">=</span><span class="n">num_mcmc_simulations</span><span class="p">,</span>
<span class="linenos" data-linenos="135 "></span>                 <span class="n">piegraph_arrow_head_width</span><span class="o">=</span><span class="n">piegraph_arrow_head_width</span><span class="p">,</span>
<span class="linenos" data-linenos="136 "></span>                 <span class="n">piegraph_edgeweight_scalingfactor</span><span class="o">=</span><span class="n">piegraph_edgeweight_scalingfactor</span><span class="p">,</span><span class="n">max_visual_outgoing_edges</span><span class="o">=</span><span class="n">max_visual_outgoing_edges</span><span class="p">,</span><span class="n">via_coarse</span><span class="o">=</span><span class="n">via_coarse</span><span class="p">,</span><span class="n">velocity_matrix</span><span class="o">=</span><span class="n">velocity_matrix</span><span class="p">,</span>
<span class="linenos" data-linenos="137 "></span>                 <span class="n">gene_matrix</span><span class="o">=</span><span class="n">gene_matrix</span><span class="p">,</span><span class="n">velo_weight</span><span class="o">=</span><span class="n">velo_weight</span><span class="p">,</span><span class="n">edgebundle_pruning</span><span class="o">=</span><span class="n">edgebundle_pruning</span><span class="p">,</span><span class="n">A_velo</span><span class="o">=</span><span class="n">A_velo</span><span class="p">,</span><span class="n">CSM</span><span class="o">=</span><span class="n">CSM</span><span class="p">,</span><span class="n">edgebundle_pruning_twice</span><span class="o">=</span><span class="n">edgebundle_pruning_twice</span><span class="p">,</span><span class="n">pca_loadings</span><span class="o">=</span><span class="n">pca_loadings</span><span class="p">,</span><span class="n">time_series</span><span class="o">=</span><span class="n">time_series</span><span class="p">,</span>
<span class="linenos" data-linenos="138 "></span>                 <span class="n">time_series_labels</span><span class="o">=</span><span class="n">time_series_labels</span><span class="p">,</span><span class="n">knn_sequential</span><span class="o">=</span><span class="n">knn_sequential</span><span class="p">,</span><span class="n">knn_sequential_reverse</span><span class="o">=</span><span class="n">knn_sequential_reverse</span><span class="p">,</span><span class="n">t_diff_step</span><span class="o">=</span><span class="n">t_diff_step</span><span class="p">,</span><span class="n">single_cell_transition_matrix</span><span class="o">=</span><span class="n">single_cell_transition_matrix</span><span class="p">,</span>
<span class="linenos" data-linenos="139 "></span>                 <span class="n">embedding_type</span><span class="o">=</span><span class="n">embedding_type</span><span class="p">,</span><span class="n">do_compute_embedding</span><span class="o">=</span><span class="n">do_compute_embedding</span><span class="p">,</span><span class="n">color_dict</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span><span class="n">user_defined_terminal_cell</span><span class="o">=</span><span class="n">user_defined_terminal_cell</span><span class="p">,</span><span class="n">user_defined_terminal_group</span><span class="o">=</span><span class="n">user_defined_terminal_group</span><span class="p">,</span>
<span class="linenos" data-linenos="140 "></span>                 <span class="n">do_gaussian_kernel_edgeweights</span><span class="o">=</span><span class="n">do_gaussian_kernel_edgeweights</span><span class="p">,</span><span class="n">RW2_mode</span><span class="o">=</span><span class="n">RW2_mode</span><span class="p">,</span><span class="n">working_dir_fp</span><span class="o">=</span><span class="n">working_dir_fp</span>
<span class="linenos" data-linenos="141 "></span>                 <span class="p">)</span>
<span class="linenos" data-linenos="142 "></span>    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="143 "></span><span class="w">        </span><span class="sd">"""calculate the via graph and pseudotime</span>
<span class="linenos" data-linenos="144 "></span>
<span class="linenos" data-linenos="145 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="146 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
<span class="linenos" data-linenos="147 "></span>
<span class="linenos" data-linenos="148 "></span>    <span class="k">def</span> <span class="nf">get_piechart_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">label</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">)</span><span class="o">-&gt;</span><span class="nb">dict</span><span class="p">:</span>
<span class="linenos" data-linenos="149 "></span><span class="w">        </span><span class="sd">"""</span>
<span class="linenos" data-linenos="150 "></span><span class="sd">        Cluster composition graph</span>
<span class="linenos" data-linenos="151 "></span>
<span class="linenos" data-linenos="152 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="153 "></span><span class="sd">            label: int (default=0) cluster label of pie chart</span>
<span class="linenos" data-linenos="154 "></span><span class="sd">            clusters: the celltype you want interested</span>
<span class="linenos" data-linenos="155 "></span>
<span class="linenos" data-linenos="156 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="157 "></span><span class="sd">            res_dict: cluster composition graph</span>
<span class="linenos" data-linenos="158 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="159 "></span>        <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="160 "></span>            <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="161 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="162 "></span>        <span class="n">cluster_i_loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">label</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="163 "></span>        <span class="n">res_dict</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">cluster_i_loc</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">clusters</span><span class="p">))</span>
<span class="linenos" data-linenos="164 "></span>        <span class="k">return</span> <span class="n">res_dict</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span>    <span class="k">def</span> <span class="nf">get_pseudotime</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="167 "></span><span class="w">        </span><span class="sd">"""</span>
<span class="linenos" data-linenos="168 "></span><span class="sd">        Extract the pseudotime of VIA</span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="171 "></span><span class="sd">            adata: an adata object of you interested,if None, it will be added to `self.adata.obs['pt_via']`</span>
<span class="linenos" data-linenos="172 "></span>
<span class="linenos" data-linenos="173 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="174 "></span>
<span class="linenos" data-linenos="175 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...the pseudotime of VIA added to AnnData obs named `pt_via`'</span><span class="p">)</span>
<span class="linenos" data-linenos="176 "></span>        <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="177 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'pt_via'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">single_cell_pt_markov</span>
<span class="linenos" data-linenos="178 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="179 "></span>            <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'pt_via'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">single_cell_pt_markov</span>
<span class="linenos" data-linenos="180 "></span>
<span class="linenos" data-linenos="181 "></span>    <span class="k">def</span> <span class="nf">plot_piechart_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">'pt'</span><span class="p">,</span>
<span class="linenos" data-linenos="182 "></span>                                <span class="n">gene_exp</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> 
<span class="linenos" data-linenos="183 "></span>                                <span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="184 "></span>                                <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span><span class="n">headwidth_arrow</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> 
<span class="linenos" data-linenos="185 "></span>                                <span class="n">alpha_edge</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> 
<span class="linenos" data-linenos="186 "></span>                                <span class="n">edge_color</span><span class="o">=</span><span class="s1">'darkblue'</span><span class="p">,</span><span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
<span class="linenos" data-linenos="187 "></span>                                <span class="n">show_legend</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">fontsize</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="188 "></span>                                                                                                          <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span>
<span class="linenos" data-linenos="189 "></span>                                                                                                          <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="190 "></span><span class="w">        </span><span class="sd">"""plot two subplots with a clustergraph level representation of the viagraph showing true-label composition (lhs) and pseudotime/gene expression (rhs)</span>
<span class="linenos" data-linenos="191 "></span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">            clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">            type_data : string  default 'pt' for pseudotime colored nodes. or 'gene'</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">            gene_exp : list of values (column of dataframe) corresponding to feature or gene expression to be used to color nodes at CLUSTER level</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">            title : string</span>
<span class="linenos" data-linenos="197 "></span><span class="sd">            cmap : default None. automatically chooses coolwarm for gene expression or viridis_r for pseudotime</span>
<span class="linenos" data-linenos="198 "></span><span class="sd">            ax_text : Bool default= True. Annotates each node with cluster number and population of membership</span>
<span class="linenos" data-linenos="199 "></span><span class="sd">            dpi : int default = 150</span>
<span class="linenos" data-linenos="200 "></span><span class="sd">            headwidth_bundle : default = 0.1. width of arrowhead used to directed edges</span>
<span class="linenos" data-linenos="201 "></span><span class="sd">            reference : None or list. list of categorical (str) labels for cluster composition of the piecharts (LHS subplot) length = n_samples.</span>
<span class="linenos" data-linenos="202 "></span><span class="sd">            pie_size_scale : float default=0.8 scaling factor of the piechart nodes</span>
<span class="linenos" data-linenos="203 "></span><span class="sd">            fontsize : float default=8. fontsize of the text in the piecharts</span>
<span class="linenos" data-linenos="204 "></span><span class="sd">            figsize : tuple default=(8,4). size of the figure</span>
<span class="linenos" data-linenos="205 "></span>
<span class="linenos" data-linenos="206 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="207 "></span><span class="sd">            fig: Returns matplotlib figure with two axes that plot the clustergraph using edge bundling</span>
<span class="linenos" data-linenos="208 "></span><span class="sd">            ax: left axis shows the clustergraph with each node colored by annotated ground truth membership.</span>
<span class="linenos" data-linenos="209 "></span><span class="sd">            ax1: right axis shows the same clustergraph with each node colored by the pseudotime or gene expression</span>
<span class="linenos" data-linenos="210 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="211 "></span>
<span class="linenos" data-linenos="212 "></span>
<span class="linenos" data-linenos="213 "></span>        <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="214 "></span>            <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="215 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="216 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">draw_piechart_graph_pyomic</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="217 "></span>                                   <span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="n">type_data</span><span class="p">,</span>
<span class="linenos" data-linenos="218 "></span>                                <span class="n">gene_exp</span><span class="o">=</span><span class="n">gene_exp</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> 
<span class="linenos" data-linenos="219 "></span>                                <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="n">ax_text</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="220 "></span>                                <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span><span class="n">headwidth_arrow</span> <span class="o">=</span> <span class="n">headwidth_arrow</span><span class="p">,</span>
<span class="linenos" data-linenos="221 "></span>                                <span class="n">alpha_edge</span><span class="o">=</span><span class="n">alpha_edge</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="n">linewidth_edge</span><span class="p">,</span>
<span class="linenos" data-linenos="222 "></span>                                <span class="n">edge_color</span><span class="o">=</span><span class="n">edge_color</span><span class="p">,</span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
<span class="linenos" data-linenos="223 "></span>                                <span class="n">show_legend</span><span class="o">=</span><span class="n">show_legend</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="o">=</span><span class="n">pie_size_scale</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="224 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span>
<span class="linenos" data-linenos="225 "></span>
<span class="linenos" data-linenos="226 "></span>    <span class="k">def</span> <span class="nf">plot_stream</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos="227 "></span>                   <span class="n">density_grid</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">arrow_size</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">arrow_color</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="s1">'k'</span><span class="p">,</span>
<span class="linenos" data-linenos="228 "></span>                   <span class="n">arrow_style</span><span class="o">=</span><span class="s2">"-|&gt;"</span><span class="p">,</span>  <span class="n">max_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">min_mass</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cutoff_perc</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
<span class="linenos" data-linenos="229 "></span>                   <span class="n">scatter_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">marker_edgewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="linenos" data-linenos="230 "></span>                   <span class="n">density_stream</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="linenos" data-linenos="231 "></span>                   <span class="n">color_scheme</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="s1">'annotation'</span><span class="p">,</span> <span class="n">add_outline_clusters</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos="232 "></span>                   <span class="n">cluster_outline_edgewidth</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span><span class="n">gp_color</span> <span class="o">=</span> <span class="s1">'white'</span><span class="p">,</span> <span class="n">bg_color</span><span class="o">=</span><span class="s1">'black'</span> <span class="p">,</span>
<span class="linenos" data-linenos="233 "></span>                   <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Streamplot'</span><span class="p">,</span> <span class="n">b_bias</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_neighbors_velocity_grid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="234 "></span>                   <span class="n">other_labels</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span><span class="n">use_sequentially_augmented</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap_str</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'rainbow'</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="235 "></span>                                                                                                          <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="236 "></span><span class="w">        </span><span class="sd">"""Construct vector streamplot on the embedding to show a fine-grained view of inferred directions in the trajectory</span>
<span class="linenos" data-linenos="237 "></span>
<span class="linenos" data-linenos="238 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="239 "></span><span class="sd">            clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="240 "></span><span class="sd">            basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="241 "></span><span class="sd">            density_grid : float, default = 0.5, density of the grid on which to project the directionality of cells</span>
<span class="linenos" data-linenos="242 "></span><span class="sd">            arrow_size : float, default = 0.7, size of the arrows in the streamplot</span>
<span class="linenos" data-linenos="243 "></span><span class="sd">            arrow_color : str, default = 'k', color of the arrows in the streamplot</span>
<span class="linenos" data-linenos="244 "></span><span class="sd">            arrow_style : str, default = "-|&gt;", style of the arrows in the streamplot</span>
<span class="linenos" data-linenos="245 "></span><span class="sd">            max_length : int, default = 4, maximum length of the arrows in the streamplot</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">            linewidth : float, default = 1, width of  lines in streamplot</span>
<span class="linenos" data-linenos="247 "></span><span class="sd">            min_mass : float, default = 1, minimum mass of the arrows in the streamplot</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">            cutoff_perc : int, default = 5, cutoff percentage of the arrows in the streamplot</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">            scatter_size : int, default = 500, size of scatter points</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">            scatter_alpha : float, default = 0.5, transpsarency of scatter points</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">            marker_edgewidth : float, default = 0.1, width of outline arround each scatter point</span>
<span class="linenos" data-linenos="252 "></span><span class="sd">            density_stream : int, default = 2, density of the streamplot</span>
<span class="linenos" data-linenos="253 "></span><span class="sd">            smooth_transition : int, default = 1, smoothness of the transition between the streamplot and the scatter points</span>
<span class="linenos" data-linenos="254 "></span><span class="sd">            smooth_grid : float, default = 0.5, smoothness of the grid on which to project the directionality of cells</span>
<span class="linenos" data-linenos="255 "></span><span class="sd">            color_scheme : str, default = 'annotation' corresponds to self.true_labels. Other options are 'time' (uses single-cell pseudotime) or 'clusters' (uses self.clusters)</span>
<span class="linenos" data-linenos="256 "></span><span class="sd">            add_outline_clusters : bool, default = False, whether to add an outline to the clusters</span>
<span class="linenos" data-linenos="257 "></span><span class="sd">            cluster_outline_edgewidth : float, default = 0.001, width of the outline around the clusters</span>
<span class="linenos" data-linenos="258 "></span><span class="sd">            gp_color : str, default = 'white', color of the grid points</span>
<span class="linenos" data-linenos="259 "></span><span class="sd">            bg_color : str, default = 'black', color of the background</span>
<span class="linenos" data-linenos="260 "></span><span class="sd">            dpi : int, default = 80, dpi of the figure</span>
<span class="linenos" data-linenos="261 "></span><span class="sd">            title : str, default = 'Streamplot', title of the figure</span>
<span class="linenos" data-linenos="262 "></span><span class="sd">            b_bias : int, default = 20, higher value makes the forward bias of pseudotime stronger</span>
<span class="linenos" data-linenos="263 "></span><span class="sd">            n_neighbors_velocity_grid : int, default = None, number of neighbors to use for the velocity grid</span>
<span class="linenos" data-linenos="264 "></span><span class="sd">            other_labels : list, default = None, list of other labels to plot in the streamplot</span>
<span class="linenos" data-linenos="265 "></span><span class="sd">            use_sequentially_augmented : bool, default = False, whether to use the sequentially augmented data</span>
<span class="linenos" data-linenos="266 "></span><span class="sd">            cmap_str : str, default = 'rainbow', color map to use for the streamplot</span>
<span class="linenos" data-linenos="267 "></span>
<span class="linenos" data-linenos="268 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="270 "></span><span class="sd">            ax : matplotlib axis</span>
<span class="linenos" data-linenos="271 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="272 "></span>
<span class="linenos" data-linenos="273 "></span>        <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="274 "></span>            <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="275 "></span>        <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="276 "></span>            <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="277 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>        <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="279 "></span>        <span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">via_streamplot_pyomic</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> 
<span class="linenos" data-linenos="280 "></span>                                 <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">density_grid</span><span class="o">=</span><span class="n">density_grid</span><span class="p">,</span> <span class="n">arrow_size</span><span class="o">=</span><span class="n">arrow_size</span><span class="p">,</span>
<span class="linenos" data-linenos="281 "></span>                                 <span class="n">arrow_color</span><span class="o">=</span><span class="n">arrow_color</span><span class="p">,</span><span class="n">arrow_style</span><span class="o">=</span><span class="n">arrow_style</span><span class="p">,</span>  <span class="n">max_length</span><span class="o">=</span><span class="n">max_length</span><span class="p">,</span>
<span class="linenos" data-linenos="282 "></span>                                 <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span><span class="n">min_mass</span> <span class="o">=</span> <span class="n">min_mass</span><span class="p">,</span> <span class="n">cutoff_perc</span><span class="o">=</span><span class="n">cutoff_perc</span><span class="p">,</span>
<span class="linenos" data-linenos="283 "></span>                                 <span class="n">scatter_size</span><span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="n">scatter_alpha</span><span class="p">,</span><span class="n">marker_edgewidth</span><span class="o">=</span><span class="n">marker_edgewidth</span><span class="p">,</span>
<span class="linenos" data-linenos="284 "></span>                                 <span class="n">density_stream</span><span class="o">=</span><span class="n">density_stream</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="o">=</span><span class="n">smooth_transition</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="o">=</span><span class="n">smooth_grid</span><span class="p">,</span>
<span class="linenos" data-linenos="285 "></span>                                 <span class="n">color_scheme</span><span class="o">=</span><span class="n">color_scheme</span><span class="p">,</span> <span class="n">add_outline_clusters</span><span class="o">=</span><span class="n">add_outline_clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="286 "></span>                                 <span class="n">cluster_outline_edgewidth</span> <span class="o">=</span> <span class="n">cluster_outline_edgewidth</span><span class="p">,</span><span class="n">gp_color</span> <span class="o">=</span> <span class="n">gp_color</span><span class="p">,</span> <span class="n">bg_color</span><span class="o">=</span><span class="n">bg_color</span><span class="p">,</span>
<span class="linenos" data-linenos="287 "></span>                                 <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">b_bias</span><span class="o">=</span><span class="n">b_bias</span><span class="p">,</span> <span class="n">n_neighbors_velocity_grid</span><span class="o">=</span><span class="n">n_neighbors_velocity_grid</span><span class="p">,</span>
<span class="linenos" data-linenos="288 "></span>                                 <span class="n">other_labels</span><span class="o">=</span><span class="n">other_labels</span><span class="p">,</span><span class="n">use_sequentially_augmented</span><span class="o">=</span><span class="n">use_sequentially_augmented</span><span class="p">,</span> <span class="n">cmap_str</span><span class="o">=</span><span class="n">cmap_str</span><span class="p">)</span>
<span class="linenos" data-linenos="289 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax</span>
<span class="linenos" data-linenos="290 "></span>
<span class="linenos" data-linenos="291 "></span>    <span class="k">def</span> <span class="nf">plot_trajectory_gams</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="292 "></span>                         <span class="n">title_str</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span> <span class="s2">"Pseudotime"</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">arrow_width_scale_factor</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>
<span class="linenos" data-linenos="293 "></span>                         <span class="n">scatter_size</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="294 "></span>                         <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap_pseudotime</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis_r'</span><span class="p">,</span><span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span>
<span class="linenos" data-linenos="295 "></span>                         <span class="n">highlight_terminal_states</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_maxout_edgelist</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="296 "></span>                                                                                                 <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span>
<span class="linenos" data-linenos="297 "></span>                                                                                                 <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="298 "></span><span class="w">        </span><span class="sd">"""projects the graph based coarse trajectory onto a umap/tsne embedding</span>
<span class="linenos" data-linenos="299 "></span>
<span class="linenos" data-linenos="300 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="301 "></span><span class="sd">            clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="302 "></span><span class="sd">            basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="303 "></span><span class="sd">            via_fine : via object suggest to use via_object only unless you found that running via_fine gave better pathways</span>
<span class="linenos" data-linenos="304 "></span><span class="sd">            idx : default: None. Or List. if you had previously computed a umap/tsne (embedding) only on a subset of the total n_samples (subsampled as per idx), then the via objects and results will be indexed according to idx too</span>
<span class="linenos" data-linenos="305 "></span><span class="sd">            title_str : title of figure</span>
<span class="linenos" data-linenos="306 "></span><span class="sd">            draw_all_curves : if the clustergraph has too many edges to project in a visually interpretable way, set this to False to get a simplified view of the graph pathways</span>
<span class="linenos" data-linenos="307 "></span><span class="sd">            arrow_width_scale_factor : the width of the arrows is proportional to the edge weight. This factor scales the width of the arrows</span>
<span class="linenos" data-linenos="308 "></span><span class="sd">            scatter_size : size of the scatter points</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">            scatter_alpha : transparency of the scatter points</span>
<span class="linenos" data-linenos="310 "></span><span class="sd">            linewidth : width of the lines</span>
<span class="linenos" data-linenos="311 "></span><span class="sd">            marker_edgewidth : width of the outline around each scatter point</span>
<span class="linenos" data-linenos="312 "></span><span class="sd">            cmap_pseudotime : color map to use for the pseudotime</span>
<span class="linenos" data-linenos="313 "></span><span class="sd">            dpi : dpi of the figure</span>
<span class="linenos" data-linenos="314 "></span><span class="sd">            highlight_terminal_states :  whether or not to highlight/distinguish the clusters which are detected as the terminal states by via</span>
<span class="linenos" data-linenos="315 "></span>
<span class="linenos" data-linenos="316 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="317 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="318 "></span><span class="sd">            ax1 : matplotlib axis</span>
<span class="linenos" data-linenos="319 "></span><span class="sd">            ax2 : matplotlib axis</span>
<span class="linenos" data-linenos="320 "></span>
<span class="linenos" data-linenos="321 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="322 "></span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span>        <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="325 "></span>            <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="326 "></span>        <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="327 "></span>            <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="328 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="329 "></span>        <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="330 "></span>        <span class="n">fig</span><span class="p">,</span><span class="n">ax1</span><span class="p">,</span><span class="n">ax2</span> <span class="o">=</span> <span class="n">draw_trajectory_gams_pyomic</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> 
<span class="linenos" data-linenos="331 "></span>                                            <span class="n">via_fine</span><span class="o">=</span><span class="n">via_fine</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span>
<span class="linenos" data-linenos="332 "></span>                                            <span class="n">title_str</span><span class="o">=</span><span class="n">title_str</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="o">=</span><span class="n">draw_all_curves</span><span class="p">,</span> <span class="n">arrow_width_scale_factor</span><span class="o">=</span><span class="n">arrow_width_scale_factor</span><span class="p">,</span>
<span class="linenos" data-linenos="333 "></span>                                            <span class="n">scatter_size</span><span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="n">scatter_alpha</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="334 "></span>                                            <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="n">marker_edgewidth</span><span class="p">,</span> <span class="n">cmap_pseudotime</span><span class="o">=</span><span class="n">cmap_pseudotime</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span>
<span class="linenos" data-linenos="335 "></span>                                            <span class="n">highlight_terminal_states</span><span class="o">=</span><span class="n">highlight_terminal_states</span><span class="p">,</span> <span class="n">use_maxout_edgelist</span><span class="o">=</span><span class="n">use_maxout_edgelist</span><span class="p">)</span>
<span class="linenos" data-linenos="336 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax1</span><span class="p">,</span><span class="n">ax2</span>
<span class="linenos" data-linenos="337 "></span>
<span class="linenos" data-linenos="338 "></span>    <span class="k">def</span> <span class="nf">plot_lineage_probability</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
<span class="linenos" data-linenos="339 "></span>                                <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="340 "></span>                                <span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'plasma'</span><span class="p">,</span> <span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">scatter_size</span> <span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="341 "></span>                                <span class="n">marker_lineages</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="342 "></span>                                                                                   <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="343 "></span><span class="w">        </span><span class="sd">"""G is the igraph knn (low K) used for shortest path in high dim space. no idx needed as it's made on full sample, knn_hnsw is the knn made in the embedded space used for query to find the nearest point in the downsampled embedding that corresponds to the single cells in the full graph</span>
<span class="linenos" data-linenos="344 "></span>
<span class="linenos" data-linenos="345 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="346 "></span><span class="sd">            clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="347 "></span><span class="sd">            basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="348 "></span><span class="sd">            via_fine : usually just set to same as via_coarse unless you ran a refined run and want to link it to initial via_coarse's terminal clusters</span>
<span class="linenos" data-linenos="349 "></span><span class="sd">            idx : if one uses a downsampled embedding of the original data, then idx is the selected indices of the downsampled samples used in the visualization</span>
<span class="linenos" data-linenos="350 "></span><span class="sd">            figsize : size of the figure</span>
<span class="linenos" data-linenos="351 "></span><span class="sd">            cmap : color map to use for the lineage probability</span>
<span class="linenos" data-linenos="352 "></span><span class="sd">            dpi : dpi of the figure</span>
<span class="linenos" data-linenos="353 "></span><span class="sd">            scatter_size : size of the scatter points</span>
<span class="linenos" data-linenos="354 "></span><span class="sd">            marker_lineages : Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</span>
<span class="linenos" data-linenos="355 "></span><span class="sd">            fontsize : fontsize of the title</span>
<span class="linenos" data-linenos="356 "></span>
<span class="linenos" data-linenos="357 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="358 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="359 "></span><span class="sd">            axs : matplotlib axis</span>
<span class="linenos" data-linenos="360 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="361 "></span>
<span class="linenos" data-linenos="362 "></span>
<span class="linenos" data-linenos="363 "></span>        <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="364 "></span>            <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="365 "></span>        <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="366 "></span>            <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="367 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="368 "></span>        <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="369 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">draw_sc_lineage_probability</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="n">via_fine</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="370 "></span>                                               <span class="n">idx</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">cmap_name</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span> <span class="n">scatter_size</span> <span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span>
<span class="linenos" data-linenos="371 "></span>                                            <span class="n">marker_lineages</span> <span class="o">=</span> <span class="n">marker_lineages</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="372 "></span>        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="373 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span>
<span class="linenos" data-linenos="374 "></span>
<span class="linenos" data-linenos="375 "></span>    <span class="k">def</span> <span class="nf">plot_gene_trend</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="376 "></span>                        <span class="n">magic_steps</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">spline_order</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'jet'</span><span class="p">,</span> 
<span class="linenos" data-linenos="377 "></span>                        <span class="n">marker_genes</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="mf">2.0</span><span class="p">,</span>
<span class="linenos" data-linenos="378 "></span>                        <span class="n">n_splines</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>  <span class="n">fontsize_</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[])</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="379 "></span>                                                                                        <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="380 "></span><span class="w">        </span><span class="sd">"""plots the gene expression trend along the pseudotime</span>
<span class="linenos" data-linenos="381 "></span>
<span class="linenos" data-linenos="382 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="383 "></span><span class="sd">            gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="384 "></span><span class="sd">            figsize : size of the figure</span>
<span class="linenos" data-linenos="385 "></span><span class="sd">            magic_steps : number of magic steps to use for imputation</span>
<span class="linenos" data-linenos="386 "></span><span class="sd">            spline_order : order of the spline to use for smoothing</span>
<span class="linenos" data-linenos="387 "></span><span class="sd">            dpi : dpi of the figure</span>
<span class="linenos" data-linenos="388 "></span><span class="sd">            cmap : color map to use for the gene expression</span>
<span class="linenos" data-linenos="389 "></span><span class="sd">            marker_genes : Default is to use all genes in gene_exp. other provide a list of marker genes that will be used from gene_exp.</span>
<span class="linenos" data-linenos="390 "></span><span class="sd">            linewidth : width of the lines</span>
<span class="linenos" data-linenos="391 "></span><span class="sd">            n_splines : number of splines to use for smoothing</span>
<span class="linenos" data-linenos="392 "></span><span class="sd">            fontsize_ : fontsize of the title</span>
<span class="linenos" data-linenos="393 "></span><span class="sd">            marker_lineages : Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</span>
<span class="linenos" data-linenos="394 "></span>
<span class="linenos" data-linenos="395 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="396 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="397 "></span><span class="sd">            axs : matplotlib axis</span>
<span class="linenos" data-linenos="398 "></span>
<span class="linenos" data-linenos="399 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="400 "></span>
<span class="linenos" data-linenos="401 "></span>        <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="n">magic_steps</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="402 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span><span class="o">=</span><span class="n">get_gene_expression_pyomic</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span><span class="n">df_magic</span><span class="p">,</span><span class="n">spline_order</span><span class="o">=</span><span class="n">spline_order</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span>
<span class="linenos" data-linenos="403 "></span>                                   <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">marker_genes</span><span class="o">=</span><span class="n">marker_genes</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="404 "></span>                                   <span class="n">n_splines</span><span class="o">=</span><span class="n">n_splines</span><span class="p">,</span>  <span class="n">fontsize_</span><span class="o">=</span><span class="n">fontsize_</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="n">marker_lineages</span><span class="p">)</span>
<span class="linenos" data-linenos="405 "></span>        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="406 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span>
<span class="linenos" data-linenos="407 "></span>
<span class="linenos" data-linenos="408 "></span>    <span class="k">def</span> <span class="nf">plot_clustergraph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">arrow_head</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
<span class="linenos" data-linenos="409 "></span>                          <span class="n">edgeweight_scale</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">label_</span><span class="o">=</span><span class="kc">True</span><span class="p">,)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="410 "></span>                                                                                        <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="411 "></span><span class="w">        </span><span class="sd">"""plot the gene in pie chart for each cluster</span>
<span class="linenos" data-linenos="412 "></span>
<span class="linenos" data-linenos="413 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="414 "></span><span class="sd">            gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="415 "></span><span class="sd">            arrow_head : size of the arrow head</span>
<span class="linenos" data-linenos="416 "></span><span class="sd">            figsize : size of the figure</span>
<span class="linenos" data-linenos="417 "></span><span class="sd">            edgeweight_scale : scale of the edge weight</span>
<span class="linenos" data-linenos="418 "></span><span class="sd">            cmap : color map to use for the gene expression</span>
<span class="linenos" data-linenos="419 "></span><span class="sd">            label_ : whether to label the nodes</span>
<span class="linenos" data-linenos="420 "></span>
<span class="linenos" data-linenos="421 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="422 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="423 "></span><span class="sd">            axs : matplotlib axis</span>
<span class="linenos" data-linenos="424 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="425 "></span>        <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="n">magic_steps</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="426 "></span>        <span class="n">df_magic</span><span class="p">[</span><span class="s1">'parc'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span>
<span class="linenos" data-linenos="427 "></span>        <span class="n">df_magic_cluster</span> <span class="o">=</span> <span class="n">df_magic</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">'parc'</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="428 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">draw_clustergraph_pyomic</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">'gene'</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="n">df_magic_cluster</span><span class="p">,</span> 
<span class="linenos" data-linenos="429 "></span>                                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">arrow_head</span><span class="o">=</span><span class="n">arrow_head</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="430 "></span>                                    <span class="n">edgeweight_scale</span><span class="o">=</span><span class="n">edgeweight_scale</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">label_</span><span class="o">=</span><span class="n">label_</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>
<span class="linenos" data-linenos="431 "></span>        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="432 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">axs</span>
<span class="linenos" data-linenos="433 "></span>
<span class="linenos" data-linenos="434 "></span>    <span class="k">def</span> <span class="nf">plot_gene_trend_heatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">marker_lineages</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> 
<span class="linenos" data-linenos="435 "></span>                             <span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span><span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">normalize</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ytick_labelrotation</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> 
<span class="linenos" data-linenos="436 "></span>                             <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="437 "></span>                                                                    <span class="nb">list</span><span class="p">]:</span>
<span class="linenos" data-linenos="438 "></span><span class="w">        </span><span class="sd">"""Plot the gene trends on heatmap: a heatmap is generated for each lineage (identified by terminal cluster number). Default selects all lineages</span>
<span class="linenos" data-linenos="439 "></span>
<span class="linenos" data-linenos="440 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="441 "></span><span class="sd">            gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="442 "></span><span class="sd">            marker_lineages : list default = None and plots all detected all lineages. Optionally provide a list of integers corresponding to the cluster number of terminal cell fates</span>
<span class="linenos" data-linenos="443 "></span><span class="sd">            fontsize : int default = 8</span>
<span class="linenos" data-linenos="444 "></span><span class="sd">            cmap : str default = 'viridis'</span>
<span class="linenos" data-linenos="445 "></span><span class="sd">            normalize : bool = True</span>
<span class="linenos" data-linenos="446 "></span><span class="sd">            ytick_labelrotation : int default = 0</span>
<span class="linenos" data-linenos="447 "></span><span class="sd">            figsize : size of the figure</span>
<span class="linenos" data-linenos="448 "></span>
<span class="linenos" data-linenos="449 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="450 "></span><span class="sd">            fig : matplotlib figure</span>
<span class="linenos" data-linenos="451 "></span><span class="sd">            axs : list of matplotlib axis       </span>
<span class="linenos" data-linenos="452 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="453 "></span>
<span class="linenos" data-linenos="454 "></span>        <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="455 "></span>        <span class="n">df_magic</span><span class="p">[</span><span class="s1">'parc'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span>
<span class="linenos" data-linenos="456 "></span>        <span class="n">df_magic_cluster</span> <span class="o">=</span> <span class="n">df_magic</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">'parc'</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="457 "></span>        <span class="n">fig</span><span class="p">,</span><span class="n">axs</span><span class="o">=</span><span class="n">plot_gene_trend_heatmaps_pyomic</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">df_gene_exp</span><span class="o">=</span><span class="n">df_magic</span><span class="p">,</span> 
<span class="linenos" data-linenos="458 "></span>                                                <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="n">normalize</span><span class="p">,</span>
<span class="linenos" data-linenos="459 "></span>                                                <span class="n">ytick_labelrotation</span><span class="o">=</span><span class="n">ytick_labelrotation</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="460 "></span>                                                <span class="n">marker_lineages</span><span class="o">=</span><span class="n">marker_lineages</span><span class="p">)</span>
<span class="linenos" data-linenos="461 "></span>        <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="462 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">axs</span>
</code></pre></div>
</details>
<div class="doc doc-children">
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_key</span><span class="o">=</span><span class="s1">'X_pca'</span><span class="p">,</span> <span class="n">adata_ncomps</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">'X_umap'</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">dist_std_local</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">jac_std_global</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_all_local_dist</span><span class="o">=</span><span class="s1">'auto'</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">resolution_parameter</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">partition_type</span><span class="o">=</span><span class="s1">'ModularityVP'</span><span class="p">,</span> <span class="n">small_pop</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">jac_weighted_edges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">knn</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_iter_leiden</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span> <span class="n">num_threads</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">'l2'</span><span class="p">,</span> <span class="n">time_smallpop</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">super_cluster_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">super_node_degree_list</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">super_terminal_cells</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">x_lazy</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">alpha_teleport</span><span class="o">=</span><span class="mf">0.99</span><span class="p">,</span> <span class="n">root_user</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">super_terminal_clusters</span><span class="o">=</span><span class="p">[],</span> <span class="n">is_coarse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">csr_full_graph</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">csr_array_locally_pruned</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">ig_full_graph</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">full_neighbor_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">full_distance_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">df_annot</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">preserve_disconnected_after_pruning</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">secondary_annotations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pseudotime_threshold_TS</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cluster_graph_pruning_std</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">visual_cluster_graph_pruning</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">neighboring_terminal_states_threshold</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">num_mcmc_simulations</span><span class="o">=</span><span class="mi">1300</span><span class="p">,</span> <span class="n">piegraph_arrow_head_width</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">piegraph_edgeweight_scalingfactor</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">max_visual_outgoing_edges</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">via_coarse</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velocity_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velo_weight</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">edgebundle_pruning</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">A_velo</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">CSM</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">edgebundle_pruning_twice</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pca_loadings</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">time_series</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">time_series_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">knn_sequential</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">knn_sequential_reverse</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">t_diff_step</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">single_cell_transition_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">embedding_type</span><span class="o">=</span><span class="s1">'via-mds'</span><span class="p">,</span> <span class="n">do_compute_embedding</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">color_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">user_defined_terminal_cell</span><span class="o">=</span><span class="p">[],</span> <span class="n">user_defined_terminal_group</span><span class="o">=</span><span class="p">[],</span> <span class="n">do_gaussian_kernel_edgeweights</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">RW2_mode</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">working_dir_fp</span><span class="o">=</span><span class="s1">'/home/shobi/Trajectory/Datasets/'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.__init__" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Initialize a pyVIA object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>An AnnData object containing the scRNA-seq.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>adata_key</code></td>
<td>
<code>str</code>
</td>
<td><p>the key of the AnnData in obsm to perform VIA on. default: 'X_pca'</p></td>
<td>
<code>'X_pca'</code>
</td>
</tr>
<tr>
<td><code>adata_ncomps</code></td>
<td>
<code>int</code>
</td>
<td><p>the number of components to use from the AnnData in obsm to perform VIA on. default: 80</p></td>
<td>
<code>80</code>
</td>
</tr>
<tr>
<td><code>basis</code></td>
<td>
<code>str</code>
</td>
<td><p>the key of the AnnData in obsm to use as the basis for the embedding. default: 'X_umap'</p></td>
<td>
<code>'X_umap'</code>
</td>
</tr>
<tr>
<td><code>clusters</code></td>
<td>
<code>str</code>
</td>
<td><p>the clusters to use for the VIA analysis. default: ''</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>dist_std_local</code></td>
<td>
<code>float</code>
</td>
<td><p>local level of pruning for PARC graph clustering stage. Range (0.1,3) higher numbers mean more edge retention</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>jac_std_global</code></td>
<td>
</td>
<td><p>(optional, default = 0.15, can also set as 'median') global level graph pruning for PARC clustering stage. Number of standard deviations below the network’s mean-jaccard-weighted edges. 0.1-1 provide reasonable pruning.higher value means less pruning (more edges retained). e.g. a value of 0.15 means all edges that are above mean(edgeweight)-0.15*std(edge-weights) are retained. We find both 0.15 and ‘median’ to yield good results/starting point and resulting in pruning away ~ 50-60% edges</p></td>
<td>
<code>0.15</code>
</td>
</tr>
<tr>
<td><code>labels</code></td>
<td>
<code><span title="omicverse.via.plotting_via.np">np</span>.<span title="omicverse.via.plotting_via.np.ndarray">ndarray</span></code>
</td>
<td><p>default is None. and PARC clusters are used for the viagraph. alternatively provide a list of clustermemberships that are integer values (not strings) to construct the viagraph using another clustering method or available annotations</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>keep_all_local_dist</code></td>
<td>
</td>
<td><p>default value of 'auto' means that for smaller datasets local-pruning is done prior to clustering, but for large datasets local pruning is set to False for speed.</p></td>
<td>
<code>'auto'</code>
</td>
</tr>
<tr>
<td><code>too_big_factor</code></td>
<td>
<code>float</code>
</td>
<td><p>(optional, default=0.4). Forces clusters &gt; 0.4*n_cells to be re-clustered</p></td>
<td>
<code>0.4</code>
</td>
</tr>
<tr>
<td><code>resolution_parameter</code></td>
<td>
<code>float</code>
</td>
<td><p>(float) the resolution parameter for the Louvain algorithm.</p></td>
<td>
<code>1.0</code>
</td>
</tr>
<tr>
<td><code>partition_type</code></td>
<td>
<code>str</code>
</td>
<td><p>(str, default: "ModularityVP") the partitioning algorithm to use.</p></td>
<td>
<code>'ModularityVP'</code>
</td>
</tr>
<tr>
<td><code>small_pop</code></td>
<td>
<code>int</code>
</td>
<td><p>(int, default: 10) the number of cells to be considered in a small population.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>jac_weighted_edges</code></td>
<td>
<code>bool</code>
</td>
<td><p>(bool, default: True) whether to use weighted edges in the PARC clustering step.</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>knn</code></td>
<td>
<code>int</code>
</td>
<td><p>(int, optional, default: 30) the number of K-Nearest Neighbors for HNSWlib KNN graph. Larger knn means more graph connectivity. Lower knn means more loosely connected clusters/cells.</p></td>
<td>
<code>30</code>
</td>
</tr>
<tr>
<td><code>n_iter_leiden</code></td>
<td>
<code>int</code>
</td>
<td><p>(int) the number of iterations for the Leiden algorithm.</p></td>
<td>
<code>5</code>
</td>
</tr>
<tr>
<td><code>random_seed</code></td>
<td>
<code>int</code>
</td>
<td><p>(int) the random seed to pass to the clustering algorithm.</p></td>
<td>
<code>42</code>
</td>
</tr>
<tr>
<td><code>num_threads</code></td>
<td>
</td>
<td><p>(int) the number of threads to use for the clustering algorithm.</p></td>
<td>
<code>-1</code>
</td>
</tr>
<tr>
<td><code>distance</code></td>
<td>
</td>
<td><p>(str, default: 'l2') the distance metric to use for graph construction and similarity. Options are 'l2', 'ip', and 'cosine'.</p></td>
<td>
<code>'l2'</code>
</td>
</tr>
<tr>
<td><code>visual_cluster_graph_pruning</code></td>
<td>
<code>float</code>
</td>
<td><p>(float, default: 0.15) the pruning level for the cluster graph. This only comes into play if the user deliberately chooses not to use the default edge-bundling method of visualizating edges (draw_piechart_graph()) and instead calls draw_piechart_graph_nobundle(). It controls the number of edges plotted for visual effect. This does not impact computation of terminal states, pseudotime, or lineage likelihoods.</p></td>
<td>
<code>0.15</code>
</td>
</tr>
<tr>
<td><code>cluster_graph_pruning_std</code></td>
<td>
<code>float</code>
</td>
<td><p>(float, default: 0.15) the pruning level of the cluster graph. Often set to the same value as the PARC clustering level of jac_std_global. Reasonable range is [0.1, 1]. To retain more connectivity in the clustergraph underlying the trajectory computations, increase the value.</p></td>
<td>
<code>0.15</code>
</td>
</tr>
<tr>
<td><code>time_smallpop</code></td>
<td>
</td>
<td><p>(max time to be allowed handling singletons) the maximum time allowed to handle singletons.</p></td>
<td>
<code>15</code>
</td>
</tr>
<tr>
<td><code>super_cluster_labels</code></td>
<td>
<code>bool</code>
</td>
<td></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>super_node_degree_list</code></td>
<td>
<code>bool</code>
</td>
<td></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>super_terminal_cells</code></td>
<td>
<code>bool</code>
</td>
<td></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>x_lazy</code></td>
<td>
<code>float</code>
</td>
<td><p>(float, default: 0.95) 1-x = probability of staying in the same node (lazy). Values between 0.9-0.99 are reasonable.</p></td>
<td>
<code>0.95</code>
</td>
</tr>
<tr>
<td><code>alpha_teleport</code></td>
<td>
<code>float</code>
</td>
<td><p>(float, default: 0.99) 1-alpha is probability of jumping. Values between 0.95-0.99 are reasonable unless prior knowledge of teleportation.</p></td>
<td>
<code>0.99</code>
</td>
</tr>
<tr>
<td><code>root_user</code></td>
<td>
</td>
<td><p>(list, None) the root user list. Can be a list of strings, a list of int, or None. The default is None. When the root_user is set as None and an RNA velocity matrix is available, a root will be automatically computed. If the root_user is None and no velocity matrix is provided, then an arbitrary root is selected. If the root_user is ['celltype_earlystage'] where the str corresponds to an item in true_label, then a suitable starting point will be selected corresponding to this group.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>preserve_disconnected</code></td>
<td>
<code>bool</code>
</td>
<td><p>bool (default = True) If you believe there may be disconnected trajectories then set this to False</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>dataset</code></td>
<td>
<code>str</code>
</td>
<td><p>str Can be set to 'group' or '' (default). this refers to the type of root label (group level root or single cell index) you are going to provide. if your true_label has a sensible group of cells for a root then you can set dataset to 'group' and make the root parameter ['labelname_root_cell_type'] if your root corresponds to one particular cell then set dataset = '' (default)</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>embedding</code></td>
<td>
</td>
<td><p>ndarray (optional, default = None) embedding (e.g. precomputed tsne, umap, phate, via-umap) for plotting data. Size n_cells x 2 If an embedding is provided when running VIA, then a scatterplot colored by pseudotime, highlighting terminal fates</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>velo_weight</code></td>
<td>
</td>
<td><p>float (optional, default = 0.5) #float between [0,1]. the weight assigned to directionality and connectivity derived from scRNA-velocity</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>neighboring_terminal_states_threshold</code></td>
<td>
</td>
<td><p>int (default = 3). Candidates for terminal states that are neighbors of each other may be removed from the list if they have this number of more of terminal states as neighbors</p></td>
<td>
<code>3</code>
</td>
</tr>
<tr>
<td><code>knn_sequential</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default =10) number of knn in the adjacent time-point for time-series data (t_i and t_i+1)</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>knn_sequential_reverse</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default = 0) number of knn enforced from current to previous time point</p></td>
<td>
<code>0</code>
</td>
</tr>
<tr>
<td><code>t_diff_step</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default =1) Number of permitted temporal intervals between connected nodes. If time data is labeled as [0,25,50,75,100,..] then t_diff_step=1 corresponds to '25' and only edges within t_diff_steps are retained</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>is_coarse</code></td>
<td>
</td>
<td><p>bool (default = True) If running VIA in two iterations where you wish to link the second fine-grained iteration with the initial iteration, then you set to False</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>via_coarse</code></td>
<td>
</td>
<td><p>VIA (default = None) If instantiating a second iteration of VIA that needs to be linked to a previous iteration (e.g. via0), then set via_coarse to the previous via0 object</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>df_annot</code></td>
<td>
</td>
<td><p>DataFrame (default None) used for the Mouse Organ data</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>preserve_disconnected_after_pruning</code></td>
<td>
<code>bool</code>
</td>
<td><p>bool (default = True) If you believe there are disconnected trajectories then set this to False and test your hypothesis</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>A_velo</code></td>
<td>
</td>
<td><p>ndarray Cluster Graph Transition matrix based on rna velocity [n_clus x n_clus]</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>velocity_matrix</code></td>
<td>
</td>
<td><p>matrix (default None) matrix of size [n_samples x n_genes]. this is the velocity matrix computed by scVelo (or similar package) and stored in adata.layers['velocity']. The genes used for computing velocity should correspond to those useing in gene_matrix Requires gene_matrix to be provided too.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>gene_matrix</code></td>
<td>
</td>
<td><p>matrix (default None) Only used if Velocity_matrix is available. matrix of size [n_samples x n_genes]. We recommend using a subset like HVGs rather than full set of genes. (need to densify input if taking from adata = adata.X.todense())</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>time_series</code></td>
<td>
</td>
<td><p>if the data has time-series labels then set to True</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>time_series_labels</code></td>
<td>
<code>list</code>
</td>
<td><p>list (default None) list of integer values of temporal annotations corresponding to e.g. hours (post fert), days, or sequential ordering</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>pca_loadings</code></td>
<td>
</td>
<td><p>array (default None) the loadings of the pcs used to project the cells (to projected euclidean location based on velocity). n_cells x n_pcs</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>secondary_annotations</code></td>
<td>
<code>list</code>
</td>
<td><p>None (default None)</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>edgebundle_pruning</code></td>
<td>
</td>
<td><p>float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>edgebundle_pruning_twice</code></td>
<td>
</td>
<td><p>bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>piegraph_arrow_head_width</code></td>
<td>
</td>
<td><p>float (default = 0.1) size of arrow heads in via cluster graph</p></td>
<td>
<code>0.1</code>
</td>
</tr>
<tr>
<td><code>piegraph_edgeweight_scalingfactor</code></td>
<td>
</td>
<td><p>(defaulf = 1.5) scaling factor for edge thickness in via cluster graph</p></td>
<td>
<code>1.5</code>
</td>
</tr>
<tr>
<td><code>max_visual_outgoing_edges</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default =2) Rarely comes into play. Only used if the user later chooses to plot the via-graph without edgebunding using draw_piechart_graph_nobundle() Only allows max_visual_outgoing_edges to come out of any given node.</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>edgebundle_pruning</code></td>
<td>
</td>
<td><p>float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>edgebundle_pruning_twice</code></td>
<td>
</td>
<td><p>bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>pseudotime_threshold_TS</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default = 30) corresponds to the criteria for a state to be considered a candidate terminal cell fate to be 30% or later of the computed pseudotime range</p></td>
<td>
<code>30</code>
</td>
</tr>
<tr>
<td><code>num_mcmc_simulations</code></td>
<td>
</td>
<td><p>int (default = 1300) number of random walk simulations conducted</p></td>
<td>
<code>1300</code>
</td>
</tr>
<tr>
<td><code>embedding_type</code></td>
<td>
<code>str</code>
</td>
<td><p>str (default = 'via-mds', other options are 'via-umap' and 'via-force'</p></td>
<td>
<code>'via-mds'</code>
</td>
</tr>
<tr>
<td><code>do_compute_embedding</code></td>
<td>
<code>bool</code>
</td>
<td><p>bool (default = False) If you want an embedding (n_samples x2) to be computed on the basis of the via sc graph then set this to True</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>do_gaussian_kernel_edgeweights</code></td>
<td>
<code>bool</code>
</td>
<td><p>bool (default = False) Type of edgeweighting on the graph edges</p></td>
<td>
<code>False</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 23 "></span><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">adata_key</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'X_pca'</span><span class="p">,</span><span class="n">adata_ncomps</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'X_umap'</span><span class="p">,</span>
<span class="linenos" data-linenos=" 24 "></span>             <span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">dist_std_local</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">jac_std_global</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">labels</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 25 "></span>             <span class="n">keep_all_local_dist</span><span class="o">=</span><span class="s1">'auto'</span><span class="p">,</span> <span class="n">too_big_factor</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">resolution_parameter</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">partition_type</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s2">"ModularityVP"</span><span class="p">,</span> <span class="n">small_pop</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="linenos" data-linenos=" 26 "></span>             <span class="n">jac_weighted_edges</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">knn</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_iter_leiden</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_seed</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
<span class="linenos" data-linenos=" 27 "></span>             <span class="n">num_threads</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s1">'l2'</span><span class="p">,</span> <span class="n">time_smallpop</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
<span class="linenos" data-linenos=" 28 "></span>             <span class="n">super_cluster_labels</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>                 <span class="n">super_node_degree_list</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">super_terminal_cells</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">x_lazy</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">alpha_teleport</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.99</span><span class="p">,</span>
<span class="linenos" data-linenos=" 29 "></span>             <span class="n">root_user</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">preserve_disconnected</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dataset</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">super_terminal_clusters</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span>
<span class="linenos" data-linenos=" 30 "></span>             <span class="n">is_coarse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">csr_full_graph</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">csr_array_locally_pruned</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">ig_full_graph</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos=" 31 "></span>             <span class="n">full_neighbor_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">full_distance_array</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>  <span class="n">df_annot</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 32 "></span>             <span class="n">preserve_disconnected_after_pruning</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos=" 33 "></span>             <span class="n">secondary_annotations</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pseudotime_threshold_TS</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cluster_graph_pruning_std</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span>
<span class="linenos" data-linenos=" 34 "></span>             <span class="n">visual_cluster_graph_pruning</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.15</span><span class="p">,</span> <span class="n">neighboring_terminal_states_threshold</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">num_mcmc_simulations</span><span class="o">=</span><span class="mi">1300</span><span class="p">,</span>
<span class="linenos" data-linenos=" 35 "></span>             <span class="n">piegraph_arrow_head_width</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="linenos" data-linenos=" 36 "></span>             <span class="n">piegraph_edgeweight_scalingfactor</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">max_visual_outgoing_edges</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">via_coarse</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velocity_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 37 "></span>             <span class="n">gene_matrix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">velo_weight</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">edgebundle_pruning</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">A_velo</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">CSM</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">edgebundle_pruning_twice</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pca_loadings</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">time_series</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos=" 38 "></span>             <span class="n">time_series_labels</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">knn_sequential</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">knn_sequential_reverse</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span><span class="n">t_diff_step</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">single_cell_transition_matrix</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 39 "></span>             <span class="n">embedding_type</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'via-mds'</span><span class="p">,</span><span class="n">do_compute_embedding</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">color_dict</span><span class="p">:</span><span class="nb">dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">user_defined_terminal_cell</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span> <span class="n">user_defined_terminal_group</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span>
<span class="linenos" data-linenos=" 40 "></span>             <span class="n">do_gaussian_kernel_edgeweights</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">RW2_mode</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">working_dir_fp</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span><span class="s1">'/home/shobi/Trajectory/Datasets/'</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos=" 41 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos=" 42 "></span><span class="sd">    Initialize a pyVIA object.</span>
<span class="linenos" data-linenos=" 43 "></span>
<span class="linenos" data-linenos=" 44 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos=" 45 "></span><span class="sd">        adata: An AnnData object containing the scRNA-seq.</span>
<span class="linenos" data-linenos=" 46 "></span><span class="sd">        adata_key: the key of the AnnData in obsm to perform VIA on. default: 'X_pca'</span>
<span class="linenos" data-linenos=" 47 "></span><span class="sd">        adata_ncomps: the number of components to use from the AnnData in obsm to perform VIA on. default: 80</span>
<span class="linenos" data-linenos=" 48 "></span><span class="sd">        basis: the key of the AnnData in obsm to use as the basis for the embedding. default: 'X_umap'</span>
<span class="linenos" data-linenos=" 49 "></span><span class="sd">        clusters: the clusters to use for the VIA analysis. default: ''</span>
<span class="linenos" data-linenos=" 50 "></span><span class="sd">        dist_std_local: local level of pruning for PARC graph clustering stage. Range (0.1,3) higher numbers mean more edge retention</span>
<span class="linenos" data-linenos=" 51 "></span><span class="sd">        jac_std_global: (optional, default = 0.15, can also set as 'median') global level graph pruning for PARC clustering stage. Number of standard deviations below the network’s mean-jaccard-weighted edges. 0.1-1 provide reasonable pruning.higher value means less pruning (more edges retained). e.g. a value of 0.15 means all edges that are above mean(edgeweight)-0.15*std(edge-weights) are retained. We find both 0.15 and ‘median’ to yield good results/starting point and resulting in pruning away ~ 50-60% edges</span>
<span class="linenos" data-linenos=" 52 "></span><span class="sd">        labels: default is None. and PARC clusters are used for the viagraph. alternatively provide a list of clustermemberships that are integer values (not strings) to construct the viagraph using another clustering method or available annotations</span>
<span class="linenos" data-linenos=" 53 "></span><span class="sd">        keep_all_local_dist: default value of 'auto' means that for smaller datasets local-pruning is done prior to clustering, but for large datasets local pruning is set to False for speed.</span>
<span class="linenos" data-linenos=" 54 "></span>
<span class="linenos" data-linenos=" 55 "></span><span class="sd">        too_big_factor: (optional, default=0.4). Forces clusters &gt; 0.4*n_cells to be re-clustered</span>
<span class="linenos" data-linenos=" 56 "></span><span class="sd">        resolution_parameter: (float) the resolution parameter for the Louvain algorithm.</span>
<span class="linenos" data-linenos=" 57 "></span><span class="sd">        partition_type: (str, default: "ModularityVP") the partitioning algorithm to use.</span>
<span class="linenos" data-linenos=" 58 "></span><span class="sd">        small_pop: (int, default: 10) the number of cells to be considered in a small population.</span>
<span class="linenos" data-linenos=" 59 "></span><span class="sd">        jac_weighted_edges: (bool, default: True) whether to use weighted edges in the PARC clustering step.</span>
<span class="linenos" data-linenos=" 60 "></span><span class="sd">        knn: (int, optional, default: 30) the number of K-Nearest Neighbors for HNSWlib KNN graph. Larger knn means more graph connectivity. Lower knn means more loosely connected clusters/cells.</span>
<span class="linenos" data-linenos=" 61 "></span><span class="sd">        n_iter_leiden: (int) the number of iterations for the Leiden algorithm.</span>
<span class="linenos" data-linenos=" 62 "></span><span class="sd">        random_seed: (int) the random seed to pass to the clustering algorithm.</span>
<span class="linenos" data-linenos=" 63 "></span><span class="sd">        num_threads: (int) the number of threads to use for the clustering algorithm.</span>
<span class="linenos" data-linenos=" 64 "></span><span class="sd">        distance: (str, default: 'l2') the distance metric to use for graph construction and similarity. Options are 'l2', 'ip', and 'cosine'.</span>
<span class="linenos" data-linenos=" 65 "></span><span class="sd">        visual_cluster_graph_pruning: (float, default: 0.15) the pruning level for the cluster graph. This only comes into play if the user deliberately chooses not to use the default edge-bundling method of visualizating edges (draw_piechart_graph()) and instead calls draw_piechart_graph_nobundle(). It controls the number of edges plotted for visual effect. This does not impact computation of terminal states, pseudotime, or lineage likelihoods.</span>
<span class="linenos" data-linenos=" 66 "></span><span class="sd">        cluster_graph_pruning_std: (float, default: 0.15) the pruning level of the cluster graph. Often set to the same value as the PARC clustering level of jac_std_global. Reasonable range is [0.1, 1]. To retain more connectivity in the clustergraph underlying the trajectory computations, increase the value.</span>
<span class="linenos" data-linenos=" 67 "></span><span class="sd">        time_smallpop: (max time to be allowed handling singletons) the maximum time allowed to handle singletons.</span>
<span class="linenos" data-linenos=" 68 "></span><span class="sd">        super_cluster_labels:</span>
<span class="linenos" data-linenos=" 69 "></span><span class="sd">        super_node_degree_list:</span>
<span class="linenos" data-linenos=" 70 "></span><span class="sd">        super_terminal_cells:</span>
<span class="linenos" data-linenos=" 71 "></span><span class="sd">        x_lazy: (float, default: 0.95) 1-x = probability of staying in the same node (lazy). Values between 0.9-0.99 are reasonable.</span>
<span class="linenos" data-linenos=" 72 "></span><span class="sd">        alpha_teleport: (float, default: 0.99) 1-alpha is probability of jumping. Values between 0.95-0.99 are reasonable unless prior knowledge of teleportation.</span>
<span class="linenos" data-linenos=" 73 "></span><span class="sd">        root_user: (list, None) the root user list. Can be a list of strings, a list of int, or None. The default is None. When the root_user is set as None and an RNA velocity matrix is available, a root will be automatically computed. If the root_user is None and no velocity matrix is provided, then an arbitrary root is selected. If the root_user is ['celltype_earlystage'] where the str corresponds to an item in true_label, then a suitable starting point will be selected corresponding to this group.</span>
<span class="linenos" data-linenos=" 74 "></span><span class="sd">        preserve_disconnected: bool (default = True) If you believe there may be disconnected trajectories then set this to False</span>
<span class="linenos" data-linenos=" 75 "></span><span class="sd">        dataset: str Can be set to 'group' or '' (default). this refers to the type of root label (group level root or single cell index) you are going to provide. if your true_label has a sensible group of cells for a root then you can set dataset to 'group' and make the root parameter ['labelname_root_cell_type'] if your root corresponds to one particular cell then set dataset = '' (default)</span>
<span class="linenos" data-linenos=" 76 "></span><span class="sd">        embedding: ndarray (optional, default = None) embedding (e.g. precomputed tsne, umap, phate, via-umap) for plotting data. Size n_cells x 2 If an embedding is provided when running VIA, then a scatterplot colored by pseudotime, highlighting terminal fates</span>
<span class="linenos" data-linenos=" 77 "></span><span class="sd">        velo_weight: float (optional, default = 0.5) #float between [0,1]. the weight assigned to directionality and connectivity derived from scRNA-velocity</span>
<span class="linenos" data-linenos=" 78 "></span><span class="sd">        neighboring_terminal_states_threshold:int (default = 3). Candidates for terminal states that are neighbors of each other may be removed from the list if they have this number of more of terminal states as neighbors</span>
<span class="linenos" data-linenos=" 79 "></span><span class="sd">        knn_sequential:int (default =10) number of knn in the adjacent time-point for time-series data (t_i and t_i+1)</span>
<span class="linenos" data-linenos=" 80 "></span><span class="sd">        knn_sequential_reverse: int (default = 0) number of knn enforced from current to previous time point</span>
<span class="linenos" data-linenos=" 81 "></span><span class="sd">        t_diff_step: int (default =1) Number of permitted temporal intervals between connected nodes. If time data is labeled as [0,25,50,75,100,..] then t_diff_step=1 corresponds to '25' and only edges within t_diff_steps are retained</span>
<span class="linenos" data-linenos=" 82 "></span><span class="sd">        is_coarse:bool (default = True) If running VIA in two iterations where you wish to link the second fine-grained iteration with the initial iteration, then you set to False</span>
<span class="linenos" data-linenos=" 83 "></span><span class="sd">        via_coarse: VIA (default = None) If instantiating a second iteration of VIA that needs to be linked to a previous iteration (e.g. via0), then set via_coarse to the previous via0 object</span>
<span class="linenos" data-linenos=" 84 "></span><span class="sd">        df_annot: DataFrame (default None) used for the Mouse Organ data</span>
<span class="linenos" data-linenos=" 85 "></span><span class="sd">        preserve_disconnected_after_pruning:bool (default = True) If you believe there are disconnected trajectories then set this to False and test your hypothesis</span>
<span class="linenos" data-linenos=" 86 "></span><span class="sd">        A_velo: ndarray Cluster Graph Transition matrix based on rna velocity [n_clus x n_clus]</span>
<span class="linenos" data-linenos=" 87 "></span><span class="sd">        velocity_matrix: matrix (default None) matrix of size [n_samples x n_genes]. this is the velocity matrix computed by scVelo (or similar package) and stored in adata.layers['velocity']. The genes used for computing velocity should correspond to those useing in gene_matrix Requires gene_matrix to be provided too.</span>
<span class="linenos" data-linenos=" 88 "></span><span class="sd">        gene_matrix: matrix (default None) Only used if Velocity_matrix is available. matrix of size [n_samples x n_genes]. We recommend using a subset like HVGs rather than full set of genes. (need to densify input if taking from adata = adata.X.todense())</span>
<span class="linenos" data-linenos=" 89 "></span><span class="sd">        time_series: if the data has time-series labels then set to True</span>
<span class="linenos" data-linenos=" 90 "></span><span class="sd">        time_series_labels:list (default None) list of integer values of temporal annotations corresponding to e.g. hours (post fert), days, or sequential ordering</span>
<span class="linenos" data-linenos=" 91 "></span><span class="sd">        pca_loadings: array (default None) the loadings of the pcs used to project the cells (to projected euclidean location based on velocity). n_cells x n_pcs</span>
<span class="linenos" data-linenos=" 92 "></span><span class="sd">        secondary_annotations: None (default None)</span>
<span class="linenos" data-linenos=" 93 "></span><span class="sd">        edgebundle_pruning:float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</span>
<span class="linenos" data-linenos=" 94 "></span><span class="sd">        edgebundle_pruning_twice:bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</span>
<span class="linenos" data-linenos=" 95 "></span><span class="sd">        piegraph_arrow_head_width: float (default = 0.1) size of arrow heads in via cluster graph</span>
<span class="linenos" data-linenos=" 96 "></span><span class="sd">        piegraph_edgeweight_scalingfactor: (defaulf = 1.5) scaling factor for edge thickness in via cluster graph</span>
<span class="linenos" data-linenos=" 97 "></span><span class="sd">        max_visual_outgoing_edges: int (default =2) Rarely comes into play. Only used if the user later chooses to plot the via-graph without edgebunding using draw_piechart_graph_nobundle() Only allows max_visual_outgoing_edges to come out of any given node.</span>
<span class="linenos" data-linenos=" 98 "></span><span class="sd">        edgebundle_pruning:float (default=None) will by default be set to the same as the cluster_graph_pruning_std and influences the visualized level of pruning of edges. Typical values can be between [0,1] with higher numbers retaining more edges</span>
<span class="linenos" data-linenos=" 99 "></span><span class="sd">        edgebundle_pruning_twice:bool default: False. When True, the edgebundling is applied to a further visually pruned (visual_cluster_graph_pruning) and can sometimes simplify the visualization. it does not impact the pseudotime and lineage computations</span>
<span class="linenos" data-linenos="100 "></span><span class="sd">        pseudotime_threshold_TS: int (default = 30) corresponds to the criteria for a state to be considered a candidate terminal cell fate to be 30% or later of the computed pseudotime range</span>
<span class="linenos" data-linenos="101 "></span><span class="sd">        num_mcmc_simulations:int (default = 1300) number of random walk simulations conducted</span>
<span class="linenos" data-linenos="102 "></span><span class="sd">        embedding_type: str (default = 'via-mds', other options are 'via-umap' and 'via-force'</span>
<span class="linenos" data-linenos="103 "></span><span class="sd">        do_compute_embedding: bool (default = False) If you want an embedding (n_samples x2) to be computed on the basis of the via sc graph then set this to True</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">        do_gaussian_kernel_edgeweights: bool (default = False) Type of edgeweighting on the graph edges</span>
<span class="linenos" data-linenos="105 "></span>
<span class="linenos" data-linenos="106 "></span>
<span class="linenos" data-linenos="107 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="108 "></span>
<span class="linenos" data-linenos="109 "></span>
<span class="linenos" data-linenos="110 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span>
<span class="linenos" data-linenos="111 "></span>    <span class="c1">#self.adata_key = adata_key</span>
<span class="linenos" data-linenos="112 "></span>    <span class="n">data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">adata_key</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="n">adata_ncomps</span><span class="p">]</span>
<span class="linenos" data-linenos="113 "></span>    <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="114 "></span>    <span class="n">true_label</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span>
<span class="linenos" data-linenos="115 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="116 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">basis</span><span class="o">=</span><span class="n">basis</span>
<span class="linenos" data-linenos="117 "></span>
<span class="linenos" data-linenos="118 "></span>    <span class="k">if</span> <span class="n">root_user</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="119 "></span>         <span class="n">dataset</span><span class="o">=</span><span class="s1">'group'</span>
<span class="linenos" data-linenos="120 "></span>
<span class="linenos" data-linenos="121 "></span>
<span class="linenos" data-linenos="122 "></span>
<span class="linenos" data-linenos="123 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">=</span><span class="n">VIA</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span><span class="n">true_label</span><span class="o">=</span><span class="n">true_label</span><span class="p">,</span>
<span class="linenos" data-linenos="124 "></span>             <span class="n">dist_std_local</span><span class="o">=</span><span class="n">dist_std_local</span><span class="p">,</span><span class="n">jac_std_global</span><span class="o">=</span><span class="n">jac_std_global</span><span class="p">,</span><span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span>
<span class="linenos" data-linenos="125 "></span>             <span class="n">keep_all_local_dist</span><span class="o">=</span><span class="n">keep_all_local_dist</span><span class="p">,</span><span class="n">too_big_factor</span><span class="o">=</span><span class="n">too_big_factor</span><span class="p">,</span><span class="n">resolution_parameter</span><span class="o">=</span><span class="n">resolution_parameter</span><span class="p">,</span><span class="n">partition_type</span><span class="o">=</span><span class="n">partition_type</span><span class="p">,</span><span class="n">small_pop</span><span class="o">=</span><span class="n">small_pop</span><span class="p">,</span>
<span class="linenos" data-linenos="126 "></span>             <span class="n">jac_weighted_edges</span><span class="o">=</span><span class="n">jac_weighted_edges</span><span class="p">,</span><span class="n">knn</span><span class="o">=</span><span class="n">knn</span><span class="p">,</span><span class="n">n_iter_leiden</span><span class="o">=</span><span class="n">n_iter_leiden</span><span class="p">,</span><span class="n">random_seed</span><span class="o">=</span><span class="n">random_seed</span><span class="p">,</span>
<span class="linenos" data-linenos="127 "></span>             <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span><span class="n">distance</span><span class="o">=</span><span class="n">distance</span><span class="p">,</span><span class="n">time_smallpop</span><span class="o">=</span><span class="n">time_smallpop</span><span class="p">,</span>
<span class="linenos" data-linenos="128 "></span>             <span class="n">super_cluster_labels</span><span class="o">=</span><span class="n">super_cluster_labels</span><span class="p">,</span><span class="n">super_node_degree_list</span><span class="o">=</span><span class="n">super_node_degree_list</span><span class="p">,</span><span class="n">super_terminal_cells</span><span class="o">=</span><span class="n">super_terminal_cells</span><span class="p">,</span><span class="n">x_lazy</span><span class="o">=</span><span class="n">x_lazy</span><span class="p">,</span><span class="n">alpha_teleport</span><span class="o">=</span><span class="n">alpha_teleport</span><span class="p">,</span>
<span class="linenos" data-linenos="129 "></span>             <span class="n">root_user</span><span class="o">=</span><span class="n">root_user</span><span class="p">,</span><span class="n">preserve_disconnected</span><span class="o">=</span><span class="n">preserve_disconnected</span><span class="p">,</span><span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span><span class="n">super_terminal_clusters</span><span class="o">=</span><span class="n">super_terminal_clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="130 "></span>             <span class="n">is_coarse</span><span class="o">=</span><span class="n">is_coarse</span><span class="p">,</span><span class="n">csr_full_graph</span><span class="o">=</span><span class="n">csr_full_graph</span><span class="p">,</span><span class="n">csr_array_locally_pruned</span><span class="o">=</span><span class="n">csr_array_locally_pruned</span><span class="p">,</span><span class="n">ig_full_graph</span><span class="o">=</span><span class="n">ig_full_graph</span><span class="p">,</span>
<span class="linenos" data-linenos="131 "></span>             <span class="n">full_neighbor_array</span><span class="o">=</span><span class="n">full_neighbor_array</span><span class="p">,</span><span class="n">full_distance_array</span><span class="o">=</span><span class="n">full_distance_array</span><span class="p">,</span><span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">df_annot</span><span class="o">=</span><span class="n">df_annot</span><span class="p">,</span>
<span class="linenos" data-linenos="132 "></span>             <span class="n">preserve_disconnected_after_pruning</span><span class="o">=</span><span class="n">preserve_disconnected_after_pruning</span><span class="p">,</span>
<span class="linenos" data-linenos="133 "></span>             <span class="n">secondary_annotations</span><span class="o">=</span><span class="n">secondary_annotations</span><span class="p">,</span><span class="n">pseudotime_threshold_TS</span><span class="o">=</span><span class="n">pseudotime_threshold_TS</span><span class="p">,</span><span class="n">cluster_graph_pruning_std</span><span class="o">=</span><span class="n">cluster_graph_pruning_std</span><span class="p">,</span>
<span class="linenos" data-linenos="134 "></span>             <span class="n">visual_cluster_graph_pruning</span><span class="o">=</span><span class="n">visual_cluster_graph_pruning</span><span class="p">,</span><span class="n">neighboring_terminal_states_threshold</span><span class="o">=</span><span class="n">neighboring_terminal_states_threshold</span><span class="p">,</span><span class="n">num_mcmc_simulations</span><span class="o">=</span><span class="n">num_mcmc_simulations</span><span class="p">,</span>
<span class="linenos" data-linenos="135 "></span>             <span class="n">piegraph_arrow_head_width</span><span class="o">=</span><span class="n">piegraph_arrow_head_width</span><span class="p">,</span>
<span class="linenos" data-linenos="136 "></span>             <span class="n">piegraph_edgeweight_scalingfactor</span><span class="o">=</span><span class="n">piegraph_edgeweight_scalingfactor</span><span class="p">,</span><span class="n">max_visual_outgoing_edges</span><span class="o">=</span><span class="n">max_visual_outgoing_edges</span><span class="p">,</span><span class="n">via_coarse</span><span class="o">=</span><span class="n">via_coarse</span><span class="p">,</span><span class="n">velocity_matrix</span><span class="o">=</span><span class="n">velocity_matrix</span><span class="p">,</span>
<span class="linenos" data-linenos="137 "></span>             <span class="n">gene_matrix</span><span class="o">=</span><span class="n">gene_matrix</span><span class="p">,</span><span class="n">velo_weight</span><span class="o">=</span><span class="n">velo_weight</span><span class="p">,</span><span class="n">edgebundle_pruning</span><span class="o">=</span><span class="n">edgebundle_pruning</span><span class="p">,</span><span class="n">A_velo</span><span class="o">=</span><span class="n">A_velo</span><span class="p">,</span><span class="n">CSM</span><span class="o">=</span><span class="n">CSM</span><span class="p">,</span><span class="n">edgebundle_pruning_twice</span><span class="o">=</span><span class="n">edgebundle_pruning_twice</span><span class="p">,</span><span class="n">pca_loadings</span><span class="o">=</span><span class="n">pca_loadings</span><span class="p">,</span><span class="n">time_series</span><span class="o">=</span><span class="n">time_series</span><span class="p">,</span>
<span class="linenos" data-linenos="138 "></span>             <span class="n">time_series_labels</span><span class="o">=</span><span class="n">time_series_labels</span><span class="p">,</span><span class="n">knn_sequential</span><span class="o">=</span><span class="n">knn_sequential</span><span class="p">,</span><span class="n">knn_sequential_reverse</span><span class="o">=</span><span class="n">knn_sequential_reverse</span><span class="p">,</span><span class="n">t_diff_step</span><span class="o">=</span><span class="n">t_diff_step</span><span class="p">,</span><span class="n">single_cell_transition_matrix</span><span class="o">=</span><span class="n">single_cell_transition_matrix</span><span class="p">,</span>
<span class="linenos" data-linenos="139 "></span>             <span class="n">embedding_type</span><span class="o">=</span><span class="n">embedding_type</span><span class="p">,</span><span class="n">do_compute_embedding</span><span class="o">=</span><span class="n">do_compute_embedding</span><span class="p">,</span><span class="n">color_dict</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span><span class="n">user_defined_terminal_cell</span><span class="o">=</span><span class="n">user_defined_terminal_cell</span><span class="p">,</span><span class="n">user_defined_terminal_group</span><span class="o">=</span><span class="n">user_defined_terminal_group</span><span class="p">,</span>
<span class="linenos" data-linenos="140 "></span>             <span class="n">do_gaussian_kernel_edgeweights</span><span class="o">=</span><span class="n">do_gaussian_kernel_edgeweights</span><span class="p">,</span><span class="n">RW2_mode</span><span class="o">=</span><span class="n">RW2_mode</span><span class="p">,</span><span class="n">working_dir_fp</span><span class="o">=</span><span class="n">working_dir_fp</span>
<span class="linenos" data-linenos="141 "></span>             <span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.run">
<code class="highlight language-python"><span class="n">run</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.run" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>calculate the via graph and pseudotime</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="142 "></span><span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="143 "></span><span class="w">    </span><span class="sd">"""calculate the via graph and pseudotime</span>
<span class="linenos" data-linenos="144 "></span>
<span class="linenos" data-linenos="145 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="146 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">run_VIA</span><span class="p">()</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.get_piechart_dict">
<code class="highlight language-python"><span class="n">get_piechart_dict</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.get_piechart_dict" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Cluster composition graph</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>label</code></td>
<td>
<code>int</code>
</td>
<td><p>int (default=0) cluster label of pie chart</p></td>
<td>
<code>0</code>
</td>
</tr>
<tr>
<td><code>clusters</code></td>
<td>
<code>str</code>
</td>
<td><p>the celltype you want interested</p></td>
<td>
<code>''</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>res_dict</code></td> <td>
<code>dict</code>
</td>
<td><p>cluster composition graph</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="148 "></span><span class="k">def</span> <span class="nf">get_piechart_dict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">label</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">)</span><span class="o">-&gt;</span><span class="nb">dict</span><span class="p">:</span>
<span class="linenos" data-linenos="149 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="150 "></span><span class="sd">    Cluster composition graph</span>
<span class="linenos" data-linenos="151 "></span>
<span class="linenos" data-linenos="152 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="153 "></span><span class="sd">        label: int (default=0) cluster label of pie chart</span>
<span class="linenos" data-linenos="154 "></span><span class="sd">        clusters: the celltype you want interested</span>
<span class="linenos" data-linenos="155 "></span>
<span class="linenos" data-linenos="156 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="157 "></span><span class="sd">        res_dict: cluster composition graph</span>
<span class="linenos" data-linenos="158 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="159 "></span>    <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="160 "></span>        <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="161 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="162 "></span>    <span class="n">cluster_i_loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span><span class="p">)</span> <span class="o">==</span> <span class="n">label</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="163 "></span>    <span class="n">res_dict</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">cluster_i_loc</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">clusters</span><span class="p">))</span>
<span class="linenos" data-linenos="164 "></span>    <span class="k">return</span> <span class="n">res_dict</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.get_pseudotime">
<code class="highlight language-python"><span class="n">get_pseudotime</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.get_pseudotime" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Extract the pseudotime of VIA</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>an adata object of you interested,if None, it will be added to <code>self.adata.obs['pt_via']</code></p></td>
<td>
<code>None</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="166 "></span><span class="k">def</span> <span class="nf">get_pseudotime</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="167 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="168 "></span><span class="sd">    Extract the pseudotime of VIA</span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="171 "></span><span class="sd">        adata: an adata object of you interested,if None, it will be added to `self.adata.obs['pt_via']`</span>
<span class="linenos" data-linenos="172 "></span>
<span class="linenos" data-linenos="173 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="174 "></span>
<span class="linenos" data-linenos="175 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'...the pseudotime of VIA added to AnnData obs named `pt_via`'</span><span class="p">)</span>
<span class="linenos" data-linenos="176 "></span>    <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="177 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'pt_via'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">single_cell_pt_markov</span>
<span class="linenos" data-linenos="178 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="179 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'pt_via'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">single_cell_pt_markov</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_piechart_graph">
<code class="highlight language-python"><span class="n">plot_piechart_graph</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">'pt'</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="p">[],</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">headwidth_arrow</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">alpha_edge</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">edge_color</span><span class="o">=</span><span class="s1">'darkblue'</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_piechart_graph" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>plot two subplots with a clustergraph level representation of the viagraph showing true-label composition (lhs) and pseudotime/gene expression (rhs)</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>column name of the adata.obs dataframe that contains the cluster labels</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>type_data</code></td>
<td>
</td>
<td><p>string  default 'pt' for pseudotime colored nodes. or 'gene'</p></td>
<td>
<code>'pt'</code>
</td>
</tr>
<tr>
<td><code>gene_exp</code></td>
<td>
</td>
<td><p>list of values (column of dataframe) corresponding to feature or gene expression to be used to color nodes at CLUSTER level</p></td>
<td>
<code>[]</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
</td>
<td><p>string</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>default None. automatically chooses coolwarm for gene expression or viridis_r for pseudotime</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>ax_text</code></td>
<td>
</td>
<td><p>Bool default= True. Annotates each node with cluster number and population of membership</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>dpi</code></td>
<td>
</td>
<td><p>int default = 150</p></td>
<td>
<code>150</code>
</td>
</tr>
<tr>
<td><code>headwidth_bundle</code></td>
<td>
</td>
<td><p>default = 0.1. width of arrowhead used to directed edges</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>reference</code></td>
<td>
</td>
<td><p>None or list. list of categorical (str) labels for cluster composition of the piecharts (LHS subplot) length = n_samples.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>pie_size_scale</code></td>
<td>
</td>
<td><p>float default=0.8 scaling factor of the piechart nodes</p></td>
<td>
<code>0.8</code>
</td>
</tr>
<tr>
<td><code>fontsize</code></td>
<td>
</td>
<td><p>float default=8. fontsize of the text in the piecharts</p></td>
<td>
<code>8</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>tuple default=(8,4). size of the figure</p></td>
<td>
<code>(8, 4)</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>Returns matplotlib figure with two axes that plot the clustergraph using edge bundling</p></td>
</tr>
<tr>
<td><code>ax</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>left axis shows the clustergraph with each node colored by annotated ground truth membership.</p></td>
</tr>
<tr>
<td><code>ax1</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>right axis shows the same clustergraph with each node colored by the pseudotime or gene expression</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="181 "></span><span class="k">def</span> <span class="nf">plot_piechart_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">'pt'</span><span class="p">,</span>
<span class="linenos" data-linenos="182 "></span>                            <span class="n">gene_exp</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[],</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> 
<span class="linenos" data-linenos="183 "></span>                            <span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="184 "></span>                            <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span><span class="n">headwidth_arrow</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> 
<span class="linenos" data-linenos="185 "></span>                            <span class="n">alpha_edge</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> 
<span class="linenos" data-linenos="186 "></span>                            <span class="n">edge_color</span><span class="o">=</span><span class="s1">'darkblue'</span><span class="p">,</span><span class="n">reference</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
<span class="linenos" data-linenos="187 "></span>                            <span class="n">show_legend</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">fontsize</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="188 "></span>                                                                                                      <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span>
<span class="linenos" data-linenos="189 "></span>                                                                                                      <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="190 "></span><span class="w">    </span><span class="sd">"""plot two subplots with a clustergraph level representation of the viagraph showing true-label composition (lhs) and pseudotime/gene expression (rhs)</span>
<span class="linenos" data-linenos="191 "></span>
<span class="linenos" data-linenos="192 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">        clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        type_data : string  default 'pt' for pseudotime colored nodes. or 'gene'</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">        gene_exp : list of values (column of dataframe) corresponding to feature or gene expression to be used to color nodes at CLUSTER level</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        title : string</span>
<span class="linenos" data-linenos="197 "></span><span class="sd">        cmap : default None. automatically chooses coolwarm for gene expression or viridis_r for pseudotime</span>
<span class="linenos" data-linenos="198 "></span><span class="sd">        ax_text : Bool default= True. Annotates each node with cluster number and population of membership</span>
<span class="linenos" data-linenos="199 "></span><span class="sd">        dpi : int default = 150</span>
<span class="linenos" data-linenos="200 "></span><span class="sd">        headwidth_bundle : default = 0.1. width of arrowhead used to directed edges</span>
<span class="linenos" data-linenos="201 "></span><span class="sd">        reference : None or list. list of categorical (str) labels for cluster composition of the piecharts (LHS subplot) length = n_samples.</span>
<span class="linenos" data-linenos="202 "></span><span class="sd">        pie_size_scale : float default=0.8 scaling factor of the piechart nodes</span>
<span class="linenos" data-linenos="203 "></span><span class="sd">        fontsize : float default=8. fontsize of the text in the piecharts</span>
<span class="linenos" data-linenos="204 "></span><span class="sd">        figsize : tuple default=(8,4). size of the figure</span>
<span class="linenos" data-linenos="205 "></span>
<span class="linenos" data-linenos="206 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="207 "></span><span class="sd">        fig: Returns matplotlib figure with two axes that plot the clustergraph using edge bundling</span>
<span class="linenos" data-linenos="208 "></span><span class="sd">        ax: left axis shows the clustergraph with each node colored by annotated ground truth membership.</span>
<span class="linenos" data-linenos="209 "></span><span class="sd">        ax1: right axis shows the same clustergraph with each node colored by the pseudotime or gene expression</span>
<span class="linenos" data-linenos="210 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="211 "></span>
<span class="linenos" data-linenos="212 "></span>
<span class="linenos" data-linenos="213 "></span>    <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="214 "></span>        <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="215 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="216 "></span>    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span> <span class="o">=</span> <span class="n">draw_piechart_graph_pyomic</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="217 "></span>                               <span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="n">type_data</span><span class="p">,</span>
<span class="linenos" data-linenos="218 "></span>                            <span class="n">gene_exp</span><span class="o">=</span><span class="n">gene_exp</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> 
<span class="linenos" data-linenos="219 "></span>                            <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">ax_text</span><span class="o">=</span><span class="n">ax_text</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="220 "></span>                            <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span><span class="n">headwidth_arrow</span> <span class="o">=</span> <span class="n">headwidth_arrow</span><span class="p">,</span>
<span class="linenos" data-linenos="221 "></span>                            <span class="n">alpha_edge</span><span class="o">=</span><span class="n">alpha_edge</span><span class="p">,</span> <span class="n">linewidth_edge</span><span class="o">=</span><span class="n">linewidth_edge</span><span class="p">,</span>
<span class="linenos" data-linenos="222 "></span>                            <span class="n">edge_color</span><span class="o">=</span><span class="n">edge_color</span><span class="p">,</span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">,</span>
<span class="linenos" data-linenos="223 "></span>                            <span class="n">show_legend</span><span class="o">=</span><span class="n">show_legend</span><span class="p">,</span> <span class="n">pie_size_scale</span><span class="o">=</span><span class="n">pie_size_scale</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="224 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">ax1</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_stream">
<code class="highlight language-python"><span class="n">plot_stream</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">density_grid</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">arrow_size</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">arrow_color</span><span class="o">=</span><span class="s1">'k'</span><span class="p">,</span> <span class="n">arrow_style</span><span class="o">=</span><span class="s1">'-|&gt;'</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">min_mass</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cutoff_perc</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">density_stream</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">'annotation'</span><span class="p">,</span> <span class="n">add_outline_clusters</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cluster_outline_edgewidth</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span> <span class="n">gp_color</span><span class="o">=</span><span class="s1">'white'</span><span class="p">,</span> <span class="n">bg_color</span><span class="o">=</span><span class="s1">'black'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Streamplot'</span><span class="p">,</span> <span class="n">b_bias</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_neighbors_velocity_grid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">other_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">use_sequentially_augmented</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap_str</span><span class="o">=</span><span class="s1">'rainbow'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_stream" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Construct vector streamplot on the embedding to show a fine-grained view of inferred directions in the trajectory</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>column name of the adata.obs dataframe that contains the cluster labels</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>basis</code></td>
<td>
</td>
<td><p>str, default = 'X_umap', which to use for the embedding</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>density_grid</code></td>
<td>
</td>
<td><p>float, default = 0.5, density of the grid on which to project the directionality of cells</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>arrow_size</code></td>
<td>
</td>
<td><p>float, default = 0.7, size of the arrows in the streamplot</p></td>
<td>
<code>0.7</code>
</td>
</tr>
<tr>
<td><code>arrow_color</code></td>
<td>
</td>
<td><p>str, default = 'k', color of the arrows in the streamplot</p></td>
<td>
<code>'k'</code>
</td>
</tr>
<tr>
<td><code>arrow_style</code></td>
<td>
</td>
<td><p>str, default = "-|&gt;", style of the arrows in the streamplot</p></td>
<td>
<code>'-|&gt;'</code>
</td>
</tr>
<tr>
<td><code>max_length</code></td>
<td>
</td>
<td><p>int, default = 4, maximum length of the arrows in the streamplot</p></td>
<td>
<code>4</code>
</td>
</tr>
<tr>
<td><code>linewidth</code></td>
<td>
</td>
<td><p>float, default = 1, width of  lines in streamplot</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>min_mass</code></td>
<td>
</td>
<td><p>float, default = 1, minimum mass of the arrows in the streamplot</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>cutoff_perc</code></td>
<td>
</td>
<td><p>int, default = 5, cutoff percentage of the arrows in the streamplot</p></td>
<td>
<code>5</code>
</td>
</tr>
<tr>
<td><code>scatter_size</code></td>
<td>
</td>
<td><p>int, default = 500, size of scatter points</p></td>
<td>
<code>500</code>
</td>
</tr>
<tr>
<td><code>scatter_alpha</code></td>
<td>
</td>
<td><p>float, default = 0.5, transpsarency of scatter points</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>marker_edgewidth</code></td>
<td>
</td>
<td><p>float, default = 0.1, width of outline arround each scatter point</p></td>
<td>
<code>0.1</code>
</td>
</tr>
<tr>
<td><code>density_stream</code></td>
<td>
</td>
<td><p>int, default = 2, density of the streamplot</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>smooth_transition</code></td>
<td>
</td>
<td><p>int, default = 1, smoothness of the transition between the streamplot and the scatter points</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>smooth_grid</code></td>
<td>
</td>
<td><p>float, default = 0.5, smoothness of the grid on which to project the directionality of cells</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>color_scheme</code></td>
<td>
</td>
<td><p>str, default = 'annotation' corresponds to self.true_labels. Other options are 'time' (uses single-cell pseudotime) or 'clusters' (uses self.clusters)</p></td>
<td>
<code>'annotation'</code>
</td>
</tr>
<tr>
<td><code>add_outline_clusters</code></td>
<td>
</td>
<td><p>bool, default = False, whether to add an outline to the clusters</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>cluster_outline_edgewidth</code></td>
<td>
</td>
<td><p>float, default = 0.001, width of the outline around the clusters</p></td>
<td>
<code>0.001</code>
</td>
</tr>
<tr>
<td><code>gp_color</code></td>
<td>
</td>
<td><p>str, default = 'white', color of the grid points</p></td>
<td>
<code>'white'</code>
</td>
</tr>
<tr>
<td><code>bg_color</code></td>
<td>
</td>
<td><p>str, default = 'black', color of the background</p></td>
<td>
<code>'black'</code>
</td>
</tr>
<tr>
<td><code>dpi</code></td>
<td>
</td>
<td><p>int, default = 80, dpi of the figure</p></td>
<td>
<code>80</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
</td>
<td><p>str, default = 'Streamplot', title of the figure</p></td>
<td>
<code>'Streamplot'</code>
</td>
</tr>
<tr>
<td><code>b_bias</code></td>
<td>
</td>
<td><p>int, default = 20, higher value makes the forward bias of pseudotime stronger</p></td>
<td>
<code>20</code>
</td>
</tr>
<tr>
<td><code>n_neighbors_velocity_grid</code></td>
<td>
</td>
<td><p>int, default = None, number of neighbors to use for the velocity grid</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>other_labels</code></td>
<td>
</td>
<td><p>list, default = None, list of other labels to plot in the streamplot</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>use_sequentially_augmented</code></td>
<td>
</td>
<td><p>bool, default = False, whether to use the sequentially augmented data</p></td>
<td>
<code>False</code>
</td>
</tr>
<tr>
<td><code>cmap_str</code></td>
<td>
</td>
<td><p>str, default = 'rainbow', color map to use for the streamplot</p></td>
<td>
<code>'rainbow'</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>ax</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="226 "></span><span class="k">def</span> <span class="nf">plot_stream</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos="227 "></span>               <span class="n">density_grid</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">arrow_size</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">arrow_color</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="s1">'k'</span><span class="p">,</span>
<span class="linenos" data-linenos="228 "></span>               <span class="n">arrow_style</span><span class="o">=</span><span class="s2">"-|&gt;"</span><span class="p">,</span>  <span class="n">max_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">min_mass</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cutoff_perc</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
<span class="linenos" data-linenos="229 "></span>               <span class="n">scatter_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">marker_edgewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
<span class="linenos" data-linenos="230 "></span>               <span class="n">density_stream</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
<span class="linenos" data-linenos="231 "></span>               <span class="n">color_scheme</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span> <span class="s1">'annotation'</span><span class="p">,</span> <span class="n">add_outline_clusters</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos="232 "></span>               <span class="n">cluster_outline_edgewidth</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span><span class="n">gp_color</span> <span class="o">=</span> <span class="s1">'white'</span><span class="p">,</span> <span class="n">bg_color</span><span class="o">=</span><span class="s1">'black'</span> <span class="p">,</span>
<span class="linenos" data-linenos="233 "></span>               <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Streamplot'</span><span class="p">,</span> <span class="n">b_bias</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">n_neighbors_velocity_grid</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="234 "></span>               <span class="n">other_labels</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span><span class="n">use_sequentially_augmented</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap_str</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'rainbow'</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="235 "></span>                                                                                                      <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="236 "></span><span class="w">    </span><span class="sd">"""Construct vector streamplot on the embedding to show a fine-grained view of inferred directions in the trajectory</span>
<span class="linenos" data-linenos="237 "></span>
<span class="linenos" data-linenos="238 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="239 "></span><span class="sd">        clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="240 "></span><span class="sd">        basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="241 "></span><span class="sd">        density_grid : float, default = 0.5, density of the grid on which to project the directionality of cells</span>
<span class="linenos" data-linenos="242 "></span><span class="sd">        arrow_size : float, default = 0.7, size of the arrows in the streamplot</span>
<span class="linenos" data-linenos="243 "></span><span class="sd">        arrow_color : str, default = 'k', color of the arrows in the streamplot</span>
<span class="linenos" data-linenos="244 "></span><span class="sd">        arrow_style : str, default = "-|&gt;", style of the arrows in the streamplot</span>
<span class="linenos" data-linenos="245 "></span><span class="sd">        max_length : int, default = 4, maximum length of the arrows in the streamplot</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">        linewidth : float, default = 1, width of  lines in streamplot</span>
<span class="linenos" data-linenos="247 "></span><span class="sd">        min_mass : float, default = 1, minimum mass of the arrows in the streamplot</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">        cutoff_perc : int, default = 5, cutoff percentage of the arrows in the streamplot</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">        scatter_size : int, default = 500, size of scatter points</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">        scatter_alpha : float, default = 0.5, transpsarency of scatter points</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">        marker_edgewidth : float, default = 0.1, width of outline arround each scatter point</span>
<span class="linenos" data-linenos="252 "></span><span class="sd">        density_stream : int, default = 2, density of the streamplot</span>
<span class="linenos" data-linenos="253 "></span><span class="sd">        smooth_transition : int, default = 1, smoothness of the transition between the streamplot and the scatter points</span>
<span class="linenos" data-linenos="254 "></span><span class="sd">        smooth_grid : float, default = 0.5, smoothness of the grid on which to project the directionality of cells</span>
<span class="linenos" data-linenos="255 "></span><span class="sd">        color_scheme : str, default = 'annotation' corresponds to self.true_labels. Other options are 'time' (uses single-cell pseudotime) or 'clusters' (uses self.clusters)</span>
<span class="linenos" data-linenos="256 "></span><span class="sd">        add_outline_clusters : bool, default = False, whether to add an outline to the clusters</span>
<span class="linenos" data-linenos="257 "></span><span class="sd">        cluster_outline_edgewidth : float, default = 0.001, width of the outline around the clusters</span>
<span class="linenos" data-linenos="258 "></span><span class="sd">        gp_color : str, default = 'white', color of the grid points</span>
<span class="linenos" data-linenos="259 "></span><span class="sd">        bg_color : str, default = 'black', color of the background</span>
<span class="linenos" data-linenos="260 "></span><span class="sd">        dpi : int, default = 80, dpi of the figure</span>
<span class="linenos" data-linenos="261 "></span><span class="sd">        title : str, default = 'Streamplot', title of the figure</span>
<span class="linenos" data-linenos="262 "></span><span class="sd">        b_bias : int, default = 20, higher value makes the forward bias of pseudotime stronger</span>
<span class="linenos" data-linenos="263 "></span><span class="sd">        n_neighbors_velocity_grid : int, default = None, number of neighbors to use for the velocity grid</span>
<span class="linenos" data-linenos="264 "></span><span class="sd">        other_labels : list, default = None, list of other labels to plot in the streamplot</span>
<span class="linenos" data-linenos="265 "></span><span class="sd">        use_sequentially_augmented : bool, default = False, whether to use the sequentially augmented data</span>
<span class="linenos" data-linenos="266 "></span><span class="sd">        cmap_str : str, default = 'rainbow', color map to use for the streamplot</span>
<span class="linenos" data-linenos="267 "></span>
<span class="linenos" data-linenos="268 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="270 "></span><span class="sd">        ax : matplotlib axis</span>
<span class="linenos" data-linenos="271 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="272 "></span>
<span class="linenos" data-linenos="273 "></span>    <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="274 "></span>        <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="275 "></span>    <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="276 "></span>        <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="277 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>    <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="279 "></span>    <span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">via_streamplot_pyomic</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> 
<span class="linenos" data-linenos="280 "></span>                             <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">density_grid</span><span class="o">=</span><span class="n">density_grid</span><span class="p">,</span> <span class="n">arrow_size</span><span class="o">=</span><span class="n">arrow_size</span><span class="p">,</span>
<span class="linenos" data-linenos="281 "></span>                             <span class="n">arrow_color</span><span class="o">=</span><span class="n">arrow_color</span><span class="p">,</span><span class="n">arrow_style</span><span class="o">=</span><span class="n">arrow_style</span><span class="p">,</span>  <span class="n">max_length</span><span class="o">=</span><span class="n">max_length</span><span class="p">,</span>
<span class="linenos" data-linenos="282 "></span>                             <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span><span class="n">min_mass</span> <span class="o">=</span> <span class="n">min_mass</span><span class="p">,</span> <span class="n">cutoff_perc</span><span class="o">=</span><span class="n">cutoff_perc</span><span class="p">,</span>
<span class="linenos" data-linenos="283 "></span>                             <span class="n">scatter_size</span><span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="n">scatter_alpha</span><span class="p">,</span><span class="n">marker_edgewidth</span><span class="o">=</span><span class="n">marker_edgewidth</span><span class="p">,</span>
<span class="linenos" data-linenos="284 "></span>                             <span class="n">density_stream</span><span class="o">=</span><span class="n">density_stream</span><span class="p">,</span> <span class="n">smooth_transition</span><span class="o">=</span><span class="n">smooth_transition</span><span class="p">,</span> <span class="n">smooth_grid</span><span class="o">=</span><span class="n">smooth_grid</span><span class="p">,</span>
<span class="linenos" data-linenos="285 "></span>                             <span class="n">color_scheme</span><span class="o">=</span><span class="n">color_scheme</span><span class="p">,</span> <span class="n">add_outline_clusters</span><span class="o">=</span><span class="n">add_outline_clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="286 "></span>                             <span class="n">cluster_outline_edgewidth</span> <span class="o">=</span> <span class="n">cluster_outline_edgewidth</span><span class="p">,</span><span class="n">gp_color</span> <span class="o">=</span> <span class="n">gp_color</span><span class="p">,</span> <span class="n">bg_color</span><span class="o">=</span><span class="n">bg_color</span><span class="p">,</span>
<span class="linenos" data-linenos="287 "></span>                             <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">b_bias</span><span class="o">=</span><span class="n">b_bias</span><span class="p">,</span> <span class="n">n_neighbors_velocity_grid</span><span class="o">=</span><span class="n">n_neighbors_velocity_grid</span><span class="p">,</span>
<span class="linenos" data-linenos="288 "></span>                             <span class="n">other_labels</span><span class="o">=</span><span class="n">other_labels</span><span class="p">,</span><span class="n">use_sequentially_augmented</span><span class="o">=</span><span class="n">use_sequentially_augmented</span><span class="p">,</span> <span class="n">cmap_str</span><span class="o">=</span><span class="n">cmap_str</span><span class="p">)</span>
<span class="linenos" data-linenos="289 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_trajectory_gams">
<code class="highlight language-python"><span class="n">plot_trajectory_gams</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">title_str</span><span class="o">=</span><span class="s1">'Pseudotime'</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">arrow_width_scale_factor</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap_pseudotime</span><span class="o">=</span><span class="s1">'viridis_r'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">highlight_terminal_states</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_maxout_edgelist</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_trajectory_gams" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>projects the graph based coarse trajectory onto a umap/tsne embedding</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>column name of the adata.obs dataframe that contains the cluster labels</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>basis</code></td>
<td>
</td>
<td><p>str, default = 'X_umap', which to use for the embedding</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>via_fine</code></td>
<td>
</td>
<td><p>via object suggest to use via_object only unless you found that running via_fine gave better pathways</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>idx</code></td>
<td>
</td>
<td><p>default: None. Or List. if you had previously computed a umap/tsne (embedding) only on a subset of the total n_samples (subsampled as per idx), then the via objects and results will be indexed according to idx too</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>title_str</code></td>
<td>
</td>
<td><p>title of figure</p></td>
<td>
<code>'Pseudotime'</code>
</td>
</tr>
<tr>
<td><code>draw_all_curves</code></td>
<td>
</td>
<td><p>if the clustergraph has too many edges to project in a visually interpretable way, set this to False to get a simplified view of the graph pathways</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>arrow_width_scale_factor</code></td>
<td>
</td>
<td><p>the width of the arrows is proportional to the edge weight. This factor scales the width of the arrows</p></td>
<td>
<code>15.0</code>
</td>
</tr>
<tr>
<td><code>scatter_size</code></td>
<td>
</td>
<td><p>size of the scatter points</p></td>
<td>
<code>50</code>
</td>
</tr>
<tr>
<td><code>scatter_alpha</code></td>
<td>
</td>
<td><p>transparency of the scatter points</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>linewidth</code></td>
<td>
</td>
<td><p>width of the lines</p></td>
<td>
<code>1.5</code>
</td>
</tr>
<tr>
<td><code>marker_edgewidth</code></td>
<td>
</td>
<td><p>width of the outline around each scatter point</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>cmap_pseudotime</code></td>
<td>
</td>
<td><p>color map to use for the pseudotime</p></td>
<td>
<code>'viridis_r'</code>
</td>
</tr>
<tr>
<td><code>dpi</code></td>
<td>
</td>
<td><p>dpi of the figure</p></td>
<td>
<code>80</code>
</td>
</tr>
<tr>
<td><code>highlight_terminal_states</code></td>
<td>
</td>
<td><p>whether or not to highlight/distinguish the clusters which are detected as the terminal states by via</p></td>
<td>
<code>True</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>ax1</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
<tr>
<td><code>ax2</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="291 "></span><span class="k">def</span> <span class="nf">plot_trajectory_gams</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="292 "></span>                     <span class="n">title_str</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span> <span class="s2">"Pseudotime"</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">arrow_width_scale_factor</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>
<span class="linenos" data-linenos="293 "></span>                     <span class="n">scatter_size</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="294 "></span>                     <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cmap_pseudotime</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis_r'</span><span class="p">,</span><span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span>
<span class="linenos" data-linenos="295 "></span>                     <span class="n">highlight_terminal_states</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">use_maxout_edgelist</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="296 "></span>                                                                                             <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span>
<span class="linenos" data-linenos="297 "></span>                                                                                             <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="298 "></span><span class="w">    </span><span class="sd">"""projects the graph based coarse trajectory onto a umap/tsne embedding</span>
<span class="linenos" data-linenos="299 "></span>
<span class="linenos" data-linenos="300 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="301 "></span><span class="sd">        clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="302 "></span><span class="sd">        basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="303 "></span><span class="sd">        via_fine : via object suggest to use via_object only unless you found that running via_fine gave better pathways</span>
<span class="linenos" data-linenos="304 "></span><span class="sd">        idx : default: None. Or List. if you had previously computed a umap/tsne (embedding) only on a subset of the total n_samples (subsampled as per idx), then the via objects and results will be indexed according to idx too</span>
<span class="linenos" data-linenos="305 "></span><span class="sd">        title_str : title of figure</span>
<span class="linenos" data-linenos="306 "></span><span class="sd">        draw_all_curves : if the clustergraph has too many edges to project in a visually interpretable way, set this to False to get a simplified view of the graph pathways</span>
<span class="linenos" data-linenos="307 "></span><span class="sd">        arrow_width_scale_factor : the width of the arrows is proportional to the edge weight. This factor scales the width of the arrows</span>
<span class="linenos" data-linenos="308 "></span><span class="sd">        scatter_size : size of the scatter points</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">        scatter_alpha : transparency of the scatter points</span>
<span class="linenos" data-linenos="310 "></span><span class="sd">        linewidth : width of the lines</span>
<span class="linenos" data-linenos="311 "></span><span class="sd">        marker_edgewidth : width of the outline around each scatter point</span>
<span class="linenos" data-linenos="312 "></span><span class="sd">        cmap_pseudotime : color map to use for the pseudotime</span>
<span class="linenos" data-linenos="313 "></span><span class="sd">        dpi : dpi of the figure</span>
<span class="linenos" data-linenos="314 "></span><span class="sd">        highlight_terminal_states :  whether or not to highlight/distinguish the clusters which are detected as the terminal states by via</span>
<span class="linenos" data-linenos="315 "></span>
<span class="linenos" data-linenos="316 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="317 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="318 "></span><span class="sd">        ax1 : matplotlib axis</span>
<span class="linenos" data-linenos="319 "></span><span class="sd">        ax2 : matplotlib axis</span>
<span class="linenos" data-linenos="320 "></span>
<span class="linenos" data-linenos="321 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="322 "></span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span>    <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="325 "></span>        <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="326 "></span>    <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="327 "></span>        <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="328 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="329 "></span>    <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="330 "></span>    <span class="n">fig</span><span class="p">,</span><span class="n">ax1</span><span class="p">,</span><span class="n">ax2</span> <span class="o">=</span> <span class="n">draw_trajectory_gams_pyomic</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> 
<span class="linenos" data-linenos="331 "></span>                                        <span class="n">via_fine</span><span class="o">=</span><span class="n">via_fine</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span>
<span class="linenos" data-linenos="332 "></span>                                        <span class="n">title_str</span><span class="o">=</span><span class="n">title_str</span><span class="p">,</span> <span class="n">draw_all_curves</span><span class="o">=</span><span class="n">draw_all_curves</span><span class="p">,</span> <span class="n">arrow_width_scale_factor</span><span class="o">=</span><span class="n">arrow_width_scale_factor</span><span class="p">,</span>
<span class="linenos" data-linenos="333 "></span>                                        <span class="n">scatter_size</span><span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span> <span class="n">scatter_alpha</span><span class="o">=</span><span class="n">scatter_alpha</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="334 "></span>                                        <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span> <span class="n">marker_edgewidth</span><span class="o">=</span><span class="n">marker_edgewidth</span><span class="p">,</span> <span class="n">cmap_pseudotime</span><span class="o">=</span><span class="n">cmap_pseudotime</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span>
<span class="linenos" data-linenos="335 "></span>                                        <span class="n">highlight_terminal_states</span><span class="o">=</span><span class="n">highlight_terminal_states</span><span class="p">,</span> <span class="n">use_maxout_edgelist</span><span class="o">=</span><span class="n">use_maxout_edgelist</span><span class="p">)</span>
<span class="linenos" data-linenos="336 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax1</span><span class="p">,</span><span class="n">ax2</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_lineage_probability">
<code class="highlight language-python"><span class="n">plot_lineage_probability</span><span class="p">(</span><span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'plasma'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">scatter_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[],</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_lineage_probability" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>G is the igraph knn (low K) used for shortest path in high dim space. no idx needed as it's made on full sample, knn_hnsw is the knn made in the embedded space used for query to find the nearest point in the downsampled embedding that corresponds to the single cells in the full graph</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>column name of the adata.obs dataframe that contains the cluster labels</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>basis</code></td>
<td>
</td>
<td><p>str, default = 'X_umap', which to use for the embedding</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>via_fine</code></td>
<td>
</td>
<td><p>usually just set to same as via_coarse unless you ran a refined run and want to link it to initial via_coarse's terminal clusters</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>idx</code></td>
<td>
</td>
<td><p>if one uses a downsampled embedding of the original data, then idx is the selected indices of the downsampled samples used in the visualization</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>size of the figure</p></td>
<td>
<code>(8, 4)</code>
</td>
</tr>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>color map to use for the lineage probability</p></td>
<td>
<code>'plasma'</code>
</td>
</tr>
<tr>
<td><code>dpi</code></td>
<td>
</td>
<td><p>dpi of the figure</p></td>
<td>
<code>80</code>
</td>
</tr>
<tr>
<td><code>scatter_size</code></td>
<td>
</td>
<td><p>size of the scatter points</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>marker_lineages</code></td>
<td>
</td>
<td><p>Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</p></td>
<td>
<code>[]</code>
</td>
</tr>
<tr>
<td><code>fontsize</code></td>
<td>
</td>
<td><p>fontsize of the title</p></td>
<td>
<code>12</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>axs</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="338 "></span><span class="k">def</span> <span class="nf">plot_lineage_probability</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">clusters</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">basis</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
<span class="linenos" data-linenos="339 "></span>                            <span class="n">idx</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="340 "></span>                            <span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'plasma'</span><span class="p">,</span> <span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">scatter_size</span> <span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="341 "></span>                            <span class="n">marker_lineages</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="342 "></span>                                                                               <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="343 "></span><span class="w">    </span><span class="sd">"""G is the igraph knn (low K) used for shortest path in high dim space. no idx needed as it's made on full sample, knn_hnsw is the knn made in the embedded space used for query to find the nearest point in the downsampled embedding that corresponds to the single cells in the full graph</span>
<span class="linenos" data-linenos="344 "></span>
<span class="linenos" data-linenos="345 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="346 "></span><span class="sd">        clusters : column name of the adata.obs dataframe that contains the cluster labels</span>
<span class="linenos" data-linenos="347 "></span><span class="sd">        basis : str, default = 'X_umap', which to use for the embedding</span>
<span class="linenos" data-linenos="348 "></span><span class="sd">        via_fine : usually just set to same as via_coarse unless you ran a refined run and want to link it to initial via_coarse's terminal clusters</span>
<span class="linenos" data-linenos="349 "></span><span class="sd">        idx : if one uses a downsampled embedding of the original data, then idx is the selected indices of the downsampled samples used in the visualization</span>
<span class="linenos" data-linenos="350 "></span><span class="sd">        figsize : size of the figure</span>
<span class="linenos" data-linenos="351 "></span><span class="sd">        cmap : color map to use for the lineage probability</span>
<span class="linenos" data-linenos="352 "></span><span class="sd">        dpi : dpi of the figure</span>
<span class="linenos" data-linenos="353 "></span><span class="sd">        scatter_size : size of the scatter points</span>
<span class="linenos" data-linenos="354 "></span><span class="sd">        marker_lineages : Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</span>
<span class="linenos" data-linenos="355 "></span><span class="sd">        fontsize : fontsize of the title</span>
<span class="linenos" data-linenos="356 "></span>
<span class="linenos" data-linenos="357 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="358 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="359 "></span><span class="sd">        axs : matplotlib axis</span>
<span class="linenos" data-linenos="360 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="361 "></span>
<span class="linenos" data-linenos="362 "></span>
<span class="linenos" data-linenos="363 "></span>    <span class="k">if</span> <span class="n">clusters</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="364 "></span>        <span class="n">clusters</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span>
<span class="linenos" data-linenos="365 "></span>    <span class="k">if</span> <span class="n">basis</span><span class="o">==</span><span class="s1">''</span><span class="p">:</span>
<span class="linenos" data-linenos="366 "></span>        <span class="n">basis</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">basis</span>
<span class="linenos" data-linenos="367 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">clusters</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
<span class="linenos" data-linenos="368 "></span>    <span class="n">embedding</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">basis</span><span class="p">]</span>
<span class="linenos" data-linenos="369 "></span>    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">draw_sc_lineage_probability</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span><span class="n">via_fine</span><span class="o">=</span><span class="n">via_fine</span><span class="p">,</span> <span class="n">embedding</span><span class="o">=</span><span class="n">embedding</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="370 "></span>                                           <span class="n">idx</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">cmap_name</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span> <span class="n">scatter_size</span> <span class="o">=</span><span class="n">scatter_size</span><span class="p">,</span>
<span class="linenos" data-linenos="371 "></span>                                        <span class="n">marker_lineages</span> <span class="o">=</span> <span class="n">marker_lineages</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="372 "></span>    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="373 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_gene_trend">
<code class="highlight language-python"><span class="n">plot_gene_trend</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">spline_order</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'jet'</span><span class="p">,</span> <span class="n">marker_genes</span><span class="o">=</span><span class="p">[],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">n_splines</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">fontsize_</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[])</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_gene_trend" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>plots the gene expression trend along the pseudotime</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>gene_list</code></td>
<td>
</td>
<td><p>list of genes to plot</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>size of the figure</p></td>
<td>
<code>(8, 4)</code>
</td>
</tr>
<tr>
<td><code>magic_steps</code></td>
<td>
</td>
<td><p>number of magic steps to use for imputation</p></td>
<td>
<code>3</code>
</td>
</tr>
<tr>
<td><code>spline_order</code></td>
<td>
</td>
<td><p>order of the spline to use for smoothing</p></td>
<td>
<code>5</code>
</td>
</tr>
<tr>
<td><code>dpi</code></td>
<td>
</td>
<td><p>dpi of the figure</p></td>
<td>
<code>80</code>
</td>
</tr>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>color map to use for the gene expression</p></td>
<td>
<code>'jet'</code>
</td>
</tr>
<tr>
<td><code>marker_genes</code></td>
<td>
</td>
<td><p>Default is to use all genes in gene_exp. other provide a list of marker genes that will be used from gene_exp.</p></td>
<td>
<code>[]</code>
</td>
</tr>
<tr>
<td><code>linewidth</code></td>
<td>
</td>
<td><p>width of the lines</p></td>
<td>
<code>2.0</code>
</td>
</tr>
<tr>
<td><code>n_splines</code></td>
<td>
</td>
<td><p>number of splines to use for smoothing</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>fontsize_</code></td>
<td>
</td>
<td><p>fontsize of the title</p></td>
<td>
<code>12</code>
</td>
</tr>
<tr>
<td><code>marker_lineages</code></td>
<td>
</td>
<td><p>Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</p></td>
<td>
<code>[]</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>axs</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="375 "></span><span class="k">def</span> <span class="nf">plot_gene_trend</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span>
<span class="linenos" data-linenos="376 "></span>                    <span class="n">magic_steps</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">spline_order</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dpi</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'jet'</span><span class="p">,</span> 
<span class="linenos" data-linenos="377 "></span>                    <span class="n">marker_genes</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">linewidth</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="mf">2.0</span><span class="p">,</span>
<span class="linenos" data-linenos="378 "></span>                    <span class="n">n_splines</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>  <span class="n">fontsize_</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[])</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="379 "></span>                                                                                    <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="380 "></span><span class="w">    </span><span class="sd">"""plots the gene expression trend along the pseudotime</span>
<span class="linenos" data-linenos="381 "></span>
<span class="linenos" data-linenos="382 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="383 "></span><span class="sd">        gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="384 "></span><span class="sd">        figsize : size of the figure</span>
<span class="linenos" data-linenos="385 "></span><span class="sd">        magic_steps : number of magic steps to use for imputation</span>
<span class="linenos" data-linenos="386 "></span><span class="sd">        spline_order : order of the spline to use for smoothing</span>
<span class="linenos" data-linenos="387 "></span><span class="sd">        dpi : dpi of the figure</span>
<span class="linenos" data-linenos="388 "></span><span class="sd">        cmap : color map to use for the gene expression</span>
<span class="linenos" data-linenos="389 "></span><span class="sd">        marker_genes : Default is to use all genes in gene_exp. other provide a list of marker genes that will be used from gene_exp.</span>
<span class="linenos" data-linenos="390 "></span><span class="sd">        linewidth : width of the lines</span>
<span class="linenos" data-linenos="391 "></span><span class="sd">        n_splines : number of splines to use for smoothing</span>
<span class="linenos" data-linenos="392 "></span><span class="sd">        fontsize_ : fontsize of the title</span>
<span class="linenos" data-linenos="393 "></span><span class="sd">        marker_lineages : Default is to use all lineage pathways. other provide a list of lineage number (terminal cluster number).</span>
<span class="linenos" data-linenos="394 "></span>
<span class="linenos" data-linenos="395 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="396 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="397 "></span><span class="sd">        axs : matplotlib axis</span>
<span class="linenos" data-linenos="398 "></span>
<span class="linenos" data-linenos="399 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="400 "></span>
<span class="linenos" data-linenos="401 "></span>    <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="n">magic_steps</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="402 "></span>    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span><span class="o">=</span><span class="n">get_gene_expression_pyomic</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span><span class="n">df_magic</span><span class="p">,</span><span class="n">spline_order</span><span class="o">=</span><span class="n">spline_order</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">,</span>
<span class="linenos" data-linenos="403 "></span>                               <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">marker_genes</span><span class="o">=</span><span class="n">marker_genes</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="n">linewidth</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="404 "></span>                               <span class="n">n_splines</span><span class="o">=</span><span class="n">n_splines</span><span class="p">,</span>  <span class="n">fontsize_</span><span class="o">=</span><span class="n">fontsize_</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="n">marker_lineages</span><span class="p">)</span>
<span class="linenos" data-linenos="405 "></span>    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="406 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_gene_trend_heatmap">
<code class="highlight language-python"><span class="n">plot_gene_trend_heatmap</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">marker_lineages</span><span class="o">=</span><span class="p">[],</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ytick_labelrotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_gene_trend_heatmap" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Plot the gene trends on heatmap: a heatmap is generated for each lineage (identified by terminal cluster number). Default selects all lineages</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>gene_list</code></td>
<td>
</td>
<td><p>list of genes to plot</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>marker_lineages</code></td>
<td>
</td>
<td><p>list default = None and plots all detected all lineages. Optionally provide a list of integers corresponding to the cluster number of terminal cell fates</p></td>
<td>
<code>[]</code>
</td>
</tr>
<tr>
<td><code>fontsize</code></td>
<td>
</td>
<td><p>int default = 8</p></td>
<td>
<code>8</code>
</td>
</tr>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>str default = 'viridis'</p></td>
<td>
<code>'viridis'</code>
</td>
</tr>
<tr>
<td><code>normalize</code></td>
<td>
</td>
<td><p>bool = True</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>ytick_labelrotation</code></td>
<td>
</td>
<td><p>int default = 0</p></td>
<td>
<code>0</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>size of the figure</p></td>
<td>
<code>(2, 4)</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>axs</code></td> <td>
<code>list</code>
</td>
<td><p>list of matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="434 "></span><span class="k">def</span> <span class="nf">plot_gene_trend_heatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">marker_lineages</span><span class="p">:</span><span class="nb">list</span> <span class="o">=</span> <span class="p">[],</span> 
<span class="linenos" data-linenos="435 "></span>                         <span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span><span class="n">cmap</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">normalize</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">ytick_labelrotation</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> 
<span class="linenos" data-linenos="436 "></span>                         <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="437 "></span>                                                                <span class="nb">list</span><span class="p">]:</span>
<span class="linenos" data-linenos="438 "></span><span class="w">    </span><span class="sd">"""Plot the gene trends on heatmap: a heatmap is generated for each lineage (identified by terminal cluster number). Default selects all lineages</span>
<span class="linenos" data-linenos="439 "></span>
<span class="linenos" data-linenos="440 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="441 "></span><span class="sd">        gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="442 "></span><span class="sd">        marker_lineages : list default = None and plots all detected all lineages. Optionally provide a list of integers corresponding to the cluster number of terminal cell fates</span>
<span class="linenos" data-linenos="443 "></span><span class="sd">        fontsize : int default = 8</span>
<span class="linenos" data-linenos="444 "></span><span class="sd">        cmap : str default = 'viridis'</span>
<span class="linenos" data-linenos="445 "></span><span class="sd">        normalize : bool = True</span>
<span class="linenos" data-linenos="446 "></span><span class="sd">        ytick_labelrotation : int default = 0</span>
<span class="linenos" data-linenos="447 "></span><span class="sd">        figsize : size of the figure</span>
<span class="linenos" data-linenos="448 "></span>
<span class="linenos" data-linenos="449 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="450 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="451 "></span><span class="sd">        axs : list of matplotlib axis       </span>
<span class="linenos" data-linenos="452 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="453 "></span>
<span class="linenos" data-linenos="454 "></span>    <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="455 "></span>    <span class="n">df_magic</span><span class="p">[</span><span class="s1">'parc'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span>
<span class="linenos" data-linenos="456 "></span>    <span class="n">df_magic_cluster</span> <span class="o">=</span> <span class="n">df_magic</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">'parc'</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="457 "></span>    <span class="n">fig</span><span class="p">,</span><span class="n">axs</span><span class="o">=</span><span class="n">plot_gene_trend_heatmaps_pyomic</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">df_gene_exp</span><span class="o">=</span><span class="n">df_magic</span><span class="p">,</span> 
<span class="linenos" data-linenos="458 "></span>                                            <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span><span class="n">normalize</span><span class="o">=</span><span class="n">normalize</span><span class="p">,</span>
<span class="linenos" data-linenos="459 "></span>                                            <span class="n">ytick_labelrotation</span><span class="o">=</span><span class="n">ytick_labelrotation</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="460 "></span>                                            <span class="n">marker_lineages</span><span class="o">=</span><span class="n">marker_lineages</span><span class="p">)</span>
<span class="linenos" data-linenos="461 "></span>    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="462 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">axs</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._via.pyVIA.plot_clustergraph">
<code class="highlight language-python"><span class="n">plot_clustergraph</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">arrow_head</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">edgeweight_scale</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">label_</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._via.pyVIA.plot_clustergraph" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>plot the gene in pie chart for each cluster</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>gene_list</code></td>
<td>
</td>
<td><p>list of genes to plot</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>arrow_head</code></td>
<td>
</td>
<td><p>size of the arrow head</p></td>
<td>
<code>0.1</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p>size of the figure</p></td>
<td>
<code>(8, 4)</code>
</td>
</tr>
<tr>
<td><code>edgeweight_scale</code></td>
<td>
</td>
<td><p>scale of the edge weight</p></td>
<td>
<code>1.5</code>
</td>
</tr>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>color map to use for the gene expression</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>label_</code></td>
<td>
</td>
<td><p>whether to label the nodes</p></td>
<td>
<code>True</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>matplotlib figure</p></td>
</tr>
<tr>
<td><code>axs</code></td> <td>
<code><span title="omicverse.via.plotting_via.matplotlib">matplotlib</span>.<span title="omicverse.via.plotting_via.matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>matplotlib axis</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_via.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="408 "></span><span class="k">def</span> <span class="nf">plot_clustergraph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">gene_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">arrow_head</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span><span class="n">magic_steps</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
<span class="linenos" data-linenos="409 "></span>                      <span class="n">edgeweight_scale</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">label_</span><span class="o">=</span><span class="kc">True</span><span class="p">,)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span>
<span class="linenos" data-linenos="410 "></span>                                                                                    <span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="411 "></span><span class="w">    </span><span class="sd">"""plot the gene in pie chart for each cluster</span>
<span class="linenos" data-linenos="412 "></span>
<span class="linenos" data-linenos="413 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="414 "></span><span class="sd">        gene_list : list of genes to plot</span>
<span class="linenos" data-linenos="415 "></span><span class="sd">        arrow_head : size of the arrow head</span>
<span class="linenos" data-linenos="416 "></span><span class="sd">        figsize : size of the figure</span>
<span class="linenos" data-linenos="417 "></span><span class="sd">        edgeweight_scale : scale of the edge weight</span>
<span class="linenos" data-linenos="418 "></span><span class="sd">        cmap : color map to use for the gene expression</span>
<span class="linenos" data-linenos="419 "></span><span class="sd">        label_ : whether to label the nodes</span>
<span class="linenos" data-linenos="420 "></span>
<span class="linenos" data-linenos="421 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="422 "></span><span class="sd">        fig : matplotlib figure</span>
<span class="linenos" data-linenos="423 "></span><span class="sd">        axs : matplotlib axis</span>
<span class="linenos" data-linenos="424 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="425 "></span>    <span class="n">df_magic</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">do_impute</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">magic_steps</span><span class="o">=</span><span class="n">magic_steps</span><span class="p">,</span> <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="426 "></span>    <span class="n">df_magic</span><span class="p">[</span><span class="s1">'parc'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">labels</span>
<span class="linenos" data-linenos="427 "></span>    <span class="n">df_magic_cluster</span> <span class="o">=</span> <span class="n">df_magic</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">'parc'</span><span class="p">,</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="428 "></span>    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">draw_clustergraph_pyomic</span><span class="p">(</span><span class="n">via_object</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="n">type_data</span><span class="o">=</span><span class="s1">'gene'</span><span class="p">,</span> <span class="n">gene_exp</span><span class="o">=</span><span class="n">df_magic_cluster</span><span class="p">,</span> 
<span class="linenos" data-linenos="429 "></span>                                <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">arrow_head</span><span class="o">=</span><span class="n">arrow_head</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="430 "></span>                                <span class="n">edgeweight_scale</span><span class="o">=</span><span class="n">edgeweight_scale</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">label_</span><span class="o">=</span><span class="n">label_</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>
<span class="linenos" data-linenos="431 "></span>    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="linenos" data-linenos="432 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">axs</span>
</code></pre></div>
</details>
</div>
</div>
</div>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2024, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>