
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../alignment/kb_api/" rel="prev"/>
<link href="../bulk2single/" rel="next"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.1.2" name="generator"/>
<title>bulk - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../css/mkapi-common.css" rel="stylesheet"/>
<link href="../../../css/mkapi-material.css" rel="stylesheet"/>
<link href="../../../css/tags.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.6.0/css/all.min.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.bulk">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              bulk
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_1" id="__nav_3_1_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_2" id="__nav_3_1_2_label" tabindex="0">
          Downstream
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_2">
<span class="md-nav__icon md-icon"></span>
          Downstream
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_3" id="__nav_3_1_3_label" tabindex="0">
          Deconvolution
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_3">
<span class="md-nav__icon md-icon"></span>
          Deconvolution
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_decov_bulk/">
        Bulk deconvolution with reference scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_4" id="__nav_3_1_4_label" tabindex="0">
          Others
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_4">
<span class="md-nav__icon md-icon"></span>
          Others
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Alignment
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Alignment
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_alignment_1k/">
        Alignment and analysis of single-cell RNA-seq data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_alignment_velocity/">
        Alignment and RNA velocity analysis of single-cell RNA-seq data.
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        GeneModule Identified
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_noref/">
        Reference-free automated single-cell cell type annotation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_ref/">
        Reference automated single-cell cell type annotation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology/Taxonomy
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scenic/">
        Gene Regulatory Network Analysis with SCENIC
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell-cell communication using CellPhoneDBViz
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_6" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_6" id="__nav_3_2_6_label" tabindex="0">
          Velocity
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_6_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_6">
<span class="md-nav__icon md-icon"></span>
          Velocity
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-velo/t_velo/">
        Velocity Basic Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-velo/t_graphvelo/">
        Velocity Optimization
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_7" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_7" id="__nav_3_2_7_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_7_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_7">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_1" id="__nav_3_4_1_label" tabindex="0">
          Preprocess
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_1">
<span class="md-nav__icon md-icon"></span>
          Preprocess
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cellpose/">
        Cell Segmemtation (10x HD)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_2" id="__nav_3_4_2_label" tabindex="0">
          Deconvolution
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_2">
<span class="md-nav__icon md-icon"></span>
          Deconvolution
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_decov/">
        Spatial deconvolution with reference scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_flashdeconv/">
        FlashDeconv: Fast Spatial Deconvolution via Structure-Preserving Sketching
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh_new/">
        Spatial deconvolution without reference scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_3" id="__nav_3_4_3_label" tabindex="0">
          Downstream
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_3">
<span class="md-nav__icon md-icon"></span>
          Downstream
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          LLM
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          LLM
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_dr/">
        Domain Research (dr): Comprehensive Report Tutorial
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_dr_web/">
        Domain Research (dr) — Live Web
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_ov_agent_pbmc3k/">
        ov.Agent with Skills: Natural‑Language Single‑Cell Analysis (PBMC3k)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_uce/">
        UCE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_scgpt/">
        scGPT
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_scfoundation/">
        scFoundation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_geneformer/">
        GeneFormer
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_cellplm/">
        CellPLM
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_6" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_6_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_6">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../GPU_supported_registry_functions/">
        GPU Support
      </a>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_5" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../">API documentation</a>
<label for="__nav_5">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="true" aria-labelledby="__nav_5_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_5">
<span class="md-nav__icon md-icon"></span>
          API documentation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_2" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../agent/">agent</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_2">
<span class="md-nav__icon md-icon"></span>
          agent
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_3" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../alignment/">alignment</a>
<label for="__nav_5_3">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_3">
<span class="md-nav__icon md-icon"></span>
          alignment
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../alignment/kb_api/">
        kb_api
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_5_4" type="checkbox"/>
<div class="md-nav__link md-nav__link--index md-nav__link--active">
<a href="./">bulk</a>
</div>
<nav aria-expanded="true" aria-labelledby="__nav_5_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_4">
<span class="md-nav__icon md-icon"></span>
          bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_5" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../bulk2single/">bulk2single</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_5">
<span class="md-nav__icon md-icon"></span>
          bulk2single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_6" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../datasets/">datasets</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_6_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_6">
<span class="md-nav__icon md-icon"></span>
          datasets
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_7" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pl/">pl</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_7_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_7">
<span class="md-nav__icon md-icon"></span>
          pl
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_8" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pp/">pp</a>
<label for="__nav_5_8">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_8_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_8">
<span class="md-nav__icon md-icon"></span>
          pp
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_8_2" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pp/experimental/">experimental</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_8_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_5_8_2">
<span class="md-nav__icon md-icon"></span>
          experimental
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../pp/pyg_knn_implementation/">
        pyg_knn_implementation
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_9" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../single/">single</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_9_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_9">
<span class="md-nav__icon md-icon"></span>
          single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_10" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../space/">space</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_10_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_10">
<span class="md-nav__icon md-icon"></span>
          space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_11" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../utils/">utils</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_11_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_11">
<span class="md-nav__icon md-icon"></span>
          utils
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.geo_data_preprocess">
    geo_data_preprocess
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.fq_data_preprocess">
    fq_data_preprocess
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSEA">
    pyGSEA
  </a>
<nav aria-label="pyGSEA" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSEA.enrichment">
    enrichment
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSEA.plot_gsea">
    plot_gsea
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSEA.plot_enrichment">
    plot_enrichment
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSE">
    pyGSE
  </a>
<nav aria-label="pyGSE" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSE.enrichment">
    enrichment
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyGSE.plot_enrichment">
    plot_enrichment
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.geneset_enrichment">
    geneset_enrichment
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.geneset_plot">
    geneset_plot
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.geneset_enrichment_GSEA">
    geneset_enrichment_GSEA
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.geneset_plot_multi">
    geneset_plot_multi
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.enrichment_multi_concat">
    enrichment_multi_concat
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyPPI">
    pyPPI
  </a>
<nav aria-label="pyPPI" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyPPI.interaction_analysis">
    interaction_analysis
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyPPI.plot_network">
    plot_network
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.string_interaction">
    string_interaction
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.string_map">
    string_map
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.generate_G">
    generate_G
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.get_chm13_gene">
    get_chm13_gene
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.find_chm13_gene">
    find_chm13_gene
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG">
    pyDEG
  </a>
<nav aria-label="pyDEG" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.drop_duplicates_index">
    drop_duplicates_index
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.normalize">
    normalize
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.foldchange_set">
    foldchange_set
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.plot_volcano">
    plot_volcano
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.plot_boxplot">
    plot_boxplot
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.ranking2gsea">
    ranking2gsea
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG.deg_analysis">
    deg_analysis
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.deseq2_normalize">
    deseq2_normalize
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.estimateSizeFactors">
    estimateSizeFactors
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.estimateDispersions">
    estimateDispersions
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.Matrix_ID_mapping">
    Matrix_ID_mapping
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.data_drop_duplicates_index">
    data_drop_duplicates_index
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA">
    pyTCGA
  </a>
<nav aria-label="pyTCGA" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.adata_read">
    adata_read
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.adata_init">
    adata_init
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.adata_meta_init">
    adata_meta_init
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.survial_init">
    survial_init
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.index_init">
    index_init
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.expression_init">
    expression_init
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.matrix_construct">
    matrix_construct
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.matrix_normalize">
    matrix_normalize
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.survival_analysis">
    survival_analysis
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyTCGA.survial_analysis_all">
    survial_analysis_all
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.batch_correction">
    batch_correction
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.Deconvolution">
    Deconvolution
  </a>
<nav aria-label="Deconvolution" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.Deconvolution.deconvolution">
    deconvolution
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 style="display: none;">omicverse.bulk<a class="headerlink" href="#omicverse.bulk" title="Permanent link">¶</a></h1><p class="mkapi-object mkapi-page-source" id="omicverse.bulk">
<span class="mkapi-object-link">
<a href="../../../src/omicverse/bulk/#omicverse.bulk" title="Go to source">source</a>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">package</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk" title="omicverse.bulk">omicverse.bulk</a></span>
</p>
<div class="mkapi-document">
<p>Bulk omics analysis utilities.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this module provides comprehensive tools for bulk rna-seq analysis including">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This module provides comprehensive tools for bulk RNA-seq analysis including</span>
</p>
<div class="mkapi-section-content">
<ul>
<li>Differential expression analysis with DESeq2</li>
<li>Gene co-expression network analysis (WGCNA)</li>
<li>Pathway enrichment analysis (GSEA)</li>
<li>Protein-protein interaction networks</li>
<li>Batch effect correction</li>
<li>Multi-omics integration</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle classes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Classes</span>
</p>
<div class="mkapi-section-content">
<p>pyDEG: Differential expression analysis using DESeq2
pyWGCNA: Weighted Gene Co-expression Network Analysis
pyGSEA: Gene Set Enrichment Analysis
pyPPI: Protein-Protein Interaction analysis
pyTCGA: TCGA data analysis utilities</p>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle functions">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Functions</span>
</p>
<div class="mkapi-section-content">
<p>geneset_enrichment: Perform pathway enrichment analysis
string_interaction: Retrieve STRING database interactions
batch_correction: Correct for batch effects
get_chm13_gene: Map genes to CHM13 reference</p>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Differential expression analysis</span>
<span class="n">dds</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">pyDEG</span><span class="p">(</span><span class="n">count_matrix</span><span class="p">,</span> <span class="n">design_matrix</span><span class="p">)</span>
<span class="n">dds</span><span class="o">.</span><span class="n">drop_duplicates_index</span><span class="p">()</span>
<span class="n">dds</span><span class="o">.</span><span class="n">deseq2</span><span class="p">()</span>
<span class="n">deg_results</span> <span class="o">=</span> <span class="n">dds</span><span class="o">.</span><span class="n">get_results</span><span class="p">()</span>

<span class="c1"># WGCNA analysis</span>
<span class="n">wgcna</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">pyWGCNA</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">wgcna</span><span class="o">.</span><span class="n">calculate_correlation</span><span class="p">()</span>
<span class="n">wgcna</span><span class="o">.</span><span class="n">module_detection</span><span class="p">()</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle classes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Classes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSEA" title="omicverse.bulk.pyGSEA">pyGSEA</a></span>
<span class="mkapi-dash">—</span>
Initialize the pyGSEA class.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSE" title="omicverse.bulk.pyGSE">pyGSE</a></span>
<span class="mkapi-dash">—</span>
Initialize the pyGSE class.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyPPI" title="omicverse.bulk.pyPPI">pyPPI</a></span>
<span class="mkapi-dash">—</span>
Initialize protein-protein interaction analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a></span>
<span class="mkapi-dash">—</span>
Initialize the pyDEG class.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a></span>
<span class="mkapi-dash">—</span>
TCGA (The Cancer Genome Atlas) data analysis module.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.Deconvolution" title="omicverse.bulk.Deconvolution">Deconvolution</a></span></p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle functions">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Functions</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.geo_data_preprocess" title="omicverse.bulk.geo_data_preprocess">geo_data_preprocess</a></span>
<span class="mkapi-dash">—</span>
Convenience helper that runs the analysis end-to-end.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.fq_data_preprocess" title="omicverse.bulk.fq_data_preprocess">fq_data_preprocess</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.geneset_enrichment" title="omicverse.bulk.geneset_enrichment">geneset_enrichment</a></span>
<span class="mkapi-dash">—</span>
Perform gene set enrichment analysis using Enrichr API.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.geneset_plot" title="omicverse.bulk.geneset_plot">geneset_plot</a></span>
<span class="mkapi-dash">—</span>
Plot the gene set enrichment result.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.geneset_enrichment_GSEA" title="omicverse.bulk.geneset_enrichment_GSEA">geneset_enrichment_GSEA</a></span>
<span class="mkapi-dash">—</span>
Enrichment analysis using GSEA.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.geneset_plot_multi" title="omicverse.bulk.geneset_plot_multi">geneset_plot_multi</a></span>
<span class="mkapi-dash">—</span>
Enrichment multi genesets analysis using GSEA.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.enrichment_multi_concat" title="omicverse.bulk.enrichment_multi_concat">enrichment_multi_concat</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.string_interaction" title="omicverse.bulk.string_interaction">string_interaction</a></span>
<span class="mkapi-dash">—</span>
Analyze protein-protein interaction network using STRING database.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.string_map" title="omicverse.bulk.string_map">string_map</a></span>
<span class="mkapi-dash">—</span>
Map gene names to STRING database identifiers.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.generate_G" title="omicverse.bulk.generate_G">generate_G</a></span>
<span class="mkapi-dash">—</span>
Generate protein-protein interaction network from STRING database.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.get_chm13_gene" title="omicverse.bulk.get_chm13_gene">get_chm13_gene</a></span>
<span class="mkapi-dash">—</span>
Get T2T-CHM13 genome annotation and mapping data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.find_chm13_gene" title="omicverse.bulk.find_chm13_gene">find_chm13_gene</a></span>
<span class="mkapi-dash">—</span>
Identify T2T-CHM13 novel and GRCh38 issue genes from input gene list.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.deseq2_normalize" title="omicverse.bulk.deseq2_normalize">deseq2_normalize</a></span>
<span class="mkapi-dash">—</span>
Normalize the data using DESeq2 method.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.estimateSizeFactors" title="omicverse.bulk.estimateSizeFactors">estimateSizeFactors</a></span>
<span class="mkapi-dash">—</span>
Estimate size factors for data normalization.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.estimateDispersions" title="omicverse.bulk.estimateDispersions">estimateDispersions</a></span>
<span class="mkapi-dash">—</span>
Estimate the dispersion parameter of the Negative Binomial distribution for each gene in the input count matrix.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.Matrix_ID_mapping" title="omicverse.bulk.Matrix_ID_mapping">Matrix_ID_mapping</a></span>
<span class="mkapi-dash">—</span>
Map gene IDs in the input data to gene symbols using a reference table.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.data_drop_duplicates_index" title="omicverse.bulk.data_drop_duplicates_index">data_drop_duplicates_index</a></span>
<span class="mkapi-dash">—</span>
Drop the duplicated index of data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.batch_correction" title="omicverse.bulk.batch_correction">batch_correction</a></span>
<span class="mkapi-dash">—</span>
Perform batch effect correction using ComBat algorithm.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.geo_data_preprocess">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.geo_data_preprocess" title="omicverse.bulk.geo_data_preprocess">geo_data_preprocess</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">input_data</span><span class="mkapi-comma">, </span><span class="mkapi-star">*</span><span class="mkapi-comma">, </span><span class="mkapi-arg">config</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">input_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'auto'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">library_layout</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'auto'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">with_align</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">work_dir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'work'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">threads</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">genome</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'human'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sample_prefix</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">download_method</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'prefetch'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_gzip</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_aspera</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_database</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'sra'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_protocol</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'ftp'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_parallel</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">4</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iseq_threads</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">dict</span></span>
</p>
<div class="mkapi-document">
<p>Convenience helper that runs the analysis end-to-end.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">input_data</span>
<span class="mkapi-dash">—</span>
Input payload.</p>
</li>
<li>
<p><span class="mkapi-item-name">config</span>
<span class="mkapi-dash">—</span>
Configuration object or source.</p>
</li>
<li>
<p><span class="mkapi-item-name">input_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Input type string.</p>
</li>
<li>
<p><span class="mkapi-item-name">library_layout</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Library layout - "auto" (detect automatically), "single" (single-end), "paired" (paired-end).</p>
</li>
<li>
<p><span class="mkapi-item-name">with_align</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
Whether to perform alignment.</p>
</li>
<li>
<p><span class="mkapi-item-name">work_dir</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Working directory.</p>
</li>
<li>
<p><span class="mkapi-item-name">threads</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Thread count.</p>
</li>
<li>
<p><span class="mkapi-item-name">genome</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Genome identifier.</p>
</li>
<li>
<p><span class="mkapi-item-name">sample_prefix</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Sample ID prefix (vendor data only).</p>
</li>
<li>
<p><span class="mkapi-item-name">download_method</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Download method, "prefetch" (default) or "iseq".</p>
<ul>
<li>prefetch: Use NCBI SRA Toolkit (prefetch + fasterq-dump).</li>
<li>iseq: Use the iseq tool (multi-database, Aspera acceleration, gzip downloads, etc.).</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">iseq_gzip</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
Download FASTQ as gzip (iseq mode only, default True).</p>
</li>
<li>
<p><span class="mkapi-item-name">iseq_aspera</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
Enable Aspera acceleration (iseq mode only, default False).</p>
</li>
<li>
<p><span class="mkapi-item-name">iseq_database</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Database to target: ena or sra (iseq mode only, default ena).</p>
</li>
<li>
<p><span class="mkapi-item-name">iseq_protocol</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Protocol to use: ftp or https (iseq mode only, default ftp).</p>
</li>
<li>
<p><span class="mkapi-item-name">iseq_parallel</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of parallel downloads (iseq mode only, default 4).</p>
</li>
<li>
<p><span class="mkapi-item-name">iseq_threads</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Threads for iseq processing (iseq mode only, default 8).</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Analysis result dictionary.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.fq_data_preprocess">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.fq_data_preprocess" title="omicverse.bulk.fq_data_preprocess">fq_data_preprocess</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">fastq_files</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.List">List</span>[str]</span><span class="mkapi-comma">, </span><span class="mkapi-star">*</span><span class="mkapi-comma">, </span><span class="mkapi-arg">config</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="typing.Union">Union</span>[<span class="mkapi-tooltip" title="omicverse.bulk._alignment.AlignmentConfig">AlignmentConfig</span>, <span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, <span class="mkapi-tooltip" title="typing.Any">Any</span>], str, <span class="mkapi-tooltip" title="pathlib.Path">Path</span>]]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">input_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'fastq'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">with_align</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">work_dir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'work'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">threads</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">genome</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'human'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sample_prefix</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">dict</span></span>
</p>
<div class="mkapi-document">
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle fastq entry point following the same style as geo_data_preprocess">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">FASTQ entry point following the same style as geo_data_preprocess</span>
</p>
<div class="mkapi-section-content">
<ul>
<li>Skip prefetch / fasterq-dump</li>
<li>Accept a flat FASTQ list and auto pair</li>
<li>Subsequent steps (fastp / STAR / featureCounts) match geo</li>
<li>Returns the same structure as geo_data_preprocess</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSEA">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSEA" title="omicverse.bulk.pyGSEA">pyGSEA</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_rnk</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">pathways_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">processes</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">permutation_num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">outdir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'./enrichr_gsea'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cutoff</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.5</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Initialize the pyGSEA class.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_rnk</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
pre-ranked correlation table or pandas DataFrame. Same input with <code>GSEA</code> .rnk file.</p>
</li>
<li>
<p><span class="mkapi-item-name">pathways_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Dictionary of pathway library names and corresponding Enrichr API URLs.</p>
</li>
<li>
<p><span class="mkapi-item-name">processes</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of Processes you are going to use. Default: 8.</p>
</li>
<li>
<p><span class="mkapi-item-name">permutation_num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of permutations for significance computation. Default: 100.</p>
</li>
<li>
<p><span class="mkapi-item-name">outdir</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Output directory for Enrichr results. Default is './enrichr_gsea'.</p>
</li>
<li>
<p><span class="mkapi-item-name">cutoff</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
The cutoff for enrichment. Default is 0.5.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSEA.enrichment" title="omicverse.bulk.pyGSEA.enrichment">enrichment</a></span>
<span class="mkapi-dash">—</span>
gene set enrichment analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSEA.plot_gsea" title="omicverse.bulk.pyGSEA.plot_gsea">plot_gsea</a></span>
<span class="mkapi-dash">—</span>
Plot the gene set enrichment result.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSEA.plot_enrichment" title="omicverse.bulk.pyGSEA.plot_enrichment">plot_enrichment</a></span>
<span class="mkapi-dash">—</span>
Plot the gene set enrichment result.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSEA.enrichment">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyGSEA" title="omicverse.bulk.pyGSEA">pyGSEA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSEA.enrichment" title="omicverse.bulk.pyGSEA.enrichment">enrichment</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">format</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'png'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pval</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">seed</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">112</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>gene set enrichment analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">format</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Matplotlib figure format. Default: 'png'.</p>
</li>
<li>
<p><span class="mkapi-item-name">seed</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Random seed. Default: 112.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Enrichment.pyGSEA.enrichment.enrich_res">enrich_res</span></span> <span class="mkapi-dash">—</span>
A pandas.DataFrame object containing the enrichment results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSEA.plot_gsea">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyGSEA" title="omicverse.bulk.pyGSEA">pyGSEA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSEA.plot_gsea" title="omicverse.bulk.pyGSEA.plot_gsea">plot_gsea</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">term_num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_set_title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(3, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cmap</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'RdBu_r'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">title_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">title_y</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.95</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.figure">figure</span>.<span class="mkapi-tooltip" title="matplotlib.figure.Figure">Figure</span></span></span>
</p>
<div class="mkapi-document">
<p>Plot the gene set enrichment result.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">term_num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of enriched terms to plot. Default is 0.</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_set_title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. Default is an empty string.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the plot. Default is (3,4).</p>
</li>
<li>
<p><span class="mkapi-item-name">cmap</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The colormap to use for the plot. Default is 'RdBu_r'.</p>
</li>
<li>
<p><span class="mkapi-item-name">title_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the title. Default is 12.</p>
</li>
<li>
<p><span class="mkapi-item-name">title_y</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
The y coordinate of the title. Default is 0.95.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">fig</span> <span class="mkapi-dash">—</span>
A matplotlib.figure.Figure object.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSEA.plot_enrichment">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyGSEA" title="omicverse.bulk.pyGSEA">pyGSEA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSEA.plot_enrichment" title="omicverse.bulk.pyGSEA.plot_enrichment">plot_enrichment</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">node_size</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[5, 10, 15]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_loc</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_xlabel</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Fractions of genes'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(2, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cmap</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'YlGnBu'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_knock</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_maxsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span></span>
</p>
<div class="mkapi-document">
<p>Plot the gene set enrichment result.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of enriched terms to plot. Default is 10.</p>
</li>
<li>
<p><span class="mkapi-item-name">node_size</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
A list of integers defining the size of nodes in the plot. Default is [5,10,15].</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_loc</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The location of the colorbar on the plot. Default is 2.</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the colorbar label. Default is 12.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. Default is an empty string.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_xlabel</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The label of the x-axis. Default is 'Fractions of genes'.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the plot. Default is (2,4).</p>
</li>
<li>
<p><span class="mkapi-item-name">cmap</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The colormap to use for the plot. Default is 'YlGnBu'.</p>
</li>
<li>
<p><span class="mkapi-item-name">text_knock</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of terms to knock out for text labels. Default is 2.</p>
</li>
<li>
<p><span class="mkapi-item-name">text_maxsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The maximum fontsize of text labels. Default is 20.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ax</span> <span class="mkapi-dash">—</span>
A matplotlib.axes.Axes object.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSE">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSE" title="omicverse.bulk.pyGSE">pyGSE</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_list</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pathways_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pvalue_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pvalue_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'auto'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">background</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">organism</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Human'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">description</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'None'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">outdir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'./enrichr'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cutoff</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.5</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Initialize the pyGSE class.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_list</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
A list of genes.</p>
</li>
<li>
<p><span class="mkapi-item-name">pathways_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
A dictionary of pathways.</p>
</li>
<li>
<p><span class="mkapi-item-name">pvalue_threshold</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
The p-value threshold for enrichment. Default is 0.05.</p>
</li>
<li>
<p><span class="mkapi-item-name">pvalue_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The p-value type. Default is 'auto'.</p>
</li>
<li>
<p><span class="mkapi-item-name">organism</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The organism. Default is 'Human'.</p>
</li>
<li>
<p><span class="mkapi-item-name">description</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The description. Default is 'None'.</p>
</li>
<li>
<p><span class="mkapi-item-name">outdir</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The output directory. Default is './enrichr'.</p>
</li>
<li>
<p><span class="mkapi-item-name">cutoff</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
The cutoff for enrichment. Default is 0.5.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSE.enrichment" title="omicverse.bulk.pyGSE.enrichment">enrichment</a></span>
<span class="mkapi-dash">—</span>
gene set enrichment analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyGSE.plot_enrichment" title="omicverse.bulk.pyGSE.plot_enrichment">plot_enrichment</a></span>
<span class="mkapi-dash">—</span>
Plot the gene set enrichment result.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSE.enrichment">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyGSE" title="omicverse.bulk.pyGSE">pyGSE</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSE.enrichment" title="omicverse.bulk.pyGSE.enrichment">enrichment</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>gene set enrichment analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>A pandas.DataFrame object containing the enrichment results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyGSE.plot_enrichment">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyGSE" title="omicverse.bulk.pyGSE">pyGSE</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyGSE.plot_enrichment" title="omicverse.bulk.pyGSE.plot_enrichment">plot_enrichment</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">node_size</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[5, 10, 15]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_loc</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_xlabel</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Fractions of genes'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(2, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cmap</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'YlGnBu'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_knock</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_maxsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span></span>
</p>
<div class="mkapi-document">
<p>Plot the gene set enrichment result.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of enriched terms to plot. Default is 10.</p>
</li>
<li>
<p><span class="mkapi-item-name">node_size</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
A list of integers defining the size of nodes in the plot. Default is [5,10,15].</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_loc</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The location of the colorbar on the plot. Default is 2.</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the colorbar label. Default is 12.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. Default is an empty string.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_xlabel</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The label of the x-axis. Default is 'Fractions of genes'.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the plot. Default is (2,4).</p>
</li>
<li>
<p><span class="mkapi-item-name">cmap</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The colormap to use for the plot. Default is 'YlGnBu'.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span> <span class="mkapi-dash">—</span>
A matplotlib.axes.Axes object.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.geneset_enrichment">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.geneset_enrichment" title="omicverse.bulk.geneset_enrichment">geneset_enrichment</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_list</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pathways_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pvalue_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pvalue_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'auto'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">organism</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Human'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">description</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'None'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">background</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">outdir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'./enrichr'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cutoff</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.5</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Perform gene set enrichment analysis using Enrichr API.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_list</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
List of gene symbols to be tested for enrichment.</p>
</li>
<li>
<p><span class="mkapi-item-name">pathways_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Dictionary of pathway library names and corresponding Enrichr API URLs.</p>
</li>
<li>
<p><span class="mkapi-item-name">pvalue_threshold</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
P-value threshold for significant pathways. (0.05)</p>
</li>
<li>
<p><span class="mkapi-item-name">pvalue_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Type of p-value correction to use. 'auto' uses Benjamini-Hochberg correction for small gene sets (&lt;500 genes) and Bonferroni correction for larger gene sets. 'bh' uses only Benjamini-Hochberg correction. 'bonferroni' uses only Bonferroni correction. ('auto')</p>
</li>
<li>
<p><span class="mkapi-item-name">organism</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Organism of the input gene list. ('Human')</p>
</li>
<li>
<p><span class="mkapi-item-name">description</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Description of the input gene list. ('None')</p>
</li>
<li>
<p><span class="mkapi-item-name">background</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
Background gene list to use for enrichment analysis. If None, the background gene list is automatically set to the organism-specific gene list. (None)</p>
</li>
<li>
<p><span class="mkapi-item-name">outdir</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Output directory for Enrichr results. ('./enrichr')</p>
</li>
<li>
<p><span class="mkapi-item-name">cutoff</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
Show enriched terms which Adjusted P-value &lt; cutoff. (0.5)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">enrich_res</span> <span class="mkapi-dash">—</span>
A pandas DataFrame containing the enrichment results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.geneset_plot">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.geneset_plot" title="omicverse.bulk.geneset_plot">geneset_plot</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">enrich_res</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">node_size</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[5, 10, 15]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_loc</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[2, 0.55, 0.5, 0.02]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cax_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_xlabel</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Fractions of genes'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(2, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cmap</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'YlGnBu'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_knock</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_maxsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">bbox_to_anchor_used</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(-0.45, -13)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">node_diameter</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">custom_ticks</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[5, 10]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ax</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span></span>
</p>
<div class="mkapi-document">
<p>Plot the gene set enrichment result.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">enrich_res</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
Enrichment results DataFrame.</p>
</li>
<li>
<p><span class="mkapi-item-name">num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of enriched terms to plot. Default: 10.</p>
</li>
<li>
<p><span class="mkapi-item-name">node_size</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
A list of integers defining the size of nodes in the plot. Default: [5,10,15].</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_loc</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The location, width and height of the colorbar on the plot. Default: [2, 0.55, 0.5, 0.02].</p>
</li>
<li>
<p><span class="mkapi-item-name">cax_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the colorbar label. Default: 12.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. Default: ''.</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_xlabel</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The label of the x-axis. Default: 'Fractions of genes'.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the plot. Default: (2,4).</p>
</li>
<li>
<p><span class="mkapi-item-name">cmap</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The colormap to use for the plot. Default: 'YlGnBu'.</p>
</li>
<li>
<p><span class="mkapi-item-name">text_knock</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of characters to knock off the end of the term name. Default: 5.</p>
</li>
<li>
<p><span class="mkapi-item-name">text_maxsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The maximum fontsize of the term names. Default: 20.</p>
</li>
<li>
<p><span class="mkapi-item-name">bbox_to_anchor_used</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The anchor point for placing the legend. Default: (-0.45, -13).</p>
</li>
<li>
<p><span class="mkapi-item-name">node_diameter</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The base size for nodes in the plot. Default: 10.</p>
</li>
<li>
<p><span class="mkapi-item-name">custom_ticks</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
Custom tick marks for the plot. Default: [5,10].</p>
</li>
<li>
<p><span class="mkapi-item-name">ax</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span> <span class="mkapi-dash">—</span>
Matplotlib axes object. Default: None.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ax</span> <span class="mkapi-dash">—</span>
A matplotlib.axes.Axes object containing the enrichment bubble plot.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.geneset_enrichment_GSEA">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.geneset_enrichment_GSEA" title="omicverse.bulk.geneset_enrichment_GSEA">geneset_enrichment_GSEA</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_rnk</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">pathways_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">processes</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">permutation_num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">outdir</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'./enrichr_gsea'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">format</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'png'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">seed</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">112</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">dict</span></span>
</p>
<div class="mkapi-document">
<p>Enrichment analysis using GSEA.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_rnk</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
Pre-ranked correlation table or pandas DataFrame. Same input with <code>GSEA</code> .rnk file.</p>
</li>
<li>
<p><span class="mkapi-item-name">pathways_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Dictionary of pathway library names and corresponding Enrichr API URLs.</p>
</li>
<li>
<p><span class="mkapi-item-name">processes</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of Processes you are going to use. (8)</p>
</li>
<li>
<p><span class="mkapi-item-name">permutation_num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of permutations for significance computation. (100)</p>
</li>
<li>
<p><span class="mkapi-item-name">outdir</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Output directory for Enrichr results. ('./enrichr_gsea')</p>
</li>
<li>
<p><span class="mkapi-item-name">format</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Matplotlib figure format. ('png')</p>
</li>
<li>
<p><span class="mkapi-item-name">seed</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Random seed. (112)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Enrichment.geneset_enrichment_GSEA.pre_res">pre_res</span></span> <span class="mkapi-dash">—</span>
A prerank object containing the enrichment results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.geneset_plot_multi">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.geneset_plot_multi" title="omicverse.bulk.geneset_plot_multi">geneset_plot_multi</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">enr_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, <span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span>]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">colors_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, str]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fig_xlabel</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Fractions of genes'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(2, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cmap</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'YlGnBu'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_knock</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">text_maxsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ax</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span></span>
</p>
<div class="mkapi-document">
<p>Enrichment multi genesets analysis using GSEA.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">enr_dict</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, <span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span>]</span> <span class="mkapi-dash">—</span>
A dictionary of enrichment results.</p>
</li>
<li>
<p><span class="mkapi-item-name">colors_dict</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, str]</span> <span class="mkapi-dash">—</span>
A dictionary of colors for each gene set.</p>
</li>
<li>
<p><span class="mkapi-item-name">num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of enriched terms to plot. (5)</p>
</li>
<li>
<p><span class="mkapi-item-name">fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the plot. (10)</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. ('')</p>
</li>
<li>
<p><span class="mkapi-item-name">fig_xlabel</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The label of the x-axis. ('Fractions of genes')</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the plot. ((2,4))</p>
</li>
<li>
<p><span class="mkapi-item-name">cmap</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The colormap to use for the plot. ('YlGnBu')</p>
</li>
<li>
<p><span class="mkapi-item-name">text_knock</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of characters to knock off the end of the term name. (5)</p>
</li>
<li>
<p><span class="mkapi-item-name">text_maxsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The maximum fontsize of the term names. (20)</p>
</li>
<li>
<p><span class="mkapi-item-name">ax</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span> <span class="mkapi-dash">—</span>
A matplotlib.axes.Axes object. (None)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ax</span> <span class="mkapi-dash">—</span>
The matplotlib axes object</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.enrichment_multi_concat">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.enrichment_multi_concat" title="omicverse.bulk.enrichment_multi_concat">enrichment_multi_concat</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">enr_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Dict">Dict</span>[str, <span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span>]</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyPPI">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.pyPPI" title="omicverse.bulk.pyPPI">pyPPI</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">species</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_type_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_color_dict</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">score</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.4</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Initialize protein-protein interaction analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
List of gene names for PPI analysis</p>
</li>
<li>
<p><span class="mkapi-item-name">species</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
NCBI taxon identifiers (e.g. Human is 9606, see STRING organisms)</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_type_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Dictionary mapping gene names to gene types</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_color_dict</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
Dictionary mapping gene names to colors for visualization</p>
</li>
<li>
<p><span class="mkapi-item-name">score</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
Threshold for protein interaction confidence (default: 0.4)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyPPI.interaction_analysis" title="omicverse.bulk.pyPPI.interaction_analysis">interaction_analysis</a></span>
<span class="mkapi-dash">—</span>
Perform protein-protein interaction analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyPPI.plot_network" title="omicverse.bulk.pyPPI.plot_network">plot_network</a></span>
<span class="mkapi-dash">—</span>
Plot protein-protein interaction network.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyPPI.interaction_analysis">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyPPI" title="omicverse.bulk.pyPPI">pyPPI</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyPPI.interaction_analysis" title="omicverse.bulk.pyPPI.interaction_analysis">interaction_analysis</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="networkx">nx</span>.<span class="mkapi-tooltip" title="networkx.classes.graph.Graph">Graph</span></span></span>
</p>
<div class="mkapi-document">
<p>Perform protein-protein interaction analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._network.pyPPI.interaction_analysis.G">G</span></span> <span class="mkapi-dash">—</span>
NetworkX Graph object containing PPI network for query genes</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyPPI.plot_network">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyPPI" title="omicverse.bulk.pyPPI">pyPPI</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyPPI.plot_network" title="omicverse.bulk.pyPPI.plot_network">plot_network</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="typing.Tuple">Tuple</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.figure">figure</span>.<span class="mkapi-tooltip" title="matplotlib.figure.Figure">Figure</span>, <span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span></span>
</p>
<div class="mkapi-document">
<p>Plot protein-protein interaction network.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Additional keyword arguments passed to plot_network function</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">fig</span> <span class="mkapi-dash">—</span>
Figure object containing PPI network plot
ax: Axes object containing PPI network plot</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.string_interaction">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.string_interaction" title="omicverse.bulk.string_interaction">string_interaction</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">species</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Analyze protein-protein interaction network using STRING database.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
List of gene names for PPI analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name">species</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
NCBI taxon identifiers (e.g. Human is 9606, see STRING organisms).</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._network.string_interaction.res">res</span></span> <span class="mkapi-dash">—</span>
DataFrame containing protein-protein interaction data with columns stringId_A, stringId_B, preferredName_A, preferredName_B, ncbiTaxonId, score, nscore, fscore, pscore, ascore, escore, dscore, tscore.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.string_map">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.string_map" title="omicverse.bulk.string_map">string_map</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">species</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Map gene names to STRING database identifiers.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
List of gene names for PPI analysis</p>
</li>
<li>
<p><span class="mkapi-item-name">species</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
NCBI taxon identifiers (e.g. Human is 9606, see STRING organisms)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._network.string_map.res">res</span></span> <span class="mkapi-dash">—</span>
DataFrame containing gene mapping information with columns:
     queryItem, queryIndex, stringId, ncbiTaxonId, taxonName,
     preferredName, annotation</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.generate_G">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.generate_G" title="omicverse.bulk.generate_G">generate_G</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">species</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-comma">, </span><span class="mkapi-arg">score</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.4</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="networkx">nx</span>.<span class="mkapi-tooltip" title="networkx.classes.graph.Graph">Graph</span></span></span>
</p>
<div class="mkapi-document">
<p>Generate protein-protein interaction network from STRING database.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
List of gene names for PPI analysis</p>
</li>
<li>
<p><span class="mkapi-item-name">species</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
NCBI taxon identifiers (e.g. Human is 9606, see STRING organisms)</p>
</li>
<li>
<p><span class="mkapi-item-name">score</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
Threshold for protein interaction confidence (default: 0.4)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._network.generate_G.G">G</span></span> <span class="mkapi-dash">—</span>
NetworkX Graph object containing PPI network</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.get_chm13_gene">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.get_chm13_gene" title="omicverse.bulk.get_chm13_gene">get_chm13_gene</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Get T2T-CHM13 genome annotation and mapping data.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._chm13.get_chm13_gene.data">data</span></span> <span class="mkapi-dash">—</span>
DataFrame with CHM13 gene annotations containing columns:
      'closest GENCODE ID', 'Gene name', 'biotype', 'Novel region',
      'GRCh38 issue', 'Either novel or issue', 
      'Intersects medically relevant name list'</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.find_chm13_gene">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.find_chm13_gene" title="omicverse.bulk.find_chm13_gene">find_chm13_gene</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">search_gene_list</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Identify T2T-CHM13 novel and GRCh38 issue genes from input gene list.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">search_gene_list</span>
<span class="mkapi-dash">—</span>
List of gene IDs to search for CHM13 annotations</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._chm13.find_chm13_gene.res">res</span></span> <span class="mkapi-dash">—</span>
Dictionary containing:
     'Novel' - List of T2T-CHM13 novel region genes found in input
     'GRCh38' - List of GRCh38 issue genes found in input</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">raw_data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Initialize the pyDEG class.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">raw_data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
The raw data to be processed.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.drop_duplicates_index" title="omicverse.bulk.pyDEG.drop_duplicates_index">drop_duplicates_index</a></span>
<span class="mkapi-dash">—</span>
Drop the duplicated index of data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.normalize" title="omicverse.bulk.pyDEG.normalize">normalize</a></span>
<span class="mkapi-dash">—</span>
Normalize the data using DESeq2 method.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.foldchange_set" title="omicverse.bulk.pyDEG.foldchange_set">foldchange_set</a></span>
<span class="mkapi-dash">—</span>
Set fold-change and p-value thresholds to classify differentially expressed genes as up-regulated, down-regulated, or not significant.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.plot_volcano" title="omicverse.bulk.pyDEG.plot_volcano">plot_volcano</a></span>
<span class="mkapi-dash">—</span>
Generate a volcano plot for the differential gene expression analysis results.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.plot_boxplot" title="omicverse.bulk.pyDEG.plot_boxplot">plot_boxplot</a></span>
<span class="mkapi-dash">—</span>
Plot the boxplot of genes from dds data</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.ranking2gsea" title="omicverse.bulk.pyDEG.ranking2gsea">ranking2gsea</a></span>
<span class="mkapi-dash">—</span>
Ranking the result of dds data for gsea analysis</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyDEG.deg_analysis" title="omicverse.bulk.pyDEG.deg_analysis">deg_analysis</a></span>
<span class="mkapi-dash">—</span>
Differential expression analysis.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.drop_duplicates_index">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.drop_duplicates_index" title="omicverse.bulk.pyDEG.drop_duplicates_index">drop_duplicates_index</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Drop the duplicated index of data.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.pyDEG.data">data</span></span> <span class="mkapi-dash">—</span>
The data after dropping the duplicated index.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.normalize">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.normalize" title="omicverse.bulk.pyDEG.normalize">normalize</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Normalize the data using DESeq2 method.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.pyDEG.data">data</span></span> <span class="mkapi-dash">—</span>
The normalized data.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.foldchange_set">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.foldchange_set" title="omicverse.bulk.pyDEG.foldchange_set">foldchange_set</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">fc_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">-1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pval_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">logp_max</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">6</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fold_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">None</span></span>
</p>
<div class="mkapi-document">
<p>Set fold-change and p-value thresholds to classify differentially expressed genes as up-regulated, down-regulated, or not significant.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">fc_threshold</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Absolute fold-change threshold. If set to -1, the threshold is calculated based on the histogram of log2 fold-changes. (-1)</p>
</li>
<li>
<p><span class="mkapi-item-name">pval_threshold</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
p-value threshold for determining significance. (0.05)</p>
</li>
<li>
<p><span class="mkapi-item-name">logp_max</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Maximum value for log-transformed p-values. (6)</p>
</li>
<li>
<p><span class="mkapi-item-name">fold_threshold</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Index of the histogram bin corresponding to the fold-change threshold (only applicable if fc_threshold=-1). (0)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">None</span> <span class="mkapi-dash">—</span>
None</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.plot_volcano">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.plot_volcano" title="omicverse.bulk.pyDEG.plot_volcano">plot_volcano</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(4, 4)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pval_name</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'qvalue'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fc_name</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'log2FC'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">titlefont</span><span class="mkapi-colon">: </span><span class="mkapi-ann">dict</span><span class="mkapi-equal"> = </span><span class="mkapi-default">{'weight': 'normal', 'size': 14}</span><span class="mkapi-comma">, </span><span class="mkapi-arg">up_color</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#e25d5d'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">down_color</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#7388c1'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">normal_color</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#d7d7d7'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">up_fontcolor</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#e25d5d'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">down_fontcolor</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#7388c1'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">normal_fontcolor</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'#d7d7d7'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">legend_bbox</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(0.8, -0.2)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">legend_ncol</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">legend_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plot_genes</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[list]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plot_genes_num</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plot_genes_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ticks_fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ax</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span></span></span>
</p>
<div class="mkapi-document">
<p>Generate a volcano plot for the differential gene expression analysis results.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The size of the generated figure. ((4,4))</p>
</li>
<li>
<p><span class="mkapi-item-name">pval_name</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Column name for p-values. ('qvalue')</p>
</li>
<li>
<p><span class="mkapi-item-name">fc_name</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Column name for fold changes. ('log2FC')</p>
</li>
<li>
<p><span class="mkapi-item-name">title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot. ('')</p>
</li>
<li>
<p><span class="mkapi-item-name">titlefont</span> : 
<span class="mkapi-item-type">dict</span> <span class="mkapi-dash">—</span>
A dictionary of font properties for the plot title. ({'weight':'normal','size':14,})</p>
</li>
<li>
<p><span class="mkapi-item-name">up_color</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The color of the up-regulated genes in the plot. ('#e25d5d')</p>
</li>
<li>
<p><span class="mkapi-item-name">down_color</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The color of the down-regulated genes in the plot. ('#7388c1')</p>
</li>
<li>
<p><span class="mkapi-item-name">normal_color</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The color of the non-significant genes in the plot. ('#d7d7d7')</p>
</li>
<li>
<p><span class="mkapi-item-name">up_fontcolor</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Font color for up-regulated gene labels. ('#e25d5d')</p>
</li>
<li>
<p><span class="mkapi-item-name">down_fontcolor</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Font color for down-regulated gene labels. ('#7388c1')</p>
</li>
<li>
<p><span class="mkapi-item-name">normal_fontcolor</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Font color for normal gene labels. ('#d7d7d7')</p>
</li>
<li>
<p><span class="mkapi-item-name">legend_bbox</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
A tuple containing the coordinates of the legend's bounding box. ((0.8, -0.2))</p>
</li>
<li>
<p><span class="mkapi-item-name">legend_ncol</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of columns in the legend. (2)</p>
</li>
<li>
<p><span class="mkapi-item-name">legend_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The font size of the legend. (12)</p>
</li>
<li>
<p><span class="mkapi-item-name">plot_genes</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[list]</span> <span class="mkapi-dash">—</span>
A list of genes to be plotted on the volcano plot. (None)</p>
</li>
<li>
<p><span class="mkapi-item-name">plot_genes_num</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of genes to be plotted on the volcano plot. (10)</p>
</li>
<li>
<p><span class="mkapi-item-name">plot_genes_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The font size of the genes to be plotted on the volcano plot. (10)</p>
</li>
<li>
<p><span class="mkapi-item-name">ticks_fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The font size of the ticks. (12)</p>
</li>
<li>
<p><span class="mkapi-item-name">ax</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Optional">Optional</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span> <span class="mkapi-dash">—</span>
Matplotlib axis object. (None)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.pyDEG.plot_volcano.ax">ax</span></span> <span class="mkapi-dash">—</span>
The generated volcano plot.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.plot_boxplot">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.plot_boxplot" title="omicverse.bulk.pyDEG.plot_boxplot">plot_boxplot</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">genes</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">treatment_groups</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">control_groups</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">log</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">treatment_name</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Treatment'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">control_name</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Control'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(4, 3)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">palette</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">['#a64d79', '#674ea7']</span><span class="mkapi-comma">, </span><span class="mkapi-arg">title</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'Gene Expression'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fontsize</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">12</span><span class="mkapi-comma">, </span><span class="mkapi-arg">legend_bbox</span><span class="mkapi-colon">: </span><span class="mkapi-ann">tuple</span><span class="mkapi-equal"> = </span><span class="mkapi-default">(1, 0.55)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">legend_ncol</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwarg</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="typing.Tuple">Tuple</span>[<span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.figure">figure</span>.<span class="mkapi-tooltip" title="matplotlib.figure.Figure">Figure</span>, <span class="mkapi-tooltip" title="matplotlib">matplotlib</span>.<span class="mkapi-tooltip" title="matplotlib.axes">axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes">_axes</span>.<span class="mkapi-tooltip" title="matplotlib.axes._axes.Axes">Axes</span>]</span></span>
</p>
<div class="mkapi-document">
<p>Plot the boxplot of genes from dds data</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">genes</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The genes to plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">treatment_groups</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The treatment groups.</p>
</li>
<li>
<p><span class="mkapi-item-name">control_groups</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The control groups.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The figure size.</p>
</li>
<li>
<p><span class="mkapi-item-name">palette</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The color palette.</p>
</li>
<li>
<p><span class="mkapi-item-name">title</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The title of the plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">fontsize</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The fontsize of the plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">legend_bbox</span> : 
<span class="mkapi-item-type">tuple</span> <span class="mkapi-dash">—</span>
The bbox of the legend.</p>
</li>
<li>
<p><span class="mkapi-item-name">legend_ncol</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The number of columns of the legend.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwarg</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Other arguments for plot_boxplot function.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">fig</span> <span class="mkapi-dash">—</span>
The figure of the plot.
ax: The axis of the plot.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.ranking2gsea">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.ranking2gsea" title="omicverse.bulk.pyDEG.ranking2gsea">ranking2gsea</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">rank_max</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">200</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rank_min</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">274</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Ranking the result of dds data for gsea analysis</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">rank_max</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The max rank of the result.</p>
</li>
<li>
<p><span class="mkapi-item-name">rank_min</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
The min rank of the result.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.pyDEG.ranking2gsea.rnk">rnk</span></span> <span class="mkapi-dash">—</span>
The ranking result.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyDEG.deg_analysis">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyDEG" title="omicverse.bulk.pyDEG">pyDEG</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyDEG.deg_analysis" title="omicverse.bulk.pyDEG.deg_analysis">deg_analysis</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">group1</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">group2</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-comma">, </span><span class="mkapi-arg">method</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'DEseq2'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">alpha</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">multipletests_method</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'fdr_bh'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_cpus</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cooks_filter</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">independent_filter</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Differential expression analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">group1</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The first group to be compared.</p>
</li>
<li>
<p><span class="mkapi-item-name">group2</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
The second group to be compared.</p>
</li>
<li>
<p><span class="mkapi-item-name">method</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The method to be used for differential expression analysis.</p>
<ul>
<li><code>DEseq2</code>: DEseq2</li>
<li><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.pyDEG.deg_analysis.ttest"><code>ttest</code></span>: ttest</li>
<li><code>wilcox</code>: wilconx test</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">alpha</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
The threshold of p-value.</p>
</li>
<li>
<p><span class="mkapi-item-name">multipletests_method</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
- <code>bonferroni</code> : one-step correction
- <code>sidak</code> : one-step correction
- <code>holm-sidak</code> : step down method using Sidak adjustments
- <code>holm</code> : step-down method using Bonferroni adjustments
- <code>simes-hochberg</code> : step-up method  (independent)
- <code>hommel</code> : closed method based on Simes tests (non-negative)
- <code>fdr_bh</code> : Benjamini/Hochberg  (non-negative)
- <code>fdr_by</code> : Benjamini/Yekutieli (negative)
- <code>fdr_tsbh</code> : two stage fdr correction (non-negative)
- <code>fdr_tsbky</code> : two stage fdr correction (non-negative)</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<p>Returns
    result: The result of differential expression analysis.</p>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
<li>
<p><span class="mkapi-item-type">ImportError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.deseq2_normalize">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.deseq2_normalize" title="omicverse.bulk.deseq2_normalize">deseq2_normalize</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Normalize the data using DESeq2 method.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
The data to be normalized.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">data</span> <span class="mkapi-dash">—</span>
The normalized data.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.estimateSizeFactors">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.estimateSizeFactors" title="omicverse.bulk.estimateSizeFactors">estimateSizeFactors</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.series.Series">Series</span></span></span>
</p>
<div class="mkapi-document">
<p>Estimate size factors for data normalization.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
A pandas DataFrame of gene expression data where rows correspond to samples and columns correspond to genes.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.estimateSizeFactors.scale">scale</span></span> <span class="mkapi-dash">—</span>
A pandas Series of size factors, one for each sample.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="s1">'abcdefghij'</span><span class="p">))</span>
<span class="n">size_factors</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.estimateDispersions">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.estimateDispersions" title="omicverse.bulk.estimateDispersions">estimateDispersions</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">counts</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.series.Series">Series</span></span></span>
</p>
<div class="mkapi-document">
<p>Estimate the dispersion parameter of the Negative Binomial distribution for each gene in the input count matrix.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">counts</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
Input count matrix with shape (n_genes, n_samples).</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.estimateDispersions.disp">disp</span></span> <span class="mkapi-dash">—</span>
Array of dispersion values for each gene in the input count matrix.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.Matrix_ID_mapping">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.Matrix_ID_mapping" title="omicverse.bulk.Matrix_ID_mapping">Matrix_ID_mapping</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_ref_path</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-comma">, </span><span class="mkapi-arg">keep_unmapped</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Map gene IDs in the input data to gene symbols using a reference table.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
The input data containing gene IDs as index.</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_ref_path</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
The path to the reference table containing the mapping from gene IDs to gene symbols.</p>
</li>
<li>
<p><span class="mkapi-item-name">keep_unmapped</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
Whether to keep genes that are not found in the mapping table. If True, unmapped genes retain their original IDs. If False, unmapped genes are removed (original behavior). Default: True.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">data</span> <span class="mkapi-dash">—</span>
The input data with gene IDs mapped to gene symbols.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.data_drop_duplicates_index">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.data_drop_duplicates_index" title="omicverse.bulk.data_drop_duplicates_index">data_drop_duplicates_index</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Drop the duplicated index of data.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
The data to be processed.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._Deseq2.data_drop_duplicates_index.data">data</span></span> <span class="mkapi-dash">—</span>
The data after dropping the duplicated index.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gdc_sample_sheep</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gdc_download_files</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-comma">, </span><span class="mkapi-arg">clinical_cart</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>TCGA (The Cancer Genome Atlas) data analysis module.</p>
<p>This class provides comprehensive functionality for downloading, processing,
and analyzing TCGA genomic and clinical data.</p>
<p>Initialize TCGA analysis module.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gdc_sample_sheep</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Path to TCGA Sample Sheet TSV file</p>
</li>
<li>
<p><span class="mkapi-item-name">gdc_download_files</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Path to downloaded TCGA data files directory</p>
</li>
<li>
<p><span class="mkapi-item-name">clinical_cart</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Path to TCGA clinical data tar.gz file</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.adata_read" title="omicverse.bulk.pyTCGA.adata_read">adata_read</a></span>
<span class="mkapi-dash">—</span>
Read AnnData object from file.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.adata_init" title="omicverse.bulk.pyTCGA.adata_init">adata_init</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.adata_meta_init" title="omicverse.bulk.pyTCGA.adata_meta_init">adata_meta_init</a></span>
<span class="mkapi-dash">—</span>
Initialize AnnData metadata.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.survial_init" title="omicverse.bulk.pyTCGA.survial_init">survial_init</a></span>
<span class="mkapi-dash">—</span>
Initialize survival analysis data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.index_init" title="omicverse.bulk.pyTCGA.index_init">index_init</a></span>
<span class="mkapi-dash">—</span>
Initialize gene indices for AnnData construction.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.expression_init" title="omicverse.bulk.pyTCGA.expression_init">expression_init</a></span>
<span class="mkapi-dash">—</span>
Initialize expression matrices for TCGA data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.matrix_construct" title="omicverse.bulk.pyTCGA.matrix_construct">matrix_construct</a></span>
<span class="mkapi-dash">—</span>
Construct AnnData object from expression matrices.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.matrix_normalize" title="omicverse.bulk.pyTCGA.matrix_normalize">matrix_normalize</a></span>
<span class="mkapi-dash">—</span>
Normalize expression matrix using DESeq2 method.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.survival_analysis" title="omicverse.bulk.pyTCGA.survival_analysis">survival_analysis</a></span>
<span class="mkapi-dash">—</span>
Perform survival analysis for a specific gene.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.pyTCGA.survial_analysis_all" title="omicverse.bulk.pyTCGA.survial_analysis_all">survial_analysis_all</a></span>
<span class="mkapi-dash">—</span>
Perform survival analysis for all genes in the dataset.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.adata_read">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.adata_read" title="omicverse.bulk.pyTCGA.adata_read">adata_read</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">path</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Read AnnData object from file.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">path</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Path to AnnData file</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.adata_init">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.adata_init" title="omicverse.bulk.pyTCGA.adata_init">adata_init</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.adata_meta_init">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.adata_meta_init" title="omicverse.bulk.pyTCGA.adata_meta_init">adata_meta_init</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">var_names</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">['gene_name', 'gene_type']</span><span class="mkapi-comma">, </span><span class="mkapi-arg">obs_names</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">['Case ID', 'Sample Type']</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="anndata">anndata</span>.<span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span></span>
</p>
<div class="mkapi-document">
<p>Initialize AnnData metadata.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">var_names</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
Column names for variable (gene) metadata (default: ['gene_name','gene_type'])</p>
</li>
<li>
<p><span class="mkapi-item-name">obs_names</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
Column names for observation (sample) metadata (default: ['Case ID','Sample Type'])</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._tcga.pyTCGA.adata_meta_init.adata">adata</span></span> <span class="mkapi-dash">—</span>
AnnData object with initialized metadata</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.survial_init">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.survial_init" title="omicverse.bulk.pyTCGA.survial_init">survial_init</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Initialize survival analysis data.</p>
<p>Processes clinical data to extract survival information including
vital status and survival days.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.index_init">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.index_init" title="omicverse.bulk.pyTCGA.index_init">index_init</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">list</span></span>
</p>
<div class="mkapi-document">
<p>Initialize gene indices for AnnData construction.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="omicverse.bulk._tcga.pyTCGA.index_init.all_lncRNA_index">all_lncRNA_index</span></span> <span class="mkapi-dash">—</span>
List of all gene indices from TCGA samples</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.expression_init">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.expression_init" title="omicverse.bulk.pyTCGA.expression_init">expression_init</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Initialize expression matrices for TCGA data.</p>
<p>Creates count, TPM, and FPKM expression matrices from TCGA files.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.matrix_construct">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.matrix_construct" title="omicverse.bulk.pyTCGA.matrix_construct">matrix_construct</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Construct AnnData object from expression matrices.</p>
<p>Creates AnnData object with multiple layers including raw counts,
TPM, FPKM, and DESeq2-normalized expression.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.matrix_normalize">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.matrix_normalize" title="omicverse.bulk.pyTCGA.matrix_normalize">matrix_normalize</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">data</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Normalize expression matrix using DESeq2 method.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">data</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
Raw count expression matrix to normalize</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">data</span> <span class="mkapi-dash">—</span>
DESeq2-normalized expression matrix</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.survival_analysis">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.survival_analysis" title="omicverse.bulk.pyTCGA.survival_analysis">survival_analysis</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-comma">, </span><span class="mkapi-arg">layer</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'raw'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plot</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_threshold</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'median'</span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="typing.Tuple">Tuple</span>[float, float]</span></span>
</p>
<div class="mkapi-document">
<p>Perform survival analysis for a specific gene.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Gene name for survival analysis</p>
</li>
<li>
<p><span class="mkapi-item-name">layer</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
AnnData layer to use for expression values (default: 'raw')</p>
</li>
<li>
<p><span class="mkapi-item-name">plot</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
Whether to generate Kaplan-Meier survival plot (default: False)</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_threshold</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Method to split samples into high/low expression groups
(default: 'median', options: 'median', 'mean', or numeric value)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">test_statistic</span> <span class="mkapi-dash">—</span>
Log-rank test statistic
pvalue: Log-rank test p-value</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.pyTCGA.survial_analysis_all">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.pyTCGA" title="omicverse.bulk.pyTCGA">pyTCGA</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.pyTCGA.survial_analysis_all" title="omicverse.bulk.pyTCGA.survial_analysis_all">survial_analysis_all</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Perform survival analysis for all genes in the dataset.</p>
<p>Calculates survival statistics for every gene and stores results
in AnnData.var as 'survial_test_statistic' and 'survial_p' columns.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.batch_correction">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.batch_correction" title="omicverse.bulk.batch_correction">batch_correction</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata">anndata</span>.<span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">key_added</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'batch_correction'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Perform batch effect correction using ComBat algorithm.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata">anndata</span>.<span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData object containing expression data.</p>
</li>
<li>
<p><span class="mkapi-item-name">batch_key</span>
<span class="mkapi-dash">—</span>
Key in adata.obs containing batch information.</p>
</li>
<li>
<p><span class="mkapi-item-name">key_added</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Name for the corrected data layer. Default: 'batch_correction'.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">None</span> <span class="mkapi-dash">—</span>
The function modifies adata.layers[key_added] in place with batch-corrected expression data.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ImportError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.Deconvolution">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.bulk.Deconvolution" title="omicverse.bulk.Deconvolution">Deconvolution</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_bulk</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_single</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_single_cells</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">celltype_key</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'celltype'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cellstate_key</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gpu</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Union">Union</span>[int, str]</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.bulk.Deconvolution.deconvolution" title="omicverse.bulk.Deconvolution.deconvolution">deconvolution</a></span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.bulk.Deconvolution.deconvolution">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.bulk.Deconvolution" title="omicverse.bulk.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.bulk.Deconvolution.deconvolution" title="omicverse.bulk.Deconvolution.deconvolution">deconvolution</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">method</span><span class="mkapi-equal">=</span><span class="mkapi-default">'tape'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sep</span><span class="mkapi-equal">=</span><span class="mkapi-default">'\t'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">scaler</span><span class="mkapi-equal">=</span><span class="mkapi-default">'mms'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">datatype</span><span class="mkapi-equal">=</span><span class="mkapi-default">'counts'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">genelenfile</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mode</span><span class="mkapi-equal">=</span><span class="mkapi-default">'overall'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adaptive</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">variance_threshold</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.98</span><span class="mkapi-comma">, </span><span class="mkapi-arg">save_model_name</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_size</span><span class="mkapi-equal">=</span><span class="mkapi-default">128</span><span class="mkapi-comma">, </span><span class="mkapi-arg">epochs</span><span class="mkapi-equal">=</span><span class="mkapi-default">128</span><span class="mkapi-comma">, </span><span class="mkapi-arg">seed</span><span class="mkapi-equal">=</span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">scale_size</span><span class="mkapi-equal">=</span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">scale</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_cores</span><span class="mkapi-equal">=</span><span class="mkapi-default">4</span><span class="mkapi-comma">, </span><span class="mkapi-arg">fast_mode</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pseudobulk_size</span><span class="mkapi-equal">=</span><span class="mkapi-default">2000</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2025, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascript/mkapi.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>