
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api deseq - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.bulk.pyDEG">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api deseq
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis in single cell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyDEG">
    pyDEG
  </a>
<nav aria-label="pyDEG" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.__init__">
    __init__()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.drop_duplicates_index">
    drop_duplicates_index()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.normalize">
    normalize()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.foldchange_set">
    foldchange_set()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.deg_analysis">
    deg_analysis()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.ranking2gsea">
    ranking2gsea()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.plot_volcano">
    plot_volcano()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk._Deseq2.pyDEG.plot_boxplot">
    plot_boxplot()
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.deseq2_normalize">
    deseq2_normalize()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.estimateSizeFactors">
    estimateSizeFactors()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.estimateDispersions">
    estimateDispersions()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.Matrix_ID_mapping">
    Matrix_ID_mapping()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api deseq</h1>
<div class="doc doc-object doc-class">
<h2 class="doc doc-heading" id="omicverse.bulk.pyDEG">
<code>omicverse.bulk.pyDEG</code>
<a class="headerlink" href="#omicverse.bulk.pyDEG" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p class="doc doc-class-bases">
        Bases: <code>object</code></p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="167 "></span><span class="k">class</span> <span class="nc">pyDEG</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="linenos" data-linenos="168 "></span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">raw_data</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="171 "></span><span class="w">        </span><span class="sd">"""Initialize the pyDEG class.</span>
<span class="linenos" data-linenos="172 "></span>
<span class="linenos" data-linenos="173 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="174 "></span><span class="sd">            raw_data: The raw data to be processed.</span>
<span class="linenos" data-linenos="175 "></span>
<span class="linenos" data-linenos="176 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="177 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">raw_data</span><span class="o">=</span><span class="n">raw_data</span>
<span class="linenos" data-linenos="178 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">raw_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="179 "></span>
<span class="linenos" data-linenos="180 "></span>    <span class="k">def</span> <span class="nf">drop_duplicates_index</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="181 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">        Drop the duplicated index of data.</span>
<span class="linenos" data-linenos="183 "></span>
<span class="linenos" data-linenos="184 "></span><span class="sd">        Returns</span>
<span class="linenos" data-linenos="185 "></span><span class="sd">            data: The data after dropping the duplicated index.</span>
<span class="linenos" data-linenos="186 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="187 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">data_drop_duplicates_index</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="188 "></span>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="189 "></span>
<span class="linenos" data-linenos="190 "></span>    <span class="k">def</span> <span class="nf">normalize</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="191 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        Normalize the data using DESeq2 method.</span>
<span class="linenos" data-linenos="193 "></span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        Returns</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">            data: The normalized data.</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="197 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">size_factors</span><span class="o">=</span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="198 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">deseq2_normalize</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="199 "></span>        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="200 "></span>
<span class="linenos" data-linenos="201 "></span>    <span class="k">def</span> <span class="nf">foldchange_set</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">fc_threshold</span><span class="p">:</span><span class="nb">int</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="n">pval_threshold</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">logp_max</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span><span class="n">fold_threshold</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="linenos" data-linenos="202 "></span><span class="w">        </span><span class="sd">"""</span>
<span class="linenos" data-linenos="203 "></span><span class="sd">        Sets fold-change and p-value thresholds to classify differentially expressed genes as up-regulated, down-regulated, or not significant.</span>
<span class="linenos" data-linenos="204 "></span>
<span class="linenos" data-linenos="205 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="206 "></span><span class="sd">            fc_threshold: Absolute fold-change threshold. If set to -1, the threshold is calculated based on the histogram of log2 fold-changes.</span>
<span class="linenos" data-linenos="207 "></span><span class="sd">            pval_threshold: p-value threshold for determining significance.</span>
<span class="linenos" data-linenos="208 "></span><span class="sd">            logp_max: Maximum value for log-transformed p-values.</span>
<span class="linenos" data-linenos="209 "></span><span class="sd">            fold_threshold: Index of the histogram bin corresponding to the fold-change threshold (only applicable if fc_threshold=-1).</span>
<span class="linenos" data-linenos="210 "></span>
<span class="linenos" data-linenos="211 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="212 "></span>        <span class="k">if</span> <span class="n">fc_threshold</span><span class="o">==-</span><span class="mi">1</span><span class="p">:</span>
<span class="linenos" data-linenos="213 "></span>            <span class="n">foldp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">())</span>
<span class="linenos" data-linenos="214 "></span>            <span class="n">foldchange</span><span class="o">=</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="n">fold_threshold</span><span class="p">]]</span><span class="o">+</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="n">fold_threshold</span><span class="o">+</span><span class="mi">1</span><span class="p">]])</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="215 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="216 "></span>            <span class="n">foldchange</span><span class="o">=</span><span class="n">fc_threshold</span>
<span class="linenos" data-linenos="217 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'... Fold change threshold: </span><span class="si">%s</span><span class="s1">'</span><span class="o">%</span><span class="n">foldchange</span><span class="p">)</span>
<span class="linenos" data-linenos="218 "></span>        <span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span><span class="o">=</span><span class="n">foldchange</span><span class="p">,</span><span class="mi">0</span><span class="o">-</span><span class="n">foldchange</span>
<span class="linenos" data-linenos="219 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">fc_max</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_min</span><span class="o">=</span><span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span>
<span class="linenos" data-linenos="220 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">pval_threshold</span><span class="o">=</span><span class="n">pval_threshold</span>
<span class="linenos" data-linenos="221 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="222 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">&gt;</span><span class="n">fc_max</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pval_threshold</span><span class="p">)),</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'up'</span>
<span class="linenos" data-linenos="223 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">fc_min</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pval_threshold</span><span class="p">)),</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'down'</span>
<span class="linenos" data-linenos="224 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span><span class="o">&gt;</span><span class="n">logp_max</span><span class="p">,</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span><span class="o">=</span><span class="n">logp_max</span>
<span class="linenos" data-linenos="225 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">logp_max</span><span class="o">=</span><span class="n">logp_max</span>
<span class="linenos" data-linenos="226 "></span>
<span class="linenos" data-linenos="227 "></span>
<span class="linenos" data-linenos="228 "></span>    <span class="k">def</span> <span class="nf">plot_volcano</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span><span class="n">pval_name</span><span class="o">=</span><span class="s1">'qvalue'</span><span class="p">,</span><span class="n">fc_name</span><span class="o">=</span><span class="s1">'log2FC'</span><span class="p">,</span>
<span class="linenos" data-linenos="229 "></span>                     <span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">titlefont</span><span class="p">:</span><span class="nb">dict</span><span class="o">=</span><span class="p">{</span><span class="s1">'weight'</span><span class="p">:</span><span class="s1">'normal'</span><span class="p">,</span><span class="s1">'size'</span><span class="p">:</span><span class="mi">14</span><span class="p">,},</span>
<span class="linenos" data-linenos="230 "></span>                     <span class="n">up_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span><span class="n">down_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span><span class="n">normal_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span>
<span class="linenos" data-linenos="231 "></span>                     <span class="n">up_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span><span class="n">down_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span><span class="n">normal_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span>
<span class="linenos" data-linenos="232 "></span>                     <span class="n">legend_bbox</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">),</span><span class="n">legend_ncol</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">legend_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
<span class="linenos" data-linenos="233 "></span>                     <span class="n">plot_genes</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">plot_genes_num</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">plot_genes_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="linenos" data-linenos="234 "></span>                     <span class="n">ticks_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="235 "></span><span class="w">        </span><span class="sd">"""</span>
<span class="linenos" data-linenos="236 "></span><span class="sd">        Generate a volcano plot for the differential gene expression analysis results.</span>
<span class="linenos" data-linenos="237 "></span>
<span class="linenos" data-linenos="238 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="239 "></span><span class="sd">            figsize: The size of the generated figure, by default (4,4).</span>
<span class="linenos" data-linenos="240 "></span><span class="sd">            title: The title of the plot, by default ''.</span>
<span class="linenos" data-linenos="241 "></span><span class="sd">            titlefont: A dictionary of font properties for the plot title, by default {'weight':'normal','size':14,}.</span>
<span class="linenos" data-linenos="242 "></span><span class="sd">            up_color: The color of the up-regulated genes in the plot, by default '#e25d5d'.</span>
<span class="linenos" data-linenos="243 "></span><span class="sd">            down_color: The color of the down-regulated genes in the plot, by default '#7388c1'.</span>
<span class="linenos" data-linenos="244 "></span><span class="sd">            normal_color: The color of the non-significant genes in the plot, by default '#d7d7d7'.</span>
<span class="linenos" data-linenos="245 "></span><span class="sd">            legend_bbox: A tuple containing the coordinates of the legend's bounding box, by default (0.8, -0.2).</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">            legend_ncol: The number of columns in the legend, by default 2.</span>
<span class="linenos" data-linenos="247 "></span><span class="sd">            legend_fontsize: The font size of the legend, by default 12.</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">            plot_genes: A list of genes to be plotted on the volcano plot, by default None.</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">            plot_genes_num: The number of genes to be plotted on the volcano plot, by default 10.</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">            plot_genes_fontsize: The font size of the genes to be plotted on the volcano plot, by default 10.</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">            ticks_fontsize: The font size of the ticks, by default 12.</span>
<span class="linenos" data-linenos="252 "></span>
<span class="linenos" data-linenos="253 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="254 "></span><span class="sd">            ax: The generated volcano plot.</span>
<span class="linenos" data-linenos="255 "></span>
<span class="linenos" data-linenos="256 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="257 "></span>
<span class="linenos" data-linenos="258 "></span>        <span class="n">ax</span><span class="o">=</span><span class="n">volcano</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">,</span><span class="n">pval_name</span><span class="o">=</span><span class="n">pval_name</span><span class="p">,</span><span class="n">fc_name</span><span class="o">=</span><span class="n">fc_name</span><span class="p">,</span><span class="n">pval_max</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">logp_max</span><span class="p">,</span>
<span class="linenos" data-linenos="259 "></span>                       <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span><span class="n">titlefont</span><span class="o">=</span><span class="n">titlefont</span><span class="p">,</span>
<span class="linenos" data-linenos="260 "></span>                       <span class="n">up_color</span><span class="o">=</span><span class="n">up_color</span><span class="p">,</span><span class="n">down_color</span><span class="o">=</span><span class="n">down_color</span><span class="p">,</span><span class="n">normal_color</span><span class="o">=</span><span class="n">normal_color</span><span class="p">,</span>
<span class="linenos" data-linenos="261 "></span>                       <span class="n">up_fontcolor</span><span class="o">=</span><span class="n">up_fontcolor</span><span class="p">,</span><span class="n">down_fontcolor</span><span class="o">=</span><span class="n">down_fontcolor</span><span class="p">,</span><span class="n">normal_fontcolor</span><span class="o">=</span><span class="n">normal_fontcolor</span><span class="p">,</span>
<span class="linenos" data-linenos="262 "></span>                       <span class="n">legend_bbox</span><span class="o">=</span><span class="n">legend_bbox</span><span class="p">,</span><span class="n">legend_ncol</span><span class="o">=</span><span class="n">legend_ncol</span><span class="p">,</span><span class="n">legend_fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span><span class="n">plot_genes</span><span class="o">=</span><span class="n">plot_genes</span><span class="p">,</span>
<span class="linenos" data-linenos="263 "></span>                       <span class="n">plot_genes_num</span><span class="o">=</span><span class="n">plot_genes_num</span><span class="p">,</span><span class="n">plot_genes_fontsize</span><span class="o">=</span><span class="n">plot_genes_fontsize</span><span class="p">,</span>
<span class="linenos" data-linenos="264 "></span>                       <span class="n">ticks_fontsize</span><span class="o">=</span><span class="n">ticks_fontsize</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<span class="linenos" data-linenos="265 "></span>                       <span class="n">pval_threshold</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pval_threshold</span><span class="p">,</span><span class="n">fc_max</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_min</span><span class="p">)</span>
<span class="linenos" data-linenos="266 "></span>        <span class="k">return</span> <span class="n">ax</span>
<span class="linenos" data-linenos="267 "></span><span class="w">        </span><span class="sd">'''</span>
<span class="linenos" data-linenos="268 "></span><span class="sd">        fig, ax = plt.subplots(figsize=figsize)</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">        result=self.result.copy()</span>
<span class="linenos" data-linenos="270 "></span><span class="sd">        #首先绘制正常基因</span>
<span class="linenos" data-linenos="271 "></span><span class="sd">        ax.scatter(x=result[result['sig']=='normal']['log2FC'],</span>
<span class="linenos" data-linenos="272 "></span><span class="sd">                y=result[result['sig']=='normal']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="273 "></span><span class="sd">                color=normal_color,#颜色</span>
<span class="linenos" data-linenos="274 "></span><span class="sd">                alpha=.5,#透明度</span>
<span class="linenos" data-linenos="275 "></span><span class="sd">                )</span>
<span class="linenos" data-linenos="276 "></span><span class="sd">        #接着绘制上调基因</span>
<span class="linenos" data-linenos="277 "></span><span class="sd">        ax.scatter(x=result[result['sig']=='up']['log2FC'],</span>
<span class="linenos" data-linenos="278 "></span><span class="sd">                y=result[result['sig']=='up']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="279 "></span><span class="sd">                color=up_color,#选择色卡第15个颜色</span>
<span class="linenos" data-linenos="280 "></span><span class="sd">                alpha=.5,#透明度</span>
<span class="linenos" data-linenos="281 "></span><span class="sd">                )</span>
<span class="linenos" data-linenos="282 "></span><span class="sd">        #绘制下调基因</span>
<span class="linenos" data-linenos="283 "></span><span class="sd">        ax.scatter(x=result[result['sig']=='down']['log2FC'],</span>
<span class="linenos" data-linenos="284 "></span><span class="sd">                y=result[result['sig']=='down']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="285 "></span><span class="sd">                color=down_color,#颜色</span>
<span class="linenos" data-linenos="286 "></span><span class="sd">                alpha=.5,#透明度</span>
<span class="linenos" data-linenos="287 "></span><span class="sd">                )</span>
<span class="linenos" data-linenos="288 "></span>
<span class="linenos" data-linenos="289 "></span><span class="sd">        ax.plot([result['log2FC'].min(),result['log2FC'].max()],#辅助线的x值起点与终点</span>
<span class="linenos" data-linenos="290 "></span><span class="sd">                [-np.log10(self.pval_threshold),-np.log10(self.pval_threshold)],#辅助线的y值起点与终点</span>
<span class="linenos" data-linenos="291 "></span><span class="sd">                linewidth=2,#辅助线的宽度</span>
<span class="linenos" data-linenos="292 "></span><span class="sd">                linestyle="--",#辅助线类型：虚线</span>
<span class="linenos" data-linenos="293 "></span><span class="sd">                color='black'#辅助线的颜色</span>
<span class="linenos" data-linenos="294 "></span><span class="sd">        )</span>
<span class="linenos" data-linenos="295 "></span><span class="sd">        ax.plot([self.fc_max,self.fc_max],</span>
<span class="linenos" data-linenos="296 "></span><span class="sd">                [result['-log(qvalue)'].min(),result['-log(qvalue)'].max()],</span>
<span class="linenos" data-linenos="297 "></span><span class="sd">                linewidth=2, </span>
<span class="linenos" data-linenos="298 "></span><span class="sd">                linestyle="--",</span>
<span class="linenos" data-linenos="299 "></span><span class="sd">                color='black')</span>
<span class="linenos" data-linenos="300 "></span><span class="sd">        ax.plot([self.fc_min,self.fc_min],</span>
<span class="linenos" data-linenos="301 "></span><span class="sd">                [result['-log(qvalue)'].min(),result['-log(qvalue)'].max()],</span>
<span class="linenos" data-linenos="302 "></span><span class="sd">                linewidth=2, </span>
<span class="linenos" data-linenos="303 "></span><span class="sd">                linestyle="--",</span>
<span class="linenos" data-linenos="304 "></span><span class="sd">                color='black')</span>
<span class="linenos" data-linenos="305 "></span><span class="sd">        #设置横标签与纵标签</span>
<span class="linenos" data-linenos="306 "></span><span class="sd">        ax.set_ylabel(r'$-log_{10}(qvalue)$',titlefont)                                    </span>
<span class="linenos" data-linenos="307 "></span><span class="sd">        ax.set_xlabel(r'$log_{2}FC$',titlefont)</span>
<span class="linenos" data-linenos="308 "></span><span class="sd">        #设置标题</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">        ax.set_title(title,titlefont)</span>
<span class="linenos" data-linenos="310 "></span>
<span class="linenos" data-linenos="311 "></span><span class="sd">        #绘制图注</span>
<span class="linenos" data-linenos="312 "></span><span class="sd">        #legend标签列表，上面的color即是颜色列表</span>
<span class="linenos" data-linenos="313 "></span><span class="sd">        labels = ['up:{0}'.format(len(result[result['sig']=='up'])),</span>
<span class="linenos" data-linenos="314 "></span><span class="sd">                'down:{0}'.format(len(result[result['sig']=='down']))]  </span>
<span class="linenos" data-linenos="315 "></span><span class="sd">        #用label和color列表生成mpatches.Patch对象，它将作为句柄来生成legend</span>
<span class="linenos" data-linenos="316 "></span><span class="sd">        color = [up_color,down_color]</span>
<span class="linenos" data-linenos="317 "></span><span class="sd">        patches = [mpatches.Patch(color=color[i], label="{:s}".format(labels[i]) ) for i in range(len(color))] </span>
<span class="linenos" data-linenos="318 "></span>
<span class="linenos" data-linenos="319 "></span><span class="sd">        ax.legend(handles=patches,</span>
<span class="linenos" data-linenos="320 "></span><span class="sd">            bbox_to_anchor=legend_bbox, </span>
<span class="linenos" data-linenos="321 "></span><span class="sd">            ncol=legend_ncol,</span>
<span class="linenos" data-linenos="322 "></span><span class="sd">            fontsize=legend_fontsize)</span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span><span class="sd">        ax.spines['top'].set_visible(False)</span>
<span class="linenos" data-linenos="325 "></span><span class="sd">        ax.spines['right'].set_visible(False)</span>
<span class="linenos" data-linenos="326 "></span><span class="sd">        ax.spines['bottom'].set_visible(True)</span>
<span class="linenos" data-linenos="327 "></span><span class="sd">        ax.spines['left'].set_visible(True)</span>
<span class="linenos" data-linenos="328 "></span>
<span class="linenos" data-linenos="329 "></span><span class="sd">        from adjustText import adjust_text</span>
<span class="linenos" data-linenos="330 "></span><span class="sd">        import adjustText</span>
<span class="linenos" data-linenos="331 "></span>
<span class="linenos" data-linenos="332 "></span><span class="sd">        if plot_genes is not None:</span>
<span class="linenos" data-linenos="333 "></span><span class="sd">            hub_gene=plot_genes</span>
<span class="linenos" data-linenos="334 "></span><span class="sd">        else:</span>
<span class="linenos" data-linenos="335 "></span><span class="sd">            up_result=result.loc[result['sig']=='up']</span>
<span class="linenos" data-linenos="336 "></span><span class="sd">            down_result=result.loc[result['sig']=='down']</span>
<span class="linenos" data-linenos="337 "></span><span class="sd">            hub_gene=up_result.sort_values('qvalue').index[:plot_genes_num//2].tolist()+down_result.sort_values('qvalue').index[:plot_genes_num//2].tolist()</span>
<span class="linenos" data-linenos="338 "></span>
<span class="linenos" data-linenos="339 "></span><span class="sd">        color_dict={</span>
<span class="linenos" data-linenos="340 "></span><span class="sd">        'up':up_fontcolor,</span>
<span class="linenos" data-linenos="341 "></span><span class="sd">            'down':down_fontcolor,</span>
<span class="linenos" data-linenos="342 "></span><span class="sd">            'normal':normal_fontcolor</span>
<span class="linenos" data-linenos="343 "></span><span class="sd">        }</span>
<span class="linenos" data-linenos="344 "></span>
<span class="linenos" data-linenos="345 "></span><span class="sd">        texts=[ax.text(result.loc[i,'log2FC'], </span>
<span class="linenos" data-linenos="346 "></span><span class="sd">               result.loc[i,'-log(qvalue)'],</span>
<span class="linenos" data-linenos="347 "></span><span class="sd">               i,</span>
<span class="linenos" data-linenos="348 "></span><span class="sd">               fontdict={'size':plot_genes_fontsize,'weight':'bold','color':color_dict[result.loc[i,'sig']]}</span>
<span class="linenos" data-linenos="349 "></span><span class="sd">               ) for i in hub_gene]</span>
<span class="linenos" data-linenos="350 "></span>
<span class="linenos" data-linenos="351 "></span><span class="sd">        if adjustText.__version__&lt;='0.8':</span>
<span class="linenos" data-linenos="352 "></span><span class="sd">            adjust_text(texts,only_move={'text': 'xy'},arrowprops=dict(arrowstyle='-&gt;', color='red'),)</span>
<span class="linenos" data-linenos="353 "></span><span class="sd">        else:</span>
<span class="linenos" data-linenos="354 "></span><span class="sd">            adjust_text(texts,only_move={"text": "xy", "static": "xy", "explode": "xy", "pull": "xy"},</span>
<span class="linenos" data-linenos="355 "></span><span class="sd">                        arrowprops=dict(arrowstyle='-&gt;', color='red'))</span>
<span class="linenos" data-linenos="356 "></span>
<span class="linenos" data-linenos="357 "></span><span class="sd">        ax.set_xticks([round(i,2) for i in ax.get_xticks()[1:-1]],#获取x坐标轴内容</span>
<span class="linenos" data-linenos="358 "></span><span class="sd">              [round(i,2) for i in ax.get_xticks()[1:-1]],#更新x坐标轴内容</span>
<span class="linenos" data-linenos="359 "></span><span class="sd">              fontsize=ticks_fontsize,</span>
<span class="linenos" data-linenos="360 "></span><span class="sd">              fontweight='normal'</span>
<span class="linenos" data-linenos="361 "></span><span class="sd">              )</span>
<span class="linenos" data-linenos="362 "></span><span class="sd">        return fig,ax</span>
<span class="linenos" data-linenos="363 "></span><span class="sd">        '''</span>
<span class="linenos" data-linenos="364 "></span>
<span class="linenos" data-linenos="365 "></span>    <span class="k">def</span> <span class="nf">plot_boxplot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">genes</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">treatment_groups</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">control_groups</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
<span class="linenos" data-linenos="366 "></span>                     <span class="n">log</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="367 "></span>                     <span class="n">treatment_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Treatment'</span><span class="p">,</span><span class="n">control_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Control'</span><span class="p">,</span>
<span class="linenos" data-linenos="368 "></span>                     <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span><span class="n">palette</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[</span><span class="s2">"#a64d79"</span><span class="p">,</span><span class="s2">"#674ea7"</span><span class="p">],</span>
<span class="linenos" data-linenos="369 "></span>                     <span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Gene Expression'</span><span class="p">,</span><span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">legend_bbox</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.55</span><span class="p">),</span><span class="n">legend_ncol</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="linenos" data-linenos="370 "></span>                     <span class="o">**</span><span class="n">kwarg</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="371 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="372 "></span><span class="sd">        Plot the boxplot of genes from dds data</span>
<span class="linenos" data-linenos="373 "></span>
<span class="linenos" data-linenos="374 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="375 "></span><span class="sd">            genes: The genes to plot.</span>
<span class="linenos" data-linenos="376 "></span><span class="sd">            treatment_groups: The treatment groups.</span>
<span class="linenos" data-linenos="377 "></span><span class="sd">            control_groups: The control groups.</span>
<span class="linenos" data-linenos="378 "></span><span class="sd">            figsize: The figure size.</span>
<span class="linenos" data-linenos="379 "></span><span class="sd">            palette: The color palette.</span>
<span class="linenos" data-linenos="380 "></span><span class="sd">            title: The title of the plot.</span>
<span class="linenos" data-linenos="381 "></span><span class="sd">            fontsize: The fontsize of the plot.</span>
<span class="linenos" data-linenos="382 "></span><span class="sd">            legend_bbox: The bbox of the legend.</span>
<span class="linenos" data-linenos="383 "></span><span class="sd">            legend_ncol: The number of columns of the legend.</span>
<span class="linenos" data-linenos="384 "></span><span class="sd">            **kwarg: Other arguments for plot_boxplot function.</span>
<span class="linenos" data-linenos="385 "></span>
<span class="linenos" data-linenos="386 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="387 "></span><span class="sd">            fig: The figure of the plot.</span>
<span class="linenos" data-linenos="388 "></span><span class="sd">            ax: The axis of the plot.</span>
<span class="linenos" data-linenos="389 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="390 "></span>        <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">,</span><span class="s1">'Gene'</span><span class="p">,</span><span class="s1">'Type'</span><span class="p">])</span>
<span class="linenos" data-linenos="391 "></span>        <span class="k">if</span> <span class="n">log</span><span class="p">:</span>
<span class="linenos" data-linenos="392 "></span>            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
<span class="linenos" data-linenos="393 "></span>                <span class="n">plot_data1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="394 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">treatment_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="395 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="396 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">treatment_name</span>
<span class="linenos" data-linenos="397 "></span>
<span class="linenos" data-linenos="398 "></span>                <span class="n">plot_data2</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="399 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">control_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="400 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="401 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">control_name</span>
<span class="linenos" data-linenos="402 "></span>
<span class="linenos" data-linenos="403 "></span>                <span class="n">plot_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_data1</span><span class="p">,</span><span class="n">plot_data2</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="404 "></span>                <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">p_data</span><span class="p">,</span><span class="n">plot_data</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="405 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="406 "></span>            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
<span class="linenos" data-linenos="407 "></span>                <span class="n">plot_data1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="408 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">treatment_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="409 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="410 "></span>                <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">treatment_name</span>
<span class="linenos" data-linenos="411 "></span>
<span class="linenos" data-linenos="412 "></span>                <span class="n">plot_data2</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="413 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">control_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="414 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="415 "></span>                <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">control_name</span>
<span class="linenos" data-linenos="416 "></span>
<span class="linenos" data-linenos="417 "></span>                <span class="n">plot_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_data1</span><span class="p">,</span><span class="n">plot_data2</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="418 "></span>                <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">p_data</span><span class="p">,</span><span class="n">plot_data</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="419 "></span>
<span class="linenos" data-linenos="420 "></span>        <span class="n">fig</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">plot_boxplot</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="s1">'Type'</span><span class="p">,</span><span class="n">x_value</span><span class="o">=</span><span class="s1">'Gene'</span><span class="p">,</span><span class="n">y_value</span><span class="o">=</span><span class="s1">'Value'</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
<span class="linenos" data-linenos="421 "></span>                          <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<span class="linenos" data-linenos="422 "></span>                          <span class="n">legend_bbox</span><span class="o">=</span><span class="n">legend_bbox</span><span class="p">,</span><span class="n">legend_ncol</span><span class="o">=</span><span class="n">legend_ncol</span><span class="p">,</span> <span class="o">**</span><span class="n">kwarg</span><span class="p">)</span>
<span class="linenos" data-linenos="423 "></span>        <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax</span>
<span class="linenos" data-linenos="424 "></span>
<span class="linenos" data-linenos="425 "></span>    <span class="k">def</span> <span class="nf">ranking2gsea</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">rank_max</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">rank_min</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">274</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="426 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="427 "></span><span class="sd">        Ranking the result of dds data for gsea analysis</span>
<span class="linenos" data-linenos="428 "></span>
<span class="linenos" data-linenos="429 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="430 "></span><span class="sd">            rank_max: The max rank of the result.</span>
<span class="linenos" data-linenos="431 "></span><span class="sd">            rank_min: The min rank of the result.</span>
<span class="linenos" data-linenos="432 "></span>
<span class="linenos" data-linenos="433 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="434 "></span><span class="sd">            rnk: The ranking result.</span>
<span class="linenos" data-linenos="435 "></span>
<span class="linenos" data-linenos="436 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="437 "></span>
<span class="linenos" data-linenos="438 "></span>
<span class="linenos" data-linenos="439 "></span>        <span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="440 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'fcsign'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">])</span>
<span class="linenos" data-linenos="441 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="442 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'metric'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">/</span><span class="n">result</span><span class="p">[</span><span class="s1">'fcsign'</span><span class="p">]</span>
<span class="linenos" data-linenos="443 "></span>        <span class="n">rnk</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="444 "></span>        <span class="n">rnk</span><span class="p">[</span><span class="s1">'gene_name'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">index</span>
<span class="linenos" data-linenos="445 "></span>        <span class="n">rnk</span><span class="p">[</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'metric'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="446 "></span>        <span class="n">rnk</span><span class="o">=</span><span class="n">rnk</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">'rnk'</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="447 "></span>        <span class="n">k</span><span class="o">=</span><span class="mi">1</span>
<span class="linenos" data-linenos="448 "></span>        <span class="n">total</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="449 "></span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="450 "></span>            <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="451 "></span>                <span class="n">total</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="452 "></span>        <span class="c1">#200跟274根据你的数据进行更改，保证inf比你数据最大的大，-inf比数据最小的小就好</span>
<span class="linenos" data-linenos="453 "></span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="454 "></span>            <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="455 "></span>                <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=</span><span class="n">rank_max</span><span class="o">+</span><span class="p">(</span><span class="n">total</span><span class="o">-</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="456 "></span>                <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="457 "></span>            <span class="k">elif</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="458 "></span>                <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=-</span><span class="p">(</span><span class="n">rank_min</span><span class="o">+</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="459 "></span>                <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="460 "></span>        <span class="k">return</span> <span class="n">rnk</span>
<span class="linenos" data-linenos="461 "></span>
<span class="linenos" data-linenos="462 "></span>    <span class="k">def</span> <span class="nf">deg_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">group1</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">group2</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
<span class="linenos" data-linenos="463 "></span>                     <span class="n">method</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'DEseq2'</span><span class="p">,</span><span class="n">alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="linenos" data-linenos="464 "></span>                     <span class="n">multipletests_method</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'fdr_bh'</span><span class="p">,</span><span class="n">n_cpus</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="linenos" data-linenos="465 "></span>                     <span class="n">cooks_filter</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">independent_filter</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="466 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="467 "></span><span class="sd">        Differential expression analysis.</span>
<span class="linenos" data-linenos="468 "></span>
<span class="linenos" data-linenos="469 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="470 "></span><span class="sd">            group1: The first group to be compared.</span>
<span class="linenos" data-linenos="471 "></span><span class="sd">            group2: The second group to be compared.</span>
<span class="linenos" data-linenos="472 "></span><span class="sd">            method: The method to be used for differential expression analysis.</span>
<span class="linenos" data-linenos="473 "></span><span class="sd">                - `DEseq2`: DEseq2</span>
<span class="linenos" data-linenos="474 "></span><span class="sd">                - `ttest`: ttest</span>
<span class="linenos" data-linenos="475 "></span><span class="sd">                - `wilcox`: wilconx test</span>
<span class="linenos" data-linenos="476 "></span><span class="sd">            alpha: The threshold of p-value.</span>
<span class="linenos" data-linenos="477 "></span><span class="sd">            multipletests_method:</span>
<span class="linenos" data-linenos="478 "></span><span class="sd">                - `bonferroni` : one-step correction</span>
<span class="linenos" data-linenos="479 "></span><span class="sd">                - `sidak` : one-step correction</span>
<span class="linenos" data-linenos="480 "></span><span class="sd">                - `holm-sidak` : step down method using Sidak adjustments</span>
<span class="linenos" data-linenos="481 "></span><span class="sd">                - `holm` : step-down method using Bonferroni adjustments</span>
<span class="linenos" data-linenos="482 "></span><span class="sd">                - `simes-hochberg` : step-up method  (independent)</span>
<span class="linenos" data-linenos="483 "></span><span class="sd">                - `hommel` : closed method based on Simes tests (non-negative)</span>
<span class="linenos" data-linenos="484 "></span><span class="sd">                - `fdr_bh` : Benjamini/Hochberg  (non-negative)</span>
<span class="linenos" data-linenos="485 "></span><span class="sd">                - `fdr_by` : Benjamini/Yekutieli (negative)</span>
<span class="linenos" data-linenos="486 "></span><span class="sd">                - `fdr_tsbh` : two stage fdr correction (non-negative)</span>
<span class="linenos" data-linenos="487 "></span><span class="sd">                - `fdr_tsbky` : two stage fdr correction (non-negative)</span>
<span class="linenos" data-linenos="488 "></span>
<span class="linenos" data-linenos="489 "></span><span class="sd">        Returns</span>
<span class="linenos" data-linenos="490 "></span><span class="sd">            result: The result of differential expression analysis.</span>
<span class="linenos" data-linenos="491 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="492 "></span>        <span class="kn">from</span> <span class="nn">pydeseq2.dds</span> <span class="kn">import</span> <span class="n">DeseqDataSet</span>
<span class="linenos" data-linenos="493 "></span>        <span class="kn">from</span> <span class="nn">pydeseq2.ds</span> <span class="kn">import</span> <span class="n">DeseqStats</span>
<span class="linenos" data-linenos="494 "></span>        <span class="k">if</span> <span class="n">method</span><span class="o">==</span><span class="s1">'ttest'</span><span class="p">:</span>
<span class="linenos" data-linenos="495 "></span>            <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ttest_ind</span>
<span class="linenos" data-linenos="496 "></span>            <span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>
<span class="linenos" data-linenos="497 "></span>            <span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="498 "></span>
<span class="linenos" data-linenos="499 "></span>            <span class="n">g1_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="500 "></span>            <span class="n">g2_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="501 "></span>            <span class="n">g</span><span class="o">=</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="n">g1_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="502 "></span>            <span class="n">g</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">g</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="linenos" data-linenos="503 "></span>            <span class="n">fold</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="n">g</span><span class="p">)</span>
<span class="linenos" data-linenos="504 "></span>            <span class="c1">#log2fold=np.log2(fold)</span>
<span class="linenos" data-linenos="505 "></span>            <span class="n">ttest</span> <span class="o">=</span> <span class="n">ttest_ind</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="506 "></span>            <span class="n">pvalue</span><span class="o">=</span><span class="n">ttest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="linenos" data-linenos="507 "></span>            <span class="n">qvalue</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pvalue</span><span class="p">),</span><span class="mi">0</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
<span class="linenos" data-linenos="508 "></span>                               <span class="n">method</span><span class="o">=</span><span class="n">multipletests_method</span><span class="p">,</span> <span class="n">is_sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">returnsorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="509 "></span>            <span class="c1">#qvalue=fdrcorrection(np.nan_to_num(np.array(pvalue),0), alpha=0.05, method='indep', is_sorted=False)</span>
<span class="linenos" data-linenos="510 "></span>            <span class="n">genearray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
<span class="linenos" data-linenos="511 "></span>            <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'pvalue'</span><span class="p">:</span><span class="n">genearray</span><span class="p">,</span><span class="s1">'qvalue'</span><span class="p">:</span><span class="n">qvalue</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s1">'FoldChange'</span><span class="p">:</span><span class="n">fold</span><span class="p">})</span>
<span class="linenos" data-linenos="512 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'MaxBaseMean'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">g1_mean</span><span class="p">,</span><span class="n">g2_mean</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="513 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="514 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">((</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="515 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span>
<span class="linenos" data-linenos="516 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">]))</span>
<span class="linenos" data-linenos="517 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'size'</span><span class="p">]</span>  <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span><span class="o">/</span><span class="mi">10</span>
<span class="linenos" data-linenos="518 "></span>            <span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
<span class="linenos" data-linenos="519 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="520 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="521 "></span>            <span class="c1">#max mean of between each value in group1 and group2</span>
<span class="linenos" data-linenos="522 "></span>            <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="523 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="524 "></span>            <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="525 "></span>
<span class="linenos" data-linenos="526 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="527 "></span>            <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="528 "></span>        <span class="k">elif</span> <span class="n">method</span><span class="o">==</span><span class="s1">'wilcox'</span><span class="p">:</span>
<span class="linenos" data-linenos="529 "></span>            <span class="c1">#raise ValueError('The method is not supported.')</span>
<span class="linenos" data-linenos="530 "></span>            <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ranksums</span>
<span class="linenos" data-linenos="531 "></span>            <span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>
<span class="linenos" data-linenos="532 "></span>            <span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="533 "></span>
<span class="linenos" data-linenos="534 "></span>            <span class="n">g1_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="535 "></span>            <span class="n">g2_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="536 "></span>            <span class="n">fold</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="mf">0.00001</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="mf">0.00001</span><span class="p">)</span>
<span class="linenos" data-linenos="537 "></span>            <span class="c1">#log2fold=np.log2(fold)</span>
<span class="linenos" data-linenos="538 "></span>            <span class="n">wilcox</span> <span class="o">=</span> <span class="n">ranksums</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="539 "></span>            <span class="n">pvalue</span><span class="o">=</span><span class="n">wilcox</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="linenos" data-linenos="540 "></span>
<span class="linenos" data-linenos="541 "></span>            <span class="c1">#qvalue=fdrcorrection(np.nan_to_num(np.array(pvalue),0), alpha=0.05, method='indep', is_sorted=False)</span>
<span class="linenos" data-linenos="542 "></span>            <span class="n">qvalue</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pvalue</span><span class="p">),</span><span class="mi">0</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
<span class="linenos" data-linenos="543 "></span>                               <span class="n">method</span><span class="o">=</span><span class="n">multipletests_method</span><span class="p">,</span> <span class="n">is_sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">returnsorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="544 "></span>            <span class="n">genearray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
<span class="linenos" data-linenos="545 "></span>            <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'pvalue'</span><span class="p">:</span><span class="n">genearray</span><span class="p">,</span><span class="s1">'qvalue'</span><span class="p">:</span><span class="n">qvalue</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s1">'FoldChange'</span><span class="p">:</span><span class="n">fold</span><span class="p">})</span>
<span class="linenos" data-linenos="546 "></span>            <span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
<span class="linenos" data-linenos="547 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="548 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="549 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="550 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">((</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="551 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span>
<span class="linenos" data-linenos="552 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">]))</span>
<span class="linenos" data-linenos="553 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'size'</span><span class="p">]</span>  <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span><span class="o">/</span><span class="mi">10</span>
<span class="linenos" data-linenos="554 "></span>            <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="555 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="556 "></span>            <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="557 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="558 "></span>            <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="559 "></span>        <span class="k">elif</span> <span class="n">method</span><span class="o">==</span><span class="s1">'DEseq2'</span><span class="p">:</span>
<span class="linenos" data-linenos="560 "></span>            <span class="kn">import</span> <span class="nn">pydeseq2</span>
<span class="linenos" data-linenos="561 "></span>            <span class="n">counts_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="o">+</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="562 "></span>            <span class="n">clinical_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">group1</span><span class="o">+</span><span class="n">group2</span><span class="p">)</span>
<span class="linenos" data-linenos="563 "></span>            <span class="n">clinical_df</span><span class="p">[</span><span class="s1">'condition'</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="s1">'Treatment'</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">group1</span><span class="p">)</span><span class="o">+</span><span class="p">[</span><span class="s1">'Control'</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">group2</span><span class="p">)</span>
<span class="linenos" data-linenos="564 "></span>
<span class="linenos" data-linenos="565 "></span>            <span class="c1">#Determining the version of pydeseq2 smaller than 0.4</span>
<span class="linenos" data-linenos="566 "></span>            <span class="k">if</span> <span class="n">pydeseq2</span><span class="o">.</span><span class="n">__version__</span><span class="o">&lt;=</span><span class="s1">'0.3.5'</span><span class="p">:</span>
<span class="linenos" data-linenos="567 "></span>                <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="568 "></span>                    <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="569 "></span>                    <span class="n">clinical</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="570 "></span>                    <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="571 "></span>                    <span class="n">ref_level</span><span class="o">=</span><span class="p">[</span><span class="s2">"condition"</span><span class="p">,</span> <span class="s2">"Control"</span><span class="p">],</span>
<span class="linenos" data-linenos="572 "></span>                    <span class="c1"># column ("B" vs "A")</span>
<span class="linenos" data-linenos="573 "></span>                    <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="574 "></span>                    <span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">,</span>
<span class="linenos" data-linenos="575 "></span>                <span class="p">)</span>
<span class="linenos" data-linenos="576 "></span>            <span class="k">elif</span> <span class="n">pydeseq2</span><span class="o">.</span><span class="n">__version__</span><span class="o">&lt;=</span><span class="s1">'0.4.1'</span><span class="p">:</span>
<span class="linenos" data-linenos="577 "></span>                <span class="k">if</span> <span class="n">ad</span><span class="o">.</span><span class="n">__version__</span><span class="o">&gt;</span><span class="s1">'0.10.8'</span><span class="p">:</span>
<span class="linenos" data-linenos="578 "></span>                    <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span>
<span class="linenos" data-linenos="579 "></span>                            <span class="s1">'Please install the 0.10.8 version of anndata: `pip install anndata==0.10.8`.'</span>
<span class="linenos" data-linenos="580 "></span>                        <span class="p">)</span>
<span class="linenos" data-linenos="581 "></span>                <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="582 "></span>                    <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="583 "></span>                    <span class="n">metadata</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="584 "></span>                    <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="585 "></span>                    <span class="c1">#ref_level=["condition", "Control"],</span>
<span class="linenos" data-linenos="586 "></span>                    <span class="c1"># column ("B" vs "A")</span>
<span class="linenos" data-linenos="587 "></span>                    <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="588 "></span>                    <span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">,</span>
<span class="linenos" data-linenos="589 "></span>                <span class="p">)</span>
<span class="linenos" data-linenos="590 "></span>            <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="591 "></span>                <span class="kn">from</span> <span class="nn">pydeseq2.default_inference</span> <span class="kn">import</span> <span class="n">DefaultInference</span>
<span class="linenos" data-linenos="592 "></span>                <span class="n">inference</span> <span class="o">=</span> <span class="n">DefaultInference</span><span class="p">(</span><span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">)</span>
<span class="linenos" data-linenos="593 "></span>                <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="594 "></span>                    <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="595 "></span>                    <span class="n">metadata</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="596 "></span>                    <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="597 "></span>                    <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="598 "></span>                    <span class="n">inference</span><span class="o">=</span><span class="n">inference</span><span class="p">,</span>
<span class="linenos" data-linenos="599 "></span>                <span class="p">)</span>
<span class="linenos" data-linenos="600 "></span>
<span class="linenos" data-linenos="601 "></span>
<span class="linenos" data-linenos="602 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_size_factors</span><span class="p">()</span>
<span class="linenos" data-linenos="603 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_genewise_dispersions</span><span class="p">()</span>
<span class="linenos" data-linenos="604 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_dispersion_trend</span><span class="p">()</span>
<span class="linenos" data-linenos="605 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_dispersion_prior</span><span class="p">()</span>
<span class="linenos" data-linenos="606 "></span>            <span class="nb">print</span><span class="p">(</span>
<span class="linenos" data-linenos="607 "></span>                <span class="sa">f</span><span class="s2">"logres_prior=</span><span class="si">{</span><span class="n">dds</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'_squared_logres'</span><span class="p">]</span><span class="si">}</span><span class="s2">, sigma_prior=</span><span class="si">{</span><span class="n">dds</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'prior_disp_var'</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span>
<span class="linenos" data-linenos="608 "></span>            <span class="p">)</span>
<span class="linenos" data-linenos="609 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_MAP_dispersions</span><span class="p">()</span>
<span class="linenos" data-linenos="610 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">fit_LFC</span><span class="p">()</span>
<span class="linenos" data-linenos="611 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">calculate_cooks</span><span class="p">()</span>
<span class="linenos" data-linenos="612 "></span>            <span class="k">if</span> <span class="n">dds</span><span class="o">.</span><span class="n">refit_cooks</span><span class="p">:</span>
<span class="linenos" data-linenos="613 "></span>                <span class="c1"># Replace outlier counts</span>
<span class="linenos" data-linenos="614 "></span>                <span class="n">dds</span><span class="o">.</span><span class="n">refit</span><span class="p">()</span>
<span class="linenos" data-linenos="615 "></span>            <span class="n">stat_res</span> <span class="o">=</span> <span class="n">DeseqStats</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span> <span class="n">cooks_filter</span><span class="o">=</span><span class="n">cooks_filter</span><span class="p">,</span> <span class="n">independent_filter</span><span class="o">=</span><span class="n">independent_filter</span><span class="p">)</span>
<span class="linenos" data-linenos="616 "></span>            <span class="n">stat_res</span><span class="o">.</span><span class="n">run_wald_test</span><span class="p">()</span>
<span class="linenos" data-linenos="617 "></span>            <span class="k">if</span> <span class="n">stat_res</span><span class="o">.</span><span class="n">cooks_filter</span><span class="p">:</span>
<span class="linenos" data-linenos="618 "></span>                <span class="n">stat_res</span><span class="o">.</span><span class="n">_cooks_filtering</span><span class="p">()</span>
<span class="linenos" data-linenos="619 "></span>            <span class="k">if</span> <span class="n">stat_res</span><span class="o">.</span><span class="n">independent_filter</span><span class="p">:</span>
<span class="linenos" data-linenos="620 "></span>                <span class="n">stat_res</span><span class="o">.</span><span class="n">_independent_filtering</span><span class="p">()</span>
<span class="linenos" data-linenos="621 "></span>            <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="622 "></span>                <span class="n">stat_res</span><span class="o">.</span><span class="n">_p_value_adjustment</span><span class="p">()</span>
<span class="linenos" data-linenos="623 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">stat_res</span><span class="o">=</span><span class="n">stat_res</span>
<span class="linenos" data-linenos="624 "></span>            <span class="n">stat_res</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="linenos" data-linenos="625 "></span>            <span class="n">result</span><span class="o">=</span><span class="n">stat_res</span><span class="o">.</span><span class="n">results_df</span>
<span class="linenos" data-linenos="626 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'padj'</span><span class="p">]</span>
<span class="linenos" data-linenos="627 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="628 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'padj'</span><span class="p">])</span>
<span class="linenos" data-linenos="629 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'baseMean'</span><span class="p">]</span>
<span class="linenos" data-linenos="630 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'baseMean'</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="631 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s1">'log2FoldChange'</span><span class="p">]</span>
<span class="linenos" data-linenos="632 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">])</span>
<span class="linenos" data-linenos="633 "></span>            <span class="c1">#result['size']  =np.abs(result['FoldChange'])/10</span>
<span class="linenos" data-linenos="634 "></span>            <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="635 "></span>            <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="636 "></span>            <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="637 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="638 "></span>            <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="639 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="640 "></span>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">'The method is not supported.'</span><span class="p">)</span>
</code></pre></div>
</details>
<div class="doc doc-children">
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">raw_data</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.__init__" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Initialize the pyDEG class.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>raw_data</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>The raw data to be processed.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="170 "></span><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">raw_data</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="171 "></span><span class="w">    </span><span class="sd">"""Initialize the pyDEG class.</span>
<span class="linenos" data-linenos="172 "></span>
<span class="linenos" data-linenos="173 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="174 "></span><span class="sd">        raw_data: The raw data to be processed.</span>
<span class="linenos" data-linenos="175 "></span>
<span class="linenos" data-linenos="176 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="177 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">raw_data</span><span class="o">=</span><span class="n">raw_data</span>
<span class="linenos" data-linenos="178 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">raw_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.drop_duplicates_index">
<code class="highlight language-python"><span class="n">drop_duplicates_index</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.drop_duplicates_index" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Drop the duplicated index of data.</p>
<p>Returns
    data: The data after dropping the duplicated index.</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="180 "></span><span class="k">def</span> <span class="nf">drop_duplicates_index</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="181 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">    Drop the duplicated index of data.</span>
<span class="linenos" data-linenos="183 "></span>
<span class="linenos" data-linenos="184 "></span><span class="sd">    Returns</span>
<span class="linenos" data-linenos="185 "></span><span class="sd">        data: The data after dropping the duplicated index.</span>
<span class="linenos" data-linenos="186 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="187 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">data_drop_duplicates_index</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="188 "></span>    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.normalize">
<code class="highlight language-python"><span class="n">normalize</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.normalize" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Normalize the data using DESeq2 method.</p>
<p>Returns
    data: The normalized data.</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="190 "></span><span class="k">def</span> <span class="nf">normalize</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="191 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">    Normalize the data using DESeq2 method.</span>
<span class="linenos" data-linenos="193 "></span>
<span class="linenos" data-linenos="194 "></span><span class="sd">    Returns</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">        data: The normalized data.</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="197 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">size_factors</span><span class="o">=</span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="198 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">=</span><span class="n">deseq2_normalize</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="199 "></span>    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">data</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.foldchange_set">
<code class="highlight language-python"><span class="n">foldchange_set</span><span class="p">(</span><span class="n">fc_threshold</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">pval_threshold</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">logp_max</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">fold_threshold</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.foldchange_set" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Sets fold-change and p-value thresholds to classify differentially expressed genes as up-regulated, down-regulated, or not significant.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fc_threshold</code></td>
<td>
<code>int</code>
</td>
<td><p>Absolute fold-change threshold. If set to -1, the threshold is calculated based on the histogram of log2 fold-changes.</p></td>
<td>
<code>-1</code>
</td>
</tr>
<tr>
<td><code>pval_threshold</code></td>
<td>
<code>float</code>
</td>
<td><p>p-value threshold for determining significance.</p></td>
<td>
<code>0.05</code>
</td>
</tr>
<tr>
<td><code>logp_max</code></td>
<td>
<code>int</code>
</td>
<td><p>Maximum value for log-transformed p-values.</p></td>
<td>
<code>6</code>
</td>
</tr>
<tr>
<td><code>fold_threshold</code></td>
<td>
<code>int</code>
</td>
<td><p>Index of the histogram bin corresponding to the fold-change threshold (only applicable if fc_threshold=-1).</p></td>
<td>
<code>0</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="201 "></span><span class="k">def</span> <span class="nf">foldchange_set</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">fc_threshold</span><span class="p">:</span><span class="nb">int</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="n">pval_threshold</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">logp_max</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span><span class="n">fold_threshold</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="linenos" data-linenos="202 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="203 "></span><span class="sd">    Sets fold-change and p-value thresholds to classify differentially expressed genes as up-regulated, down-regulated, or not significant.</span>
<span class="linenos" data-linenos="204 "></span>
<span class="linenos" data-linenos="205 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="206 "></span><span class="sd">        fc_threshold: Absolute fold-change threshold. If set to -1, the threshold is calculated based on the histogram of log2 fold-changes.</span>
<span class="linenos" data-linenos="207 "></span><span class="sd">        pval_threshold: p-value threshold for determining significance.</span>
<span class="linenos" data-linenos="208 "></span><span class="sd">        logp_max: Maximum value for log-transformed p-values.</span>
<span class="linenos" data-linenos="209 "></span><span class="sd">        fold_threshold: Index of the histogram bin corresponding to the fold-change threshold (only applicable if fc_threshold=-1).</span>
<span class="linenos" data-linenos="210 "></span>
<span class="linenos" data-linenos="211 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="212 "></span>    <span class="k">if</span> <span class="n">fc_threshold</span><span class="o">==-</span><span class="mi">1</span><span class="p">:</span>
<span class="linenos" data-linenos="213 "></span>        <span class="n">foldp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">())</span>
<span class="linenos" data-linenos="214 "></span>        <span class="n">foldchange</span><span class="o">=</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="n">fold_threshold</span><span class="p">]]</span><span class="o">+</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">foldp</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="n">fold_threshold</span><span class="o">+</span><span class="mi">1</span><span class="p">]])</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="215 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="216 "></span>        <span class="n">foldchange</span><span class="o">=</span><span class="n">fc_threshold</span>
<span class="linenos" data-linenos="217 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'... Fold change threshold: </span><span class="si">%s</span><span class="s1">'</span><span class="o">%</span><span class="n">foldchange</span><span class="p">)</span>
<span class="linenos" data-linenos="218 "></span>    <span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span><span class="o">=</span><span class="n">foldchange</span><span class="p">,</span><span class="mi">0</span><span class="o">-</span><span class="n">foldchange</span>
<span class="linenos" data-linenos="219 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">fc_max</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_min</span><span class="o">=</span><span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span>
<span class="linenos" data-linenos="220 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">pval_threshold</span><span class="o">=</span><span class="n">pval_threshold</span>
<span class="linenos" data-linenos="221 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="222 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">&gt;</span><span class="n">fc_max</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pval_threshold</span><span class="p">)),</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'up'</span>
<span class="linenos" data-linenos="223 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[((</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">fc_min</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">pval_threshold</span><span class="p">)),</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'down'</span>
<span class="linenos" data-linenos="224 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span><span class="o">&gt;</span><span class="n">logp_max</span><span class="p">,</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span><span class="o">=</span><span class="n">logp_max</span>
<span class="linenos" data-linenos="225 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">logp_max</span><span class="o">=</span><span class="n">logp_max</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.deg_analysis">
<code class="highlight language-python"><span class="n">deg_analysis</span><span class="p">(</span><span class="n">group1</span><span class="p">,</span> <span class="n">group2</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">'DEseq2'</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">multipletests_method</span><span class="o">=</span><span class="s1">'fdr_bh'</span><span class="p">,</span> <span class="n">n_cpus</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">cooks_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">independent_filter</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.deg_analysis" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Differential expression analysis.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>group1</code></td>
<td>
<code>list</code>
</td>
<td><p>The first group to be compared.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>group2</code></td>
<td>
<code>list</code>
</td>
<td><p>The second group to be compared.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>method</code></td>
<td>
<code>str</code>
</td>
<td><p>The method to be used for differential expression analysis.
- <code>DEseq2</code>: DEseq2
- <code>ttest</code>: ttest
- <code>wilcox</code>: wilconx test</p></td>
<td>
<code>'DEseq2'</code>
</td>
</tr>
<tr>
<td><code>alpha</code></td>
<td>
<code>float</code>
</td>
<td><p>The threshold of p-value.</p></td>
<td>
<code>0.05</code>
</td>
</tr>
<tr>
<td><code>multipletests_method</code></td>
<td>
<code>str</code>
</td>
<td><ul>
<li><code>bonferroni</code> : one-step correction</li>
<li><code>sidak</code> : one-step correction</li>
<li><code>holm-sidak</code> : step down method using Sidak adjustments</li>
<li><code>holm</code> : step-down method using Bonferroni adjustments</li>
<li><code>simes-hochberg</code> : step-up method  (independent)</li>
<li><code>hommel</code> : closed method based on Simes tests (non-negative)</li>
<li><code>fdr_bh</code> : Benjamini/Hochberg  (non-negative)</li>
<li><code>fdr_by</code> : Benjamini/Yekutieli (negative)</li>
<li><code>fdr_tsbh</code> : two stage fdr correction (non-negative)</li>
<li><code>fdr_tsbky</code> : two stage fdr correction (non-negative)</li>
</ul></td>
<td>
<code>'fdr_bh'</code>
</td>
</tr>
</tbody>
</table>
<p>Returns
    result: The result of differential expression analysis.</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="462 "></span><span class="k">def</span> <span class="nf">deg_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">group1</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">group2</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
<span class="linenos" data-linenos="463 "></span>                 <span class="n">method</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'DEseq2'</span><span class="p">,</span><span class="n">alpha</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="linenos" data-linenos="464 "></span>                 <span class="n">multipletests_method</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'fdr_bh'</span><span class="p">,</span><span class="n">n_cpus</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="linenos" data-linenos="465 "></span>                 <span class="n">cooks_filter</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">independent_filter</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="466 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="467 "></span><span class="sd">    Differential expression analysis.</span>
<span class="linenos" data-linenos="468 "></span>
<span class="linenos" data-linenos="469 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="470 "></span><span class="sd">        group1: The first group to be compared.</span>
<span class="linenos" data-linenos="471 "></span><span class="sd">        group2: The second group to be compared.</span>
<span class="linenos" data-linenos="472 "></span><span class="sd">        method: The method to be used for differential expression analysis.</span>
<span class="linenos" data-linenos="473 "></span><span class="sd">            - `DEseq2`: DEseq2</span>
<span class="linenos" data-linenos="474 "></span><span class="sd">            - `ttest`: ttest</span>
<span class="linenos" data-linenos="475 "></span><span class="sd">            - `wilcox`: wilconx test</span>
<span class="linenos" data-linenos="476 "></span><span class="sd">        alpha: The threshold of p-value.</span>
<span class="linenos" data-linenos="477 "></span><span class="sd">        multipletests_method:</span>
<span class="linenos" data-linenos="478 "></span><span class="sd">            - `bonferroni` : one-step correction</span>
<span class="linenos" data-linenos="479 "></span><span class="sd">            - `sidak` : one-step correction</span>
<span class="linenos" data-linenos="480 "></span><span class="sd">            - `holm-sidak` : step down method using Sidak adjustments</span>
<span class="linenos" data-linenos="481 "></span><span class="sd">            - `holm` : step-down method using Bonferroni adjustments</span>
<span class="linenos" data-linenos="482 "></span><span class="sd">            - `simes-hochberg` : step-up method  (independent)</span>
<span class="linenos" data-linenos="483 "></span><span class="sd">            - `hommel` : closed method based on Simes tests (non-negative)</span>
<span class="linenos" data-linenos="484 "></span><span class="sd">            - `fdr_bh` : Benjamini/Hochberg  (non-negative)</span>
<span class="linenos" data-linenos="485 "></span><span class="sd">            - `fdr_by` : Benjamini/Yekutieli (negative)</span>
<span class="linenos" data-linenos="486 "></span><span class="sd">            - `fdr_tsbh` : two stage fdr correction (non-negative)</span>
<span class="linenos" data-linenos="487 "></span><span class="sd">            - `fdr_tsbky` : two stage fdr correction (non-negative)</span>
<span class="linenos" data-linenos="488 "></span>
<span class="linenos" data-linenos="489 "></span><span class="sd">    Returns</span>
<span class="linenos" data-linenos="490 "></span><span class="sd">        result: The result of differential expression analysis.</span>
<span class="linenos" data-linenos="491 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="492 "></span>    <span class="kn">from</span> <span class="nn">pydeseq2.dds</span> <span class="kn">import</span> <span class="n">DeseqDataSet</span>
<span class="linenos" data-linenos="493 "></span>    <span class="kn">from</span> <span class="nn">pydeseq2.ds</span> <span class="kn">import</span> <span class="n">DeseqStats</span>
<span class="linenos" data-linenos="494 "></span>    <span class="k">if</span> <span class="n">method</span><span class="o">==</span><span class="s1">'ttest'</span><span class="p">:</span>
<span class="linenos" data-linenos="495 "></span>        <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ttest_ind</span>
<span class="linenos" data-linenos="496 "></span>        <span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>
<span class="linenos" data-linenos="497 "></span>        <span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="498 "></span>
<span class="linenos" data-linenos="499 "></span>        <span class="n">g1_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="500 "></span>        <span class="n">g2_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="501 "></span>        <span class="n">g</span><span class="o">=</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="n">g1_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="502 "></span>        <span class="n">g</span><span class="o">=</span><span class="n">g</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">g</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="linenos" data-linenos="503 "></span>        <span class="n">fold</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="n">g</span><span class="p">)</span>
<span class="linenos" data-linenos="504 "></span>        <span class="c1">#log2fold=np.log2(fold)</span>
<span class="linenos" data-linenos="505 "></span>        <span class="n">ttest</span> <span class="o">=</span> <span class="n">ttest_ind</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="506 "></span>        <span class="n">pvalue</span><span class="o">=</span><span class="n">ttest</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="linenos" data-linenos="507 "></span>        <span class="n">qvalue</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pvalue</span><span class="p">),</span><span class="mi">0</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
<span class="linenos" data-linenos="508 "></span>                           <span class="n">method</span><span class="o">=</span><span class="n">multipletests_method</span><span class="p">,</span> <span class="n">is_sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">returnsorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="509 "></span>        <span class="c1">#qvalue=fdrcorrection(np.nan_to_num(np.array(pvalue),0), alpha=0.05, method='indep', is_sorted=False)</span>
<span class="linenos" data-linenos="510 "></span>        <span class="n">genearray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
<span class="linenos" data-linenos="511 "></span>        <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'pvalue'</span><span class="p">:</span><span class="n">genearray</span><span class="p">,</span><span class="s1">'qvalue'</span><span class="p">:</span><span class="n">qvalue</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s1">'FoldChange'</span><span class="p">:</span><span class="n">fold</span><span class="p">})</span>
<span class="linenos" data-linenos="512 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'MaxBaseMean'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">g1_mean</span><span class="p">,</span><span class="n">g2_mean</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="513 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="514 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">((</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="515 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span>
<span class="linenos" data-linenos="516 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">]))</span>
<span class="linenos" data-linenos="517 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'size'</span><span class="p">]</span>  <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span><span class="o">/</span><span class="mi">10</span>
<span class="linenos" data-linenos="518 "></span>        <span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
<span class="linenos" data-linenos="519 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="520 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="521 "></span>        <span class="c1">#max mean of between each value in group1 and group2</span>
<span class="linenos" data-linenos="522 "></span>        <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="523 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="524 "></span>        <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="525 "></span>
<span class="linenos" data-linenos="526 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="527 "></span>        <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="528 "></span>    <span class="k">elif</span> <span class="n">method</span><span class="o">==</span><span class="s1">'wilcox'</span><span class="p">:</span>
<span class="linenos" data-linenos="529 "></span>        <span class="c1">#raise ValueError('The method is not supported.')</span>
<span class="linenos" data-linenos="530 "></span>        <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ranksums</span>
<span class="linenos" data-linenos="531 "></span>        <span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>
<span class="linenos" data-linenos="532 "></span>        <span class="n">data</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span>
<span class="linenos" data-linenos="533 "></span>
<span class="linenos" data-linenos="534 "></span>        <span class="n">g1_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="535 "></span>        <span class="n">g2_mean</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="536 "></span>        <span class="n">fold</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="mf">0.00001</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">g2_mean</span><span class="o">+</span><span class="mf">0.00001</span><span class="p">)</span>
<span class="linenos" data-linenos="537 "></span>        <span class="c1">#log2fold=np.log2(fold)</span>
<span class="linenos" data-linenos="538 "></span>        <span class="n">wilcox</span> <span class="o">=</span> <span class="n">ranksums</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="539 "></span>        <span class="n">pvalue</span><span class="o">=</span><span class="n">wilcox</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="linenos" data-linenos="540 "></span>
<span class="linenos" data-linenos="541 "></span>        <span class="c1">#qvalue=fdrcorrection(np.nan_to_num(np.array(pvalue),0), alpha=0.05, method='indep', is_sorted=False)</span>
<span class="linenos" data-linenos="542 "></span>        <span class="n">qvalue</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pvalue</span><span class="p">),</span><span class="mi">0</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
<span class="linenos" data-linenos="543 "></span>                           <span class="n">method</span><span class="o">=</span><span class="n">multipletests_method</span><span class="p">,</span> <span class="n">is_sorted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">returnsorted</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="544 "></span>        <span class="n">genearray</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)</span>
<span class="linenos" data-linenos="545 "></span>        <span class="n">result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'pvalue'</span><span class="p">:</span><span class="n">genearray</span><span class="p">,</span><span class="s1">'qvalue'</span><span class="p">:</span><span class="n">qvalue</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s1">'FoldChange'</span><span class="p">:</span><span class="n">fold</span><span class="p">})</span>
<span class="linenos" data-linenos="546 "></span>        <span class="n">result</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
<span class="linenos" data-linenos="547 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="548 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="549 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="p">(</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="linenos" data-linenos="550 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">((</span><span class="n">g1_mean</span><span class="o">+</span><span class="n">g2_mean</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="551 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span>
<span class="linenos" data-linenos="552 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">]))</span>
<span class="linenos" data-linenos="553 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'size'</span><span class="p">]</span>  <span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'FoldChange'</span><span class="p">])</span><span class="o">/</span><span class="mi">10</span>
<span class="linenos" data-linenos="554 "></span>        <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="555 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="556 "></span>        <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="557 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="558 "></span>        <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="559 "></span>    <span class="k">elif</span> <span class="n">method</span><span class="o">==</span><span class="s1">'DEseq2'</span><span class="p">:</span>
<span class="linenos" data-linenos="560 "></span>        <span class="kn">import</span> <span class="nn">pydeseq2</span>
<span class="linenos" data-linenos="561 "></span>        <span class="n">counts_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">group1</span><span class="o">+</span><span class="n">group2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="562 "></span>        <span class="n">clinical_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">group1</span><span class="o">+</span><span class="n">group2</span><span class="p">)</span>
<span class="linenos" data-linenos="563 "></span>        <span class="n">clinical_df</span><span class="p">[</span><span class="s1">'condition'</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="s1">'Treatment'</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">group1</span><span class="p">)</span><span class="o">+</span><span class="p">[</span><span class="s1">'Control'</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">group2</span><span class="p">)</span>
<span class="linenos" data-linenos="564 "></span>
<span class="linenos" data-linenos="565 "></span>        <span class="c1">#Determining the version of pydeseq2 smaller than 0.4</span>
<span class="linenos" data-linenos="566 "></span>        <span class="k">if</span> <span class="n">pydeseq2</span><span class="o">.</span><span class="n">__version__</span><span class="o">&lt;=</span><span class="s1">'0.3.5'</span><span class="p">:</span>
<span class="linenos" data-linenos="567 "></span>            <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="568 "></span>                <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="569 "></span>                <span class="n">clinical</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="570 "></span>                <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="571 "></span>                <span class="n">ref_level</span><span class="o">=</span><span class="p">[</span><span class="s2">"condition"</span><span class="p">,</span> <span class="s2">"Control"</span><span class="p">],</span>
<span class="linenos" data-linenos="572 "></span>                <span class="c1"># column ("B" vs "A")</span>
<span class="linenos" data-linenos="573 "></span>                <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="574 "></span>                <span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">,</span>
<span class="linenos" data-linenos="575 "></span>            <span class="p">)</span>
<span class="linenos" data-linenos="576 "></span>        <span class="k">elif</span> <span class="n">pydeseq2</span><span class="o">.</span><span class="n">__version__</span><span class="o">&lt;=</span><span class="s1">'0.4.1'</span><span class="p">:</span>
<span class="linenos" data-linenos="577 "></span>            <span class="k">if</span> <span class="n">ad</span><span class="o">.</span><span class="n">__version__</span><span class="o">&gt;</span><span class="s1">'0.10.8'</span><span class="p">:</span>
<span class="linenos" data-linenos="578 "></span>                <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span>
<span class="linenos" data-linenos="579 "></span>                        <span class="s1">'Please install the 0.10.8 version of anndata: `pip install anndata==0.10.8`.'</span>
<span class="linenos" data-linenos="580 "></span>                    <span class="p">)</span>
<span class="linenos" data-linenos="581 "></span>            <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="582 "></span>                <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="583 "></span>                <span class="n">metadata</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="584 "></span>                <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="585 "></span>                <span class="c1">#ref_level=["condition", "Control"],</span>
<span class="linenos" data-linenos="586 "></span>                <span class="c1"># column ("B" vs "A")</span>
<span class="linenos" data-linenos="587 "></span>                <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="588 "></span>                <span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">,</span>
<span class="linenos" data-linenos="589 "></span>            <span class="p">)</span>
<span class="linenos" data-linenos="590 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="591 "></span>            <span class="kn">from</span> <span class="nn">pydeseq2.default_inference</span> <span class="kn">import</span> <span class="n">DefaultInference</span>
<span class="linenos" data-linenos="592 "></span>            <span class="n">inference</span> <span class="o">=</span> <span class="n">DefaultInference</span><span class="p">(</span><span class="n">n_cpus</span><span class="o">=</span><span class="n">n_cpus</span><span class="p">)</span>
<span class="linenos" data-linenos="593 "></span>            <span class="n">dds</span> <span class="o">=</span> <span class="n">DeseqDataSet</span><span class="p">(</span>
<span class="linenos" data-linenos="594 "></span>                <span class="n">counts</span><span class="o">=</span><span class="n">counts_df</span><span class="p">,</span>
<span class="linenos" data-linenos="595 "></span>                <span class="n">metadata</span><span class="o">=</span><span class="n">clinical_df</span><span class="p">,</span>
<span class="linenos" data-linenos="596 "></span>                <span class="n">design_factors</span><span class="o">=</span><span class="s2">"condition"</span><span class="p">,</span>  <span class="c1"># compare samples based on the "condition"</span>
<span class="linenos" data-linenos="597 "></span>                <span class="n">refit_cooks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="598 "></span>                <span class="n">inference</span><span class="o">=</span><span class="n">inference</span><span class="p">,</span>
<span class="linenos" data-linenos="599 "></span>            <span class="p">)</span>
<span class="linenos" data-linenos="600 "></span>
<span class="linenos" data-linenos="601 "></span>
<span class="linenos" data-linenos="602 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_size_factors</span><span class="p">()</span>
<span class="linenos" data-linenos="603 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_genewise_dispersions</span><span class="p">()</span>
<span class="linenos" data-linenos="604 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_dispersion_trend</span><span class="p">()</span>
<span class="linenos" data-linenos="605 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_dispersion_prior</span><span class="p">()</span>
<span class="linenos" data-linenos="606 "></span>        <span class="nb">print</span><span class="p">(</span>
<span class="linenos" data-linenos="607 "></span>            <span class="sa">f</span><span class="s2">"logres_prior=</span><span class="si">{</span><span class="n">dds</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'_squared_logres'</span><span class="p">]</span><span class="si">}</span><span class="s2">, sigma_prior=</span><span class="si">{</span><span class="n">dds</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'prior_disp_var'</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span>
<span class="linenos" data-linenos="608 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="609 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_MAP_dispersions</span><span class="p">()</span>
<span class="linenos" data-linenos="610 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">fit_LFC</span><span class="p">()</span>
<span class="linenos" data-linenos="611 "></span>        <span class="n">dds</span><span class="o">.</span><span class="n">calculate_cooks</span><span class="p">()</span>
<span class="linenos" data-linenos="612 "></span>        <span class="k">if</span> <span class="n">dds</span><span class="o">.</span><span class="n">refit_cooks</span><span class="p">:</span>
<span class="linenos" data-linenos="613 "></span>            <span class="c1"># Replace outlier counts</span>
<span class="linenos" data-linenos="614 "></span>            <span class="n">dds</span><span class="o">.</span><span class="n">refit</span><span class="p">()</span>
<span class="linenos" data-linenos="615 "></span>        <span class="n">stat_res</span> <span class="o">=</span> <span class="n">DeseqStats</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span> <span class="n">cooks_filter</span><span class="o">=</span><span class="n">cooks_filter</span><span class="p">,</span> <span class="n">independent_filter</span><span class="o">=</span><span class="n">independent_filter</span><span class="p">)</span>
<span class="linenos" data-linenos="616 "></span>        <span class="n">stat_res</span><span class="o">.</span><span class="n">run_wald_test</span><span class="p">()</span>
<span class="linenos" data-linenos="617 "></span>        <span class="k">if</span> <span class="n">stat_res</span><span class="o">.</span><span class="n">cooks_filter</span><span class="p">:</span>
<span class="linenos" data-linenos="618 "></span>            <span class="n">stat_res</span><span class="o">.</span><span class="n">_cooks_filtering</span><span class="p">()</span>
<span class="linenos" data-linenos="619 "></span>        <span class="k">if</span> <span class="n">stat_res</span><span class="o">.</span><span class="n">independent_filter</span><span class="p">:</span>
<span class="linenos" data-linenos="620 "></span>            <span class="n">stat_res</span><span class="o">.</span><span class="n">_independent_filtering</span><span class="p">()</span>
<span class="linenos" data-linenos="621 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="622 "></span>            <span class="n">stat_res</span><span class="o">.</span><span class="n">_p_value_adjustment</span><span class="p">()</span>
<span class="linenos" data-linenos="623 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">stat_res</span><span class="o">=</span><span class="n">stat_res</span>
<span class="linenos" data-linenos="624 "></span>        <span class="n">stat_res</span><span class="o">.</span><span class="n">summary</span><span class="p">()</span>
<span class="linenos" data-linenos="625 "></span>        <span class="n">result</span><span class="o">=</span><span class="n">stat_res</span><span class="o">.</span><span class="n">results_df</span>
<span class="linenos" data-linenos="626 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'padj'</span><span class="p">]</span>
<span class="linenos" data-linenos="627 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(pvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="628 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'-log(qvalue)'</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'padj'</span><span class="p">])</span>
<span class="linenos" data-linenos="629 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'BaseMean'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'baseMean'</span><span class="p">]</span>
<span class="linenos" data-linenos="630 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2(BaseMean)'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'baseMean'</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="631 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s1">'log2FoldChange'</span><span class="p">]</span>
<span class="linenos" data-linenos="632 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'abs(log2FC)'</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">])</span>
<span class="linenos" data-linenos="633 "></span>        <span class="c1">#result['size']  =np.abs(result['FoldChange'])/10</span>
<span class="linenos" data-linenos="634 "></span>        <span class="c1">#result=result[result['padj']&lt;alpha]</span>
<span class="linenos" data-linenos="635 "></span>        <span class="n">result</span><span class="p">[</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'normal'</span>
<span class="linenos" data-linenos="636 "></span>        <span class="n">result</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="s1">'qvalue'</span><span class="p">]</span><span class="o">&lt;</span><span class="n">alpha</span><span class="p">,</span><span class="s1">'sig'</span><span class="p">]</span><span class="o">=</span><span class="s1">'sig'</span>
<span class="linenos" data-linenos="637 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">=</span><span class="n">result</span>
<span class="linenos" data-linenos="638 "></span>        <span class="k">return</span> <span class="n">result</span>
<span class="linenos" data-linenos="639 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="640 "></span>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">'The method is not supported.'</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.ranking2gsea">
<code class="highlight language-python"><span class="n">ranking2gsea</span><span class="p">(</span><span class="n">rank_max</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">rank_min</span><span class="o">=</span><span class="mi">274</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.ranking2gsea" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Ranking the result of dds data for gsea analysis</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>rank_max</code></td>
<td>
<code>int</code>
</td>
<td><p>The max rank of the result.</p></td>
<td>
<code>200</code>
</td>
</tr>
<tr>
<td><code>rank_min</code></td>
<td>
<code>int</code>
</td>
<td><p>The min rank of the result.</p></td>
<td>
<code>274</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>rnk</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>The ranking result.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="425 "></span><span class="k">def</span> <span class="nf">ranking2gsea</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">rank_max</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span><span class="n">rank_min</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">274</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="426 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="427 "></span><span class="sd">    Ranking the result of dds data for gsea analysis</span>
<span class="linenos" data-linenos="428 "></span>
<span class="linenos" data-linenos="429 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="430 "></span><span class="sd">        rank_max: The max rank of the result.</span>
<span class="linenos" data-linenos="431 "></span><span class="sd">        rank_min: The min rank of the result.</span>
<span class="linenos" data-linenos="432 "></span>
<span class="linenos" data-linenos="433 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="434 "></span><span class="sd">        rnk: The ranking result.</span>
<span class="linenos" data-linenos="435 "></span>
<span class="linenos" data-linenos="436 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="437 "></span>
<span class="linenos" data-linenos="438 "></span>
<span class="linenos" data-linenos="439 "></span>    <span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="440 "></span>    <span class="n">result</span><span class="p">[</span><span class="s1">'fcsign'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'log2FC'</span><span class="p">])</span>
<span class="linenos" data-linenos="441 "></span>    <span class="n">result</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">'pvalue'</span><span class="p">])</span>
<span class="linenos" data-linenos="442 "></span>    <span class="n">result</span><span class="p">[</span><span class="s1">'metric'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'logp'</span><span class="p">]</span><span class="o">/</span><span class="n">result</span><span class="p">[</span><span class="s1">'fcsign'</span><span class="p">]</span>
<span class="linenos" data-linenos="443 "></span>    <span class="n">rnk</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="444 "></span>    <span class="n">rnk</span><span class="p">[</span><span class="s1">'gene_name'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="o">.</span><span class="n">index</span>
<span class="linenos" data-linenos="445 "></span>    <span class="n">rnk</span><span class="p">[</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=</span><span class="n">result</span><span class="p">[</span><span class="s1">'metric'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="446 "></span>    <span class="n">rnk</span><span class="o">=</span><span class="n">rnk</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">'rnk'</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="447 "></span>    <span class="n">k</span><span class="o">=</span><span class="mi">1</span>
<span class="linenos" data-linenos="448 "></span>    <span class="n">total</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="449 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="450 "></span>        <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="451 "></span>            <span class="n">total</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="452 "></span>    <span class="c1">#200跟274根据你的数据进行更改，保证inf比你数据最大的大，-inf比数据最小的小就好</span>
<span class="linenos" data-linenos="453 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="454 "></span>        <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="455 "></span>            <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=</span><span class="n">rank_max</span><span class="o">+</span><span class="p">(</span><span class="n">total</span><span class="o">-</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="456 "></span>            <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="457 "></span>        <span class="k">elif</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">==-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="458 "></span>            <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'rnk'</span><span class="p">]</span><span class="o">=-</span><span class="p">(</span><span class="n">rank_min</span><span class="o">+</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="459 "></span>            <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="460 "></span>    <span class="k">return</span> <span class="n">rnk</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.plot_volcano">
<code class="highlight language-python"><span class="n">plot_volcano</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">pval_name</span><span class="o">=</span><span class="s1">'qvalue'</span><span class="p">,</span> <span class="n">fc_name</span><span class="o">=</span><span class="s1">'log2FC'</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">titlefont</span><span class="o">=</span><span class="p">{</span><span class="s1">'weight'</span><span class="p">:</span> <span class="s1">'normal'</span><span class="p">,</span> <span class="s1">'size'</span><span class="p">:</span> <span class="mi">14</span><span class="p">},</span> <span class="n">up_color</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span> <span class="n">down_color</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span> <span class="n">normal_color</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span> <span class="n">up_fontcolor</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span> <span class="n">down_fontcolor</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span> <span class="n">normal_fontcolor</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span> <span class="n">legend_bbox</span><span class="o">=</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">),</span> <span class="n">legend_ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">legend_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">plot_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">plot_genes_num</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">plot_genes_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ticks_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.plot_volcano" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Generate a volcano plot for the differential gene expression analysis results.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The size of the generated figure, by default (4,4).</p></td>
<td>
<code>(4, 4)</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
<code>str</code>
</td>
<td><p>The title of the plot, by default ''.</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>titlefont</code></td>
<td>
<code>dict</code>
</td>
<td><p>A dictionary of font properties for the plot title, by default {'weight':'normal','size':14,}.</p></td>
<td>
<code>{'weight': 'normal', 'size': 14}</code>
</td>
</tr>
<tr>
<td><code>up_color</code></td>
<td>
<code>str</code>
</td>
<td><p>The color of the up-regulated genes in the plot, by default '#e25d5d'.</p></td>
<td>
<code>'#e25d5d'</code>
</td>
</tr>
<tr>
<td><code>down_color</code></td>
<td>
<code>str</code>
</td>
<td><p>The color of the down-regulated genes in the plot, by default '#7388c1'.</p></td>
<td>
<code>'#7388c1'</code>
</td>
</tr>
<tr>
<td><code>normal_color</code></td>
<td>
<code>str</code>
</td>
<td><p>The color of the non-significant genes in the plot, by default '#d7d7d7'.</p></td>
<td>
<code>'#d7d7d7'</code>
</td>
</tr>
<tr>
<td><code>legend_bbox</code></td>
<td>
<code>tuple</code>
</td>
<td><p>A tuple containing the coordinates of the legend's bounding box, by default (0.8, -0.2).</p></td>
<td>
<code>(0.8, -0.2)</code>
</td>
</tr>
<tr>
<td><code>legend_ncol</code></td>
<td>
<code>int</code>
</td>
<td><p>The number of columns in the legend, by default 2.</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>legend_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The font size of the legend, by default 12.</p></td>
<td>
<code>12</code>
</td>
</tr>
<tr>
<td><code>plot_genes</code></td>
<td>
<code>list</code>
</td>
<td><p>A list of genes to be plotted on the volcano plot, by default None.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>plot_genes_num</code></td>
<td>
<code>int</code>
</td>
<td><p>The number of genes to be plotted on the volcano plot, by default 10.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>plot_genes_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The font size of the genes to be plotted on the volcano plot, by default 10.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>ticks_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The font size of the ticks, by default 12.</p></td>
<td>
<code>12</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ax</code></td> <td>
</td>
<td><p>The generated volcano plot.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="228 "></span><span class="k">def</span> <span class="nf">plot_volcano</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span><span class="n">pval_name</span><span class="o">=</span><span class="s1">'qvalue'</span><span class="p">,</span><span class="n">fc_name</span><span class="o">=</span><span class="s1">'log2FC'</span><span class="p">,</span>
<span class="linenos" data-linenos="229 "></span>                 <span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span><span class="n">titlefont</span><span class="p">:</span><span class="nb">dict</span><span class="o">=</span><span class="p">{</span><span class="s1">'weight'</span><span class="p">:</span><span class="s1">'normal'</span><span class="p">,</span><span class="s1">'size'</span><span class="p">:</span><span class="mi">14</span><span class="p">,},</span>
<span class="linenos" data-linenos="230 "></span>                 <span class="n">up_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span><span class="n">down_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span><span class="n">normal_color</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span>
<span class="linenos" data-linenos="231 "></span>                 <span class="n">up_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#e25d5d'</span><span class="p">,</span><span class="n">down_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#7388c1'</span><span class="p">,</span><span class="n">normal_fontcolor</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'#d7d7d7'</span><span class="p">,</span>
<span class="linenos" data-linenos="232 "></span>                 <span class="n">legend_bbox</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">),</span><span class="n">legend_ncol</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">legend_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
<span class="linenos" data-linenos="233 "></span>                 <span class="n">plot_genes</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">plot_genes_num</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">plot_genes_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="linenos" data-linenos="234 "></span>                 <span class="n">ticks_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="235 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="236 "></span><span class="sd">    Generate a volcano plot for the differential gene expression analysis results.</span>
<span class="linenos" data-linenos="237 "></span>
<span class="linenos" data-linenos="238 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="239 "></span><span class="sd">        figsize: The size of the generated figure, by default (4,4).</span>
<span class="linenos" data-linenos="240 "></span><span class="sd">        title: The title of the plot, by default ''.</span>
<span class="linenos" data-linenos="241 "></span><span class="sd">        titlefont: A dictionary of font properties for the plot title, by default {'weight':'normal','size':14,}.</span>
<span class="linenos" data-linenos="242 "></span><span class="sd">        up_color: The color of the up-regulated genes in the plot, by default '#e25d5d'.</span>
<span class="linenos" data-linenos="243 "></span><span class="sd">        down_color: The color of the down-regulated genes in the plot, by default '#7388c1'.</span>
<span class="linenos" data-linenos="244 "></span><span class="sd">        normal_color: The color of the non-significant genes in the plot, by default '#d7d7d7'.</span>
<span class="linenos" data-linenos="245 "></span><span class="sd">        legend_bbox: A tuple containing the coordinates of the legend's bounding box, by default (0.8, -0.2).</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">        legend_ncol: The number of columns in the legend, by default 2.</span>
<span class="linenos" data-linenos="247 "></span><span class="sd">        legend_fontsize: The font size of the legend, by default 12.</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">        plot_genes: A list of genes to be plotted on the volcano plot, by default None.</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">        plot_genes_num: The number of genes to be plotted on the volcano plot, by default 10.</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">        plot_genes_fontsize: The font size of the genes to be plotted on the volcano plot, by default 10.</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">        ticks_fontsize: The font size of the ticks, by default 12.</span>
<span class="linenos" data-linenos="252 "></span>
<span class="linenos" data-linenos="253 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="254 "></span><span class="sd">        ax: The generated volcano plot.</span>
<span class="linenos" data-linenos="255 "></span>
<span class="linenos" data-linenos="256 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="257 "></span>
<span class="linenos" data-linenos="258 "></span>    <span class="n">ax</span><span class="o">=</span><span class="n">volcano</span><span class="p">(</span><span class="n">result</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">result</span><span class="p">,</span><span class="n">pval_name</span><span class="o">=</span><span class="n">pval_name</span><span class="p">,</span><span class="n">fc_name</span><span class="o">=</span><span class="n">fc_name</span><span class="p">,</span><span class="n">pval_max</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">logp_max</span><span class="p">,</span>
<span class="linenos" data-linenos="259 "></span>                   <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span><span class="n">titlefont</span><span class="o">=</span><span class="n">titlefont</span><span class="p">,</span>
<span class="linenos" data-linenos="260 "></span>                   <span class="n">up_color</span><span class="o">=</span><span class="n">up_color</span><span class="p">,</span><span class="n">down_color</span><span class="o">=</span><span class="n">down_color</span><span class="p">,</span><span class="n">normal_color</span><span class="o">=</span><span class="n">normal_color</span><span class="p">,</span>
<span class="linenos" data-linenos="261 "></span>                   <span class="n">up_fontcolor</span><span class="o">=</span><span class="n">up_fontcolor</span><span class="p">,</span><span class="n">down_fontcolor</span><span class="o">=</span><span class="n">down_fontcolor</span><span class="p">,</span><span class="n">normal_fontcolor</span><span class="o">=</span><span class="n">normal_fontcolor</span><span class="p">,</span>
<span class="linenos" data-linenos="262 "></span>                   <span class="n">legend_bbox</span><span class="o">=</span><span class="n">legend_bbox</span><span class="p">,</span><span class="n">legend_ncol</span><span class="o">=</span><span class="n">legend_ncol</span><span class="p">,</span><span class="n">legend_fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">,</span><span class="n">plot_genes</span><span class="o">=</span><span class="n">plot_genes</span><span class="p">,</span>
<span class="linenos" data-linenos="263 "></span>                   <span class="n">plot_genes_num</span><span class="o">=</span><span class="n">plot_genes_num</span><span class="p">,</span><span class="n">plot_genes_fontsize</span><span class="o">=</span><span class="n">plot_genes_fontsize</span><span class="p">,</span>
<span class="linenos" data-linenos="264 "></span>                   <span class="n">ticks_fontsize</span><span class="o">=</span><span class="n">ticks_fontsize</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
<span class="linenos" data-linenos="265 "></span>                   <span class="n">pval_threshold</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pval_threshold</span><span class="p">,</span><span class="n">fc_max</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_max</span><span class="p">,</span><span class="n">fc_min</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fc_min</span><span class="p">)</span>
<span class="linenos" data-linenos="266 "></span>    <span class="k">return</span> <span class="n">ax</span>
<span class="linenos" data-linenos="267 "></span><span class="w">    </span><span class="sd">'''</span>
<span class="linenos" data-linenos="268 "></span><span class="sd">    fig, ax = plt.subplots(figsize=figsize)</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">    result=self.result.copy()</span>
<span class="linenos" data-linenos="270 "></span><span class="sd">    #首先绘制正常基因</span>
<span class="linenos" data-linenos="271 "></span><span class="sd">    ax.scatter(x=result[result['sig']=='normal']['log2FC'],</span>
<span class="linenos" data-linenos="272 "></span><span class="sd">            y=result[result['sig']=='normal']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="273 "></span><span class="sd">            color=normal_color,#颜色</span>
<span class="linenos" data-linenos="274 "></span><span class="sd">            alpha=.5,#透明度</span>
<span class="linenos" data-linenos="275 "></span><span class="sd">            )</span>
<span class="linenos" data-linenos="276 "></span><span class="sd">    #接着绘制上调基因</span>
<span class="linenos" data-linenos="277 "></span><span class="sd">    ax.scatter(x=result[result['sig']=='up']['log2FC'],</span>
<span class="linenos" data-linenos="278 "></span><span class="sd">            y=result[result['sig']=='up']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="279 "></span><span class="sd">            color=up_color,#选择色卡第15个颜色</span>
<span class="linenos" data-linenos="280 "></span><span class="sd">            alpha=.5,#透明度</span>
<span class="linenos" data-linenos="281 "></span><span class="sd">            )</span>
<span class="linenos" data-linenos="282 "></span><span class="sd">    #绘制下调基因</span>
<span class="linenos" data-linenos="283 "></span><span class="sd">    ax.scatter(x=result[result['sig']=='down']['log2FC'],</span>
<span class="linenos" data-linenos="284 "></span><span class="sd">            y=result[result['sig']=='down']['-log(qvalue)'],</span>
<span class="linenos" data-linenos="285 "></span><span class="sd">            color=down_color,#颜色</span>
<span class="linenos" data-linenos="286 "></span><span class="sd">            alpha=.5,#透明度</span>
<span class="linenos" data-linenos="287 "></span><span class="sd">            )</span>
<span class="linenos" data-linenos="288 "></span>
<span class="linenos" data-linenos="289 "></span><span class="sd">    ax.plot([result['log2FC'].min(),result['log2FC'].max()],#辅助线的x值起点与终点</span>
<span class="linenos" data-linenos="290 "></span><span class="sd">            [-np.log10(self.pval_threshold),-np.log10(self.pval_threshold)],#辅助线的y值起点与终点</span>
<span class="linenos" data-linenos="291 "></span><span class="sd">            linewidth=2,#辅助线的宽度</span>
<span class="linenos" data-linenos="292 "></span><span class="sd">            linestyle="--",#辅助线类型：虚线</span>
<span class="linenos" data-linenos="293 "></span><span class="sd">            color='black'#辅助线的颜色</span>
<span class="linenos" data-linenos="294 "></span><span class="sd">    )</span>
<span class="linenos" data-linenos="295 "></span><span class="sd">    ax.plot([self.fc_max,self.fc_max],</span>
<span class="linenos" data-linenos="296 "></span><span class="sd">            [result['-log(qvalue)'].min(),result['-log(qvalue)'].max()],</span>
<span class="linenos" data-linenos="297 "></span><span class="sd">            linewidth=2, </span>
<span class="linenos" data-linenos="298 "></span><span class="sd">            linestyle="--",</span>
<span class="linenos" data-linenos="299 "></span><span class="sd">            color='black')</span>
<span class="linenos" data-linenos="300 "></span><span class="sd">    ax.plot([self.fc_min,self.fc_min],</span>
<span class="linenos" data-linenos="301 "></span><span class="sd">            [result['-log(qvalue)'].min(),result['-log(qvalue)'].max()],</span>
<span class="linenos" data-linenos="302 "></span><span class="sd">            linewidth=2, </span>
<span class="linenos" data-linenos="303 "></span><span class="sd">            linestyle="--",</span>
<span class="linenos" data-linenos="304 "></span><span class="sd">            color='black')</span>
<span class="linenos" data-linenos="305 "></span><span class="sd">    #设置横标签与纵标签</span>
<span class="linenos" data-linenos="306 "></span><span class="sd">    ax.set_ylabel(r'$-log_{10}(qvalue)$',titlefont)                                    </span>
<span class="linenos" data-linenos="307 "></span><span class="sd">    ax.set_xlabel(r'$log_{2}FC$',titlefont)</span>
<span class="linenos" data-linenos="308 "></span><span class="sd">    #设置标题</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">    ax.set_title(title,titlefont)</span>
<span class="linenos" data-linenos="310 "></span>
<span class="linenos" data-linenos="311 "></span><span class="sd">    #绘制图注</span>
<span class="linenos" data-linenos="312 "></span><span class="sd">    #legend标签列表，上面的color即是颜色列表</span>
<span class="linenos" data-linenos="313 "></span><span class="sd">    labels = ['up:{0}'.format(len(result[result['sig']=='up'])),</span>
<span class="linenos" data-linenos="314 "></span><span class="sd">            'down:{0}'.format(len(result[result['sig']=='down']))]  </span>
<span class="linenos" data-linenos="315 "></span><span class="sd">    #用label和color列表生成mpatches.Patch对象，它将作为句柄来生成legend</span>
<span class="linenos" data-linenos="316 "></span><span class="sd">    color = [up_color,down_color]</span>
<span class="linenos" data-linenos="317 "></span><span class="sd">    patches = [mpatches.Patch(color=color[i], label="{:s}".format(labels[i]) ) for i in range(len(color))] </span>
<span class="linenos" data-linenos="318 "></span>
<span class="linenos" data-linenos="319 "></span><span class="sd">    ax.legend(handles=patches,</span>
<span class="linenos" data-linenos="320 "></span><span class="sd">        bbox_to_anchor=legend_bbox, </span>
<span class="linenos" data-linenos="321 "></span><span class="sd">        ncol=legend_ncol,</span>
<span class="linenos" data-linenos="322 "></span><span class="sd">        fontsize=legend_fontsize)</span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span><span class="sd">    ax.spines['top'].set_visible(False)</span>
<span class="linenos" data-linenos="325 "></span><span class="sd">    ax.spines['right'].set_visible(False)</span>
<span class="linenos" data-linenos="326 "></span><span class="sd">    ax.spines['bottom'].set_visible(True)</span>
<span class="linenos" data-linenos="327 "></span><span class="sd">    ax.spines['left'].set_visible(True)</span>
<span class="linenos" data-linenos="328 "></span>
<span class="linenos" data-linenos="329 "></span><span class="sd">    from adjustText import adjust_text</span>
<span class="linenos" data-linenos="330 "></span><span class="sd">    import adjustText</span>
<span class="linenos" data-linenos="331 "></span>
<span class="linenos" data-linenos="332 "></span><span class="sd">    if plot_genes is not None:</span>
<span class="linenos" data-linenos="333 "></span><span class="sd">        hub_gene=plot_genes</span>
<span class="linenos" data-linenos="334 "></span><span class="sd">    else:</span>
<span class="linenos" data-linenos="335 "></span><span class="sd">        up_result=result.loc[result['sig']=='up']</span>
<span class="linenos" data-linenos="336 "></span><span class="sd">        down_result=result.loc[result['sig']=='down']</span>
<span class="linenos" data-linenos="337 "></span><span class="sd">        hub_gene=up_result.sort_values('qvalue').index[:plot_genes_num//2].tolist()+down_result.sort_values('qvalue').index[:plot_genes_num//2].tolist()</span>
<span class="linenos" data-linenos="338 "></span>
<span class="linenos" data-linenos="339 "></span><span class="sd">    color_dict={</span>
<span class="linenos" data-linenos="340 "></span><span class="sd">    'up':up_fontcolor,</span>
<span class="linenos" data-linenos="341 "></span><span class="sd">        'down':down_fontcolor,</span>
<span class="linenos" data-linenos="342 "></span><span class="sd">        'normal':normal_fontcolor</span>
<span class="linenos" data-linenos="343 "></span><span class="sd">    }</span>
<span class="linenos" data-linenos="344 "></span>
<span class="linenos" data-linenos="345 "></span><span class="sd">    texts=[ax.text(result.loc[i,'log2FC'], </span>
<span class="linenos" data-linenos="346 "></span><span class="sd">           result.loc[i,'-log(qvalue)'],</span>
<span class="linenos" data-linenos="347 "></span><span class="sd">           i,</span>
<span class="linenos" data-linenos="348 "></span><span class="sd">           fontdict={'size':plot_genes_fontsize,'weight':'bold','color':color_dict[result.loc[i,'sig']]}</span>
<span class="linenos" data-linenos="349 "></span><span class="sd">           ) for i in hub_gene]</span>
<span class="linenos" data-linenos="350 "></span>
<span class="linenos" data-linenos="351 "></span><span class="sd">    if adjustText.__version__&lt;='0.8':</span>
<span class="linenos" data-linenos="352 "></span><span class="sd">        adjust_text(texts,only_move={'text': 'xy'},arrowprops=dict(arrowstyle='-&gt;', color='red'),)</span>
<span class="linenos" data-linenos="353 "></span><span class="sd">    else:</span>
<span class="linenos" data-linenos="354 "></span><span class="sd">        adjust_text(texts,only_move={"text": "xy", "static": "xy", "explode": "xy", "pull": "xy"},</span>
<span class="linenos" data-linenos="355 "></span><span class="sd">                    arrowprops=dict(arrowstyle='-&gt;', color='red'))</span>
<span class="linenos" data-linenos="356 "></span>
<span class="linenos" data-linenos="357 "></span><span class="sd">    ax.set_xticks([round(i,2) for i in ax.get_xticks()[1:-1]],#获取x坐标轴内容</span>
<span class="linenos" data-linenos="358 "></span><span class="sd">          [round(i,2) for i in ax.get_xticks()[1:-1]],#更新x坐标轴内容</span>
<span class="linenos" data-linenos="359 "></span><span class="sd">          fontsize=ticks_fontsize,</span>
<span class="linenos" data-linenos="360 "></span><span class="sd">          fontweight='normal'</span>
<span class="linenos" data-linenos="361 "></span><span class="sd">          )</span>
<span class="linenos" data-linenos="362 "></span><span class="sd">    return fig,ax</span>
<span class="linenos" data-linenos="363 "></span><span class="sd">    '''</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.bulk._Deseq2.pyDEG.plot_boxplot">
<code class="highlight language-python"><span class="n">plot_boxplot</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">treatment_groups</span><span class="p">,</span> <span class="n">control_groups</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">treatment_name</span><span class="o">=</span><span class="s1">'Treatment'</span><span class="p">,</span> <span class="n">control_name</span><span class="o">=</span><span class="s1">'Control'</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">palette</span><span class="o">=</span><span class="p">[</span><span class="s1">'#a64d79'</span><span class="p">,</span> <span class="s1">'#674ea7'</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Gene Expression'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">legend_bbox</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.55</span><span class="p">),</span> <span class="n">legend_ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="o">**</span><span class="n">kwarg</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk._Deseq2.pyDEG.plot_boxplot" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Plot the boxplot of genes from dds data</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>genes</code></td>
<td>
<code>list</code>
</td>
<td><p>The genes to plot.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>treatment_groups</code></td>
<td>
<code>list</code>
</td>
<td><p>The treatment groups.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>control_groups</code></td>
<td>
<code>list</code>
</td>
<td><p>The control groups.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The figure size.</p></td>
<td>
<code>(4, 3)</code>
</td>
</tr>
<tr>
<td><code>palette</code></td>
<td>
<code>list</code>
</td>
<td><p>The color palette.</p></td>
<td>
<code>['#a64d79', '#674ea7']</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
<code>str</code>
</td>
<td><p>The title of the plot.</p></td>
<td>
<code>'Gene Expression'</code>
</td>
</tr>
<tr>
<td><code>fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The fontsize of the plot.</p></td>
<td>
<code>12</code>
</td>
</tr>
<tr>
<td><code>legend_bbox</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The bbox of the legend.</p></td>
<td>
<code>(1, 0.55)</code>
</td>
</tr>
<tr>
<td><code>legend_ncol</code></td>
<td>
<code>int</code>
</td>
<td><p>The number of columns of the legend.</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>**kwarg</code></td>
<td>
</td>
<td><p>Other arguments for plot_boxplot function.</p></td>
<td>
<code>{}</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
<code>matplotlib.<span title="matplotlib.figure">figure</span>.<span title="matplotlib.figure.Figure">Figure</span></code>
</td>
<td><p>The figure of the plot.</p></td>
</tr>
<tr>
<td><code>ax</code></td> <td>
<code>matplotlib.<span title="matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>The axis of the plot.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="365 "></span><span class="k">def</span> <span class="nf">plot_boxplot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">genes</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">treatment_groups</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">control_groups</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
<span class="linenos" data-linenos="366 "></span>                 <span class="n">log</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="367 "></span>                 <span class="n">treatment_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Treatment'</span><span class="p">,</span><span class="n">control_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Control'</span><span class="p">,</span>
<span class="linenos" data-linenos="368 "></span>                 <span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span><span class="n">palette</span><span class="p">:</span><span class="nb">list</span><span class="o">=</span><span class="p">[</span><span class="s2">"#a64d79"</span><span class="p">,</span><span class="s2">"#674ea7"</span><span class="p">],</span>
<span class="linenos" data-linenos="369 "></span>                 <span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'Gene Expression'</span><span class="p">,</span><span class="n">fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">legend_bbox</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.55</span><span class="p">),</span><span class="n">legend_ncol</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="linenos" data-linenos="370 "></span>                 <span class="o">**</span><span class="n">kwarg</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">Tuple</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">Figure</span><span class="p">,</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">]:</span>
<span class="linenos" data-linenos="371 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="372 "></span><span class="sd">    Plot the boxplot of genes from dds data</span>
<span class="linenos" data-linenos="373 "></span>
<span class="linenos" data-linenos="374 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="375 "></span><span class="sd">        genes: The genes to plot.</span>
<span class="linenos" data-linenos="376 "></span><span class="sd">        treatment_groups: The treatment groups.</span>
<span class="linenos" data-linenos="377 "></span><span class="sd">        control_groups: The control groups.</span>
<span class="linenos" data-linenos="378 "></span><span class="sd">        figsize: The figure size.</span>
<span class="linenos" data-linenos="379 "></span><span class="sd">        palette: The color palette.</span>
<span class="linenos" data-linenos="380 "></span><span class="sd">        title: The title of the plot.</span>
<span class="linenos" data-linenos="381 "></span><span class="sd">        fontsize: The fontsize of the plot.</span>
<span class="linenos" data-linenos="382 "></span><span class="sd">        legend_bbox: The bbox of the legend.</span>
<span class="linenos" data-linenos="383 "></span><span class="sd">        legend_ncol: The number of columns of the legend.</span>
<span class="linenos" data-linenos="384 "></span><span class="sd">        **kwarg: Other arguments for plot_boxplot function.</span>
<span class="linenos" data-linenos="385 "></span>
<span class="linenos" data-linenos="386 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="387 "></span><span class="sd">        fig: The figure of the plot.</span>
<span class="linenos" data-linenos="388 "></span><span class="sd">        ax: The axis of the plot.</span>
<span class="linenos" data-linenos="389 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="390 "></span>    <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">,</span><span class="s1">'Gene'</span><span class="p">,</span><span class="s1">'Type'</span><span class="p">])</span>
<span class="linenos" data-linenos="391 "></span>    <span class="k">if</span> <span class="n">log</span><span class="p">:</span>
<span class="linenos" data-linenos="392 "></span>        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
<span class="linenos" data-linenos="393 "></span>            <span class="n">plot_data1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="394 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">treatment_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="395 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="396 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">treatment_name</span>
<span class="linenos" data-linenos="397 "></span>
<span class="linenos" data-linenos="398 "></span>            <span class="n">plot_data2</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="399 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">control_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="linenos" data-linenos="400 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="401 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">control_name</span>
<span class="linenos" data-linenos="402 "></span>
<span class="linenos" data-linenos="403 "></span>            <span class="n">plot_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_data1</span><span class="p">,</span><span class="n">plot_data2</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="404 "></span>            <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">p_data</span><span class="p">,</span><span class="n">plot_data</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="405 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="406 "></span>        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
<span class="linenos" data-linenos="407 "></span>            <span class="n">plot_data1</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="408 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">treatment_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="409 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="410 "></span>            <span class="n">plot_data1</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">treatment_name</span>
<span class="linenos" data-linenos="411 "></span>
<span class="linenos" data-linenos="412 "></span>            <span class="n">plot_data2</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos="413 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Value'</span><span class="p">]</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">control_groups</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos="414 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">=</span><span class="n">gene</span>
<span class="linenos" data-linenos="415 "></span>            <span class="n">plot_data2</span><span class="p">[</span><span class="s1">'Type'</span><span class="p">]</span><span class="o">=</span><span class="n">control_name</span>
<span class="linenos" data-linenos="416 "></span>
<span class="linenos" data-linenos="417 "></span>            <span class="n">plot_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">plot_data1</span><span class="p">,</span><span class="n">plot_data2</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="418 "></span>            <span class="n">p_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">p_data</span><span class="p">,</span><span class="n">plot_data</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="419 "></span>
<span class="linenos" data-linenos="420 "></span>    <span class="n">fig</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">plot_boxplot</span><span class="p">(</span><span class="n">p_data</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="s1">'Type'</span><span class="p">,</span><span class="n">x_value</span><span class="o">=</span><span class="s1">'Gene'</span><span class="p">,</span><span class="n">y_value</span><span class="o">=</span><span class="s1">'Value'</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
<span class="linenos" data-linenos="421 "></span>                      <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="n">fontsize</span><span class="p">,</span><span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
<span class="linenos" data-linenos="422 "></span>                      <span class="n">legend_bbox</span><span class="o">=</span><span class="n">legend_bbox</span><span class="p">,</span><span class="n">legend_ncol</span><span class="o">=</span><span class="n">legend_ncol</span><span class="p">,</span> <span class="o">**</span><span class="n">kwarg</span><span class="p">)</span>
<span class="linenos" data-linenos="423 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="p">,</span><span class="n">ax</span>
</code></pre></div>
</details>
</div>
</div>
</div>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.bulk.deseq2_normalize">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">deseq2_normalize</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk.deseq2_normalize" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Normalize the data using DESeq2 method:</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>the data to be normalized.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>the normalized data.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="45 "></span><span class="k">def</span> <span class="nf">deseq2_normalize</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="46 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="47 "></span><span class="sd">    Normalize the data using DESeq2 method:</span>
<span class="linenos" data-linenos="48 "></span>
<span class="linenos" data-linenos="49 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="50 "></span><span class="sd">        data: the data to be normalized.</span>
<span class="linenos" data-linenos="51 "></span>
<span class="linenos" data-linenos="52 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="53 "></span><span class="sd">        data: the normalized data.</span>
<span class="linenos" data-linenos="54 "></span>
<span class="linenos" data-linenos="55 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="56 "></span>    <span class="n">avg1</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="linenos" data-linenos="57 "></span>    <span class="n">data1</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">avg1</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="linenos" data-linenos="58 "></span>    <span class="n">data_log</span><span class="o">=</span><span class="n">data1</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="59 "></span>    <span class="n">scale</span><span class="o">=</span><span class="n">data_log</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">avg1</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">)</span>
<span class="linenos" data-linenos="60 "></span>    <span class="k">return</span> <span class="n">data</span><span class="o">/</span><span class="n">scale</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.bulk.estimateSizeFactors">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="n">data</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk.estimateSizeFactors" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Estimate size factors for data normalization.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>A pandas DataFrame of gene expression data where rows correspond to samples and columns correspond to genes.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>scale</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.Series">Series</span></code>
</td>
<td><p>A pandas Series of size factors, one for each sample.</p></td>
</tr>
</tbody>
</table>
<p><strong>Examples:</strong></p>
<hr/>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">Pyomic</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="s1">'abcdefghij'</span><span class="p">))</span>
<span class="n">size_factors</span> <span class="o">=</span> <span class="n">Pyomic</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">estimateSizeFactors</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 98 "></span><span class="k">def</span> <span class="nf">estimateSizeFactors</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="linenos" data-linenos=" 99 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="100 "></span><span class="sd">    Estimate size factors for data normalization.</span>
<span class="linenos" data-linenos="101 "></span>
<span class="linenos" data-linenos="102 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="103 "></span><span class="sd">        data:  A pandas DataFrame of gene expression data where rows correspond to samples and columns correspond to genes.</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="105 "></span><span class="sd">        scale: A pandas Series of size factors, one for each sample.</span>
<span class="linenos" data-linenos="106 "></span>
<span class="linenos" data-linenos="107 "></span><span class="sd">    Examples:</span>
<span class="linenos" data-linenos="108 "></span><span class="sd">    --------</span>
<span class="linenos" data-linenos="109 "></span><span class="sd">    ```python</span>
<span class="linenos" data-linenos="110 "></span><span class="sd">    import pandas as pd</span>
<span class="linenos" data-linenos="111 "></span><span class="sd">    import numpy as np</span>
<span class="linenos" data-linenos="112 "></span><span class="sd">    import Pyomic</span>
<span class="linenos" data-linenos="113 "></span><span class="sd">    data = pd.DataFrame(np.random.rand(100, 10), columns=list('abcdefghij'))</span>
<span class="linenos" data-linenos="114 "></span><span class="sd">    size_factors = Pyomic.bulk.estimateSizeFactors(data)</span>
<span class="linenos" data-linenos="115 "></span><span class="sd">    ```</span>
<span class="linenos" data-linenos="116 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="117 "></span>    <span class="n">avg1</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">],</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="linenos" data-linenos="118 "></span>    <span class="n">data1</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">avg1</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="linenos" data-linenos="119 "></span>    <span class="n">data_log</span><span class="o">=</span><span class="n">data1</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="120 "></span>    <span class="n">scale</span><span class="o">=</span><span class="n">data_log</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">avg1</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">)</span>
<span class="linenos" data-linenos="121 "></span>    <span class="k">return</span> <span class="n">scale</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.bulk.estimateDispersions">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">estimateDispersions</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk.estimateDispersions" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Estimates the dispersion parameter of the Negative Binomial distribution
for each gene in the input count matrix.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>counts</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>Input count matrix with shape (n_genes, n_samples).</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>disp</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.Series">Series</span></code>
</td>
<td><p>Array of dispersion values for each gene in the input count matrix.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="124 "></span><span class="k">def</span> <span class="nf">estimateDispersions</span><span class="p">(</span><span class="n">counts</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="linenos" data-linenos="125 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="126 "></span><span class="sd">    Estimates the dispersion parameter of the Negative Binomial distribution</span>
<span class="linenos" data-linenos="127 "></span><span class="sd">    for each gene in the input count matrix.</span>
<span class="linenos" data-linenos="128 "></span>
<span class="linenos" data-linenos="129 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="130 "></span><span class="sd">        counts:Input count matrix with shape (n_genes, n_samples).</span>
<span class="linenos" data-linenos="131 "></span>
<span class="linenos" data-linenos="132 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="133 "></span><span class="sd">        disp: Array of dispersion values for each gene in the input count matrix.</span>
<span class="linenos" data-linenos="134 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="135 "></span>    <span class="c1"># Step 1: Calculate mean and variance of counts for each gene</span>
<span class="linenos" data-linenos="136 "></span>    <span class="n">mean_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="137 "></span>    <span class="n">var_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="138 "></span>
<span class="linenos" data-linenos="139 "></span>    <span class="c1"># Step 2: Fit trend line to variance-mean relationship using GLM</span>
<span class="linenos" data-linenos="140 "></span>    <span class="n">mean_expr</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">add_constant</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">mean_counts</span><span class="p">))</span>
<span class="linenos" data-linenos="141 "></span>    <span class="n">mod</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">GLM</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">var_counts</span><span class="p">),</span> <span class="n">mean_expr</span><span class="p">,</span> <span class="n">family</span><span class="o">=</span><span class="n">sm</span><span class="o">.</span><span class="n">families</span><span class="o">.</span><span class="n">Gamma</span><span class="p">())</span>
<span class="linenos" data-linenos="142 "></span>    <span class="n">res</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="linenos" data-linenos="143 "></span>    <span class="n">fitted_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">fittedvalues</span><span class="p">)</span>
<span class="linenos" data-linenos="144 "></span>
<span class="linenos" data-linenos="145 "></span>    <span class="c1"># Step 3: Calculate residual variance for each gene</span>
<span class="linenos" data-linenos="146 "></span>    <span class="n">disp</span> <span class="o">=</span> <span class="n">fitted_var</span> <span class="o">/</span> <span class="n">var_counts</span>
<span class="linenos" data-linenos="147 "></span>
<span class="linenos" data-linenos="148 "></span>    <span class="k">return</span> <span class="n">disp</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.bulk.Matrix_ID_mapping">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">bulk</span><span class="o">.</span><span class="n">Matrix_ID_mapping</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">gene_ref_path</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.bulk.Matrix_ID_mapping" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Maps gene IDs in the input data to gene symbols using a reference table.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>The input data containing gene IDs as index.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>gene_ref_path</code></td>
<td>
<code>str</code>
</td>
<td><p>The path to the reference table containing the mapping from gene IDs to gene symbols.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>The input data with gene IDs mapped to gene symbols.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/bulk/_Deseq2.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="17 "></span><span class="k">def</span> <span class="nf">Matrix_ID_mapping</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">gene_ref_path</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="18 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="19 "></span><span class="sd">    Maps gene IDs in the input data to gene symbols using a reference table.</span>
<span class="linenos" data-linenos="20 "></span>
<span class="linenos" data-linenos="21 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="22 "></span><span class="sd">        data: The input data containing gene IDs as index.</span>
<span class="linenos" data-linenos="23 "></span><span class="sd">        gene_ref_path: The path to the reference table containing the mapping from gene IDs to gene symbols.</span>
<span class="linenos" data-linenos="24 "></span>
<span class="linenos" data-linenos="25 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="26 "></span><span class="sd">        data: The input data with gene IDs mapped to gene symbols.</span>
<span class="linenos" data-linenos="27 "></span>
<span class="linenos" data-linenos="28 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="29 "></span>
<span class="linenos" data-linenos="30 "></span>    <span class="n">pair</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_ref_path</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">'</span><span class="se">\t</span><span class="s1">'</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="31 "></span>    <span class="n">ret_gene</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span>
<span class="linenos" data-linenos="32 "></span>    <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ret_gene</span><span class="p">]</span>
<span class="linenos" data-linenos="33 "></span>    <span class="c1">#data=data_drop_duplicates_index(data)</span>
<span class="linenos" data-linenos="34 "></span>    <span class="n">new_index</span><span class="o">=</span><span class="p">[]</span>
<span class="linenos" data-linenos="35 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ret_gene</span><span class="p">:</span>
<span class="linenos" data-linenos="36 "></span>        <span class="n">a</span><span class="o">=</span><span class="n">pair</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'symbol'</span><span class="p">]</span>
<span class="linenos" data-linenos="37 "></span>        <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">==</span><span class="s1">'nan'</span><span class="p">:</span>
<span class="linenos" data-linenos="38 "></span>            <span class="n">new_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<span class="linenos" data-linenos="39 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="40 "></span>            <span class="n">new_index</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="linenos" data-linenos="41 "></span>    <span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">new_index</span>
<span class="linenos" data-linenos="42 "></span>    <span class="k">return</span> <span class="n">data</span>
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2024, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>