
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api module - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.bulk.pyWGCNA">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api module
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology/Taxonomy
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scenic/">
        Gene Regulatory Network Analysis with SCENIC
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell-cell communication using CellPhoneDBViz
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.bulk.pyWGCNA">
    omicverse.bulk.pyWGCNA
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.__init__">
    __init__()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_soft_threshold">
    calculate_soft_threshold()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_geneTree">
    calculate_geneTree()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_dynamicMods">
    calculate_dynamicMods()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_gene_module">
    calculate_gene_module()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_module">
    get_sub_module()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_network">
    get_sub_network()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_sub_network">
    plot_sub_network()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_matrix">
    plot_matrix()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api module</h1>
<div class="doc doc-object doc-class">
<a id="omicverse.bulk.pyWGCNA"></a>
<div class="doc doc-contents first">
<p class="doc doc-class-bases">
        Bases: <code><span title="omicverse.externel.PyWGCNA.geneExp.GeneExp">GeneExp</span></code></p>
<p>A class used to do weighted gene co-expression network analysis.</p>
<p>:param name: name of the WGCNA we used to visualize data (default: 'WGCNA')
:type name: str
:param save: indicate if you want to save result of important steps in a figure directory (default: False)
:type save: bool
:param species: species of the data you use i.e mouse, human
:type species: str
:param level: which type of data you use including gene, transcript (default: gene)
:type level: str
:param outputPath: path you want to save all you figures and object (default: '', where you rau your script)
:type outputPath: str
:param anndata: if the expression data is in anndata format you should pass it through this parameter. X should be expression matrix. var is a gene information and obs is a sample information.
:type anndata: anndata
:param geneExp: expression matrix which genes are in the rows and samples are columns
:type geneExp: pandas dataframe
:param geneExpPath: path of expression matrix
:type geneExpPath: str
:param sep: separation symbol to use for reading data in geneExpPath properly
:type sep: str
:param geneInfo: dataframe that contains genes information it should have a same index as gene expression column names (gene/transcript ID)
:type geneInfo: pandas dataframe
:param sampleInfo: dataframe that contains samples information it should have a same index as gene expression index (sample ID)
:type sampleInfo: pandas dataframe
:param TPMcutoff: cut off for removing genes that expressed under this number along samples
:type TPMcutoff: int
:param cut: number to remove outlier sample (default: 'inf') By default we don't remove any sample by hierarchical clustering
:type cut: float
:param powers: different powers to test to have scale free network (default: [1:10, 11:21:2])
:type powers: list of int
:param RsquaredCut: R squaered cut to choose power for having scale free network; between 0 to 1 (default: 0.9)
:type RsquaredCut: float
:param MeanCut: mean connectivity to choose power for having scale free network (default: 100)
:type MeanCut: int
:param networkType: Type of network we can create including "unsigned", "signed" and "signed hybrid" (default: "signed hybrid")
:type networkType: str
:param TOMType: Type of topological overlap matrix(TOM) including "unsigned", "signed" (default: "signed")
:type TOMType: str
:param minModuleSize: We like large modules, so we set the minimum module size relatively high (default: 50)
:type minModuleSize: int
:param naColor: color we used to identify genes we don't find any cluster for them (default: "grey")
:type naColor: str
:param MEDissThres:  diss similarity threshold (default: 0.2)
:type MEDissThres: float
:param figureType: extension of figure (default: "pdf")
:type figureType: str
:param MEs: eigengenes
:type MEs: ndarray
:param geneExpr: gene expression object that contains raw gene expression along with gene and sample information.
:type geneExpr: geneExp class
:param datExpr: data expression data that contains preprocessed data
:type datExpr: anndata
:param dynamicMods: name of modules by clustering similar genes together
:type dynamicMods: list
:param TOM: topological overlap measure using average linkage hierarchical clustering which inputs a measure of interconnectedness
:param TOM: ndarray
:param adjacency: adjacency matrix calculating base of the type of network
:type adjacency: ndarray
:param geneTree: average hierarchical clustering of dissTOM matrix
:type geneTree: ndarray
:param power: power to have scale free network (default: 6)
:type power: int
:param sft: soft threshold table which has information for each powers
:type sft: pandas dataframe
:param datME:
:type datME: pandas dataframe
:param signedKME:(signed) eigengene-based connectivity (module membership)
:type signedKME: pandas dataframe
:param moduleTraitCor: correlation between each module and metadata
:type moduleTraitCor: pandas dataframe
:param moduleTraitPvalue: p-value of correlation between each module and metadata
:type moduleTraitPvalue: pandas dataframe</p>
<div class="doc doc-children">
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">'WGCNA'</span><span class="p">,</span> <span class="n">TPMcutoff</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">powers</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">RsquaredCut</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">MeanCut</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">networkType</span><span class="o">=</span><span class="s1">'signed hybrid'</span><span class="p">,</span> <span class="n">TOMType</span><span class="o">=</span><span class="s1">'signed'</span><span class="p">,</span> <span class="n">minModuleSize</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">naColor</span><span class="o">=</span><span class="s1">'grey'</span><span class="p">,</span> <span class="n">cut</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="s1">'inf'</span><span class="p">),</span> <span class="n">MEDissThres</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="s1">'gene'</span><span class="p">,</span> <span class="n">anndata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">geneExp</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">geneExpPath</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">','</span><span class="p">,</span> <span class="n">geneInfo</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sampleInfo</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">outputPath</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figureType</span><span class="o">=</span><span class="s1">'pdf'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.__init__" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_soft_threshold">
<code class="highlight language-python"><span class="n">calculate_soft_threshold</span><span class="p">(</span><span class="n">colorlist</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_soft_threshold" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_geneTree">
<code class="highlight language-python"><span class="n">calculate_geneTree</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_geneTree" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_dynamicMods">
<code class="highlight language-python"><span class="n">calculate_dynamicMods</span><span class="p">(</span><span class="n">kwargs_function</span><span class="o">=</span><span class="p">{</span><span class="s1">'cutreeHybrid'</span><span class="p">:</span> <span class="p">{</span><span class="s1">'deepSplit'</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">'pamRespectsDendro'</span><span class="p">:</span> <span class="kc">False</span><span class="p">}})</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_dynamicMods" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_gene_module">
<code class="highlight language-python"><span class="n">calculate_gene_module</span><span class="p">(</span><span class="n">kwargs_function</span><span class="o">=</span><span class="p">{</span><span class="s1">'cutreeHybrid'</span><span class="p">:</span> <span class="p">{</span><span class="s1">'deepSplit'</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">'pamRespectsDendro'</span><span class="p">:</span> <span class="kc">False</span><span class="p">}})</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.calculate_gene_module" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_module">
<code class="highlight language-python"><span class="n">get_sub_module</span><span class="p">(</span><span class="n">mod_list</span><span class="p">,</span> <span class="n">mod_type</span><span class="o">=</span><span class="s1">'module_color'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_module" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
<p>Get sub-module of a module</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>mod_list</code></td>
<td>
<code>list</code>
</td>
<td><p>module number</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sub_module</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>sub-module of a module</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_network">
<code class="highlight language-python"><span class="n">get_sub_network</span><span class="p">(</span><span class="n">mod_list</span><span class="p">,</span> <span class="n">mod_type</span><span class="o">=</span><span class="s1">'module_color'</span><span class="p">,</span> <span class="n">correlation_threshold</span><span class="o">=</span><span class="mf">0.95</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.get_sub_network" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
<p>Get sub-network of a module</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>mod_list</code></td>
<td>
<code>list</code>
</td>
<td><p>module number</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sub_network</code></td> <td>
<code><span title="networkx">nx</span>.<span title="networkx.Graph">Graph</span></code>
</td>
<td><p>sub-network of a module</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_sub_network">
<code class="highlight language-python"><span class="n">plot_sub_network</span><span class="p">(</span><span class="n">mod_list</span><span class="p">,</span> <span class="n">mod_type</span><span class="o">=</span><span class="s1">'module_color'</span><span class="p">,</span> <span class="n">correlation_threshold</span><span class="o">=</span><span class="mf">0.95</span><span class="p">,</span> <span class="n">plot_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">plot_gene_num</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_sub_network" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
<p>plot sub-network of a module</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>mod_list</code></td>
<td>
<code>list</code>
</td>
<td><p>module number</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>correlation_threshold</code></td>
<td>
<code>float</code>
</td>
<td><p>correlation threshold</p></td>
<td>
<code>0.95</code>
</td>
</tr>
<tr>
<td><code>plot_genes</code></td>
<td>
</td>
<td><p>genes to plot in the sub-network. If None, the hub genes will be ploted</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>plot_gene_num</code></td>
<td>
<code>int</code>
</td>
<td><p>number of genes to plot</p></td>
<td>
<code>5</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fig</code></td> <td>
</td>
<td><p>figure</p></td>
</tr>
<tr>
<td><code>ax</code></td> <td>
</td>
<td><p>axis</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_matrix">
<code class="highlight language-python"><span class="n">plot_matrix</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s1">'RdBu_r'</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="n">legene_ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">legene_bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mf">2.95</span><span class="p">),</span> <span class="n">legene_fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">color_type</span><span class="o">=</span><span class="s1">'dynamicColors'</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.externel.PyWGCNA.wgcna.pyWGCNA.plot_matrix" title="Permanent link">¶</a></h2>
<div class="doc doc-contents">
<p>plot the matrix of correlation</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>cmap</code></td>
<td>
</td>
<td><p>The color of matrix</p></td>
<td>
<code>'RdBu_r'</code>
</td>
</tr>
<tr>
<td><code>save</code></td>
<td>
<code>bool</code>
</td>
<td><p>Whether to save the figure</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The size of figure</p></td>
<td>
<code>(8, 9)</code>
</td>
</tr>
<tr>
<td><code>legene_ncol</code></td>
<td>
<code>int</code>
</td>
<td><p>The number of column of legene</p></td>
<td>
<code>2</code>
</td>
</tr>
<tr>
<td><code>legene_bbox_to_anchor</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The position of legene</p></td>
<td>
<code>(5, 2.95)</code>
</td>
</tr>
<tr>
<td><code>legene_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The size of legene</p></td>
<td>
<code>12</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ax</code></td> <td>
</td>
<td><p>The axis of figure</p></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div><div class="highlight"><pre><span></span><code>    show_root_heading: true
    show_source: true
</code></pre></div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2025, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>