
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../single/" rel="prev"/>
<link href="../utils/" rel="next"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.1.2" name="generator"/>
<title>space - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../css/mkapi-common.css" rel="stylesheet"/>
<link href="../../../css/mkapi-material.css" rel="stylesheet"/>
<link href="../../../css/tags.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.6.0/css/all.min.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.space">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              space
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_1" id="__nav_3_1_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_2" id="__nav_3_1_2_label" tabindex="0">
          Downstream
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_2">
<span class="md-nav__icon md-icon"></span>
          Downstream
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_3" id="__nav_3_1_3_label" tabindex="0">
          Deconvolution
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_3">
<span class="md-nav__icon md-icon"></span>
          Deconvolution
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_decov_bulk/">
        Bulk deconvolution with reference scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1_4" id="__nav_3_1_4_label" tabindex="0">
          Others
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_1_4">
<span class="md-nav__icon md-icon"></span>
          Others
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Alignment
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Alignment
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_alignment_1k/">
        Alignment and analysis of single-cell RNA-seq data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_alignment_velocity/">
        Alignment and RNA velocity analysis of single-cell RNA-seq data.
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        GeneModule Identified
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_noref/">
        Reference-free automated single-cell cell type annotation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_ref/">
        Reference automated single-cell cell type annotation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology/Taxonomy
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scenic/">
        Gene Regulatory Network Analysis with SCENIC
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell-cell communication using CellPhoneDBViz
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_6" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_6" id="__nav_3_2_6_label" tabindex="0">
          Velocity
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_6_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_6">
<span class="md-nav__icon md-icon"></span>
          Velocity
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-velo/t_velo/">
        Velocity Basic Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-velo/t_graphvelo/">
        Velocity Optimization
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_7" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_7" id="__nav_3_2_7_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_7_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_7">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_1" id="__nav_3_4_1_label" tabindex="0">
          Preprocess
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_1">
<span class="md-nav__icon md-icon"></span>
          Preprocess
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cellpose/">
        Cell Segmemtation (10x HD)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_2" id="__nav_3_4_2_label" tabindex="0">
          Deconvolution
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_2">
<span class="md-nav__icon md-icon"></span>
          Deconvolution
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_decov/">
        Spatial deconvolution with reference scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_flashdeconv/">
        FlashDeconv: Fast Spatial Deconvolution via Structure-Preserving Sketching
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh_new/">
        Spatial deconvolution without reference scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4_3" id="__nav_3_4_3_label" tabindex="0">
          Downstream
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_4_3">
<span class="md-nav__icon md-icon"></span>
          Downstream
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          LLM
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          LLM
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_dr/">
        Domain Research (dr): Comprehensive Report Tutorial
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_dr_web/">
        Domain Research (dr) — Live Web
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_ov_agent_pbmc3k/">
        ov.Agent with Skills: Natural‑Language Single‑Cell Analysis (PBMC3k)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_uce/">
        UCE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_scgpt/">
        scGPT
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_scfoundation/">
        scFoundation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_geneformer/">
        GeneFormer
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-llm/t_cellplm/">
        CellPLM
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_6" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_6_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_6">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../GPU_supported_registry_functions/">
        GPU Support
      </a>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_5" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../">API documentation</a>
<label for="__nav_5">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="true" aria-labelledby="__nav_5_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_5">
<span class="md-nav__icon md-icon"></span>
          API documentation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_2" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../agent/">agent</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_2">
<span class="md-nav__icon md-icon"></span>
          agent
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_3" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../alignment/">alignment</a>
<label for="__nav_5_3">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_3">
<span class="md-nav__icon md-icon"></span>
          alignment
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../alignment/kb_api/">
        kb_api
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_4" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../bulk/">bulk</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_4">
<span class="md-nav__icon md-icon"></span>
          bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_5" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../bulk2single/">bulk2single</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_5">
<span class="md-nav__icon md-icon"></span>
          bulk2single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_6" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../datasets/">datasets</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_6_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_6">
<span class="md-nav__icon md-icon"></span>
          datasets
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_7" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pl/">pl</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_7_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_7">
<span class="md-nav__icon md-icon"></span>
          pl
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_8" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pp/">pp</a>
<label for="__nav_5_8">
<span class="md-nav__icon md-icon"></span>
</label>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_8_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_8">
<span class="md-nav__icon md-icon"></span>
          pp
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_8_2" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../pp/experimental/">experimental</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_8_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_5_8_2">
<span class="md-nav__icon md-icon"></span>
          experimental
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../pp/pyg_knn_implementation/">
        pyg_knn_implementation
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_9" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../single/">single</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_9_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_9">
<span class="md-nav__icon md-icon"></span>
          single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_5_10" type="checkbox"/>
<div class="md-nav__link md-nav__link--index md-nav__link--active">
<a href="./">space</a>
</div>
<nav aria-expanded="true" aria-labelledby="__nav_5_10_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_10">
<span class="md-nav__icon md-icon"></span>
          space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_5_11" type="checkbox"/>
<div class="md-nav__link md-nav__link--index">
<a href="../utils/">utils</a>
</div>
<nav aria-expanded="false" aria-labelledby="__nav_5_11_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_5_11">
<span class="md-nav__icon md-icon"></span>
          utils
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAGATE">
    pySTAGATE
  </a>
<nav aria-label="pySTAGATE" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAGATE.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAGATE.predicted">
    predicted
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAGATE.cal_pSM">
    cal_pSM
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.clusters">
    clusters
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.merge_cluster">
    merge_cluster
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CAST">
    CAST
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram">
    Tangram
  </a>
<nav aria-label="Tangram" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram.check_tangram">
    check_tangram
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram.cell2location">
    cell2location
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram.impute">
    impute
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAligner">
    pySTAligner
  </a>
<nav aria-label="pySTAligner" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAligner.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySTAligner.predicted">
    predicted
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Cal_Spatial_Net">
    Cal_Spatial_Net
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySpaceFlow">
    pySpaceFlow
  </a>
<nav aria-label="pySpaceFlow" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySpaceFlow.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.pySpaceFlow.cal_pSM">
    cal_pSM
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.svg">
    svg
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT">
    STT
  </a>
<nav aria-label="STT" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.stage_estimate">
    stage_estimate
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.load">
    load
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.compute_pathway">
    compute_pathway
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_pathway">
    plot_pathway
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_tensor_pathway">
    plot_tensor_pathway
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_tensor">
    plot_tensor
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.construct_landscape">
    construct_landscape
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.infer_lineage">
    infer_lineage
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_landscape">
    plot_landscape
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_sankey">
    plot_sankey
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.STT.plot_top_genes">
    plot_top_genes
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellMap">
    CellMap
  </a>
<nav aria-label="CellMap" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellMap.map">
    map
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellMap.assign_coord">
    assign_coord
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc">
    CellLoc
  </a>
<nav aria-label="CellLoc" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc.loc_map">
    loc_map
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc.load_map">
    load_map
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc.save_map">
    save_map
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc.loc_prob">
    loc_prob
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.CellLoc.loc_assign">
    loc_assign
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON">
    GASTON
  </a>
<nav aria-label="GASTON" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.get_gaston_input">
    get_gaston_input
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.get_top_pearson_residuals">
    get_top_pearson_residuals
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.load_rescale">
    load_rescale
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.train">
    train
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.get_best_model">
    get_best_model
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.cal_iso_depth">
    cal_iso_depth
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_isodepth">
    plot_isodepth
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_clusters">
    plot_clusters
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_clusters_restrict">
    plot_clusters_restrict
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.restrict_spot">
    restrict_spot
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.filter_genes">
    filter_genes
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.pw_linear_fit">
    pw_linear_fit
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.bin_data">
    bin_data
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.get_restricted_adata">
    get_restricted_adata
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_gene_pwlinear">
    plot_gene_pwlinear
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_gene_raw">
    plot_gene_raw
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.GASTON.plot_gene_gastonrex">
    plot_gene_gastonrex
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.create_communication_anndata">
    create_communication_anndata
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.update_classification_from_database">
    update_classification_from_database
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution">
    Deconvolution
  </a>
<nav aria-label="Deconvolution" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.preprocess_sc">
    preprocess_sc
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.preprocess_sp">
    preprocess_sp
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.deconvolution">
    deconvolution
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.impute">
    impute
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.load_cell2location_model">
    load_cell2location_model
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.cell2location_inference">
    cell2location_inference
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.load_tangram_model">
    load_tangram_model
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Deconvolution.tangram_inference">
    tangram_inference
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.calculate_gene_signature">
    calculate_gene_signature
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.crop_space_visium">
    crop_space_visium
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.rotate_space_visium">
    rotate_space_visium
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.find_image_offset_phase_correlation_array_input">
    find_image_offset_phase_correlation_array_input
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.find_image_offset_phase_correlation_torch">
    find_image_offset_phase_correlation_torch
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.map_spatial_auto">
    map_spatial_auto
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.map_spatial_manual">
    map_spatial_manual
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.read_visium_10x">
    read_visium_10x
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.visium_10x_hd_cellpose_he">
    visium_10x_hd_cellpose_he
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.visium_10x_hd_cellpose_expand">
    visium_10x_hd_cellpose_expand
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.visium_10x_hd_cellpose_gex">
    visium_10x_hd_cellpose_gex
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.salvage_secondary_labels">
    salvage_secondary_labels
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.bin2cell">
    bin2cell
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1 style="display: none;">omicverse.space<a class="headerlink" href="#omicverse.space" title="Permanent link">¶</a></h1><p class="mkapi-object mkapi-page-source" id="omicverse.space">
<span class="mkapi-object-link">
<a href="../../../src/omicverse/space/#omicverse.space" title="Go to source">source</a>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">package</span>
<span class="mkapi-object-name"><a href="./#omicverse.space" title="omicverse.space">omicverse.space</a></span>
</p>
<div class="mkapi-document">
<p>Spatial transcriptomics analysis utilities.</p>
<p>This module provides comprehensive tools for analyzing spatial transcriptomics data
including spatial clustering, integration, trajectory analysis, and visualization.
Supported platforms include 10x Visium, Slide-seq, MERFISH, seqFISH, and others.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle spatial clustering and segmentation">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Spatial clustering and segmentation</span>
</p>
<div class="mkapi-section-content">
<div class="highlight"><pre><span></span><code>pySTAGATE: Graph attention networks for spatial domains
clusters: Spatial domain identification
merge_cluster: Merge spatial clusters
CAST: Cellular spatial organization analysis
</code></pre></div>
<p>Spatial integration and alignment:
    pySTAligner: Multi-sample spatial alignment
    Cal_Spatial_Net: Spatial neighborhood networks
    pySpaceFlow: Spatial flow analysis</p>
<p>Spatially variable genes:
    svg: Identify spatially variable genes</p>
<p>Spatial trajectory analysis:
    STT: Spatial transition tensor analysis
    CellMap: Cell mapping in spatial context
    CellLoc: Cell localization analysis</p>
<p>Spatial deconvolution:
    Tangram: Map single-cell data to spatial coordinates
    GASTON: Spatial deconvolution and analysis</p>
<p>Utility functions:
    Various spatial analysis tools and helper functions</p>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="c1"># Load spatial data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="s1">'path/to/visium/data'</span><span class="p">)</span>

<span class="c1"># Identify spatial domains</span>
<span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">pySTAGATE</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_domains</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>

<span class="c1"># Find spatially variable genes</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">svg</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">'prost'</span><span class="p">,</span> <span class="n">n_svgs</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>

<span class="c1"># Spatial trajectory analysis</span>
<span class="n">stt</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">STT</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">spatial_loc</span><span class="o">=</span><span class="s1">'spatial'</span><span class="p">)</span>
<span class="n">stt</span><span class="o">.</span><span class="n">stage_estimate</span><span class="p">()</span>
<span class="n">stt</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">n_states</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Spatial integration</span>
<span class="n">aligner</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">pySTAligner</span><span class="p">(</span><span class="n">adata_list</span><span class="p">)</span>
<span class="n">aligner</span><span class="o">.</span><span class="n">spatial_alignment</span><span class="p">()</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle classes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Classes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAGATE" title="omicverse.space.pySTAGATE">pySTAGATE</a></span>
<span class="mkapi-dash">—</span>
A class representing the PyTorch implementation of STAGATE (Spatial Transcriptomics Analysis using Graph Attention autoEncoder).</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a></span>
<span class="mkapi-dash">—</span>
Tangram spatial deconvolution class for cell type mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAligner" title="omicverse.space.pySTAligner">pySTAligner</a></span>
<span class="mkapi-dash">—</span>
STAligner for spatial transcriptomics data integration.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySpaceFlow" title="omicverse.space.pySpaceFlow">pySpaceFlow</a></span>
<span class="mkapi-dash">—</span>
SpaceFlow spatial flow analysis class.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a></span>
<span class="mkapi-dash">—</span>
Spatial Transition Tensor (STT) analysis class.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellMap" title="omicverse.space.CellMap">CellMap</a></span>
<span class="mkapi-dash">—</span>
SpatRio CellMap class for mapping single cells to spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a></span>
<span class="mkapi-dash">—</span>
SpatRio CellLoc class for probabilistic cell localization.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a></span>
<span class="mkapi-dash">—</span>
GASTON spatial depth estimation and clustering.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a></span>
<span class="mkapi-dash">—</span>
Initialize Deconvolution object.
Parameters</p>
<hr/>
<p>adata_sp : AnnData
    Spatial transcriptomics data.
adata_sc : AnnData, optional
    Single-cell RNA-seq data, by default None.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle functions">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Functions</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.clusters" title="omicverse.space.clusters">clusters</a></span>
<span class="mkapi-dash">—</span>
Perform clustering analysis on spatial transcriptomics data using multiple methods.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.merge_cluster" title="omicverse.space.merge_cluster">merge_cluster</a></span>
<span class="mkapi-dash">—</span>
Merge clusters based on hierarchical clustering of their representation.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CAST" title="omicverse.space.CAST">CAST</a></span>
<span class="mkapi-dash">—</span>
CAST (Cell Annotation for Spatial Transcriptomics) embedding for multiple spatial samples.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Cal_Spatial_Net" title="omicverse.space.Cal_Spatial_Net">Cal_Spatial_Net</a></span>
<span class="mkapi-dash">—</span>
Construct spatial neighbor networks for spatial integration.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.svg" title="omicverse.space.svg">svg</a></span>
<span class="mkapi-dash">—</span>
Identify spatially variable genes using multiple methods.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.create_communication_anndata" title="omicverse.space.create_communication_anndata">create_communication_anndata</a></span>
<span class="mkapi-dash">—</span>
Create AnnData object with communication results in CellChat format</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.update_classification_from_database" title="omicverse.space.update_classification_from_database">update_classification_from_database</a></span>
<span class="mkapi-dash">—</span>
Update classification in comm_adata using database info from original adata</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.calculate_gene_signature" title="omicverse.space.calculate_gene_signature">calculate_gene_signature</a></span>
<span class="mkapi-dash">—</span>
Calculate the gene signature.
Parameters</p>
<hr/>
<p>adata_sc : AnnData
    Single-cell RNA-seq data.
clustertype : str
    Cell type key in scRNA-seq data.
rank : bool, optional
    Whether to rank the genes, by default True.
key : str, optional
    Key in scRNA-seq data, by default 'rank_genes_groups'.
foldchange : int, optional
    Fold change, by default 2.
topgenenumber : int, optional
    Number of top genes, by default 20.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.crop_space_visium" title="omicverse.space.crop_space_visium">crop_space_visium</a></span>
<span class="mkapi-dash">—</span>
Crop Visium spatial data to a specific region of interest.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.rotate_space_visium" title="omicverse.space.rotate_space_visium">rotate_space_visium</a></span>
<span class="mkapi-dash">—</span>
Rotate Visium spatial data image and coordinates by a specified angle.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.find_image_offset_phase_correlation_array_input" title="omicverse.space.find_image_offset_phase_correlation_array_input">find_image_offset_phase_correlation_array_input</a></span>
<span class="mkapi-dash">—</span>
Calculate image offset using phase correlation method.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.find_image_offset_phase_correlation_torch" title="omicverse.space.find_image_offset_phase_correlation_torch">find_image_offset_phase_correlation_torch</a></span>
<span class="mkapi-dash">—</span>
Calculate image offset using phase correlation method with PyTorch tensors.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.map_spatial_auto" title="omicverse.space.map_spatial_auto">map_spatial_auto</a></span>
<span class="mkapi-dash">—</span>
Automatically map and align spatial transcriptomics data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.map_spatial_manual" title="omicverse.space.map_spatial_manual">map_spatial_manual</a></span>
<span class="mkapi-dash">—</span>
Manually adjust spatial transcriptomics data alignment.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.read_visium_10x" title="omicverse.space.read_visium_10x">read_visium_10x</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_he" title="omicverse.space.visium_10x_hd_cellpose_he">visium_10x_hd_cellpose_he</a></span>
<span class="mkapi-dash">—</span>
Convert Visium 10x data to cell-level data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_expand" title="omicverse.space.visium_10x_hd_cellpose_expand">visium_10x_hd_cellpose_expand</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_gex" title="omicverse.space.visium_10x_hd_cellpose_gex">visium_10x_hd_cellpose_gex</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.salvage_secondary_labels" title="omicverse.space.salvage_secondary_labels">salvage_secondary_labels</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.bin2cell" title="omicverse.space.bin2cell">bin2cell</a></span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAGATE">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.pySTAGATE" title="omicverse.space.pySTAGATE">pySTAGATE</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_batch_x</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_batch_y</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_key</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">['X', 'Y']</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_size</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rad_cutoff</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">200</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_epoch</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lr</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.001</span><span class="mkapi-comma">, </span><span class="mkapi-arg">weight_decay</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.0001</span><span class="mkapi-comma">, </span><span class="mkapi-arg">hidden_dims</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[512, 30]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">device</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'cuda:0'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>A class representing the PyTorch implementation of STAGATE (Spatial Transcriptomics Analysis using Graph Attention autoEncoder).</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData object
Annotated data matrix containing spatial transcriptomics data.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_batch_x</span>
<span class="mkapi-dash">—</span>
int
Number of batches in x direction for spatial partitioning.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_batch_y</span>
<span class="mkapi-dash">—</span>
int
Number of batches in y direction for spatial partitioning.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatial_key</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
list, optional (default=['X','Y'])
List of keys in adata.obs containing spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">batch_size</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1)
Size of batches for training.</p>
</li>
<li>
<p><span class="mkapi-item-name">rad_cutoff</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=200)
Radius cutoff for spatial network construction.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_epoch</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1000)
Number of epochs for training.</p>
</li>
<li>
<p><span class="mkapi-item-name">lr</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.001)
Learning rate for optimization.</p>
</li>
<li>
<p><span class="mkapi-item-name">weight_decay</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=1e-4)
Weight decay (L2 penalty) for optimization.</p>
</li>
<li>
<p><span class="mkapi-item-name">hidden_dims</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
list, optional (default=[512, 30])
List of hidden dimensions for the neural network layers.</p>
</li>
<li>
<p><span class="mkapi-item-name">device</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='cuda:0')
Device to run the model on ('cuda:0' or 'cpu').</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">device</span>
<span class="mkapi-dash">—</span>
torch.device
Device where the model is running.</p>
</li>
<li>
<p><span class="mkapi-item-name">loader</span>
<span class="mkapi-dash">—</span>
DataLoader
PyTorch DataLoader for batch processing.</p>
</li>
<li>
<p><span class="mkapi-item-name">model</span>
<span class="mkapi-dash">—</span>
STAGATE
The STAGATE model instance.</p>
</li>
<li>
<p><span class="mkapi-item-name">optimizer</span>
<span class="mkapi-dash">—</span>
torch.optim.Adam
Adam optimizer for model training.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Input annotated data matrix.</p>
</li>
<li>
<p><span class="mkapi-item-name">data</span>
<span class="mkapi-dash">—</span>
torch_geometric.data.Data
PyTorch geometric data object.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial_data.h5ad'</span><span class="p">)</span>
<span class="n">stagate</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">pySTAGATE</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">num_batch_x</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">num_batch_y</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">stagate</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
<span class="n">stagate</span><span class="o">.</span><span class="n">predicted</span><span class="p">()</span>
</code></pre></div>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<p>The STAGATE model is designed for analyzing spatial transcriptomics data by incorporating
spatial information through a graph attention autoencoder architecture.</p>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAGATE.train" title="omicverse.space.pySTAGATE.train">train</a></span>
<span class="mkapi-dash">—</span>
Train the STAGATE model using the configured parameters.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAGATE.predicted" title="omicverse.space.pySTAGATE.predicted">predicted</a></span>
<span class="mkapi-dash">—</span>
Generate STAGATE representations and reconstruction values for all cells.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAGATE.cal_pSM" title="omicverse.space.pySTAGATE.cal_pSM">cal_pSM</a></span>
<span class="mkapi-dash">—</span>
Calculate the pseudo-spatial map using diffusion pseudotime (DPT) algorithm.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAGATE.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySTAGATE" title="omicverse.space.pySTAGATE">pySTAGATE</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySTAGATE.train" title="omicverse.space.pySTAGATE.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Train the STAGATE model using the configured parameters.</p>
<p>This method performs training of the STAGATE model for the specified number of epochs.
For each epoch, it processes batches of data, computes the loss using mean squared error,
and updates the model parameters through backpropagation.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">None</span></p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>The training progress is displayed using a progress bar.</li>
<li>Gradient clipping is applied with a maximum norm of 5.</li>
<li>The model is automatically moved to the specified device (GPU/CPU).</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAGATE.predicted">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySTAGATE" title="omicverse.space.pySTAGATE">pySTAGATE</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySTAGATE.predicted" title="omicverse.space.pySTAGATE.predicted">predicted</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Generate STAGATE representations and reconstruction values for all cells.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this method runs the trained model in evaluation mode to generate">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This method runs the trained model in evaluation mode to generate</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>STAGATE embeddings (stored in adata.obsm['STAGATE'])</li>
<li>Reconstructed expression values (stored in adata.layers['STAGATE_ReX'])</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">None</span></p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Negative values in reconstructed expression are set to 0</li>
<li>Results are stored directly in the AnnData object:<ul>
<li>STAGATE embeddings: adata.obsm['STAGATE']</li>
<li>Reconstructed expression: adata.layers['STAGATE_ReX']</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAGATE.cal_pSM">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySTAGATE" title="omicverse.space.pySTAGATE">pySTAGATE</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySTAGATE.cal_pSM" title="omicverse.space.pySTAGATE.cal_pSM">cal_pSM</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">n_neighbors</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">resolution</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_cell_for_subsampling</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">psm_key</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'pSM_STAGATE'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate the pseudo-spatial map using diffusion pseudotime (DPT) algorithm.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this method computes a pseudo-spatial map by">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This method computes a pseudo-spatial map by</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Constructing a kNN graph using STAGATE embeddings</li>
<li>Performing UMAP and Leiden clustering</li>
<li>Computing diffusion pseudotime</li>
<li>Storing results in the AnnData object</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">n_neighbors</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=20)
Number of neighbors for the kNN graph construction.</p>
</li>
<li>
<p><span class="mkapi-item-name">resolution</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
float, optional (default=1)
Resolution parameter for Leiden clustering.</p>
</li>
<li>
<p><span class="mkapi-item-name">max_cell_for_subsampling</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=5000)
Maximum number of cells to use for distance calculation.
If exceeded, cells will be subsampled.</p>
</li>
<li>
<p><span class="mkapi-item-name">psm_key</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='pSM_STAGATE')
Key under which to store the pseudo-spatial map in adata.obs.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>numpy.ndarray
    Array containing the computed pseudo-spatial map values.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>If number of cells exceeds max_cell_for_subsampling, random subsampling is performed</li>
<li>The root cell for pseudotime calculation is chosen as the cell with maximum
  total distance to all other cells</li>
<li>Results are stored in adata.obs[psm_key]</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.clusters">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.clusters" title="omicverse.space.clusters">clusters</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">methods</span><span class="mkapi-comma">, </span><span class="mkapi-arg">methods_kwargs</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'spatial'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lognorm</span><span class="mkapi-equal">=</span><span class="mkapi-default">50 * 10000.0</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Perform clustering analysis on spatial transcriptomics data using multiple methods.</p>
<p>This function supports multiple clustering methods including STAGATE, GraphST, CAST, and BINARY.
Each method processes the spatial data differently and stores its results in the AnnData object.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing spatial transcriptomics data.</p>
</li>
<li>
<p><span class="mkapi-item-name">methods</span>
<span class="mkapi-dash">—</span>
list
List of methods to use for clustering. Supported methods are:</p>
<ul>
<li>'STAGATE': Graph attention autoencoder-based clustering</li>
<li>'GraphST': Graph-based spatial transcriptomics clustering</li>
<li>'CAST': Clustering And Spatial Transcriptomics</li>
<li>'BINARY': Binary-based spatial clustering</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">methods_kwargs</span>
<span class="mkapi-dash">—</span>
dict
Dictionary containing method-specific parameters. Each key should correspond
to a method name and contain a dictionary of parameters for that method.</p>
</li>
<li>
<p><span class="mkapi-item-name">batch_key</span>
<span class="mkapi-dash">—</span>
str, optional (default=None)
Key in adata.obs for batch information. If None, all cells are treated as one batch.</p>
</li>
<li>
<p><span class="mkapi-item-name">lognorm</span>
<span class="mkapi-dash">—</span>
float, optional (default=50*1e4)
Normalization factor for log transformation when recovering counts.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">AnnData</span>
<span class="mkapi-dash">—</span>
The input AnnData object with added clustering results in various slots:</p>
<ul>
<li>STAGATE: Results in adata.obsm['STAGATE'] and adata.layers['STAGATE_ReX']</li>
<li>GraphST: Results in adata.obsm['GraphST_embedding']</li>
<li>CAST: Results in adata.obsm['X_cast']</li>
<li>BINARY: Results in adata.obsm['BINARY']</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>For STAGATE: If n_top_genes is specified, highly variable genes are selected</li>
<li>For GraphST: Requires counts matrix in adata.layers['counts']</li>
<li>For CAST: Requires spatial coordinates in adata.obsm['spatial']</li>
<li>For BINARY: Can handle both raw counts and normalized data</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial_data.h5ad'</span><span class="p">)</span>
<span class="n">methods</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'STAGATE'</span><span class="p">,</span> <span class="s1">'GraphST'</span><span class="p">]</span>
<span class="n">methods_kwargs</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">'STAGATE'</span><span class="p">:</span> <span class="p">{</span><span class="s1">'num_batch_x'</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="s1">'num_batch_y'</span><span class="p">:</span> <span class="mi">2</span><span class="p">},</span>
    <span class="s1">'GraphST'</span><span class="p">:</span> <span class="p">{</span><span class="s1">'device'</span><span class="p">:</span> <span class="s1">'cuda:0'</span><span class="p">,</span> <span class="s1">'n_pcs'</span><span class="p">:</span> <span class="mi">30</span><span class="p">}</span>
<span class="p">}</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">clusters</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">methods</span><span class="p">,</span> <span class="n">methods_kwargs</span><span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.merge_cluster">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.merge_cluster" title="omicverse.space.merge_cluster">merge_cluster</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">groupby</span><span class="mkapi-equal">=</span><span class="mkapi-default">'mclust'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_rep</span><span class="mkapi-equal">=</span><span class="mkapi-default">'STAGATE'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">threshold</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.05</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plot</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">start_idx</span><span class="mkapi-equal">=</span><span class="mkapi-default">0</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Merge clusters based on hierarchical clustering of their representation.</p>
<p>This function performs hierarchical clustering on existing clusters and merges them
based on a distance threshold. It can optionally visualize the dendrogram showing
the merging process.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing cluster information.</p>
</li>
<li>
<p><span class="mkapi-item-name">groupby</span>
<span class="mkapi-dash">—</span>
str, optional (default='mclust')
Key in adata.obs containing the cluster labels to be merged.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep</span>
<span class="mkapi-dash">—</span>
str, optional (default='STAGATE')
Key in adata.obsm to use for calculating distances between clusters.</p>
</li>
<li>
<p><span class="mkapi-item-name">threshold</span>
<span class="mkapi-dash">—</span>
float, optional (default=0.05)
Distance threshold for merging clusters. Lower values result in more merging.</p>
</li>
<li>
<p><span class="mkapi-item-name">plot</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to plot the dendrogram with the merging threshold line.</p>
</li>
<li>
<p><span class="mkapi-item-name">start_idx</span>
<span class="mkapi-dash">—</span>
int, optional (default=0)
Starting index for cluster numbering in the output.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to scanpy.pl.dendrogram().</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>dict
    Dictionary mapping original cluster labels to merged cluster labels.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>The function uses scipy's hierarchical clustering implementation</li>
<li>Merged clusters are stored in adata.obs[f'{groupby}_tree']</li>
<li>The dendrogram is stored in adata.uns[f'dendrogram_{groupby}']</li>
<li>Cluster labels in the output are prefixed with 'c'</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'clustered_data.h5ad'</span><span class="p">)</span>
<span class="c1"># Merge clusters using STAGATE representation</span>
<span class="n">cluster_map</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">merge_cluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> 
                                     <span class="n">groupby</span><span class="o">=</span><span class="s1">'leiden'</span><span class="p">,</span>
                                     <span class="n">use_rep</span><span class="o">=</span><span class="s1">'STAGATE'</span><span class="p">,</span>
                                     <span class="n">threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="c1"># Access merged clusters</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'leiden_tree'</span><span class="p">])</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CAST">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.CAST" title="omicverse.space.CAST">CAST</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sample_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">basis</span><span class="mkapi-equal">=</span><span class="mkapi-default">'spatial'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">layer</span><span class="mkapi-equal">=</span><span class="mkapi-default">'norm_1e4'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">output_path</span><span class="mkapi-equal">=</span><span class="mkapi-default">'output/CAST_Mark'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gpu_t</span><span class="mkapi-equal">=</span><span class="mkapi-default">0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">device</span><span class="mkapi-equal">=</span><span class="mkapi-default">'cuda:0'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>CAST (Cell Annotation for Spatial Transcriptomics) embedding for multiple spatial samples.</p>
<p>This function implements the CAST algorithm to learn unified embeddings across
multiple spatial transcriptomics samples, enabling joint analysis and integration
of spatial data from different sources.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing multiple spatial samples.
Must contain spatial coordinates in adata.obsm[basis].</p>
</li>
<li>
<p><span class="mkapi-item-name">sample_key</span>
<span class="mkapi-dash">—</span>
str, optional (default=None)
Column name in adata.obs containing sample/batch information.
Used to identify different samples for integration.</p>
</li>
<li>
<p><span class="mkapi-item-name">basis</span>
<span class="mkapi-dash">—</span>
str, optional (default='spatial')
Key in adata.obsm containing spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">layer</span>
<span class="mkapi-dash">—</span>
str, optional (default='norm_1e4')
Layer in adata.layers containing normalized expression data.
Should contain values suitable for CAST analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name">output_path</span>
<span class="mkapi-dash">—</span>
str, optional (default='output/CAST_Mark')
Directory path for saving CAST intermediate outputs and results.</p>
</li>
<li>
<p><span class="mkapi-item-name">gpu_t</span>
<span class="mkapi-dash">—</span>
int, optional (default=0)
GPU device index to use for computation.</p>
</li>
<li>
<p><span class="mkapi-item-name">device</span>
<span class="mkapi-dash">—</span>
str, optional (default='cuda:0')
PyTorch device specification ('cuda:0', 'cpu', etc.).</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to CAST_MARK function:</p>
<ul>
<li>learning_rate: float</li>
<li>n_epochs: int</li>
<li>batch_size: int</li>
<li>etc.</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">AnnData</span>
<span class="mkapi-dash">—</span>
Input AnnData object updated with CAST results:</p>
<ul>
<li>adata.obsm['X_cast']: CAST embeddings matrix (n_cells × 512)</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Requires normalized expression data in specified layer</li>
<li>GPU acceleration is enabled by default</li>
<li>Creates output directory if it doesn't exist</li>
<li>Progress is shown with tqdm progress bars</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load multiple samples</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial_samples.h5ad'</span><span class="p">)</span>
<span class="c1"># Run CAST integration</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">CAST</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">sample_key</span><span class="o">=</span><span class="s1">'sample_id'</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s1">'spatial'</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s1">'normalized_counts'</span>
<span class="p">)</span>
<span class="c1"># Access CAST embeddings</span>
<span class="n">cast_embeddings</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">'X_cast'</span><span class="p">]</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Tangram">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_sc</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_sp</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-comma">, </span><span class="mkapi-arg">clusters</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">''</span><span class="mkapi-comma">, </span><span class="mkapi-arg">marker_size</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_to_lowercase</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">False</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Tangram spatial deconvolution class for cell type mapping.</p>
<p>Tangram is a method for integrating single-cell RNA sequencing (scRNA-seq) data
with spatial transcriptomics data. It enables:
1. Mapping cell types from scRNA-seq to spatial locations
2. Deconvolving cell type proportions in spatial spots
3. Imputing gene expression in spatial data
4. Analyzing spatial organization of cell types</p>
<p>Initialize Tangram spatial deconvolution object.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle the method works by">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">The method works by</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Identifying marker genes for each cell type</li>
<li>Training a mapping model using these markers</li>
<li>Projecting cell type annotations to spatial coordinates</li>
<li>Optionally imputing full gene expression profiles</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span>
<span class="mkapi-dash">—</span>
AnnData
Single-cell RNA-seq data with:</p>
<ul>
<li>Gene expression matrix in X</li>
<li>Cell type annotations in obs[clusters]</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span>
<span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data with:</p>
<ul>
<li>Gene expression matrix in X</li>
<li>Spatial coordinates in obsm['spatial']</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">clusters</span>
<span class="mkapi-dash">—</span>
str
Column name in adata_sc.obs containing cell type labels</p>
</li>
<li>
<p><span class="mkapi-item-name">markers</span>
<span class="mkapi-dash">—</span>
list
Selected marker genes used for mapping</p>
</li>
<li>
<p><span class="mkapi-item-name">ad_map</span>
<span class="mkapi-dash">—</span>
AnnData
Mapping results after training</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">"sc_data.h5ad"</span><span class="p">)</span>
<span class="n">adata_sp</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="s2">"spatial_data"</span><span class="p">)</span>
<span class="c1"># Initialize Tangram</span>
<span class="n">tangram</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">Tangram</span><span class="p">(</span>
    <span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span><span class="p">,</span>
    <span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span><span class="p">,</span>
    <span class="n">clusters</span><span class="o">=</span><span class="s1">'cell_type'</span>
<span class="p">)</span>
<span class="c1"># Train model</span>
<span class="n">tangram</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">'clusters'</span><span class="p">,</span> <span class="n">num_epochs</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="c1"># Project cell types</span>
<span class="n">adata_spatial</span> <span class="o">=</span> <span class="n">tangram</span><span class="o">.</span><span class="n">cell2location</span><span class="p">()</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this method sets up the tangram analysis by">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This method sets up the Tangram analysis by</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Checking package installation</li>
<li>Processing input data</li>
<li>Identifying marker genes</li>
<li>Preparing data for mapping</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData
Single-cell RNA-seq data containing:</p>
<ul>
<li>Normalized gene expression matrix</li>
<li>Cell type annotations in obs[clusters]</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data containing:</p>
<ul>
<li>Normalized gene expression matrix</li>
<li>Spatial coordinates in obsm['spatial']</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">clusters</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='')
Column name in adata_sc.obs containing cell type annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">marker_size</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=100)
Number of top marker genes to select per cell type.
More markers can improve accuracy but increase computation time.</p>
</li>
<li>
<p><span class="mkapi-item-name">gene_to_lowercase</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=False)
Whether to convert gene names to lowercase for matching between
datasets. Useful when gene naming conventions differ.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Automatically filters genes present in at least one cell</li>
<li>Identifies marker genes using scanpy's rank_genes_groups</li>
<li>Prepares data structures for Tangram mapping</li>
<li>Adds reference annotation to both AnnData objects</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Tangram.check_tangram" title="omicverse.space.Tangram.check_tangram">check_tangram</a></span>
<span class="mkapi-dash">—</span>
Check if Tangram package is installed.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Tangram.train" title="omicverse.space.Tangram.train">train</a></span>
<span class="mkapi-dash">—</span>
Train the Tangram spatial mapping model.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Tangram.cell2location" title="omicverse.space.Tangram.cell2location">cell2location</a></span>
<span class="mkapi-dash">—</span>
Project cell type annotations to spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Tangram.impute" title="omicverse.space.Tangram.impute">impute</a></span>
<span class="mkapi-dash">—</span>
Impute gene expression in spatial data using trained model.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Tangram.check_tangram">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Tangram.check_tangram" title="omicverse.space.Tangram.check_tangram">check_tangram</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Check if Tangram package is installed.</p>
<p>This method verifies that the Tangram package is available and prints
its version number. If not installed, it raises an informative error.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">ImportError</span>
<span class="mkapi-dash">—</span>
If Tangram package is not installed, with instructions
for installation.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Sets global tg_install flag when successful</li>
<li>Required before any Tangram operations</li>
<li>Suggests pip installation command if missing</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Tangram.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Tangram.train" title="omicverse.space.Tangram.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">mode</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'clusters'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_epochs</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">500</span><span class="mkapi-comma">, </span><span class="mkapi-arg">device</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'cuda:0'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">None</span></span>
</p>
<div class="mkapi-document">
<p>Train the Tangram spatial mapping model.</p>
<p>This method trains a model to map cells or clusters from scRNA-seq data
to spatial locations. It optimizes the mapping to preserve both gene
expression patterns and spatial structure.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">mode</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default="clusters")
Mapping mode:</p>
<ul>
<li>"clusters": Map cell type proportions (faster)</li>
<li>"cells": Map individual cells (more detailed)</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">num_epochs</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=500)
Number of training epochs. More epochs may improve results
but increase training time.</p>
</li>
<li>
<p><span class="mkapi-item-name">device</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default="cuda:0")
Computing device to use:</p>
<ul>
<li>"cuda:0" (or other GPU index) for GPU acceleration</li>
<li>"cpu" for CPU computation</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Any
Additional arguments passed to tangram.map_cells_to_space:</p>
<ul>
<li>density_prior: Prior on spatial density</li>
<li>lambda_d: Density regularization strength</li>
<li>lambda_g1: Gene-expression regularization</li>
<li>lambda_g2: Spatial regularization</li>
<li>lambda_r: Entropy regularization</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Automatically stores mapping in self.ad_map</li>
<li>Projects cell type annotations to spatial data</li>
<li>Adds reference annotation to both AnnData objects</li>
<li>Progress is shown during training</li>
<li>GPU acceleration recommended for large datasets</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Tangram.cell2location">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Tangram.cell2location" title="omicverse.space.Tangram.cell2location">cell2location</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">annotation_list</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Project cell type annotations to spatial coordinates.</p>
<p>This method creates a visualization-ready AnnData object containing the
predicted cell type proportions for each spatial location.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">annotation_list</span>
<span class="mkapi-dash">—</span>
list, optional (default=None)
List of cell types to include in the projection.
If None, uses all cell types from training data.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
AnnData
    Modified spatial data containing:
    - Original spatial data
    - Cell type proportions in obsm['tangram_ct_pred']
    - Normalized proportions in obs for each cell type</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Automatically normalizes cell type proportions</li>
<li>Clips extreme values for better visualization</li>
<li>Adds reference annotation to both AnnData objects</li>
<li>Results can be directly used for spatial plotting</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Tangram.impute">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Tangram" title="omicverse.space.Tangram">Tangram</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Tangram.impute" title="omicverse.space.Tangram.impute">impute</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">ad_map</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ad_sc</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span><span class="mkapi-equal"> = </span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span></span>
</p>
<div class="mkapi-document">
<p>Impute gene expression in spatial data using trained model.</p>
<p>This method uses the trained mapping to predict the expression of all genes
in the spatial locations, including genes not used in the mapping.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">ad_map</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData, optional (default=None)
Mapping result from train(). If None, uses self.ad_map.</p>
</li>
<li>
<p><span class="mkapi-item-name">ad_sc</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData, optional (default=None)
Single-cell reference data. If None, uses self.adata_sc.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Any
Additional arguments passed to tangram.project_genes:</p>
<ul>
<li>scale: Whether to scale imputed values</li>
<li>filter_genes: Whether to filter genes before imputation</li>
<li>filter_threshold: Expression threshold for filtering</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="anndata._core.anndata.AnnData">AnnData</span></span> <span class="mkapi-dash">—</span>
AnnData
    Spatial data with imputed gene expression for all genes
    in the single-cell reference data.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Uses mapping weights to transfer expression</li>
<li>Can impute genes not used in original mapping</li>
<li>Useful for analyzing spatial patterns of any gene</li>
<li>Computationally intensive for large gene sets</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAligner">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.pySTAligner" title="omicverse.space.pySTAligner">pySTAligner</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">hidden_dims</span><span class="mkapi-colon">: </span><span class="mkapi-ann">list</span><span class="mkapi-equal"> = </span><span class="mkapi-default">[512, 30]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_epochs</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lr</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.001</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_key</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'batch_name'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">key_added</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'STAligner'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gradient_clipping</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">weight_decay</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.0001</span><span class="mkapi-comma">, </span><span class="mkapi-arg">margin</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">verbose</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">random_seed</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">666</span><span class="mkapi-comma">, </span><span class="mkapi-arg">iter_comb</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">knn_neigh</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">Batch_list</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">device</span><span class="mkapi-equal">=</span><span class="mkapi-default"><span class="mkapi-tooltip" title="torch">torch</span>.<span class="mkapi-tooltip" title="torch.device">device</span>('cuda:0' if <span class="mkapi-tooltip" title="torch">torch</span>.<span class="mkapi-tooltip" title="torch.cuda">cuda</span>.<span class="mkapi-tooltip" title="torch.cuda.is_available">is_available</span>() else 'cpu')</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>STAligner for spatial transcriptomics data integration.</p>
<p>STAligner is a deep learning method for integrating spatial transcriptomics
data across different experimental conditions, technologies, and developmental
stages. It combines graph neural networks with mutual nearest neighbors to
preserve both transcriptional and spatial relationships during integration.</p>
<p>Initialize STAligner spatial integration model.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle the method works by">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">The method works by</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Constructing spatial neighborhood graphs</li>
<li>Learning batch-invariant embeddings</li>
<li>Aligning similar regions across batches</li>
<li>Preserving spatial organization</li>
<li>Enabling cross-condition comparison</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Combined data containing all batches</p>
</li>
<li>
<p><span class="mkapi-item-name">model</span>
<span class="mkapi-dash">—</span>
STAligner
Neural network model for integration</p>
</li>
<li>
<p><span class="mkapi-item-name">loader</span>
<span class="mkapi-dash">—</span>
DataLoader
PyTorch geometric data loader</p>
</li>
<li>
<p><span class="mkapi-item-name">device</span>
<span class="mkapi-dash">—</span>
torch.device
Computing device (GPU/CPU)</p>
</li>
<li>
<p><span class="mkapi-item-name">optimizer</span>
<span class="mkapi-dash">—</span>
torch.optim.Optimizer
Adam optimizer for training</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data</span>
<span class="n">adata1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">adata2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Construct spatial networks</span>
<span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">Cal_Spatial_Net</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">rad_cutoff</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">Cal_Spatial_Net</span><span class="p">(</span><span class="n">adata2</span><span class="p">,</span> <span class="n">rad_cutoff</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="c1"># Combine data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata1</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">adata2</span><span class="p">)</span>
<span class="c1"># Initialize STAligner</span>
<span class="n">staligner</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">pySTAligner</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s1">'batch'</span><span class="p">,</span>
    <span class="n">Batch_list</span><span class="o">=</span><span class="p">[</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">]</span>
<span class="p">)</span>
<span class="c1"># Train model</span>
<span class="n">staligner</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
<span class="c1"># Get integrated embeddings</span>
<span class="n">embeddings</span> <span class="o">=</span> <span class="n">staligner</span><span class="o">.</span><span class="n">predicted</span><span class="p">()</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this method sets up the staligner model by">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This method sets up the STAligner model by</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Processing input data</li>
<li>Constructing graph neural networks</li>
<li>Initializing optimization parameters</li>
<li>Preparing batch alignment strategy</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Combined data containing all batches to integrate.
Must have batch information in obs[batch_key].</p>
</li>
<li>
<p><span class="mkapi-item-name">hidden_dims</span> : 
<span class="mkapi-item-type">list</span> <span class="mkapi-dash">—</span>
list, optional (default=[512, 30])
Dimensions of hidden layers in the neural network.
The final dimension determines embedding size.</p>
</li>
<li>
<p><span class="mkapi-item-name">n_epochs</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1000)
Number of training epochs.
More epochs may improve integration but take longer.</p>
</li>
<li>
<p><span class="mkapi-item-name">lr</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.001)
Learning rate for Adam optimizer.
Adjust if training is unstable.</p>
</li>
<li>
<p><span class="mkapi-item-name">batch_key</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='batch_name')
Column in adata.obs containing batch information.</p>
</li>
<li>
<p><span class="mkapi-item-name">key_added</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='STAligner')
Key for storing embeddings in adata.obsm.</p>
</li>
<li>
<p><span class="mkapi-item-name">gradient_clipping</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=5)
Maximum gradient norm for stability.</p>
</li>
<li>
<p><span class="mkapi-item-name">weight_decay</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.0001)
L2 regularization strength.</p>
</li>
<li>
<p><span class="mkapi-item-name">margin</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=1)
Margin for triplet loss function.
Larger values enforce stronger separation.</p>
</li>
<li>
<p><span class="mkapi-item-name">verbose</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=False)
Whether to print training progress.</p>
</li>
<li>
<p><span class="mkapi-item-name">random_seed</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=666)
Random seed for reproducibility.</p>
</li>
<li>
<p><span class="mkapi-item-name">iter_comb</span>
<span class="mkapi-dash">—</span>
list, optional (default=None)
List of batch pairs to compare.
If None, compares all pairs.</p>
</li>
<li>
<p><span class="mkapi-item-name">knn_neigh</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=100)
Number of neighbors for mutual nearest neighbors.</p>
</li>
<li>
<p><span class="mkapi-item-name">Batch_list</span>
<span class="mkapi-dash">—</span>
list, optional (default=None)
List of individual batch AnnData objects.
Must be in same order as in combined adata.</p>
</li>
<li>
<p><span class="mkapi-item-name">device</span>
<span class="mkapi-dash">—</span>
torch.device, optional (default=auto)
Computing device to use.
Automatically uses GPU if available.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Requires pre-computed spatial networks</li>
<li>GPU acceleration recommended for large datasets</li>
<li>Batch_list order must match batch_key order</li>
<li>Memory usage scales with dataset size</li>
<li>Consider reducing knn_neigh for large datasets</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAligner.train" title="omicverse.space.pySTAligner.train">train</a></span>
<span class="mkapi-dash">—</span>
Train the STAligner spatial integration model.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySTAligner.predicted" title="omicverse.space.pySTAligner.predicted">predicted</a></span>
<span class="mkapi-dash">—</span>
Generate and store the final embedding from trained STAligner model.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAligner.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySTAligner" title="omicverse.space.pySTAligner">pySTAligner</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySTAligner.train" title="omicverse.space.pySTAligner.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Train the STAligner spatial integration model.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle this method performs two-stage training of the staligner model">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">This method performs two-stage training of the STAligner model</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Pre-training with STAGATE to learn initial embeddings</li>
<li>Fine-tuning with STAligner using triplet loss and MNN</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle the training process">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">The training process</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Sets random seeds for reproducibility</li>
<li>Pre-trains with graph autoencoder</li>
<li>Identifies mutual nearest neighbors</li>
<li>Optimizes embeddings with triplet loss</li>
<li>Monitors training progress</li>
<li>Saves final embeddings</li>
</ol>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Progress shown if verbose=True</li>
<li>Uses GPU if available</li>
<li>Early stopping not implemented</li>
<li>Results stored in adata.obsm[key_added]</li>
<li>Memory usage increases during training</li>
<li>Consider batch size for large datasets</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySTAligner.predicted">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySTAligner" title="omicverse.space.pySTAligner">pySTAligner</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySTAligner.predicted" title="omicverse.space.pySTAligner.predicted">predicted</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Generate and store the final embedding from trained STAligner model.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>AnnData object with STAligner embedding stored in obsm[key_added].</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Cal_Spatial_Net">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.Cal_Spatial_Net" title="omicverse.space.Cal_Spatial_Net">Cal_Spatial_Net</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rad_cutoff</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">k_cutoff</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_neigh</span><span class="mkapi-equal">=</span><span class="mkapi-default">50</span><span class="mkapi-comma">, </span><span class="mkapi-arg">model</span><span class="mkapi-equal">=</span><span class="mkapi-default">'Radius'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">verbose</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Construct spatial neighbor networks for spatial integration.</p>
<p>This function builds a spatial neighborhood graph by connecting spots based
on their physical distances. It supports both radius-based and k-nearest
neighbor approaches for network construction.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
    Input spatial data containing:
    - Spatial coordinates in obsm['spatial']
    - Gene expression data in X
rad_cutoff: float, optional (default=None)
    Maximum distance between neighbors for 'Radius' model.
    Only used when model='Radius'.
k_cutoff: int, optional (default=None)
    Number of nearest neighbors to connect for 'KNN' model.
    Only used when model='KNN'.
max_neigh: int, optional (default=50)
    Maximum number of neighbors to consider during graph construction.
    Helps limit memory usage for large datasets.
model: str, optional (default='Radius')
    Network construction method:
    - 'Radius': Connect spots within rad_cutoff distance
    - 'KNN': Connect k_cutoff nearest neighbors
verbose: bool, optional (default=True)
    Whether to print network statistics.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
None
    Updates adata with:
    - adata.uns['Spatial_Net']: DataFrame of edges and distances
    - adata.uns['adj']: Sparse adjacency matrix</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>For STAligner, adjust rad_cutoff to ensure 5-10 neighbors per spot</li>
<li>Includes self-loops in adjacency matrix</li>
<li>Uses ball_tree algorithm for efficient neighbor search</li>
<li>Memory efficient implementation for large datasets</li>
<li>Critical for downstream integration tasks</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySpaceFlow">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.pySpaceFlow" title="omicverse.space.pySpaceFlow">pySpaceFlow</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>SpaceFlow spatial flow analysis class.</p>
<p>SpaceFlow is a deep learning method for analyzing spatial transcriptomics data
by learning spatially-aware cell representations. It combines graph neural networks
with spatial regularization to capture both transcriptional and spatial relationships
between cells.</p>
<p>Initialize SpaceFlow spatial analysis object.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle the method">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">The method</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Constructs a spatial neighborhood graph</li>
<li>Learns embeddings using deep graph infomax</li>
<li>Applies spatial regularization to preserve spatial structure</li>
<li>Generates pseudo-spatial maps for trajectory analysis</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Input annotated data matrix containing:</p>
<ul>
<li>Gene expression data in adata.X</li>
<li>Spatial coordinates in adata.obsm['spatial']</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">sf</span>
<span class="mkapi-dash">—</span>
SpaceFlow
Internal SpaceFlow object for computations</p>
</li>
<li>
<p><span class="mkapi-item-name">embedding</span>
<span class="mkapi-dash">—</span>
array
Learned spatial-aware embeddings after training</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load spatial data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Initialize SpaceFlow</span>
<span class="n">spaceflow</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">pySpaceFlow</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># Train model</span>
<span class="n">embedding</span> <span class="o">=</span> <span class="n">spaceflow</span><span class="o">.</span><span class="n">train</span><span class="p">(</span>
    <span class="n">spatial_regularization_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">z_dim</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">epochs</span><span class="o">=</span><span class="mi">1000</span>
<span class="p">)</span>
<span class="c1"># Calculate pseudo-spatial map</span>
<span class="n">psm</span> <span class="o">=</span> <span class="n">spaceflow</span><span class="o">.</span><span class="n">cal_pSM</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing:</p>
<ul>
<li>Gene expression data in adata.X</li>
<li>Spatial coordinates in adata.obsm['spatial']</li>
</ul>
<p>The data should be preprocessed (normalized, scaled)</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Automatically checks for SpaceFlow package installation</li>
<li>Constructs initial spatial neighborhood graph</li>
<li>Uses 10 nearest neighbors for graph construction</li>
<li>Stores SpaceFlow object in self.sf</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySpaceFlow.train" title="omicverse.space.pySpaceFlow.train">train</a></span>
<span class="mkapi-dash">—</span>
Train SpaceFlow model for spatial embedding.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.pySpaceFlow.cal_pSM" title="omicverse.space.pySpaceFlow.cal_pSM">cal_pSM</a></span>
<span class="mkapi-dash">—</span>
Calculate pseudo-spatial map using diffusion pseudotime.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySpaceFlow.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySpaceFlow" title="omicverse.space.pySpaceFlow">pySpaceFlow</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySpaceFlow.train" title="omicverse.space.pySpaceFlow.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">spatial_regularization_strength</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">z_dim</span><span class="mkapi-equal">=</span><span class="mkapi-default">50</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lr</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.001</span><span class="mkapi-comma">, </span><span class="mkapi-arg">epochs</span><span class="mkapi-equal">=</span><span class="mkapi-default">1000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_patience</span><span class="mkapi-equal">=</span><span class="mkapi-default">50</span><span class="mkapi-comma">, </span><span class="mkapi-arg">min_stop</span><span class="mkapi-equal">=</span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">random_seed</span><span class="mkapi-equal">=</span><span class="mkapi-default">42</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gpu</span><span class="mkapi-equal">=</span><span class="mkapi-default">0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">regularization_acceleration</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">edge_subset_sz</span><span class="mkapi-equal">=</span><span class="mkapi-default">1000000</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Train SpaceFlow model for spatial embedding.</p>
<p>This method trains a graph neural network to learn spatially-aware cell
representations using deep graph infomax with spatial regularization.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">spatial_regularization_strength</span>
<span class="mkapi-dash">—</span>
float, optional (default=0.1)
    Weight for spatial regularization term.
    Higher values enforce stronger spatial consistency.
z_dim: int, optional (default=50)
    Dimensionality of learned embedding space.
lr: float, optional (default=1e-3)
    Learning rate for Adam optimizer.
epochs: int, optional (default=1000)
    Maximum number of training epochs.
max_patience: int, optional (default=50)
    Number of epochs to wait for improvement before early stopping.
min_stop: int, optional (default=100)
    Minimum number of epochs before allowing early stopping.
random_seed: int, optional (default=42)
    Random seed for reproducibility.
gpu: int, optional (default=0)
    GPU device index to use. Uses CPU if GPU unavailable.
regularization_acceleration: bool, optional (default=True)
    Whether to use subsampling for faster regularization.
edge_subset_sz: int, optional (default=1000000)
    Number of edges to sample for accelerated regularization.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
numpy.ndarray
    Learned embedding matrix of shape (n_cells, z_dim).</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Uses deep graph infomax for self-supervised learning</li>
<li>Applies spatial regularization to preserve spatial structure</li>
<li>Employs early stopping based on loss convergence</li>
<li>Supports GPU acceleration when available</li>
<li>Results are stored in adata.obsm['spaceflow']</li>
<li>Progress is shown with a progress bar</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.pySpaceFlow.cal_pSM">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.pySpaceFlow" title="omicverse.space.pySpaceFlow">pySpaceFlow</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.pySpaceFlow.cal_pSM" title="omicverse.space.pySpaceFlow.cal_pSM">cal_pSM</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">n_neighbors</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">resolution</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_cell_for_subsampling</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">5000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">psm_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'pSM_spaceflow'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate pseudo-spatial map using diffusion pseudotime.</p>
<p>This method constructs a pseudo-spatial map by computing diffusion pseudotime
on the learned embeddings, useful for analyzing spatial trajectories and
organization patterns.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">n_neighbors</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=20)
Number of neighbors for kNN graph construction.
Higher values create denser connectivity.</p>
</li>
<li>
<p><span class="mkapi-item-name">resolution</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1)
Resolution parameter for Leiden clustering.
Higher values yield more fine-grained clusters.</p>
</li>
<li>
<p><span class="mkapi-item-name">max_cell_for_subsampling</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=5000)
Maximum number of cells to use for distance calculations.
Enables analysis of large datasets through subsampling.</p>
</li>
<li>
<p><span class="mkapi-item-name">psm_key</span>
<span class="mkapi-dash">—</span>
str, optional (default='pSM_spaceflow')
Key in adata.obs where pseudo-spatial map values will be stored.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>numpy.ndarray
    Pseudo-spatial map values for each cell.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Constructs neighborhood graph from embeddings</li>
<li>Performs UMAP and Leiden clustering</li>
<li>Uses PAGA for trajectory inference</li>
<li>Computes diffusion pseudotime</li>
<li>Results are stored in adata.obs[psm_key]</li>
<li>Useful for ordering cells along spatial axes</li>
<li>Handles large datasets through subsampling</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.svg">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.svg" title="omicverse.space.svg">svg</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mode</span><span class="mkapi-equal">=</span><span class="mkapi-default">'prost'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_svgs</span><span class="mkapi-equal">=</span><span class="mkapi-default">3000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">target_sum</span><span class="mkapi-equal">=</span><span class="mkapi-default">50 * 10000.0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">platform</span><span class="mkapi-equal">=</span><span class="mkapi-default">'visium'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mt_startwith</span><span class="mkapi-equal">=</span><span class="mkapi-default">'MT-'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Identify spatially variable genes using multiple methods.</p>
<p>This function identifies genes that show significant spatial variation in their
expression patterns across the tissue. It supports multiple methods including
PROST (Pattern RecognitiOn of Spatial Transcriptomics), Pearson correlation,
and Spateo-based analysis.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing spatial transcriptomics data.
Must contain:</p>
<ul>
<li>Raw counts in adata.X or adata.layers['counts']</li>
<li>Spatial coordinates in adata.obsm['spatial']</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">mode</span>
<span class="mkapi-dash">—</span>
str, optional (default='prost')
Method for identifying spatially variable genes:</p>
<ul>
<li>'prost': Pattern RecognitiOn of Spatial Transcriptomics</li>
<li>'pearsonr': Pearson correlation-based method</li>
<li>'spateo': Spateo-based analysis using Wasserstein distance</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">n_svgs</span>
<span class="mkapi-dash">—</span>
int, optional (default=3000)
Number of spatially variable genes to select.</p>
</li>
<li>
<p><span class="mkapi-item-name">target_sum</span>
<span class="mkapi-dash">—</span>
float, optional (default=50*1e4)
Target sum for library size normalization.</p>
</li>
<li>
<p><span class="mkapi-item-name">platform</span>
<span class="mkapi-dash">—</span>
str, optional (default="visium")
Spatial transcriptomics platform type.</p>
</li>
<li>
<p><span class="mkapi-item-name">mt_startwith</span>
<span class="mkapi-dash">—</span>
str, optional (default='MT-')
Prefix for mitochondrial genes to exclude from analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to specific SVG methods:
For 'spateo':
    - log2fc: Minimum log2 fold change (default: 1)
    - rank_p: Maximum rank-based p-value (default: 0.05)
    - adj_pvalue: Maximum adjusted p-value (default: 0.05)</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">AnnData</span>
<span class="mkapi-dash">—</span>
Input AnnData object updated with:</p>
<ul>
<li>adata.var['space_variable_features']: Boolean mask of selected SVGs</li>
<li>adata.var['highly_variable']: Alias for space_variable_features</li>
</ul>
<p>For 'prost' mode:
    - Additional PROST-specific metrics in adata.var
For 'spateo' mode:
    - Wasserstein distance statistics in adata.var</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>PROST mode requires opencv-python package</li>
<li>Different modes use different statistical approaches:<ul>
<li>PROST: Pattern recognition and spatial autocorrelation</li>
<li>pearsonr: Correlation between gene expression and spatial coordinates</li>
<li>spateo: Wasserstein distance-based spatial variation</li>
</ul>
</li>
<li>Mitochondrial genes are excluded by default</li>
<li>Results are normalized and log-transformed</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load spatial data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Find SVGs using PROST</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">svg</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">mode</span><span class="o">=</span><span class="s1">'prost'</span><span class="p">,</span>
    <span class="n">n_svgs</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
    <span class="n">platform</span><span class="o">=</span><span class="s1">'visium'</span>
<span class="p">)</span>
<span class="c1"># Access SVGs</span>
<span class="n">svgs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'space_variable_features'</span><span class="p">]]</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_loc</span><span class="mkapi-equal">=</span><span class="mkapi-default">'xy_loc'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">region</span><span class="mkapi-equal">=</span><span class="mkapi-default">'Region'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Spatial Transition Tensor (STT) analysis class.</p>
<p>STT models spatial dynamics and transitions by learning spatial-temporal patterns
in spatial transcriptomics data using transition tensors. This class provides methods
for analyzing cell state transitions, developmental trajectories, and spatial dynamics
in tissue organization.</p>
<p>Initialize STT spatial transition analysis object.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing spatial transcriptomics data.
Must contain:</p>
<ul>
<li>Spliced counts in adata.layers['spliced']</li>
<li>Unspliced counts in adata.layers['unspliced']</li>
<li>Spatial coordinates in adata.obsm[spatial_loc]</li>
</ul>
<p>AnnData
Annotated data matrix with spatial and velocity data.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatial_loc</span>
<span class="mkapi-dash">—</span>
str, optional (default='xy_loc')
Key in adata.obsm containing spatial coordinates.</p>
<p>str, optional (default='xy_loc')
Key for spatial coordinates in adata.obsm.</p>
</li>
<li>
<p><span class="mkapi-item-name">region</span>
<span class="mkapi-dash">—</span>
str, optional (default='Region')
Column name in adata.obs containing region annotations.</p>
<p>str, optional (default='Region')
Column name for region annotations in adata.obs.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Input annotated data matrix.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_aggr</span>
<span class="mkapi-dash">—</span>
AnnData or None
Aggregated data after training.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatial_loc</span>
<span class="mkapi-dash">—</span>
str
Key for spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">region</span>
<span class="mkapi-dash">—</span>
str
Key for region annotations.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data with spliced/unspliced counts</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial_velocity.h5ad'</span><span class="p">)</span>
<span class="c1"># Initialize STT object</span>
<span class="n">stt</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">STT</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">spatial_loc</span><span class="o">=</span><span class="s1">'spatial'</span><span class="p">,</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">'tissue_region'</span>
<span class="p">)</span>
<span class="c1"># Estimate cell stages</span>
<span class="n">stt</span><span class="o">.</span><span class="n">stage_estimate</span><span class="p">()</span>
<span class="c1"># Train the model</span>
<span class="n">stt</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">n_states</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.stage_estimate" title="omicverse.space.STT.stage_estimate">stage_estimate</a></span>
<span class="mkapi-dash">—</span>
Estimate cell stages using joint clustering of spliced and unspliced data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.train" title="omicverse.space.STT.train">train</a></span>
<span class="mkapi-dash">—</span>
Train the STT spatial transition model.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.load" title="omicverse.space.STT.load">load</a></span>
<span class="mkapi-dash">—</span>
Load pre-trained STT model data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.compute_pathway" title="omicverse.space.STT.compute_pathway">compute_pathway</a></span>
<span class="mkapi-dash">—</span>
Compute spatial pathways for cell transitions.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_pathway" title="omicverse.space.STT.plot_pathway">plot_pathway</a></span>
<span class="mkapi-dash">—</span>
Plot spatial transition pathways.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_tensor_pathway" title="omicverse.space.STT.plot_tensor_pathway">plot_tensor_pathway</a></span>
<span class="mkapi-dash">—</span>
Plot tensor-based spatial pathways.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_tensor" title="omicverse.space.STT.plot_tensor">plot_tensor</a></span>
<span class="mkapi-dash">—</span>
Plot spatial transition tensors.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.construct_landscape" title="omicverse.space.STT.construct_landscape">construct_landscape</a></span>
<span class="mkapi-dash">—</span>
Construct spatial landscape for transition analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.infer_lineage" title="omicverse.space.STT.infer_lineage">infer_lineage</a></span>
<span class="mkapi-dash">—</span>
Infer cell lineage relationships from spatial transitions.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_landscape" title="omicverse.space.STT.plot_landscape">plot_landscape</a></span>
<span class="mkapi-dash">—</span>
Plot spatial landscape visualization.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_sankey" title="omicverse.space.STT.plot_sankey">plot_sankey</a></span>
<span class="mkapi-dash">—</span>
Plot Sankey diagram for cell transitions.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.STT.plot_top_genes" title="omicverse.space.STT.plot_top_genes">plot_top_genes</a></span>
<span class="mkapi-dash">—</span>
Plot top genes driving spatial transitions.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.stage_estimate">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.stage_estimate" title="omicverse.space.STT.stage_estimate">stage_estimate</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Estimate cell stages using joint clustering of spliced and unspliced data.</p>
<p>This method performs joint analysis of spliced and unspliced RNA counts to identify
developmental stages through clustering. It uses a combination of PCA, nearest
neighbor graph construction, and Leiden clustering to identify distinct cell stages.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">None</span></p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None
Updates adata.obs['joint_leiden'] with stage assignments.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Requires spliced/unspliced counts in adata.layers</li>
<li>Uses resolution=0.15 for Leiden clustering</li>
<li>Number of stages is determined automatically</li>
<li>Results are stored in adata.obs['joint_leiden']</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.train" title="omicverse.space.STT.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">n_states</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">9</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_iter</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">15</span><span class="mkapi-comma">, </span><span class="mkapi-arg">weight_connectivities</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_neighbors</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">50</span><span class="mkapi-comma">, </span><span class="mkapi-arg">thresh_ms_gene</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spa_weight</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.3</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return">None</span></span>
</p>
<div class="mkapi-document">
<p>Train the STT spatial transition model.</p>
<p>This method trains the Spatial Transition Tensor model to learn spatial-temporal
patterns and cell state transitions. It combines connectivity constraints with
spatial information to model developmental trajectories.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">n_states</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=9)
Number of states for transition modeling.</p>
</li>
<li>
<p><span class="mkapi-item-name">n_iter</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=15)
Number of iterations for model training.</p>
</li>
<li>
<p><span class="mkapi-item-name">weight_connectivities</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.5)
Weight for connectivity constraints in range [0,1].</p>
</li>
<li>
<p><span class="mkapi-item-name">n_neighbors</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=50)
Number of neighbors for graph construction.</p>
</li>
<li>
<p><span class="mkapi-item-name">thresh_ms_gene</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.2)
Threshold for marker gene selection.</p>
</li>
<li>
<p><span class="mkapi-item-name">spa_weight</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.3)
Weight for spatial constraints in range [0,1].</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Additional arguments for dynamical iteration.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">None</span> <span class="mkapi-dash">—</span>
None
Updates self.adata_aggr with trained model results.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Higher spa_weight emphasizes spatial relationships</li>
<li>Higher weight_connectivities emphasizes transcriptional similarity</li>
<li>Results are stored in self.adata_aggr</li>
<li>Spatial coordinates are preserved in obsm</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.load">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.load" title="omicverse.space.STT.load">load</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_aggr</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Load pre-trained STT model data.</p>
<p>This method allows loading previously trained STT model results, enabling
analysis continuation or result sharing.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Original AnnData object with spatial data.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_aggr</span>
<span class="mkapi-dash">—</span>
AnnData
Aggregated AnnData object from previous training.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None
Updates self.adata and self.adata_aggr with loaded data.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Useful for sharing analysis results</li>
<li>Preserves all STT-specific annotations</li>
<li>Maintains spatial coordinate information</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.compute_pathway">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.compute_pathway" title="omicverse.space.STT.compute_pathway">compute_pathway</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">pathway_dict</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_components</span><span class="mkapi-equal">=</span><span class="mkapi-default">10</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Compute spatial pathways for cell transitions.</p>
<p>This method identifies and computes transition pathways between cell states
in spatial context, revealing developmental trajectories and cell fate decisions.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">pathway_dict</span>
<span class="mkapi-dash">—</span>
dict
Dictionary defining pathway parameters including:</p>
<ul>
<li>start_states: list of starting cell states</li>
<li>end_states: list of target cell states</li>
<li>intermediate_states: optional list of intermediate states</li>
<li>constraints: optional spatial or temporal constraints</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">dict</span>
<span class="mkapi-dash">—</span>
Computed pathway information including:</p>
<ul>
<li>transition_probabilities: array of state transition probabilities</li>
<li>path_coordinates: spatial coordinates of the pathway</li>
<li>path_states: sequence of states along the pathway</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Pathways are computed using transition tensors</li>
<li>Results can be visualized using plot_pathway()</li>
<li>Considers both spatial and transcriptional information</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_pathway">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_pathway" title="omicverse.space.STT.plot_pathway">plot_pathway</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">label_fontsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot spatial transition pathways.</p>
<p>This method visualizes computed transition pathways in spatial context,
showing how cells transition between states while maintaining spatial organization.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">label_fontsize</span>
<span class="mkapi-dash">—</span>
int, optional (default=20)
Font size for axis labels.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting function:</p>
<ul>
<li>color_by: feature to color points by</li>
<li>alpha: transparency of points</li>
<li>size: size of points</li>
<li>title: plot title</li>
<li>save: path to save figure</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure object containing the pathway visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Multiple pathways can be visualized simultaneously</li>
<li>Colors indicate transition states</li>
<li>Arrows show direction of transitions</li>
<li>Spatial coordinates are preserved in visualization</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_tensor_pathway">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_tensor_pathway" title="omicverse.space.STT.plot_tensor_pathway">plot_tensor_pathway</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">pathway_name</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot tensor-based spatial pathways.</p>
<p>This method creates a specialized visualization of transition pathways using
tensor decomposition results, highlighting the continuous nature of cell state
transitions in space.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">pathway_name</span>
<span class="mkapi-dash">—</span>
str
Name of the pathway to visualize.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional plotting arguments:</p>
<ul>
<li>show_transitions: bool, whether to show transition arrows</li>
<li>color_scheme: colormap for states</li>
<li>edge_width: width of transition edges</li>
<li>node_size: size of state nodes</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.axes.Axes
    Axes object containing the tensor pathway plot.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Tensor visualization shows transition probabilities</li>
<li>Thickness of edges indicates transition likelihood</li>
<li>Colors represent different cell states</li>
<li>Spatial arrangement reflects tissue organization</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_tensor">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_tensor" title="omicverse.space.STT.plot_tensor">plot_tensor</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">list_attractor</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot spatial transition tensors.</p>
<p>This method visualizes the learned transition tensors, showing how cells
transition between states while considering spatial constraints.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">list_attractor</span>
<span class="mkapi-dash">—</span>
list
List of attractor regions or states to visualize.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional plotting arguments:</p>
<ul>
<li>mode: visualization mode ('2D' or '3D')</li>
<li>show_labels: whether to show state labels</li>
<li>cmap: colormap for tensor values</li>
<li>edge_threshold: minimum value for showing transitions</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">dict</span>
<span class="mkapi-dash">—</span>
Dictionary containing plotting results:</p>
<ul>
<li>fig: matplotlib figure object</li>
<li>tensors: plotted tensor values</li>
<li>coordinates: spatial coordinates used</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Tensors show transition probabilities between states</li>
<li>Higher values indicate more likely transitions</li>
<li>Spatial relationships are encoded in tensor structure</li>
<li>Can be used to identify major transition paths</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.construct_landscape">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.construct_landscape" title="omicverse.space.STT.construct_landscape">construct_landscape</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">coord_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'X_xy_loc'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Construct spatial landscape for transition analysis.</p>
<p>This method builds a continuous landscape representation of cell state transitions
in spatial context, useful for understanding developmental potential and barriers.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">coord_key</span>
<span class="mkapi-dash">—</span>
str, optional (default='X_xy_loc')
Key for spatial coordinates in adata.obsm.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional landscape construction arguments:</p>
<ul>
<li>n_neighbors: number of neighbors for graph construction</li>
<li>smoothing: smoothing factor for landscape</li>
<li>resolution: grid resolution for landscape</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>None
Updates adata.obsm['trans_coord'] with landscape coordinates.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Landscape represents continuous transition space</li>
<li>Valleys indicate stable states</li>
<li>Ridges represent transition barriers</li>
<li>Coordinates preserve both spatial and state information</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.infer_lineage">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.infer_lineage" title="omicverse.space.STT.infer_lineage">infer_lineage</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Infer cell lineage relationships from spatial transitions.</p>
<p>This method reconstructs developmental lineages by analyzing transition
patterns and spatial relationships between cell states.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional lineage inference arguments:</p>
<ul>
<li>start_state: starting cell state</li>
<li>end_states: target cell states</li>
<li>min_branch_len: minimum branch length</li>
<li>max_steps: maximum steps in lineage</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">dict</span>
<span class="mkapi-dash">—</span>
Lineage inference results containing:</p>
<ul>
<li>branches: list of inferred lineage branches</li>
<li>states: ordered states in each branch</li>
<li>confidence: confidence scores for branches</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Combines spatial and transcriptional information</li>
<li>Identifies branching points in development</li>
<li>Considers transition probabilities between states</li>
<li>Results can be visualized with plot_landscape()</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_landscape">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_landscape" title="omicverse.space.STT.plot_landscape">plot_landscape</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot spatial landscape visualization.</p>
<p>This method creates a visual representation of the constructed transition
landscape, showing the continuous spectrum of cell states in spatial context.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional plotting arguments:</p>
<ul>
<li>color_by: feature to color points by</li>
<li>show_trajectory: whether to show transition trajectories</li>
<li>contour: whether to show landscape contours</li>
<li>grid_size: resolution of landscape grid</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the landscape visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Colors indicate cell states or features</li>
<li>Contours show transition barriers</li>
<li>Arrows indicate preferred transition directions</li>
<li>Spatial relationships are preserved in layout</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_sankey">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_sankey" title="omicverse.space.STT.plot_sankey">plot_sankey</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">vector1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">vector2</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot Sankey diagram for cell transitions.</p>
<p>This method creates a Sankey diagram showing the flow of cells between
different states or conditions, useful for understanding transition dynamics.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">vector1</span>
<span class="mkapi-dash">—</span>
array-like
Source state assignments for cells.</p>
</li>
<li>
<p><span class="mkapi-item-name">vector2</span>
<span class="mkapi-dash">—</span>
array-like
Target state assignments for cells.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the Sankey diagram.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Width of flows indicates transition frequency</li>
<li>Colors can represent different states</li>
<li>Useful for visualizing state changes</li>
<li>Shows conservation of cell numbers</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.STT.plot_top_genes">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.STT" title="omicverse.space.STT">STT</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.STT.plot_top_genes" title="omicverse.space.STT.plot_top_genes">plot_top_genes</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot top genes driving spatial transitions.</p>
<p>This method identifies and visualizes genes that are most important in
determining cell state transitions and spatial patterns.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional plotting arguments:</p>
<ul>
<li>n_genes: number of top genes to show</li>
<li>groupby: how to group genes</li>
<li>show_labels: whether to show gene names</li>
<li>scale: whether to scale expression values</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure showing top genes visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Highlights genes driving state transitions</li>
<li>Shows expression patterns across states</li>
<li>Useful for identifying key regulators</li>
<li>Can reveal transition mechanisms</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellMap">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.CellMap" title="omicverse.space.CellMap">CellMap</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_sc</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_sp</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_rep_sc</span><span class="mkapi-equal">=</span><span class="mkapi-default">'X_pca'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_rep_sp</span><span class="mkapi-equal">=</span><span class="mkapi-default">'X_pca'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>SpatRio CellMap class for mapping single cells to spatial coordinates.</p>
<p>This class implements optimal transport-based mapping of single cells to spatial
coordinates using expression similarity and spatial awareness. It provides methods
for both mapping and coordinate assignment.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span>
<span class="mkapi-dash">—</span>
AnnData
Single-cell RNA sequencing data object.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span>
<span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data object.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sc</span>
<span class="mkapi-dash">—</span>
str, optional (default='X_pca')
Key in adata_sc.obsm for single-cell dimensional reduction.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sp</span>
<span class="mkapi-dash">—</span>
str, optional (default='X_pca')
Key in adata_sp.obsm for spatial data dimensional reduction.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span>
<span class="mkapi-dash">—</span>
AnnData
Single-cell RNA sequencing data.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span>
<span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sc</span>
<span class="mkapi-dash">—</span>
str
Representation key for single-cell data.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sp</span>
<span class="mkapi-dash">—</span>
str
Representation key for spatial data.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatrio_decon</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame
Deconvolution results after mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatrio_map</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame
Coordinate assignment results.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'single_cell.h5ad'</span><span class="p">)</span>
<span class="n">adata_sp</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial.h5ad'</span><span class="p">)</span>
<span class="c1"># Initialize CellMap</span>
<span class="n">cm</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">CellMap</span><span class="p">(</span>
    <span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span><span class="p">,</span>
    <span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span><span class="p">,</span>
    <span class="n">use_rep_sc</span><span class="o">=</span><span class="s1">'X_pca'</span>
<span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellMap.map" title="omicverse.space.CellMap.map">map</a></span>
<span class="mkapi-dash">—</span>
Map single cells to spatial locations using optimal transport.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellMap.assign_coord" title="omicverse.space.CellMap.assign_coord">assign_coord</a></span>
<span class="mkapi-dash">—</span>
Assign spatial coordinates to mapped single cells.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellMap.map">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellMap" title="omicverse.space.CellMap">CellMap</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellMap.map" title="omicverse.space.CellMap.map">map</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">sc_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'celltype'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sp_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'leiden'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">alpha</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_power</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">resolution</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_spatial</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_multi</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_gpu</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Map single cells to spatial locations using optimal transport.</p>
<p>This method performs the core mapping operation using optimal transport,
considering both expression similarity and spatial relationships.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">sc_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='celltype')
Column in adata_sc.obs containing cell type annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">sp_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='leiden')
Column in adata_sp.obs containing spatial cluster annotations.
If not present, will be computed using Leiden clustering.</p>
</li>
<li>
<p><span class="mkapi-item-name">alpha</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.1)
Balance parameter for optimal transport cost.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_power</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=2)
Power parameter for distance-based cost.</p>
</li>
<li>
<p><span class="mkapi-item-name">resolution</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1)
Resolution parameter for Leiden clustering if needed.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_spatial</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use spatial information in mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_multi</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use multiple cell type information.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_gpu</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use GPU acceleration.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Additional arguments passed to ot_alignment.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
pandas.DataFrame
    Mapping results containing cell-to-spot assignments.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>If sp_type is 'leiden' or 'louvain' and not present in data,
  clustering will be performed automatically</li>
<li>GPU acceleration is recommended for large datasets</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellMap.assign_coord">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellMap" title="omicverse.space.CellMap">CellMap</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellMap.assign_coord" title="omicverse.space.CellMap.assign_coord">assign_coord</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Assign spatial coordinates to mapped single cells.</p>
<p>This method takes the mapping results and assigns specific spatial
coordinates to each mapped single cell, creating a spatially-aware
single-cell dataset.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to assign_coord function.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">AnnData</span>
<span class="mkapi-dash">—</span>
Copy of single-cell data with added spatial coordinates and
mapping information in:</p>
<ul>
<li>adata.obsm['spatial']: Assigned coordinates</li>
<li>adata.obs['Cell_xcoord']: X coordinates</li>
<li>adata.obs['Cell_ycoord']: Y coordinates</li>
<li>adata.obs['spot_type']: Mapped spot types</li>
<li>adata.obs['spot']: Mapped spot IDs</li>
<li>adata.obs['spot_value']: Mapping confidence scores</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Only cells that were successfully mapped will be included</li>
<li>Original spatial metadata is preserved in adata.uns['spatial']</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_sc</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_sp</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_rep_sc</span><span class="mkapi-equal">=</span><span class="mkapi-default">'X_pca'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_rep_sp</span><span class="mkapi-equal">=</span><span class="mkapi-default">'X_pca'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>SpatRio CellLoc class for probabilistic cell localization.</p>
<p>This class extends CellMap with probabilistic filtering based on cell type
proportions for more accurate spatial localization. It provides methods for
mapping, saving/loading results, and probabilistic assignment.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span>
<span class="mkapi-dash">—</span>
AnnData
Single-cell RNA sequencing data object.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span>
<span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data object.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sc</span>
<span class="mkapi-dash">—</span>
str, optional (default='X_pca')
Key in adata_sc.obsm for single-cell dimensional reduction.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sp</span>
<span class="mkapi-dash">—</span>
str, optional (default='X_pca')
Key in adata_sp.obsm for spatial data dimensional reduction.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_sc</span>
<span class="mkapi-dash">—</span>
AnnData
Single-cell RNA sequencing data.</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_sp</span>
<span class="mkapi-dash">—</span>
AnnData
Spatial transcriptomics data.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sc</span>
<span class="mkapi-dash">—</span>
str
Representation key for single-cell data.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_rep_sp</span>
<span class="mkapi-dash">—</span>
str
Representation key for spatial data.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatrio_decon</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame
Deconvolution results after mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatrio_map</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame
Coordinate assignment results.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'single_cell.h5ad'</span><span class="p">)</span>
<span class="n">adata_sp</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">'spatial.h5ad'</span><span class="p">)</span>
<span class="c1"># Initialize CellLoc</span>
<span class="n">cl</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">CellLoc</span><span class="p">(</span>
    <span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span><span class="p">,</span>
    <span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span>
<span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc.loc_map" title="omicverse.space.CellLoc.loc_map">loc_map</a></span>
<span class="mkapi-dash">—</span>
Map single cells to spatial locations with probabilistic filtering.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc.load_map" title="omicverse.space.CellLoc.load_map">load_map</a></span>
<span class="mkapi-dash">—</span>
Load pre-computed mapping results.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc.save_map" title="omicverse.space.CellLoc.save_map">save_map</a></span>
<span class="mkapi-dash">—</span>
Save mapping results to a file.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc.loc_prob" title="omicverse.space.CellLoc.loc_prob">loc_prob</a></span>
<span class="mkapi-dash">—</span>
Calculate localization probabilities for mapped cells.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.CellLoc.loc_assign" title="omicverse.space.CellLoc.loc_assign">loc_assign</a></span>
<span class="mkapi-dash">—</span>
Assign spatial coordinates to single cell data.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc.loc_map">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc.loc_map" title="omicverse.space.CellLoc.loc_map">loc_map</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">sc_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'celltype'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sp_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'leiden'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">alpha</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_power</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">resolution</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_spatial</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_multi</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_gpu</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-colon">: </span><span class="mkapi-ann"><span class="mkapi-tooltip" title="typing.Any">Any</span></span><span class="mkapi-paren">)</span><span class="mkapi-arrow"> → </span><span class="mkapi-return"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span></span>
</p>
<div class="mkapi-document">
<p>Map single cells to spatial locations with probabilistic filtering.</p>
<p>Similar to CellMap.map() but includes additional probabilistic considerations
for more accurate localization.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">sc_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='celltype')
Column in adata_sc.obs containing cell type annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">sp_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='leiden')
Column in adata_sp.obs containing spatial cluster annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">alpha</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.1)
Balance parameter for optimal transport cost.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_power</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=2)
Power parameter for distance-based cost.</p>
</li>
<li>
<p><span class="mkapi-item-name">resolution</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1)
Resolution parameter for Leiden clustering if needed.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_spatial</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use spatial information in mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_multi</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use multiple cell type information.</p>
</li>
<li>
<p><span class="mkapi-item-name">use_gpu</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to use GPU acceleration.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span> : 
<span class="mkapi-item-type"><span class="mkapi-tooltip" title="typing.Any">Any</span></span> <span class="mkapi-dash">—</span>
Additional arguments passed to ot_alignment.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type"><span class="mkapi-tooltip" title="pandas">pd</span>.<span class="mkapi-tooltip" title="pandas.core.frame.DataFrame">DataFrame</span></span> <span class="mkapi-dash">—</span>
pandas.DataFrame
    Probabilistic mapping results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc.load_map">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc.load_map" title="omicverse.space.CellLoc.load_map">load_map</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">map_info</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sc_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'celltype'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sp_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'leiden'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">resolution</span><span class="mkapi-colon">: </span><span class="mkapi-ann">int</span><span class="mkapi-equal"> = </span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_spatial</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">aware_multi</span><span class="mkapi-colon">: </span><span class="mkapi-ann">bool</span><span class="mkapi-equal"> = </span><span class="mkapi-default">True</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Load pre-computed mapping results.</p>
<p>This method allows loading previously computed mapping results while
maintaining the proper data structure and annotations.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">map_info</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame
Pre-computed mapping results to load.</p>
</li>
<li>
<p><span class="mkapi-item-name">sc_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='celltype')
Column in adata_sc.obs containing cell type annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">sp_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='leiden')
Column in adata_sp.obs containing spatial cluster annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">resolution</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
int, optional (default=1)
Resolution parameter for Leiden clustering if needed.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_spatial</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether spatial information was used in mapping.</p>
</li>
<li>
<p><span class="mkapi-item-name">aware_multi</span> : 
<span class="mkapi-item-type">bool</span> <span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether multiple cell type information was used.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>pandas.DataFrame
    Loaded mapping results.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc.save_map">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc.save_map" title="omicverse.space.CellLoc.save_map">save_map</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">path</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Save mapping results to a file.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">path</span>
<span class="mkapi-dash">—</span>
str
Path where to save the mapping results CSV file.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc.loc_prob">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc.loc_prob" title="omicverse.space.CellLoc.loc_prob">loc_prob</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">spot_cell_prob</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sc_type</span><span class="mkapi-colon">: </span><span class="mkapi-ann">str</span><span class="mkapi-equal"> = </span><span class="mkapi-default">'celltype'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sc_prop</span><span class="mkapi-colon">: </span><span class="mkapi-ann">float</span><span class="mkapi-equal"> = </span><span class="mkapi-default">0.5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_cpu</span><span class="mkapi-equal">=</span><span class="mkapi-default">6</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate localization probabilities for mapped cells.</p>
<p>This method computes probabilistic scores for cell localization based on
cell type proportions and mapping confidence.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">spot_cell_prob</span>
<span class="mkapi-dash">—</span>
float
Probability threshold for spot-cell assignments.</p>
</li>
<li>
<p><span class="mkapi-item-name">sc_type</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
str, optional (default='celltype')
Column in adata_sc.obs containing cell type annotations.</p>
</li>
<li>
<p><span class="mkapi-item-name">sc_prop</span> : 
<span class="mkapi-item-type">float</span> <span class="mkapi-dash">—</span>
float, optional (default=0.5)
Minimum proportion of cells to consider for each type.</p>
</li>
<li>
<p><span class="mkapi-item-name">n_cpu</span>
<span class="mkapi-dash">—</span>
int, optional (default=6)
Number of CPU cores to use for parallel processing.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>pandas.DataFrame
    Updated mapping results with probability scores.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.CellLoc.loc_assign">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.CellLoc" title="omicverse.space.CellLoc">CellLoc</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.CellLoc.loc_assign" title="omicverse.space.CellLoc.loc_assign">loc_assign</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Assign spatial coordinates to single cell data.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>GASTON spatial depth estimation and clustering.</p>
<p>GASTON (Geometry And Spatial Transcriptomics-based OrganizatioN) is a method
for analyzing spatial transcriptomics data by learning continuous spatial depth
functions that capture tissue organization. It uses neural networks to model
spatial patterns and identify distinct domains.</p>
<p>Initialize GASTON spatial clustering object.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle the method combines gene expression data with spatial information to">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">The method combines gene expression data with spatial information to</span>
</p>
<div class="mkapi-section-content">
<ol>
<li>Learn continuous spatial depth functions</li>
<li>Identify tissue domains and boundaries</li>
<li>Model gene expression patterns along spatial gradients</li>
<li>Characterize tissue organization and architecture</li>
</ol>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle attributes">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Attributes</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Input annotated data matrix containing:</p>
<ul>
<li>Spatial coordinates in adata.obsm['spatial']</li>
<li>Gene expression data in adata.X</li>
<li>Optional histology image in adata.uns['spatial']</li>
</ul>
</li>
<li>
<p><span class="mkapi-item-name">model</span>
<span class="mkapi-dash">—</span>
GASTON model
Trained neural network model after calling train()</p>
</li>
<li>
<p><span class="mkapi-item-name">gaston_isodepth</span>
<span class="mkapi-dash">—</span>
array
Computed isodepth values after calling cal_iso_depth()</p>
</li>
<li>
<p><span class="mkapi-item-name">gaston_labels</span>
<span class="mkapi-dash">—</span>
array
Domain labels after calling cal_iso_depth()</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load spatial data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Initialize GASTON</span>
<span class="n">gaston</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">GASTON</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1"># Prepare input data</span>
<span class="n">counts</span><span class="p">,</span> <span class="n">coords</span><span class="p">,</span> <span class="n">genes</span> <span class="o">=</span> <span class="n">gaston</span><span class="o">.</span><span class="n">get_gaston_input</span><span class="p">()</span>
<span class="c1"># Get features</span>
<span class="n">features</span> <span class="o">=</span> <span class="n">gaston</span><span class="o">.</span><span class="n">get_top_pearson_residuals</span><span class="p">(</span><span class="n">num_dims</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="c1"># Train model</span>
<span class="n">gaston</span><span class="o">.</span><span class="n">load_rescale</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
<span class="n">gaston</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">num_epochs</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
<span class="c1"># Get best model</span>
<span class="n">model</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="n">coords</span> <span class="o">=</span> <span class="n">gaston</span><span class="o">.</span><span class="n">get_best_model</span><span class="p">()</span>
<span class="c1"># Calculate domains</span>
<span class="n">isodepth</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">gaston</span><span class="o">.</span><span class="n">cal_iso_depth</span><span class="p">(</span><span class="n">num_domains</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing:</p>
<ul>
<li>Spatial coordinates in adata.obsm['spatial']</li>
<li>Gene expression data in adata.X</li>
<li>Optional histology image in adata.uns['spatial']</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.get_gaston_input" title="omicverse.space.GASTON.get_gaston_input">get_gaston_input</a></span>
<span class="mkapi-dash">—</span>
Prepare input data for GASTON analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.get_top_pearson_residuals" title="omicverse.space.GASTON.get_top_pearson_residuals">get_top_pearson_residuals</a></span>
<span class="mkapi-dash">—</span>
Get top Pearson residual features for GASTON analysis.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.load_rescale" title="omicverse.space.GASTON.load_rescale">load_rescale</a></span>
<span class="mkapi-dash">—</span>
Load and rescale input data for neural network training.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.train" title="omicverse.space.GASTON.train">train</a></span>
<span class="mkapi-dash">—</span>
Train GASTON neural network models.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.get_best_model" title="omicverse.space.GASTON.get_best_model">get_best_model</a></span>
<span class="mkapi-dash">—</span>
Select best GASTON model and determine optimal domain number.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.cal_iso_depth" title="omicverse.space.GASTON.cal_iso_depth">cal_iso_depth</a></span>
<span class="mkapi-dash">—</span>
Calculate isodepth values and domain labels.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_isodepth" title="omicverse.space.GASTON.plot_isodepth">plot_isodepth</a></span>
<span class="mkapi-dash">—</span>
Plot isodepth contours and streamlines.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_clusters" title="omicverse.space.GASTON.plot_clusters">plot_clusters</a></span>
<span class="mkapi-dash">—</span>
Plot spatial domains with cluster colors.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_clusters_restrict" title="omicverse.space.GASTON.plot_clusters_restrict">plot_clusters_restrict</a></span>
<span class="mkapi-dash">—</span>
Plot spatial domains with restricted isodepth range.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.restrict_spot" title="omicverse.space.GASTON.restrict_spot">restrict_spot</a></span>
<span class="mkapi-dash">—</span>
Restrict analysis to spots within specific isodepth range.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.filter_genes" title="omicverse.space.GASTON.filter_genes">filter_genes</a></span>
<span class="mkapi-dash">—</span>
Filter genes based on expression and name patterns.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.pw_linear_fit" title="omicverse.space.GASTON.pw_linear_fit">pw_linear_fit</a></span>
<span class="mkapi-dash">—</span>
Perform piecewise linear fitting of gene expression.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.bin_data" title="omicverse.space.GASTON.bin_data">bin_data</a></span>
<span class="mkapi-dash">—</span>
Bin data along spatial depth gradient.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.get_restricted_adata" title="omicverse.space.GASTON.get_restricted_adata">get_restricted_adata</a></span>
<span class="mkapi-dash">—</span>
Create AnnData object from restricted data.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_gene_pwlinear" title="omicverse.space.GASTON.plot_gene_pwlinear">plot_gene_pwlinear</a></span>
<span class="mkapi-dash">—</span>
Plot piecewise linear fit of gene expression.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_gene_raw" title="omicverse.space.GASTON.plot_gene_raw">plot_gene_raw</a></span>
<span class="mkapi-dash">—</span>
Plot raw gene expression in spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.GASTON.plot_gene_gastonrex" title="omicverse.space.GASTON.plot_gene_gastonrex">plot_gene_gastonrex</a></span>
<span class="mkapi-dash">—</span>
Plot gene expression with GASTON-specific visualization.</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.get_gaston_input">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.get_gaston_input" title="omicverse.space.GASTON.get_gaston_input">get_gaston_input</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">get_rgb</span><span class="mkapi-equal">=</span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spot_umi_threshold</span><span class="mkapi-equal">=</span><span class="mkapi-default">50</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Prepare input data for GASTON analysis.</p>
<p>This method processes the input AnnData object to extract necessary data
for GASTON analysis, including gene expression counts, spatial coordinates,
and optionally RGB features from histology images.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">get_rgb</span>
<span class="mkapi-dash">—</span>
bool, optional (default=False)
    Whether to extract RGB features from histology image.
spot_umi_threshold: int, optional (default=50)
    Minimum UMI count threshold for filtering spots.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
tuple
    If get_rgb=False:
        - counts_mat: Gene expression count matrix
        - coords_mat: Spatial coordinates array
        - gene_labels: Array of gene names
    If get_rgb=True:
        - counts_mat: Gene expression count matrix
        - coords_mat: Spatial coordinates array
        - gene_labels: Array of gene names
        - RGB_mean: Array of mean RGB values per spot</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Filters spots based on UMI count threshold</li>
<li>RGB features are extracted from high-resolution images if available</li>
<li>RGB values are normalized to [0,1] range</li>
<li>Coordinates are converted to float type</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.get_top_pearson_residuals">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.get_top_pearson_residuals" title="omicverse.space.GASTON.get_top_pearson_residuals">get_top_pearson_residuals</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">num_dims</span><span class="mkapi-equal">=</span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">clip</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.01</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_top_genes</span><span class="mkapi-equal">=</span><span class="mkapi-default">5000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">use_RGB</span><span class="mkapi-equal">=</span><span class="mkapi-default">False</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Get top Pearson residual features for GASTON analysis.</p>
<p>This method selects highly variable genes using Pearson residuals,
performs dimensionality reduction, and optionally combines with RGB features.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">num_dims</span>
<span class="mkapi-dash">—</span>
int, optional (default=5)
    Number of PCA dimensions to retain.
clip: float, optional (default=0.01)
    Clipping value for Pearson residuals to handle outliers.
n_top_genes: int, optional (default=5000)
    Number of highly variable genes to select.
use_RGB: bool, optional (default=False)
    Whether to include RGB features in output matrix.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
numpy.ndarray
    Feature matrix combining:
    - PCA of top Pearson residual genes
    - RGB features (if use_RGB=True)</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Uses Pearson residuals for robust feature selection</li>
<li>Applies sqrt-normalization before PCA</li>
<li>RGB features are concatenated after PCA if use_RGB=True</li>
<li>Features are used as input for neural network training</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.load_rescale">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.load_rescale" title="omicverse.space.GASTON.load_rescale">load_rescale</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">A</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Load and rescale input data for neural network training.</p>
<p>This method prepares the feature matrix and spatial coordinates for
neural network training by rescaling them to appropriate ranges.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">A</span>
<span class="mkapi-dash">—</span>
numpy.ndarray
Feature matrix from get_top_pearson_residuals().</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Converts data to PyTorch tensors</li>
<li>Rescales features to standardized range</li>
<li>Stores results in self.S_torch and self.A_torch</li>
<li>Required before calling train()</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.train">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.train" title="omicverse.space.GASTON.train">train</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">isodepth_arch</span><span class="mkapi-equal">=</span><span class="mkapi-default">[20, 20]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">expression_fn_arch</span><span class="mkapi-equal">=</span><span class="mkapi-default">[20, 20]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_epochs</span><span class="mkapi-equal">=</span><span class="mkapi-default">10000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">checkpoint</span><span class="mkapi-equal">=</span><span class="mkapi-default">500</span><span class="mkapi-comma">, </span><span class="mkapi-arg">out_dir</span><span class="mkapi-equal">=</span><span class="mkapi-default">'result/test_outputs'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">optimizer</span><span class="mkapi-equal">=</span><span class="mkapi-default">'adam'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_restarts</span><span class="mkapi-equal">=</span><span class="mkapi-default">30</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Train GASTON neural network models.</p>
<p>This method trains neural networks to learn spatial depth functions and
gene expression patterns. It performs multiple training runs with different
random initializations to ensure robust results.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">isodepth_arch</span>
<span class="mkapi-dash">—</span>
list, optional (default=[20,20])
Architecture for isodepth neural network d(x,y).
Each element specifies number of neurons in a hidden layer.</p>
</li>
<li>
<p><span class="mkapi-item-name">expression_fn_arch</span>
<span class="mkapi-dash">—</span>
list, optional (default=[20,20])
Architecture for expression function network h(w).
Each element specifies number of neurons in a hidden layer.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_epochs</span>
<span class="mkapi-dash">—</span>
int, optional (default=10000)
Number of training epochs per restart.</p>
</li>
<li>
<p><span class="mkapi-item-name">checkpoint</span>
<span class="mkapi-dash">—</span>
int, optional (default=500)
Save model checkpoint every N epochs.</p>
</li>
<li>
<p><span class="mkapi-item-name">out_dir</span>
<span class="mkapi-dash">—</span>
str, optional (default='result/test_outputs')
Directory to save model files and checkpoints.</p>
</li>
<li>
<p><span class="mkapi-item-name">optimizer</span>
<span class="mkapi-dash">—</span>
str, optional (default="adam")
Optimization algorithm to use.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_restarts</span>
<span class="mkapi-dash">—</span>
int, optional (default=30)
Number of training runs with different random seeds.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Multiple restarts help avoid local optima</li>
<li>Models are saved periodically during training</li>
<li>Each restart uses a different random initialization</li>
<li>Best model can be selected using get_best_model()</li>
<li>Training progress is shown with a progress bar</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.get_best_model">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.get_best_model" title="omicverse.space.GASTON.get_best_model">get_best_model</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">out_dir</span><span class="mkapi-equal">=</span><span class="mkapi-default">'result/test_outputs'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_domain_num</span><span class="mkapi-equal">=</span><span class="mkapi-default">8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">start_from</span><span class="mkapi-equal">=</span><span class="mkapi-default">2</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Select best GASTON model and determine optimal domain number.</p>
<p>This method analyzes trained models to select the best performing one
and helps determine the optimal number of spatial domains.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">out_dir</span>
<span class="mkapi-dash">—</span>
str, optional (default='result/test_outputs')
    Directory containing trained model files.
max_domain_num: int, optional (default=8)
    Maximum number of domains to test.
start_from: int, optional (default=2)
    Minimum number of domains to consider.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
tuple
    (best_model, features, coordinates):
    - best_model: Selected GASTON model
    - features: Feature matrix used for training
    - coordinates: Spatial coordinates</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Uses likelihood curves to determine optimal domain number</li>
<li>Plots model selection diagnostics</li>
<li>Stores selected model in self.model</li>
<li>Required before calling cal_iso_depth()</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.cal_iso_depth">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.cal_iso_depth" title="omicverse.space.GASTON.cal_iso_depth">cal_iso_depth</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">num_domains</span><span class="mkapi-equal">=</span><span class="mkapi-default">10</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate isodepth values and domain labels.</p>
<p>This method uses the trained model to compute continuous spatial depth
values and assign spots to discrete spatial domains.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">num_domains</span>
<span class="mkapi-dash">—</span>
int, optional (default=10)
Number of spatial domains to identify.</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span>
<span class="mkapi-dash">—</span>
tuple
    (isodepth_values, domain_labels):
    - isodepth_values: Continuous spatial depth per spot
    - domain_labels: Discrete domain assignments</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Isodepth values represent continuous spatial organization</li>
<li>Domain labels are discrete assignments (0 to num_domains-1)</li>
<li>Results are stored in self.gaston_isodepth and self.gaston_labels</li>
<li>Values are adjusted so higher numbers indicate deeper layers</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_isodepth">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_isodepth" title="omicverse.space.GASTON.plot_isodepth">plot_isodepth</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">show_streamlines</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">arrowsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(7, 6)</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot isodepth contours and streamlines.</p>
<p>This method creates visualizations of the learned spatial organization
including isodepth contours and optional flow streamlines.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">show_streamlines</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to show directional flow streamlines.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">arrowsize</span>
<span class="mkapi-dash">—</span>
float, optional (default=2)
Size of streamline arrows.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(7,6))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Contours show lines of constant spatial depth</li>
<li>Streamlines indicate gradient directions</li>
<li>Colors represent depth values</li>
<li>Useful for understanding tissue organization</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_clusters">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_clusters" title="omicverse.space.GASTON.plot_clusters">plot_clusters</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">domain_colors</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(6, 6)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">s</span><span class="mkapi-equal">=</span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lgd</span><span class="mkapi-equal">=</span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">show_boundary</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">boundary_lw</span><span class="mkapi-equal">=</span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot spatial domains with cluster colors.</p>
<p>This method visualizes the identified spatial domains by coloring spots
according to their domain assignments.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">domain_colors</span>
<span class="mkapi-dash">—</span>
dict or list
Colors for each spatial domain.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(6,6))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">s</span>
<span class="mkapi-dash">—</span>
float, optional (default=20)
Size of spots in scatter plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">lgd</span>
<span class="mkapi-dash">—</span>
bool, optional (default=False)
Whether to show legend.</p>
</li>
<li>
<p><span class="mkapi-item-name">show_boundary</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to show domain boundaries.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">boundary_lw</span>
<span class="mkapi-dash">—</span>
float, optional (default=5)
Line width for domain boundaries.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the domain visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Each domain is shown in a different color</li>
<li>Boundaries between domains can be highlighted</li>
<li>Spots are colored by their domain assignment</li>
<li>Useful for visualizing discrete spatial structure</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_clusters_restrict">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_clusters_restrict" title="omicverse.space.GASTON.plot_clusters_restrict">plot_clusters_restrict</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">domain_colors</span><span class="mkapi-comma">, </span><span class="mkapi-arg">isodepth_min</span><span class="mkapi-equal">=</span><span class="mkapi-default">4.5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">isodepth_max</span><span class="mkapi-equal">=</span><span class="mkapi-default">6.8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">s</span><span class="mkapi-equal">=</span><span class="mkapi-default">20</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lgd</span><span class="mkapi-equal">=</span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(6, 6)</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot spatial domains with restricted isodepth range.</p>
<p>This method visualizes spatial domains while focusing on a specific range
of isodepth values, useful for examining particular tissue layers.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">domain_colors</span>
<span class="mkapi-dash">—</span>
dict or list
Colors for each spatial domain.</p>
</li>
<li>
<p><span class="mkapi-item-name">isodepth_min</span>
<span class="mkapi-dash">—</span>
float, optional (default=4.5)
Minimum isodepth value to include.</p>
</li>
<li>
<p><span class="mkapi-item-name">isodepth_max</span>
<span class="mkapi-dash">—</span>
float, optional (default=6.8)
Maximum isodepth value to include.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">s</span>
<span class="mkapi-dash">—</span>
float, optional (default=20)
Size of spots in scatter plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">lgd</span>
<span class="mkapi-dash">—</span>
bool, optional (default=False)
Whether to show legend.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(6,6))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the restricted domain visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Only shows spots within specified isodepth range</li>
<li>Useful for focusing on specific tissue layers</li>
<li>Maintains domain color scheme from plot_clusters()</li>
<li>Helps visualize layer-specific patterns</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.restrict_spot">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.restrict_spot" title="omicverse.space.GASTON.restrict_spot">restrict_spot</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">isodepth_min</span><span class="mkapi-equal">=</span><span class="mkapi-default">4.5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">isodepth_max</span><span class="mkapi-equal">=</span><span class="mkapi-default">6.8</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adjust_physical</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">scale_factor</span><span class="mkapi-equal">=</span><span class="mkapi-default">100</span><span class="mkapi-comma">, </span><span class="mkapi-arg">plotisodepth</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">show_streamlines</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">arrowsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(6, 3)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">neg_gradient</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Restrict analysis to spots within specific isodepth range.</p>
<p>This method filters spots based on isodepth values and optionally adjusts
for physical distances, useful for layer-specific analyses.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">isodepth_min</span>
<span class="mkapi-dash">—</span>
float, optional (default=4.5)
Minimum isodepth value to include.</p>
</li>
<li>
<p><span class="mkapi-item-name">isodepth_max</span>
<span class="mkapi-dash">—</span>
float, optional (default=6.8)
Maximum isodepth value to include.</p>
</li>
<li>
<p><span class="mkapi-item-name">adjust_physical</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to adjust for physical distances.</p>
</li>
<li>
<p><span class="mkapi-item-name">scale_factor</span>
<span class="mkapi-dash">—</span>
float, optional (default=100)
Scale factor for physical distance adjustment.</p>
</li>
<li>
<p><span class="mkapi-item-name">plotisodepth</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to plot isodepth contours.</p>
</li>
<li>
<p><span class="mkapi-item-name">show_streamlines</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to show gradient streamlines.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">arrowsize</span>
<span class="mkapi-dash">—</span>
float, optional (default=1)
Size of streamline arrows.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(6,3))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">neg_gradient</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to negate gradient direction.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">tuple</span>
<span class="mkapi-dash">—</span>
(counts, coords, isodepth, labels, features):</p>
<ul>
<li>counts: Filtered count matrix</li>
<li>coords: Filtered spatial coordinates</li>
<li>isodepth: Filtered isodepth values</li>
<li>labels: Filtered domain labels</li>
<li>features: Filtered feature matrix</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Results are stored in object attributes with '_restrict' suffix</li>
<li>Physical distance adjustment helps maintain spatial relationships</li>
<li>Visualization shows restricted region context</li>
<li>Useful for focused analysis of specific tissue layers</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.filter_genes">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.filter_genes" title="omicverse.space.GASTON.filter_genes">filter_genes</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">umi_thresh</span><span class="mkapi-equal">=</span><span class="mkapi-default">1000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">exclude_prefix</span><span class="mkapi-equal">=</span><span class="mkapi-default">['Mt-', 'Rpl', 'Rps']</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Filter genes based on expression and name patterns.</p>
<p>This method removes genes with low expression and those matching specified
prefixes (e.g., mitochondrial and ribosomal genes).</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">umi_thresh</span>
<span class="mkapi-dash">—</span>
int, optional (default=1000)
Minimum total UMI count threshold for genes.</p>
</li>
<li>
<p><span class="mkapi-item-name">exclude_prefix</span>
<span class="mkapi-dash">—</span>
list, optional (default=['Mt-', 'Rpl', 'Rps'])
Gene name prefixes to exclude.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">tuple</span>
<span class="mkapi-dash">—</span>
(gene_labels, gene_indices):</p>
<ul>
<li>gene_labels: Filtered gene names</li>
<li>gene_indices: Indices of kept genes</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Removes low-expressed genes</li>
<li>Excludes common housekeeping genes</li>
<li>Useful for focusing on tissue-specific genes</li>
<li>Results can be used for downstream analysis</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.pw_linear_fit">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.pw_linear_fit" title="omicverse.space.GASTON.pw_linear_fit">pw_linear_fit</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">cell_type_df</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ct_list</span><span class="mkapi-equal">=</span><span class="mkapi-default">[]</span><span class="mkapi-comma">, </span><span class="mkapi-arg">isodepth_mult_factor</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.01</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Perform piecewise linear fitting of gene expression.</p>
<p>This method fits piecewise linear functions to gene expression patterns
along the spatial depth gradient.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">cell_type_df</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame, optional (default=None)
Cell type annotations if available.</p>
</li>
<li>
<p><span class="mkapi-item-name">ct_list</span>
<span class="mkapi-dash">—</span>
list, optional (default=[])
List of cell types to analyze.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments for piecewise fitting.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>dict
    Dictionary containing fitted parameters and statistics.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Models expression changes along spatial axis</li>
<li>Can incorporate cell type information</li>
<li>Identifies expression breakpoints</li>
<li>Useful for finding spatial transition points</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.bin_data">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.bin_data" title="omicverse.space.GASTON.bin_data">bin_data</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">cell_type_df</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">num_bins</span><span class="mkapi-equal">=</span><span class="mkapi-default">15</span><span class="mkapi-comma">, </span><span class="mkapi-arg">q_discont</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.95</span><span class="mkapi-comma">, </span><span class="mkapi-arg">q_cont</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.8</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Bin data along spatial depth gradient.</p>
<p>This method bins spots and their expression data based on isodepth values,
useful for analyzing trends along spatial axes.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">cell_type_df</span>
<span class="mkapi-dash">—</span>
pandas.DataFrame, optional (default=None)
Cell type annotations if available.</p>
</li>
<li>
<p><span class="mkapi-item-name">num_bins</span>
<span class="mkapi-dash">—</span>
int, optional (default=15)
Number of spatial bins to create.</p>
</li>
<li>
<p><span class="mkapi-item-name">q_discont</span>
<span class="mkapi-dash">—</span>
float, optional (default=0.95)
Quantile threshold for discontinuous patterns.</p>
</li>
<li>
<p><span class="mkapi-item-name">q_cont</span>
<span class="mkapi-dash">—</span>
float, optional (default=0.8)
Quantile threshold for continuous patterns.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments for binning process.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">tuple</span>
<span class="mkapi-dash">—</span>
(binned_data, bin_edges, statistics):</p>
<ul>
<li>binned_data: Expression data per bin</li>
<li>bin_edges: Isodepth values defining bins</li>
<li>statistics: Summary statistics per bin</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Creates equal-width bins along isodepth axis</li>
<li>Computes statistics within each bin</li>
<li>Can incorporate cell type information</li>
<li>Useful for trajectory analysis</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.get_restricted_adata">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.get_restricted_adata" title="omicverse.space.GASTON.get_restricted_adata">get_restricted_adata</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">offset</span><span class="mkapi-equal">=</span><span class="mkapi-default">10 ** 6</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Create AnnData object from restricted data.</p>
<p>This method constructs a new AnnData object containing only the spots
within the restricted isodepth range.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">offset</span>
<span class="mkapi-dash">—</span>
float, optional (default=10**6)
Offset for coordinate normalization.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>anndata.AnnData
    AnnData object containing restricted data.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Preserves original data structure</li>
<li>Contains only filtered spots</li>
<li>Maintains spatial coordinates</li>
<li>Useful for downstream analysis</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_gene_pwlinear">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_gene_pwlinear" title="omicverse.space.GASTON.plot_gene_pwlinear">plot_gene_pwlinear</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene</span><span class="mkapi-comma">, </span><span class="mkapi-arg">domain_colors</span><span class="mkapi-comma">, </span><span class="mkapi-arg">offset</span><span class="mkapi-equal">=</span><span class="mkapi-default">10 ** 6</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cell_type_list</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">pt_size</span><span class="mkapi-equal">=</span><span class="mkapi-default">50</span><span class="mkapi-comma">, </span><span class="mkapi-arg">linear_fit</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">ticksize</span><span class="mkapi-equal">=</span><span class="mkapi-default">15</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(4, 2.5)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">lw</span><span class="mkapi-equal">=</span><span class="mkapi-default">3</span><span class="mkapi-comma">, </span><span class="mkapi-arg">domain_boundary_plotting</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot piecewise linear fit of gene expression.</p>
<p>This method visualizes gene expression patterns and their piecewise
linear fits along the spatial depth gradient.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene</span>
<span class="mkapi-dash">—</span>
str
Name of gene to plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">domain_colors</span>
<span class="mkapi-dash">—</span>
dict or list
Colors for each spatial domain.</p>
</li>
<li>
<p><span class="mkapi-item-name">offset</span>
<span class="mkapi-dash">—</span>
float, optional (default=10**6)
Offset for coordinate normalization.</p>
</li>
<li>
<p><span class="mkapi-item-name">cell_type_list</span>
<span class="mkapi-dash">—</span>
list, optional (default=None)
List of cell types to include.</p>
</li>
<li>
<p><span class="mkapi-item-name">pt_size</span>
<span class="mkapi-dash">—</span>
float, optional (default=50)
Size of scatter points.</p>
</li>
<li>
<p><span class="mkapi-item-name">linear_fit</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to show linear fit lines.</p>
</li>
<li>
<p><span class="mkapi-item-name">ticksize</span>
<span class="mkapi-dash">—</span>
float, optional (default=15)
Size of axis tick labels.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(4,2.5))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">lw</span>
<span class="mkapi-dash">—</span>
float, optional (default=3)
Line width for fits.</p>
</li>
<li>
<p><span class="mkapi-item-name">domain_boundary_plotting</span>
<span class="mkapi-dash">—</span>
bool, optional (default=True)
Whether to show domain boundaries.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Shows expression vs isodepth relationship</li>
<li>Highlights transition points</li>
<li>Can show multiple cell types</li>
<li>Useful for understanding spatial patterns</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_gene_raw">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_gene_raw" title="omicverse.space.GASTON.plot_gene_raw">plot_gene_raw</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_name</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">vmin</span><span class="mkapi-equal">=</span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(6, 3)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">s</span><span class="mkapi-equal">=</span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot raw gene expression in spatial coordinates.</p>
<p>This method creates a spatial visualization of raw gene expression levels
across the tissue section.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_name</span>
<span class="mkapi-dash">—</span>
str
Name of gene to plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">vmin</span>
<span class="mkapi-dash">—</span>
float, optional (default=5)
Minimum value for color scale.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(6,3))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">s</span>
<span class="mkapi-dash">—</span>
float, optional (default=10)
Size of scatter points.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the expression visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Shows spatial distribution of expression</li>
<li>Uses continuous color scale</li>
<li>Maintains tissue orientation</li>
<li>Useful for examining expression patterns</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.GASTON.plot_gene_gastonrex">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.GASTON" title="omicverse.space.GASTON">GASTON</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.GASTON.plot_gene_gastonrex" title="omicverse.space.GASTON.plot_gene_gastonrex">plot_gene_gastonrex</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">gene_name</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rotate_angle</span><span class="mkapi-equal">=</span><span class="mkapi-default">-90</span><span class="mkapi-comma">, </span><span class="mkapi-arg">figsize</span><span class="mkapi-equal">=</span><span class="mkapi-default">(6, 3)</span><span class="mkapi-comma">, </span><span class="mkapi-arg">s</span><span class="mkapi-equal">=</span><span class="mkapi-default">10</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Plot gene expression with GASTON-specific visualization.</p>
<p>This method creates a specialized visualization of gene expression using
GASTON's representation enhancement.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">gene_name</span>
<span class="mkapi-dash">—</span>
str
Name of gene to plot.</p>
</li>
<li>
<p><span class="mkapi-item-name">rotate_angle</span>
<span class="mkapi-dash">—</span>
float, optional (default=-90)
Rotation angle in degrees for plot orientation.</p>
</li>
<li>
<p><span class="mkapi-item-name">figsize</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=(6,3))
Figure size in inches.</p>
</li>
<li>
<p><span class="mkapi-item-name">s</span>
<span class="mkapi-dash">—</span>
float, optional (default=10)
Size of scatter points.</p>
</li>
<li>
<p><span class="mkapi-item-name">**kwargs</span>
<span class="mkapi-dash">—</span>
Additional arguments passed to plotting functions.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>matplotlib.figure.Figure
    Figure containing the enhanced visualization.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Uses GASTON-specific representation</li>
<li>Enhances spatial patterns</li>
<li>Maintains tissue orientation</li>
<li>Useful for detailed pattern analysis</li>
</ul>
</div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.create_communication_anndata">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.create_communication_anndata" title="omicverse.space.create_communication_anndata">create_communication_anndata</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">clustering_column</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_permutations</span><span class="mkapi-equal">=</span><span class="mkapi-default">100</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Create AnnData object with communication results in CellChat format</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span> : 
<span class="mkapi-item-type">AnnData</span> <span class="mkapi-dash">—</span>
Input spatial data with commot results</p>
</li>
<li>
<p><span class="mkapi-item-name">clustering_column</span> : 
<span class="mkapi-item-type">str</span> <span class="mkapi-dash">—</span>
Column name for cell type clustering</p>
</li>
<li>
<p><span class="mkapi-item-name">n_permutations</span> : 
<span class="mkapi-item-type">int</span> <span class="mkapi-dash">—</span>
Number of permutations for p-value calculation</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span></p>
</li>
<li>
<p><span class="mkapi-item-name">-------</span></p>
</li>
<li>
<p><span class="mkapi-item-name">comm_adata</span> : 
<span class="mkapi-item-type">AnnData</span> <span class="mkapi-dash">—</span>
Communication results with structure:</p>
<ul>
<li>obs: cell type pairs ('celltype1|celltype2')  </li>
<li>var: ligand-receptor pairs with metadata</li>
<li>layers: 'pvalues' and 'means'</li>
</ul>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.update_classification_from_database">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.update_classification_from_database" title="omicverse.space.update_classification_from_database">update_classification_from_database</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">comm_adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_with_db</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Update classification in comm_adata using database info from original adata</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">comm_adata</span> : 
<span class="mkapi-item-type">AnnData</span> <span class="mkapi-dash">—</span>
Communication AnnData object to update</p>
</li>
<li>
<p><span class="mkapi-item-name">adata_with_db</span> : 
<span class="mkapi-item-type">AnnData</span> <span class="mkapi-dash">—</span>
Original adata containing commot database info</p>
</li>
<li>
<p><span class="mkapi-item-name">Returns</span></p>
</li>
<li>
<p><span class="mkapi-item-name">-------</span></p>
</li>
<li>
<p><span class="mkapi-item-name">comm_adata</span> : 
<span class="mkapi-item-type">AnnData</span> <span class="mkapi-dash">—</span>
Updated communication AnnData with proper classifications</p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">class</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_sp</span><span class="mkapi-comma">, </span><span class="mkapi-arg">adata_sc</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p><p class="mkapi-bases" markdown="1">Bases :
<span class="mkapi-base">object</span></p></p>
<p>Initialize Deconvolution object.
Parameters</p>
<hr/>
<p>adata_sp : AnnData
    Spatial transcriptomics data.
adata_sc : AnnData, optional
    Single-cell RNA-seq data, by default None.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle methods">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Methods</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.preprocess_sc" title="omicverse.space.Deconvolution.preprocess_sc">preprocess_sc</a></span>
<span class="mkapi-dash">—</span>
Preprocess the scRNA-seq data.
Parameters</p>
<hr/>
<p>mode : str, optional
    Preprocessing mode, by default 'shiftlog|pearson'.
n_HVGs : int, optional
    Number of highly variable genes, by default 3000.
target_sum : int, optional
    Target sum for preprocessing, by default 1e4.
**kwargs : dict, optional
    Additional keyword arguments for preprocessing.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.preprocess_sp" title="omicverse.space.Deconvolution.preprocess_sp">preprocess_sp</a></span>
<span class="mkapi-dash">—</span>
Preprocess the spatial transcriptomics data.
Parameters</p>
<hr/>
<p>mode : str, optional
    Preprocessing mode, by default 'pearsonr'.
n_svgs : int, optional
    Number of space variable genes, by default 3000.
target_sum : int, optional
    Target sum for preprocessing, by default 50<em>1e4.  <br/>
platform : str, optional
    Platform of the spatial transcriptomics data, by default 'visium'.
mt_startwith : str, optional
    Start with of the mitochondrial genes, by default 'MT-'.
</em>*kwargs : dict, optional
    Additional keyword arguments for preprocessing.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.deconvolution" title="omicverse.space.Deconvolution.deconvolution">deconvolution</a></span>
<span class="mkapi-dash">—</span>
Deconvolution the spatial transcriptomics data.
Parameters</p>
<hr/>
<p>method : str, optional
    Deconvolution method, by default 'Tangram'.
celltype_key_sc : str, optional
    Cell type key in scRNA-seq data, by default 'cell_type'.
batch_key_sc : str, optional
    Batch key in scRNA-seq data, by default None.
batch_key_sp : str, optional
    Batch key in spatial transcriptomics data, by default None.
tangram_kwargs : dict, optional
    Keyword arguments for Tangram, by default None.
cell2location_scrna_kwargs : dict, optional
    Keyword arguments for cell2location, by default None.
cell2location_spatial_kwargs : dict, optional
    Keyword arguments for cell2location, by default None.
N_cells_per_location : int, optional
    Number of cells per location, by default 30.
detection_alpha : int, optional
    Detection alpha, by default 200.
sample_kwargs : dict, optional
    Keyword arguments for sampling, by default None.
flashdeconv_kwargs : dict, optional
    Keyword arguments for FlashDeconv, by default None.
starfysh_kwargs : dict, optional
    Keyword arguments for starfysh, by default None.
spatial_type : str, optional
    Spatial type, by default 'visium'.
gene_sig : dict, optional
    Gene signature, by default None.</p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.impute" title="omicverse.space.Deconvolution.impute">impute</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.load_cell2location_model" title="omicverse.space.Deconvolution.load_cell2location_model">load_cell2location_model</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.cell2location_inference" title="omicverse.space.Deconvolution.cell2location_inference">cell2location_inference</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.load_tangram_model" title="omicverse.space.Deconvolution.load_tangram_model">load_tangram_model</a></span></p>
</li>
<li>
<p><span class="mkapi-item-name"><a href="./#omicverse.space.Deconvolution.tangram_inference" title="omicverse.space.Deconvolution.tangram_inference">tangram_inference</a></span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.preprocess_sc">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.preprocess_sc" title="omicverse.space.Deconvolution.preprocess_sc">preprocess_sc</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">mode</span><span class="mkapi-equal">=</span><span class="mkapi-default">'shiftlog|pearson'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_HVGs</span><span class="mkapi-equal">=</span><span class="mkapi-default">3000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">target_sum</span><span class="mkapi-equal">=</span><span class="mkapi-default">10000.0</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Preprocess the scRNA-seq data.
Parameters</p>
<hr/>
<p>mode : str, optional
    Preprocessing mode, by default 'shiftlog|pearson'.
n_HVGs : int, optional
    Number of highly variable genes, by default 3000.
target_sum : int, optional
    Target sum for preprocessing, by default 1e4.
**kwargs : dict, optional
    Additional keyword arguments for preprocessing.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.preprocess_sp">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.preprocess_sp" title="omicverse.space.Deconvolution.preprocess_sp">preprocess_sp</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">mode</span><span class="mkapi-equal">=</span><span class="mkapi-default">'pearsonr'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">n_svgs</span><span class="mkapi-equal">=</span><span class="mkapi-default">3000</span><span class="mkapi-comma">, </span><span class="mkapi-arg">target_sum</span><span class="mkapi-equal">=</span><span class="mkapi-default">50 * 10000.0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">platform</span><span class="mkapi-equal">=</span><span class="mkapi-default">'visium'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mt_startwith</span><span class="mkapi-equal">=</span><span class="mkapi-default">'MT-'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">subset_genes</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Preprocess the spatial transcriptomics data.
Parameters</p>
<hr/>
<p>mode : str, optional
    Preprocessing mode, by default 'pearsonr'.
n_svgs : int, optional
    Number of space variable genes, by default 3000.
target_sum : int, optional
    Target sum for preprocessing, by default 50<em>1e4.  <br/>
platform : str, optional
    Platform of the spatial transcriptomics data, by default 'visium'.
mt_startwith : str, optional
    Start with of the mitochondrial genes, by default 'MT-'.
</em>*kwargs : dict, optional
    Additional keyword arguments for preprocessing.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.deconvolution">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.deconvolution" title="omicverse.space.Deconvolution.deconvolution">deconvolution</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">method</span><span class="mkapi-equal">=</span><span class="mkapi-default">'Tangram'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">celltype_key_sc</span><span class="mkapi-equal">=</span><span class="mkapi-default">'cell_type'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_key_sc</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">batch_key_sp</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">tangram_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cell2location_scrna_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">cell2location_spatial_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">N_cells_per_location</span><span class="mkapi-equal">=</span><span class="mkapi-default">30</span><span class="mkapi-comma">, </span><span class="mkapi-arg">detection_alpha</span><span class="mkapi-equal">=</span><span class="mkapi-default">200</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sample_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">flashdeconv_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">starfysh_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_type</span><span class="mkapi-equal">=</span><span class="mkapi-default">'visium'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gene_sig</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Deconvolution the spatial transcriptomics data.
Parameters</p>
<hr/>
<p>method : str, optional
    Deconvolution method, by default 'Tangram'.
celltype_key_sc : str, optional
    Cell type key in scRNA-seq data, by default 'cell_type'.
batch_key_sc : str, optional
    Batch key in scRNA-seq data, by default None.
batch_key_sp : str, optional
    Batch key in spatial transcriptomics data, by default None.
tangram_kwargs : dict, optional
    Keyword arguments for Tangram, by default None.
cell2location_scrna_kwargs : dict, optional
    Keyword arguments for cell2location, by default None.
cell2location_spatial_kwargs : dict, optional
    Keyword arguments for cell2location, by default None.
N_cells_per_location : int, optional
    Number of cells per location, by default 30.
detection_alpha : int, optional
    Detection alpha, by default 200.
sample_kwargs : dict, optional
    Keyword arguments for sampling, by default None.
flashdeconv_kwargs : dict, optional
    Keyword arguments for FlashDeconv, by default None.
starfysh_kwargs : dict, optional
    Keyword arguments for starfysh, by default None.
spatial_type : str, optional
    Spatial type, by default 'visium'.
gene_sig : dict, optional
    Gene signature, by default None.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
<li>
<p><span class="mkapi-item-type">ImportError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.impute">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.impute" title="omicverse.space.Deconvolution.impute">impute</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">method</span><span class="mkapi-equal">=</span><span class="mkapi-default">'Tangram'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.load_cell2location_model">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.load_cell2location_model" title="omicverse.space.Deconvolution.load_cell2location_model">load_cell2location_model</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">mod_sp_path</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.cell2location_inference">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.cell2location_inference" title="omicverse.space.Deconvolution.cell2location_inference">cell2location_inference</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">sample_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.load_tangram_model">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.load_tangram_model" title="omicverse.space.Deconvolution.load_tangram_model">load_tangram_model</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">model_path</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.Deconvolution.tangram_inference">
<span class="mkapi-object-link">
<span class="mkapi-parent-toggle" title="Toggle class names">
<i class="fa-solid fa-square-plus"></i>
</span>
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span><span class="mkapi-object-kind">method</span>
<span class="mkapi-object-parent">
<a href="./#omicverse.space.Deconvolution" title="omicverse.space.Deconvolution">Deconvolution</a><span class="mkapi-dot">.</span></span><span class="mkapi-object-name"><a href="./#omicverse.space.Deconvolution.tangram_inference" title="omicverse.space.Deconvolution.tangram_inference">tangram_inference</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">sample_kwargs</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.calculate_gene_signature">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.calculate_gene_signature" title="omicverse.space.calculate_gene_signature">calculate_gene_signature</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_sc</span><span class="mkapi-comma">, </span><span class="mkapi-arg">clustertype</span><span class="mkapi-comma">, </span><span class="mkapi-arg">rank</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'rank_genes_groups'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">foldchange</span><span class="mkapi-equal">=</span><span class="mkapi-default">2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">topgenenumber</span><span class="mkapi-equal">=</span><span class="mkapi-default">20</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate the gene signature.
Parameters</p>
<hr/>
<p>adata_sc : AnnData
    Single-cell RNA-seq data.
clustertype : str
    Cell type key in scRNA-seq data.
rank : bool, optional
    Whether to rank the genes, by default True.
key : str, optional
    Key in scRNA-seq data, by default 'rank_genes_groups'.
foldchange : int, optional
    Fold change, by default 2.
topgenenumber : int, optional
    Number of top genes, by default 20.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.crop_space_visium">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.crop_space_visium" title="omicverse.space.crop_space_visium">crop_space_visium</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">crop_loc</span><span class="mkapi-comma">, </span><span class="mkapi-arg">crop_area</span><span class="mkapi-comma">, </span><span class="mkapi-arg">library_id</span><span class="mkapi-comma">, </span><span class="mkapi-arg">scale</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'spatial'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">res</span><span class="mkapi-equal">=</span><span class="mkapi-default">'hires'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Crop Visium spatial data to a specific region of interest.</p>
<p>This function allows cropping of Visium spatial transcriptomics data to focus on
a specific region while maintaining proper scaling and coordinate systems.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing Visium spatial data.</p>
</li>
<li>
<p><span class="mkapi-item-name">crop_loc</span>
<span class="mkapi-dash">—</span>
tuple
(x, y) coordinates for the top-left corner of the crop region.</p>
</li>
<li>
<p><span class="mkapi-item-name">crop_area</span>
<span class="mkapi-dash">—</span>
tuple
(width, height) of the cropping area in spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">library_id</span>
<span class="mkapi-dash">—</span>
str
Library ID for the spatial data in adata.uns['spatial'].</p>
</li>
<li>
<p><span class="mkapi-item-name">scale</span>
<span class="mkapi-dash">—</span>
float
Scale factor for the cropping operation.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatial_key</span>
<span class="mkapi-dash">—</span>
str, optional (default='spatial')
Key in adata.obsm containing spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">res</span>
<span class="mkapi-dash">—</span>
str, optional (default='hires')
Image resolution to use ('hires' or 'lowres').</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>AnnData
    Cropped AnnData object containing only spots within the specified region.
    The spatial coordinates and image are adjusted accordingly.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>The function preserves the original coordinate system scaling</li>
<li>The cropped image is stored in adata.uns['spatial'][library_id]['images'][res]</li>
<li>Coordinates are automatically adjusted to the new cropped region</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load Visium data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Crop a 1000x1000 region starting at (500, 500)</span>
<span class="n">adata_cropped</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">crop_space_visium</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">crop_loc</span><span class="o">=</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="mi">500</span><span class="p">),</span>
    <span class="n">crop_area</span><span class="o">=</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="n">library_id</span><span class="o">=</span><span class="s1">'V1_Human_Brain'</span><span class="p">,</span>
    <span class="n">scale</span><span class="o">=</span><span class="mf">1.0</span>
<span class="p">)</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.rotate_space_visium">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.rotate_space_visium" title="omicverse.space.rotate_space_visium">rotate_space_visium</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">angle</span><span class="mkapi-comma">, </span><span class="mkapi-arg">center</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'spatial'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">res</span><span class="mkapi-equal">=</span><span class="mkapi-default">'hires'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">library_id</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-comma">, </span><span class="mkapi-arg">interpolation_order</span><span class="mkapi-equal">=</span><span class="mkapi-default">1</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Rotate Visium spatial data image and coordinates by a specified angle.</p>
<p>This function performs rotation of both the tissue image and spot coordinates
while maintaining proper alignment and scaling.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing Visium spatial data.</p>
</li>
<li>
<p><span class="mkapi-item-name">angle</span>
<span class="mkapi-dash">—</span>
float
Rotation angle in degrees (positive for counterclockwise).</p>
</li>
<li>
<p><span class="mkapi-item-name">center</span>
<span class="mkapi-dash">—</span>
tuple, optional (default=None)
(x, y) coordinates of rotation center. If None, uses center of spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">spatial_key</span>
<span class="mkapi-dash">—</span>
str, optional (default='spatial')
Key in adata.obsm containing spatial coordinates.</p>
</li>
<li>
<p><span class="mkapi-item-name">res</span>
<span class="mkapi-dash">—</span>
str, optional (default='hires')
Image resolution to use ('hires' or 'lowres').</p>
</li>
<li>
<p><span class="mkapi-item-name">library_id</span>
<span class="mkapi-dash">—</span>
str
Library ID for the spatial data in adata.uns['spatial'].</p>
</li>
<li>
<p><span class="mkapi-item-name">interpolation_order</span>
<span class="mkapi-dash">—</span>
int, optional (default=1)
Order of interpolation for image rotation:
0: nearest neighbor, 1: bilinear, 3: cubic spline.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>AnnData
    Rotated AnnData object with transformed coordinates and image.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>The function preserves the original coordinate system scaling</li>
<li>Both image and spot coordinates are rotated around the same center</li>
<li>The rotation is performed counterclockwise for positive angles</li>
<li>Image interpolation can be adjusted for quality vs speed tradeoff</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load Visium data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Rotate 45 degrees counterclockwise</span>
<span class="n">adata_rotated</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">rotate_space_visium</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">angle</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span>
    <span class="n">library_id</span><span class="o">=</span><span class="s1">'V1_Human_Brain'</span>
<span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.find_image_offset_phase_correlation_array_input">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.find_image_offset_phase_correlation_array_input" title="omicverse.space.find_image_offset_phase_correlation_array_input">find_image_offset_phase_correlation_array_input</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">image1_array</span><span class="mkapi-comma">, </span><span class="mkapi-arg">image2_array</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate image offset using phase correlation method.</p>
<p>This function computes the relative displacement between two images using
Fourier-based phase correlation, which is robust to intensity variations
and noise.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">image1_array</span>
<span class="mkapi-dash">—</span>
numpy.ndarray
First image array (image to be aligned).
Can be grayscale or RGB.</p>
</li>
<li>
<p><span class="mkapi-item-name">image2_array</span>
<span class="mkapi-dash">—</span>
numpy.ndarray
Second image array (reference image).
Can be grayscale or RGB.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">tuple</span>
<span class="mkapi-dash">—</span>
(offset, aligned_image) where:</p>
<ul>
<li>offset: tuple (dx, dy) representing the displacement in pixels</li>
<li>aligned_image: numpy.ndarray of the aligned first image</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Images are automatically converted to grayscale if RGB</li>
<li>The method is based on frequency-domain phase correlation</li>
<li>Works best with images of similar size and content</li>
<li>Handles sub-pixel accuracy for precise alignment</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="c1"># Create sample images</span>
<span class="n">img1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">img2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">img1</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="c1"># Find offset</span>
<span class="n">offset</span><span class="p">,</span> <span class="n">aligned</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">find_image_offset_phase_correlation_array_input</span><span class="p">(</span>
    <span class="n">img1</span><span class="p">,</span> <span class="n">img2</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Detected offset: </span><span class="si">{</span><span class="n">offset</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.find_image_offset_phase_correlation_torch">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.find_image_offset_phase_correlation_torch" title="omicverse.space.find_image_offset_phase_correlation_torch">find_image_offset_phase_correlation_torch</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">image1_tensor</span><span class="mkapi-comma">, </span><span class="mkapi-arg">image2_tensor</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Calculate image offset using phase correlation method with PyTorch tensors.</p>
<p>This function is the PyTorch implementation of phase correlation for image
alignment, offering GPU acceleration when available.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">image1_tensor</span>
<span class="mkapi-dash">—</span>
torch.Tensor
First image as PyTorch tensor (image to be aligned).
Expected shape: (H, W) or (1, H, W) or (3, H, W).</p>
</li>
<li>
<p><span class="mkapi-item-name">image2_tensor</span>
<span class="mkapi-dash">—</span>
torch.Tensor
Second image as PyTorch tensor (reference image).
Expected shape: (H, W) or (1, H, W) or (3, H, W).</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">tuple</span>
<span class="mkapi-dash">—</span>
(offset, aligned_image) where:</p>
<ul>
<li>offset: tuple (dx, dy) representing the displacement in pixels</li>
<li>aligned_image: torch.Tensor of the aligned first image</li>
</ul>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Images are automatically converted to grayscale if RGB</li>
<li>Computation is performed on GPU if available</li>
<li>More efficient than NumPy version for large images when using GPU</li>
<li>Maintains sub-pixel accuracy for precise alignment</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="c1"># Create sample images</span>
<span class="n">img1</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">img2</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">img1</span><span class="p">,</span> <span class="n">shifts</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="n">dims</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="c1"># Find offset</span>
<span class="n">offset</span><span class="p">,</span> <span class="n">aligned</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">find_image_offset_phase_correlation_torch</span><span class="p">(</span>
    <span class="n">img1</span><span class="p">,</span> <span class="n">img2</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Detected offset: </span><span class="si">{</span><span class="n">offset</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.map_spatial_auto">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.map_spatial_auto" title="omicverse.space.map_spatial_auto">map_spatial_auto</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_rotated</span><span class="mkapi-comma">, </span><span class="mkapi-arg">method</span><span class="mkapi-equal">=</span><span class="mkapi-default">'phase'</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Automatically map and align spatial transcriptomics data.</p>
<p>This function performs automatic alignment of spatial transcriptomics data
with the corresponding tissue image using various alignment methods.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_rotated</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing spatial data to be aligned.</p>
</li>
<li>
<p><span class="mkapi-item-name">method</span>
<span class="mkapi-dash">—</span>
str, optional (default='phase')
Alignment method to use:</p>
<ul>
<li>'phase': Phase correlation-based alignment</li>
<li>'feature': Feature-based alignment</li>
<li>'hybrid': Combination of phase and feature methods</li>
</ul>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>AnnData
    Aligned AnnData object with updated spatial coordinates.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>The function automatically selects the best alignment</li>
<li>Results can be verified using spatial plotting functions</li>
<li>Different methods may work better for different data types</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load and preprocess data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Perform automatic alignment</span>
<span class="n">adata_aligned</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">map_spatial_auto</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">'phase'</span>
<span class="p">)</span>
</code></pre></div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle raises">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Raises</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-type">ValueError</span></p>
</li>
</ul>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.map_spatial_manual">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.map_spatial_manual" title="omicverse.space.map_spatial_manual">map_spatial_manual</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata_rotated</span><span class="mkapi-comma">, </span><span class="mkapi-arg">offset</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Manually adjust spatial transcriptomics data alignment.</p>
<p>This function allows manual adjustment of the alignment between
spatial transcriptomics data and the tissue image using specified offsets.</p>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle parameters">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Parameters</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p><span class="mkapi-item-name">adata_rotated</span>
<span class="mkapi-dash">—</span>
AnnData
Annotated data matrix containing spatial data to be aligned.</p>
</li>
<li>
<p><span class="mkapi-item-name">offset</span>
<span class="mkapi-dash">—</span>
tuple
(dx, dy) tuple specifying the manual offset to apply.</p>
</li>
</ul>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle returns">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Returns</span>
</p>
<div class="mkapi-section-content">
<ul class="mkapi-item-list">
<li>
<p>AnnData
    Aligned AnnData object with manually adjusted spatial coordinates.</p>
</li>
</ul>
</div>
<div class="mkapi-section-content">
<div class="admonition note">
<p class="admonition-title">Notes</p>
<ul>
<li>Useful for fine-tuning automatic alignment results</li>
<li>Offset values are in pixel coordinates</li>
<li>Positive dx moves spots right, positive dy moves spots down</li>
</ul>
</div>
</div>
<p class="mkapi-section">
<span class="mkapi-object-link">
<span class="mkapi-section-toggle" title="Toggle examples">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-section-name">Examples</span>
</p>
<div class="mkapi-section-content">
<div class="mkapi-example-input highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">omicverse</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ov</span>
<span class="c1"># Load data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="c1"># Apply manual offset</span>
<span class="n">adata_aligned</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">map_spatial_manual</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">offset</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">)</span>  <span class="c1"># Move 10 pixels right, 5 pixels up</span>
<span class="p">)</span>
</code></pre></div>
</div>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.read_visium_10x">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.read_visium_10x" title="omicverse.space.read_visium_10x">read_visium_10x</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.visium_10x_hd_cellpose_he">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_he" title="omicverse.space.visium_10x_hd_cellpose_he">visium_10x_hd_cellpose_he</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mpp</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.3</span><span class="mkapi-comma">, </span><span class="mkapi-arg">he_save_path</span><span class="mkapi-equal">=</span><span class="mkapi-default">'stardist/he_colon.tiff'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">prob_thresh</span><span class="mkapi-equal">=</span><span class="mkapi-default">0</span><span class="mkapi-comma">, </span><span class="mkapi-arg">flow_threshold</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.2</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gpu</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">buffer</span><span class="mkapi-equal">=</span><span class="mkapi-default">150</span><span class="mkapi-comma">, </span><span class="mkapi-arg">backend</span><span class="mkapi-equal">=</span><span class="mkapi-default">'tifffile'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document">
<p>Convert Visium 10x data to cell-level data.</p>
</div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.visium_10x_hd_cellpose_expand">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_expand" title="omicverse.space.visium_10x_hd_cellpose_expand">visium_10x_hd_cellpose_expand</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">max_bin_distance</span><span class="mkapi-equal">=</span><span class="mkapi-default">4</span><span class="mkapi-comma">, </span><span class="mkapi-arg">labels_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_he'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">expanded_labels_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_he_expanded'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.visium_10x_hd_cellpose_gex">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.visium_10x_hd_cellpose_gex" title="omicverse.space.visium_10x_hd_cellpose_gex">visium_10x_hd_cellpose_gex</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">obs_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'n_counts_adjusted'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">log1p</span><span class="mkapi-equal">=</span><span class="mkapi-default">False</span><span class="mkapi-comma">, </span><span class="mkapi-arg">mpp</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.3</span><span class="mkapi-comma">, </span><span class="mkapi-arg">sigma</span><span class="mkapi-equal">=</span><span class="mkapi-default">5</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gex_save_path</span><span class="mkapi-equal">=</span><span class="mkapi-default">'stardist/gex_colon.tiff'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">prob_thresh</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.01</span><span class="mkapi-comma">, </span><span class="mkapi-arg">nms_thresh</span><span class="mkapi-equal">=</span><span class="mkapi-default">0.1</span><span class="mkapi-comma">, </span><span class="mkapi-arg">gpu</span><span class="mkapi-equal">=</span><span class="mkapi-default">True</span><span class="mkapi-comma">, </span><span class="mkapi-arg">buffer</span><span class="mkapi-equal">=</span><span class="mkapi-default">150</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.salvage_secondary_labels">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.salvage_secondary_labels" title="omicverse.space.salvage_secondary_labels">salvage_secondary_labels</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">primary_label</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_he'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">secondary_label</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_gex'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">labels_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_joint'</span><span class="mkapi-comma">, </span><span class="mkapi-star">**</span><span class="mkapi-arg">kwargs</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
<p class="mkapi-object mkapi-page-source" id="omicverse.space.bin2cell">
<span class="mkapi-object-link">
<span class="mkapi-definition-link"></span>
<span class="mkapi-document-toggle" title="Toggle all docs">
<i class="fa-regular fa-square-minus"></i>
</span>
</span>
<span class="mkapi-object-name"><a href="./#omicverse.space.bin2cell" title="omicverse.space.bin2cell">bin2cell</a></span><span class="mkapi-signature"><span class="mkapi-paren">(</span><span class="mkapi-arg">adata</span><span class="mkapi-comma">, </span><span class="mkapi-arg">labels_key</span><span class="mkapi-equal">=</span><span class="mkapi-default">'labels_joint'</span><span class="mkapi-comma">, </span><span class="mkapi-arg">spatial_keys</span><span class="mkapi-equal">=</span><span class="mkapi-default">['spatial']</span><span class="mkapi-comma">, </span><span class="mkapi-arg">diameter_scale_factor</span><span class="mkapi-equal">=</span><span class="mkapi-default">None</span><span class="mkapi-paren">)</span></span>
</p>
<div class="mkapi-document"></div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2025, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascript/mkapi.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>