
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api tangram - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.space.Tangram">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api tangram
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology/Taxonomy
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scenic/">
        Gene Regulatory Network Analysis with SCENIC
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell-cell communication using CellPhoneDBViz
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space.Tangram">
    Tangram
  </a>
<nav aria-label="Tangram" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space._tangram.Tangram.__init__">
    __init__()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space._tangram.Tangram.check_tangram">
    check_tangram()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space._tangram.Tangram.train">
    train()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space._tangram.Tangram.cell2location">
    cell2location()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.space._tangram.Tangram.impute">
    impute()
  </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api tangram</h1>
<div class="doc doc-object doc-class">
<h2 class="doc doc-heading" id="omicverse.space.Tangram">
<code>omicverse.space.Tangram</code>
<a class="headerlink" href="#omicverse.space.Tangram" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p class="doc doc-class-bases">
        Bases: <code>object</code></p>
<p>Tangram spatial deconvolution class for cell type mapping.</p>
<p>Tangram is a method for integrating single-cell RNA sequencing (scRNA-seq) data
with spatial transcriptomics data. It enables:
1. Mapping cell types from scRNA-seq to spatial locations
2. Deconvolving cell type proportions in spatial spots
3. Imputing gene expression in spatial data
4. Analyzing spatial organization of cell types</p>
<p>The method works by:
1. Identifying marker genes for each cell type
2. Training a mapping model using these markers
3. Projecting cell type annotations to spatial coordinates
4. Optionally imputing full gene expression profiles</p>
<p><strong>Attributes:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata_sc</code></td>
<td>
</td>
<td><p>AnnData
Single-cell RNA-seq data with:
- Gene expression matrix in X
- Cell type annotations in obs[clusters]</p></td>
</tr>
<tr>
<td><code>adata_sp</code></td>
<td>
</td>
<td><p>AnnData
Spatial transcriptomics data with:
- Gene expression matrix in X
- Spatial coordinates in obsm['spatial']</p></td>
</tr>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>str
Column name in adata_sc.obs containing cell type labels</p></td>
</tr>
<tr>
<td><code>markers</code></td>
<td>
</td>
<td><p>list
Selected marker genes used for mapping</p></td>
</tr>
<tr>
<td><code>ad_map</code></td>
<td>
</td>
<td><p>AnnData
Mapping results after training</p></td>
</tr>
</tbody>
</table>
<p><strong>Examples:</strong></p>
<div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">omicverse</span> <span class="k">as</span> <span class="nn">ov</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">"sc_data.h5ad"</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_sp</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_visium</span><span class="p">(</span><span class="s2">"spatial_data"</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Initialize Tangram</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tangram</span> <span class="o">=</span> <span class="n">ov</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">Tangram</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">clusters</span><span class="o">=</span><span class="s1">'cell_type'</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Train model</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tangram</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">'clusters'</span><span class="p">,</span> <span class="n">num_epochs</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Project cell types</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_spatial</span> <span class="o">=</span> <span class="n">tangram</span><span class="o">.</span><span class="n">cell2location</span><span class="p">()</span>
</code></pre></div>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 17 "></span><span class="k">class</span> <span class="nc">Tangram</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="linenos" data-linenos=" 18 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Tangram spatial deconvolution class for cell type mapping.</span>
<span class="linenos" data-linenos=" 19 "></span>
<span class="linenos" data-linenos=" 20 "></span><span class="sd">    Tangram is a method for integrating single-cell RNA sequencing (scRNA-seq) data</span>
<span class="linenos" data-linenos=" 21 "></span><span class="sd">    with spatial transcriptomics data. It enables:</span>
<span class="linenos" data-linenos=" 22 "></span><span class="sd">    1. Mapping cell types from scRNA-seq to spatial locations</span>
<span class="linenos" data-linenos=" 23 "></span><span class="sd">    2. Deconvolving cell type proportions in spatial spots</span>
<span class="linenos" data-linenos=" 24 "></span><span class="sd">    3. Imputing gene expression in spatial data</span>
<span class="linenos" data-linenos=" 25 "></span><span class="sd">    4. Analyzing spatial organization of cell types</span>
<span class="linenos" data-linenos=" 26 "></span>
<span class="linenos" data-linenos=" 27 "></span><span class="sd">    The method works by:</span>
<span class="linenos" data-linenos=" 28 "></span><span class="sd">    1. Identifying marker genes for each cell type</span>
<span class="linenos" data-linenos=" 29 "></span><span class="sd">    2. Training a mapping model using these markers</span>
<span class="linenos" data-linenos=" 30 "></span><span class="sd">    3. Projecting cell type annotations to spatial coordinates</span>
<span class="linenos" data-linenos=" 31 "></span><span class="sd">    4. Optionally imputing full gene expression profiles</span>
<span class="linenos" data-linenos=" 32 "></span>
<span class="linenos" data-linenos=" 33 "></span><span class="sd">    Attributes:</span>
<span class="linenos" data-linenos=" 34 "></span><span class="sd">        adata_sc: AnnData</span>
<span class="linenos" data-linenos=" 35 "></span><span class="sd">            Single-cell RNA-seq data with:</span>
<span class="linenos" data-linenos=" 36 "></span><span class="sd">            - Gene expression matrix in X</span>
<span class="linenos" data-linenos=" 37 "></span><span class="sd">            - Cell type annotations in obs[clusters]</span>
<span class="linenos" data-linenos=" 38 "></span><span class="sd">        adata_sp: AnnData</span>
<span class="linenos" data-linenos=" 39 "></span><span class="sd">            Spatial transcriptomics data with:</span>
<span class="linenos" data-linenos=" 40 "></span><span class="sd">            - Gene expression matrix in X</span>
<span class="linenos" data-linenos=" 41 "></span><span class="sd">            - Spatial coordinates in obsm['spatial']</span>
<span class="linenos" data-linenos=" 42 "></span><span class="sd">        clusters: str</span>
<span class="linenos" data-linenos=" 43 "></span><span class="sd">            Column name in adata_sc.obs containing cell type labels</span>
<span class="linenos" data-linenos=" 44 "></span><span class="sd">        markers: list</span>
<span class="linenos" data-linenos=" 45 "></span><span class="sd">            Selected marker genes used for mapping</span>
<span class="linenos" data-linenos=" 46 "></span><span class="sd">        ad_map: AnnData</span>
<span class="linenos" data-linenos=" 47 "></span><span class="sd">            Mapping results after training</span>
<span class="linenos" data-linenos=" 48 "></span>
<span class="linenos" data-linenos=" 49 "></span><span class="sd">    Examples:</span>
<span class="linenos" data-linenos=" 50 "></span><span class="sd">        &gt;&gt;&gt; import scanpy as sc</span>
<span class="linenos" data-linenos=" 51 "></span><span class="sd">        &gt;&gt;&gt; import omicverse as ov</span>
<span class="linenos" data-linenos=" 52 "></span><span class="sd">        &gt;&gt;&gt; # Load data</span>
<span class="linenos" data-linenos=" 53 "></span><span class="sd">        &gt;&gt;&gt; adata_sc = sc.read_h5ad("sc_data.h5ad")</span>
<span class="linenos" data-linenos=" 54 "></span><span class="sd">        &gt;&gt;&gt; adata_sp = sc.read_visium("spatial_data")</span>
<span class="linenos" data-linenos=" 55 "></span><span class="sd">        &gt;&gt;&gt; # Initialize Tangram</span>
<span class="linenos" data-linenos=" 56 "></span><span class="sd">        &gt;&gt;&gt; tangram = ov.space.Tangram(</span>
<span class="linenos" data-linenos=" 57 "></span><span class="sd">        ...     adata_sc=adata_sc,</span>
<span class="linenos" data-linenos=" 58 "></span><span class="sd">        ...     adata_sp=adata_sp,</span>
<span class="linenos" data-linenos=" 59 "></span><span class="sd">        ...     clusters='cell_type'</span>
<span class="linenos" data-linenos=" 60 "></span><span class="sd">        ... )</span>
<span class="linenos" data-linenos=" 61 "></span><span class="sd">        &gt;&gt;&gt; # Train model</span>
<span class="linenos" data-linenos=" 62 "></span><span class="sd">        &gt;&gt;&gt; tangram.train(mode='clusters', num_epochs=500)</span>
<span class="linenos" data-linenos=" 63 "></span><span class="sd">        &gt;&gt;&gt; # Project cell types</span>
<span class="linenos" data-linenos=" 64 "></span><span class="sd">        &gt;&gt;&gt; adata_spatial = tangram.cell2location()</span>
<span class="linenos" data-linenos=" 65 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos=" 66 "></span>    <span class="k">def</span> <span class="nf">check_tangram</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos=" 67 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""Check if Tangram package is installed.</span>
<span class="linenos" data-linenos=" 68 "></span>
<span class="linenos" data-linenos=" 69 "></span><span class="sd">        This method verifies that the Tangram package is available and prints</span>
<span class="linenos" data-linenos=" 70 "></span><span class="sd">        its version number. If not installed, it raises an informative error.</span>
<span class="linenos" data-linenos=" 71 "></span>
<span class="linenos" data-linenos=" 72 "></span><span class="sd">        Raises:</span>
<span class="linenos" data-linenos=" 73 "></span><span class="sd">            ImportError: If Tangram package is not installed, with instructions</span>
<span class="linenos" data-linenos=" 74 "></span><span class="sd">                for installation.</span>
<span class="linenos" data-linenos=" 75 "></span>
<span class="linenos" data-linenos=" 76 "></span><span class="sd">        Notes:</span>
<span class="linenos" data-linenos=" 77 "></span><span class="sd">            - Sets global tg_install flag when successful</span>
<span class="linenos" data-linenos=" 78 "></span><span class="sd">            - Required before any Tangram operations</span>
<span class="linenos" data-linenos=" 79 "></span><span class="sd">            - Suggests pip installation command if missing</span>
<span class="linenos" data-linenos=" 80 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos=" 81 "></span>        <span class="k">global</span> <span class="n">tg_install</span>
<span class="linenos" data-linenos=" 82 "></span>        <span class="k">try</span><span class="p">:</span>
<span class="linenos" data-linenos=" 83 "></span>            <span class="kn">import</span> <span class="nn">tangram</span> <span class="k">as</span> <span class="nn">tg</span>
<span class="linenos" data-linenos=" 84 "></span>            <span class="n">tg_install</span><span class="o">=</span><span class="kc">True</span>
<span class="linenos" data-linenos=" 85 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'tangram have been install version:'</span><span class="p">,</span><span class="n">tg</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="linenos" data-linenos=" 86 "></span>        <span class="k">except</span> <span class="ne">ImportError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
<span class="linenos" data-linenos=" 87 "></span>            <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span>
<span class="linenos" data-linenos=" 88 "></span>                <span class="s1">'Please install the tangram: `pip install -U tangram-sc`.'</span>
<span class="linenos" data-linenos=" 89 "></span>            <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
<span class="linenos" data-linenos=" 90 "></span>
<span class="linenos" data-linenos=" 91 "></span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos=" 92 "></span>            <span class="n">adata_sc</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
<span class="linenos" data-linenos=" 93 "></span>            <span class="n">adata_sp</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
<span class="linenos" data-linenos=" 94 "></span>            <span class="n">clusters</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos=" 95 "></span>            <span class="n">marker_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="linenos" data-linenos=" 96 "></span>            <span class="n">gene_to_lowercase</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="linenos" data-linenos=" 97 "></span>        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos=" 98 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""Initialize Tangram spatial deconvolution object.</span>
<span class="linenos" data-linenos=" 99 "></span>
<span class="linenos" data-linenos="100 "></span><span class="sd">        This method sets up the Tangram analysis by:</span>
<span class="linenos" data-linenos="101 "></span><span class="sd">        1. Checking package installation</span>
<span class="linenos" data-linenos="102 "></span><span class="sd">        2. Processing input data</span>
<span class="linenos" data-linenos="103 "></span><span class="sd">        3. Identifying marker genes</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">        4. Preparing data for mapping</span>
<span class="linenos" data-linenos="105 "></span>
<span class="linenos" data-linenos="106 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="107 "></span><span class="sd">            adata_sc: AnnData</span>
<span class="linenos" data-linenos="108 "></span><span class="sd">                Single-cell RNA-seq data containing:</span>
<span class="linenos" data-linenos="109 "></span><span class="sd">                - Normalized gene expression matrix</span>
<span class="linenos" data-linenos="110 "></span><span class="sd">                - Cell type annotations in obs[clusters]</span>
<span class="linenos" data-linenos="111 "></span><span class="sd">            adata_sp: AnnData</span>
<span class="linenos" data-linenos="112 "></span><span class="sd">                Spatial transcriptomics data containing:</span>
<span class="linenos" data-linenos="113 "></span><span class="sd">                - Normalized gene expression matrix</span>
<span class="linenos" data-linenos="114 "></span><span class="sd">                - Spatial coordinates in obsm['spatial']</span>
<span class="linenos" data-linenos="115 "></span><span class="sd">            clusters: str, optional (default='')</span>
<span class="linenos" data-linenos="116 "></span><span class="sd">                Column name in adata_sc.obs containing cell type annotations.</span>
<span class="linenos" data-linenos="117 "></span><span class="sd">            marker_size: int, optional (default=100)</span>
<span class="linenos" data-linenos="118 "></span><span class="sd">                Number of top marker genes to select per cell type.</span>
<span class="linenos" data-linenos="119 "></span><span class="sd">                More markers can improve accuracy but increase computation time.</span>
<span class="linenos" data-linenos="120 "></span><span class="sd">            gene_to_lowercase: bool, optional (default=False)</span>
<span class="linenos" data-linenos="121 "></span><span class="sd">                Whether to convert gene names to lowercase for matching between</span>
<span class="linenos" data-linenos="122 "></span><span class="sd">                datasets. Useful when gene naming conventions differ.</span>
<span class="linenos" data-linenos="123 "></span>
<span class="linenos" data-linenos="124 "></span><span class="sd">        Notes:</span>
<span class="linenos" data-linenos="125 "></span><span class="sd">            - Automatically filters genes present in at least one cell</span>
<span class="linenos" data-linenos="126 "></span><span class="sd">            - Identifies marker genes using scanpy's rank_genes_groups</span>
<span class="linenos" data-linenos="127 "></span><span class="sd">            - Prepares data structures for Tangram mapping</span>
<span class="linenos" data-linenos="128 "></span><span class="sd">            - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="129 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="130 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">check_tangram</span><span class="p">()</span>
<span class="linenos" data-linenos="131 "></span>        <span class="k">global</span> <span class="n">tg_install</span>
<span class="linenos" data-linenos="132 "></span>        <span class="k">if</span> <span class="n">tg_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="133 "></span>            <span class="n">global_imports</span><span class="p">(</span><span class="s2">"tangram"</span><span class="p">,</span><span class="s2">"tg"</span><span class="p">)</span>
<span class="linenos" data-linenos="134 "></span>        <span class="n">ad_map_dict</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="135 "></span>
<span class="linenos" data-linenos="136 "></span>        <span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'log1p'</span><span class="p">]</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="137 "></span>        <span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'log1p'</span><span class="p">][</span><span class="s1">'base'</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span>
<span class="linenos" data-linenos="138 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="139 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="140 "></span>                                <span class="n">key_added</span><span class="o">=</span><span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">clusters</span><span class="si">}</span><span class="s1">_rank_genes_groups'</span><span class="p">,</span><span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="141 "></span>        <span class="n">markers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">clusters</span><span class="si">}</span><span class="s2">_rank_genes_groups"</span><span class="p">][</span><span class="s2">"names"</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">marker_size</span><span class="p">,</span> <span class="p">:]</span>
<span class="linenos" data-linenos="142 "></span>
<span class="linenos" data-linenos="143 "></span>        <span class="n">markers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">markers_df</span><span class="o">.</span><span class="n">melt</span><span class="p">()</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<span class="linenos" data-linenos="144 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Calculate The Number of Markers:'</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">markers</span><span class="p">))</span>
<span class="linenos" data-linenos="145 "></span>
<span class="linenos" data-linenos="146 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span>
<span class="linenos" data-linenos="147 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span>
<span class="linenos" data-linenos="148 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="149 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">markers</span><span class="o">=</span><span class="n">markers</span>
<span class="linenos" data-linenos="150 "></span>
<span class="linenos" data-linenos="151 "></span>        <span class="n">tg</span><span class="o">.</span><span class="n">pp_adatas</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span>
<span class="linenos" data-linenos="152 "></span>                      <span class="n">genes</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">markers</span><span class="p">,</span><span class="n">gene_to_lowercase</span><span class="o">=</span><span class="n">gene_to_lowercase</span><span class="p">)</span>
<span class="linenos" data-linenos="153 "></span>
<span class="linenos" data-linenos="154 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Model prepared successfully'</span><span class="p">)</span>
<span class="linenos" data-linenos="155 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="156 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="157 "></span>
<span class="linenos" data-linenos="158 "></span>    <span class="k">def</span> <span class="nf">train</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos="159 "></span>            <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">"clusters"</span><span class="p">,</span>
<span class="linenos" data-linenos="160 "></span>            <span class="n">num_epochs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
<span class="linenos" data-linenos="161 "></span>            <span class="n">device</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">"cuda:0"</span><span class="p">,</span>
<span class="linenos" data-linenos="162 "></span>            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span>
<span class="linenos" data-linenos="163 "></span>        <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="164 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""Train the Tangram spatial mapping model.</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span><span class="sd">        This method trains a model to map cells or clusters from scRNA-seq data</span>
<span class="linenos" data-linenos="167 "></span><span class="sd">        to spatial locations. It optimizes the mapping to preserve both gene</span>
<span class="linenos" data-linenos="168 "></span><span class="sd">        expression patterns and spatial structure.</span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="171 "></span><span class="sd">            mode: str, optional (default="clusters")</span>
<span class="linenos" data-linenos="172 "></span><span class="sd">                Mapping mode:</span>
<span class="linenos" data-linenos="173 "></span><span class="sd">                - "clusters": Map cell type proportions (faster)</span>
<span class="linenos" data-linenos="174 "></span><span class="sd">                - "cells": Map individual cells (more detailed)</span>
<span class="linenos" data-linenos="175 "></span><span class="sd">            num_epochs: int, optional (default=500)</span>
<span class="linenos" data-linenos="176 "></span><span class="sd">                Number of training epochs. More epochs may improve results</span>
<span class="linenos" data-linenos="177 "></span><span class="sd">                but increase training time.</span>
<span class="linenos" data-linenos="178 "></span><span class="sd">            device: str, optional (default="cuda:0")</span>
<span class="linenos" data-linenos="179 "></span><span class="sd">                Computing device to use:</span>
<span class="linenos" data-linenos="180 "></span><span class="sd">                - "cuda:0" (or other GPU index) for GPU acceleration</span>
<span class="linenos" data-linenos="181 "></span><span class="sd">                - "cpu" for CPU computation</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">            **kwargs: Any</span>
<span class="linenos" data-linenos="183 "></span><span class="sd">                Additional arguments passed to tangram.map_cells_to_space:</span>
<span class="linenos" data-linenos="184 "></span><span class="sd">                - density_prior: Prior on spatial density</span>
<span class="linenos" data-linenos="185 "></span><span class="sd">                - lambda_d: Density regularization strength</span>
<span class="linenos" data-linenos="186 "></span><span class="sd">                - lambda_g1: Gene-expression regularization</span>
<span class="linenos" data-linenos="187 "></span><span class="sd">                - lambda_g2: Spatial regularization</span>
<span class="linenos" data-linenos="188 "></span><span class="sd">                - lambda_r: Entropy regularization</span>
<span class="linenos" data-linenos="189 "></span>
<span class="linenos" data-linenos="190 "></span><span class="sd">        Notes:</span>
<span class="linenos" data-linenos="191 "></span><span class="sd">            - Automatically stores mapping in self.ad_map</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">            - Projects cell type annotations to spatial data</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">            - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">            - Progress is shown during training</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">            - GPU acceleration recommended for large datasets</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="197 "></span>        <span class="n">ad_map</span> <span class="o">=</span> <span class="n">tg</span><span class="o">.</span><span class="n">map_cells_to_space</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span>
<span class="linenos" data-linenos="198 "></span>            <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">,</span>
<span class="linenos" data-linenos="199 "></span>            <span class="n">cluster_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="200 "></span>            <span class="n">num_epochs</span><span class="o">=</span><span class="n">num_epochs</span><span class="p">,</span>
<span class="linenos" data-linenos="201 "></span>            <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
<span class="linenos" data-linenos="202 "></span>            <span class="o">**</span><span class="n">kwargs</span>
<span class="linenos" data-linenos="203 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="204 "></span>
<span class="linenos" data-linenos="205 "></span>        <span class="n">tg</span><span class="o">.</span><span class="n">project_cell_annotations</span><span class="p">(</span><span class="n">ad_map</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span>
<span class="linenos" data-linenos="206 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">ad_map</span><span class="o">=</span><span class="n">ad_map</span>
<span class="linenos" data-linenos="207 "></span>
<span class="linenos" data-linenos="208 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Model train successfully'</span><span class="p">)</span>
<span class="linenos" data-linenos="209 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="210 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="211 "></span>
<span class="linenos" data-linenos="212 "></span>    <span class="k">def</span> <span class="nf">cell2location</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">annotation_list</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="213 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""Project cell type annotations to spatial coordinates.</span>
<span class="linenos" data-linenos="214 "></span>
<span class="linenos" data-linenos="215 "></span><span class="sd">        This method creates a visualization-ready AnnData object containing the</span>
<span class="linenos" data-linenos="216 "></span><span class="sd">        predicted cell type proportions for each spatial location.</span>
<span class="linenos" data-linenos="217 "></span>
<span class="linenos" data-linenos="218 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="219 "></span><span class="sd">            annotation_list: list, optional (default=None)</span>
<span class="linenos" data-linenos="220 "></span><span class="sd">                List of cell types to include in the projection.</span>
<span class="linenos" data-linenos="221 "></span><span class="sd">                If None, uses all cell types from training data.</span>
<span class="linenos" data-linenos="222 "></span>
<span class="linenos" data-linenos="223 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="224 "></span><span class="sd">            AnnData</span>
<span class="linenos" data-linenos="225 "></span><span class="sd">                Modified spatial data containing:</span>
<span class="linenos" data-linenos="226 "></span><span class="sd">                - Original spatial data</span>
<span class="linenos" data-linenos="227 "></span><span class="sd">                - Cell type proportions in obsm['tangram_ct_pred']</span>
<span class="linenos" data-linenos="228 "></span><span class="sd">                - Normalized proportions in obs for each cell type</span>
<span class="linenos" data-linenos="229 "></span>
<span class="linenos" data-linenos="230 "></span><span class="sd">        Notes:</span>
<span class="linenos" data-linenos="231 "></span><span class="sd">            - Automatically normalizes cell type proportions</span>
<span class="linenos" data-linenos="232 "></span><span class="sd">            - Clips extreme values for better visualization</span>
<span class="linenos" data-linenos="233 "></span><span class="sd">            - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="234 "></span><span class="sd">            - Results can be directly used for spatial plotting</span>
<span class="linenos" data-linenos="235 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="236 "></span>        <span class="n">adata_plot</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="237 "></span>        <span class="k">if</span> <span class="n">annotation_list</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="238 "></span>            <span class="n">annotation_list</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">]))</span>
<span class="linenos" data-linenos="239 "></span>
<span class="linenos" data-linenos="240 "></span>        <span class="n">df</span> <span class="o">=</span> <span class="n">adata_plot</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">"tangram_ct_pred"</span><span class="p">][</span><span class="n">annotation_list</span><span class="p">]</span>
<span class="linenos" data-linenos="241 "></span>        <span class="n">construct_obs_plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">adata_plot</span><span class="p">,</span> <span class="n">perc</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="242 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="243 "></span>        <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="244 "></span>        <span class="k">return</span> <span class="n">adata_plot</span>
<span class="linenos" data-linenos="245 "></span>
<span class="linenos" data-linenos="246 "></span>    <span class="k">def</span> <span class="nf">impute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos="247 "></span>               <span class="n">ad_map</span><span class="p">:</span> <span class="n">AnnData</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="248 "></span>               <span class="n">ad_sc</span><span class="p">:</span> <span class="n">AnnData</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="249 "></span>                <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">AnnData</span><span class="p">:</span>
<span class="linenos" data-linenos="250 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""Impute gene expression in spatial data using trained model.</span>
<span class="linenos" data-linenos="251 "></span>
<span class="linenos" data-linenos="252 "></span><span class="sd">        This method uses the trained mapping to predict the expression of all genes</span>
<span class="linenos" data-linenos="253 "></span><span class="sd">        in the spatial locations, including genes not used in the mapping.</span>
<span class="linenos" data-linenos="254 "></span>
<span class="linenos" data-linenos="255 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="256 "></span><span class="sd">            ad_map: AnnData, optional (default=None)</span>
<span class="linenos" data-linenos="257 "></span><span class="sd">                Mapping result from train(). If None, uses self.ad_map.</span>
<span class="linenos" data-linenos="258 "></span><span class="sd">            ad_sc: AnnData, optional (default=None)</span>
<span class="linenos" data-linenos="259 "></span><span class="sd">                Single-cell reference data. If None, uses self.adata_sc.</span>
<span class="linenos" data-linenos="260 "></span><span class="sd">            **kwargs: Any</span>
<span class="linenos" data-linenos="261 "></span><span class="sd">                Additional arguments passed to tangram.project_genes:</span>
<span class="linenos" data-linenos="262 "></span><span class="sd">                - scale: Whether to scale imputed values</span>
<span class="linenos" data-linenos="263 "></span><span class="sd">                - filter_genes: Whether to filter genes before imputation</span>
<span class="linenos" data-linenos="264 "></span><span class="sd">                - filter_threshold: Expression threshold for filtering</span>
<span class="linenos" data-linenos="265 "></span>
<span class="linenos" data-linenos="266 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="267 "></span><span class="sd">            AnnData</span>
<span class="linenos" data-linenos="268 "></span><span class="sd">                Spatial data with imputed gene expression for all genes</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">                in the single-cell reference data.</span>
<span class="linenos" data-linenos="270 "></span>
<span class="linenos" data-linenos="271 "></span><span class="sd">        Notes:</span>
<span class="linenos" data-linenos="272 "></span><span class="sd">            - Uses mapping weights to transfer expression</span>
<span class="linenos" data-linenos="273 "></span><span class="sd">            - Can impute genes not used in original mapping</span>
<span class="linenos" data-linenos="274 "></span><span class="sd">            - Useful for analyzing spatial patterns of any gene</span>
<span class="linenos" data-linenos="275 "></span><span class="sd">            - Computationally intensive for large gene sets</span>
<span class="linenos" data-linenos="276 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="277 "></span>        <span class="k">if</span> <span class="n">ad_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="278 "></span>            <span class="n">ad_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ad_map</span>
<span class="linenos" data-linenos="279 "></span>        <span class="k">if</span> <span class="n">ad_sc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="280 "></span>            <span class="n">ad_sc</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span>
<span class="linenos" data-linenos="281 "></span>        <span class="n">ad_ge</span> <span class="o">=</span> <span class="n">tg</span><span class="o">.</span><span class="n">project_genes</span><span class="p">(</span><span class="n">adata_map</span><span class="o">=</span><span class="n">ad_map</span><span class="p">,</span> 
<span class="linenos" data-linenos="282 "></span>                                 <span class="n">adata_sc</span><span class="o">=</span><span class="n">ad_sc</span><span class="p">,</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="linenos" data-linenos="283 "></span>        <span class="k">return</span> <span class="n">ad_ge</span>
</code></pre></div>
</details>
<div class="doc doc-children">
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.space._tangram.Tangram.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">adata_sp</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">marker_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">gene_to_lowercase</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.space._tangram.Tangram.__init__" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Initialize Tangram spatial deconvolution object.</p>
<p>This method sets up the Tangram analysis by:
1. Checking package installation
2. Processing input data
3. Identifying marker genes
4. Preparing data for mapping</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata_sc</code></td>
<td>
<code><span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData
Single-cell RNA-seq data containing:
- Normalized gene expression matrix
- Cell type annotations in obs[clusters]</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>adata_sp</code></td>
<td>
<code><span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData
Spatial transcriptomics data containing:
- Normalized gene expression matrix
- Spatial coordinates in obsm['spatial']</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>clusters</code></td>
<td>
<code>str</code>
</td>
<td><p>str, optional (default='')
Column name in adata_sc.obs containing cell type annotations.</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>marker_size</code></td>
<td>
<code>int</code>
</td>
<td><p>int, optional (default=100)
Number of top marker genes to select per cell type.
More markers can improve accuracy but increase computation time.</p></td>
<td>
<code>100</code>
</td>
</tr>
<tr>
<td><code>gene_to_lowercase</code></td>
<td>
<code>bool</code>
</td>
<td><p>bool, optional (default=False)
Whether to convert gene names to lowercase for matching between
datasets. Useful when gene naming conventions differ.</p></td>
<td>
<code>False</code>
</td>
</tr>
</tbody>
</table>
<details class="notes">
<summary>Notes</summary>
<ul>
<li>Automatically filters genes present in at least one cell</li>
<li>Identifies marker genes using scanpy's rank_genes_groups</li>
<li>Prepares data structures for Tangram mapping</li>
<li>Adds reference annotation to both AnnData objects</li>
</ul>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 91 "></span><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos=" 92 "></span>        <span class="n">adata_sc</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
<span class="linenos" data-linenos=" 93 "></span>        <span class="n">adata_sp</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
<span class="linenos" data-linenos=" 94 "></span>        <span class="n">clusters</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos=" 95 "></span>        <span class="n">marker_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="linenos" data-linenos=" 96 "></span>        <span class="n">gene_to_lowercase</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="linenos" data-linenos=" 97 "></span>    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos=" 98 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Initialize Tangram spatial deconvolution object.</span>
<span class="linenos" data-linenos=" 99 "></span>
<span class="linenos" data-linenos="100 "></span><span class="sd">    This method sets up the Tangram analysis by:</span>
<span class="linenos" data-linenos="101 "></span><span class="sd">    1. Checking package installation</span>
<span class="linenos" data-linenos="102 "></span><span class="sd">    2. Processing input data</span>
<span class="linenos" data-linenos="103 "></span><span class="sd">    3. Identifying marker genes</span>
<span class="linenos" data-linenos="104 "></span><span class="sd">    4. Preparing data for mapping</span>
<span class="linenos" data-linenos="105 "></span>
<span class="linenos" data-linenos="106 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="107 "></span><span class="sd">        adata_sc: AnnData</span>
<span class="linenos" data-linenos="108 "></span><span class="sd">            Single-cell RNA-seq data containing:</span>
<span class="linenos" data-linenos="109 "></span><span class="sd">            - Normalized gene expression matrix</span>
<span class="linenos" data-linenos="110 "></span><span class="sd">            - Cell type annotations in obs[clusters]</span>
<span class="linenos" data-linenos="111 "></span><span class="sd">        adata_sp: AnnData</span>
<span class="linenos" data-linenos="112 "></span><span class="sd">            Spatial transcriptomics data containing:</span>
<span class="linenos" data-linenos="113 "></span><span class="sd">            - Normalized gene expression matrix</span>
<span class="linenos" data-linenos="114 "></span><span class="sd">            - Spatial coordinates in obsm['spatial']</span>
<span class="linenos" data-linenos="115 "></span><span class="sd">        clusters: str, optional (default='')</span>
<span class="linenos" data-linenos="116 "></span><span class="sd">            Column name in adata_sc.obs containing cell type annotations.</span>
<span class="linenos" data-linenos="117 "></span><span class="sd">        marker_size: int, optional (default=100)</span>
<span class="linenos" data-linenos="118 "></span><span class="sd">            Number of top marker genes to select per cell type.</span>
<span class="linenos" data-linenos="119 "></span><span class="sd">            More markers can improve accuracy but increase computation time.</span>
<span class="linenos" data-linenos="120 "></span><span class="sd">        gene_to_lowercase: bool, optional (default=False)</span>
<span class="linenos" data-linenos="121 "></span><span class="sd">            Whether to convert gene names to lowercase for matching between</span>
<span class="linenos" data-linenos="122 "></span><span class="sd">            datasets. Useful when gene naming conventions differ.</span>
<span class="linenos" data-linenos="123 "></span>
<span class="linenos" data-linenos="124 "></span><span class="sd">    Notes:</span>
<span class="linenos" data-linenos="125 "></span><span class="sd">        - Automatically filters genes present in at least one cell</span>
<span class="linenos" data-linenos="126 "></span><span class="sd">        - Identifies marker genes using scanpy's rank_genes_groups</span>
<span class="linenos" data-linenos="127 "></span><span class="sd">        - Prepares data structures for Tangram mapping</span>
<span class="linenos" data-linenos="128 "></span><span class="sd">        - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="129 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="130 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">check_tangram</span><span class="p">()</span>
<span class="linenos" data-linenos="131 "></span>    <span class="k">global</span> <span class="n">tg_install</span>
<span class="linenos" data-linenos="132 "></span>    <span class="k">if</span> <span class="n">tg_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="133 "></span>        <span class="n">global_imports</span><span class="p">(</span><span class="s2">"tangram"</span><span class="p">,</span><span class="s2">"tg"</span><span class="p">)</span>
<span class="linenos" data-linenos="134 "></span>    <span class="n">ad_map_dict</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="135 "></span>
<span class="linenos" data-linenos="136 "></span>    <span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'log1p'</span><span class="p">]</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="137 "></span>    <span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'log1p'</span><span class="p">][</span><span class="s1">'base'</span><span class="p">]</span><span class="o">=</span><span class="kc">None</span>
<span class="linenos" data-linenos="138 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="139 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="140 "></span>                            <span class="n">key_added</span><span class="o">=</span><span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">clusters</span><span class="si">}</span><span class="s1">_rank_genes_groups'</span><span class="p">,</span><span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="141 "></span>    <span class="n">markers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">clusters</span><span class="si">}</span><span class="s2">_rank_genes_groups"</span><span class="p">][</span><span class="s2">"names"</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">marker_size</span><span class="p">,</span> <span class="p">:]</span>
<span class="linenos" data-linenos="142 "></span>
<span class="linenos" data-linenos="143 "></span>    <span class="n">markers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">markers_df</span><span class="o">.</span><span class="n">melt</span><span class="p">()</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<span class="linenos" data-linenos="144 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'...Calculate The Number of Markers:'</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">markers</span><span class="p">))</span>
<span class="linenos" data-linenos="145 "></span>
<span class="linenos" data-linenos="146 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="o">=</span><span class="n">adata_sc</span>
<span class="linenos" data-linenos="147 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="o">=</span><span class="n">adata_sp</span>
<span class="linenos" data-linenos="148 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="149 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">markers</span><span class="o">=</span><span class="n">markers</span>
<span class="linenos" data-linenos="150 "></span>
<span class="linenos" data-linenos="151 "></span>    <span class="n">tg</span><span class="o">.</span><span class="n">pp_adatas</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span>
<span class="linenos" data-linenos="152 "></span>                  <span class="n">genes</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">markers</span><span class="p">,</span><span class="n">gene_to_lowercase</span><span class="o">=</span><span class="n">gene_to_lowercase</span><span class="p">)</span>
<span class="linenos" data-linenos="153 "></span>
<span class="linenos" data-linenos="154 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'...Model prepared successfully'</span><span class="p">)</span>
<span class="linenos" data-linenos="155 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="156 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.space._tangram.Tangram.check_tangram">
<code class="highlight language-python"><span class="n">check_tangram</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.space._tangram.Tangram.check_tangram" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Check if Tangram package is installed.</p>
<p>This method verifies that the Tangram package is available and prints
its version number. If not installed, it raises an informative error.</p>
<p><strong>Raises:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<code>ImportError</code>
</td>
<td><p>If Tangram package is not installed, with instructions
for installation.</p></td>
</tr>
</tbody>
</table>
<details class="notes">
<summary>Notes</summary>
<ul>
<li>Sets global tg_install flag when successful</li>
<li>Required before any Tangram operations</li>
<li>Suggests pip installation command if missing</li>
</ul>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="66 "></span><span class="k">def</span> <span class="nf">check_tangram</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="67 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Check if Tangram package is installed.</span>
<span class="linenos" data-linenos="68 "></span>
<span class="linenos" data-linenos="69 "></span><span class="sd">    This method verifies that the Tangram package is available and prints</span>
<span class="linenos" data-linenos="70 "></span><span class="sd">    its version number. If not installed, it raises an informative error.</span>
<span class="linenos" data-linenos="71 "></span>
<span class="linenos" data-linenos="72 "></span><span class="sd">    Raises:</span>
<span class="linenos" data-linenos="73 "></span><span class="sd">        ImportError: If Tangram package is not installed, with instructions</span>
<span class="linenos" data-linenos="74 "></span><span class="sd">            for installation.</span>
<span class="linenos" data-linenos="75 "></span>
<span class="linenos" data-linenos="76 "></span><span class="sd">    Notes:</span>
<span class="linenos" data-linenos="77 "></span><span class="sd">        - Sets global tg_install flag when successful</span>
<span class="linenos" data-linenos="78 "></span><span class="sd">        - Required before any Tangram operations</span>
<span class="linenos" data-linenos="79 "></span><span class="sd">        - Suggests pip installation command if missing</span>
<span class="linenos" data-linenos="80 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="81 "></span>    <span class="k">global</span> <span class="n">tg_install</span>
<span class="linenos" data-linenos="82 "></span>    <span class="k">try</span><span class="p">:</span>
<span class="linenos" data-linenos="83 "></span>        <span class="kn">import</span> <span class="nn">tangram</span> <span class="k">as</span> <span class="nn">tg</span>
<span class="linenos" data-linenos="84 "></span>        <span class="n">tg_install</span><span class="o">=</span><span class="kc">True</span>
<span class="linenos" data-linenos="85 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'tangram have been install version:'</span><span class="p">,</span><span class="n">tg</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="linenos" data-linenos="86 "></span>    <span class="k">except</span> <span class="ne">ImportError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
<span class="linenos" data-linenos="87 "></span>        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span>
<span class="linenos" data-linenos="88 "></span>            <span class="s1">'Please install the tangram: `pip install -U tangram-sc`.'</span>
<span class="linenos" data-linenos="89 "></span>        <span class="p">)</span> <span class="kn">from</span> <span class="nn">e</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.space._tangram.Tangram.train">
<code class="highlight language-python"><span class="n">train</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">'clusters'</span><span class="p">,</span> <span class="n">num_epochs</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="s1">'cuda:0'</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.space._tangram.Tangram.train" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Train the Tangram spatial mapping model.</p>
<p>This method trains a model to map cells or clusters from scRNA-seq data
to spatial locations. It optimizes the mapping to preserve both gene
expression patterns and spatial structure.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>mode</code></td>
<td>
<code>str</code>
</td>
<td><p>str, optional (default="clusters")
Mapping mode:
- "clusters": Map cell type proportions (faster)
- "cells": Map individual cells (more detailed)</p></td>
<td>
<code>'clusters'</code>
</td>
</tr>
<tr>
<td><code>num_epochs</code></td>
<td>
<code>int</code>
</td>
<td><p>int, optional (default=500)
Number of training epochs. More epochs may improve results
but increase training time.</p></td>
<td>
<code>500</code>
</td>
</tr>
<tr>
<td><code>device</code></td>
<td>
<code>str</code>
</td>
<td><p>str, optional (default="cuda:0")
Computing device to use:
- "cuda:0" (or other GPU index) for GPU acceleration
- "cpu" for CPU computation</p></td>
<td>
<code>'cuda:0'</code>
</td>
</tr>
<tr>
<td><code>**kwargs</code></td>
<td>
<code><span title="typing.Any">Any</span></code>
</td>
<td><p>Any
Additional arguments passed to tangram.map_cells_to_space:
- density_prior: Prior on spatial density
- lambda_d: Density regularization strength
- lambda_g1: Gene-expression regularization
- lambda_g2: Spatial regularization
- lambda_r: Entropy regularization</p></td>
<td>
<code>{}</code>
</td>
</tr>
</tbody>
</table>
<details class="notes">
<summary>Notes</summary>
<ul>
<li>Automatically stores mapping in self.ad_map</li>
<li>Projects cell type annotations to spatial data</li>
<li>Adds reference annotation to both AnnData objects</li>
<li>Progress is shown during training</li>
<li>GPU acceleration recommended for large datasets</li>
</ul>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="158 "></span><span class="k">def</span> <span class="nf">train</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos="159 "></span>        <span class="n">mode</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">"clusters"</span><span class="p">,</span>
<span class="linenos" data-linenos="160 "></span>        <span class="n">num_epochs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
<span class="linenos" data-linenos="161 "></span>        <span class="n">device</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">"cuda:0"</span><span class="p">,</span>
<span class="linenos" data-linenos="162 "></span>        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span>
<span class="linenos" data-linenos="163 "></span>    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="164 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Train the Tangram spatial mapping model.</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span><span class="sd">    This method trains a model to map cells or clusters from scRNA-seq data</span>
<span class="linenos" data-linenos="167 "></span><span class="sd">    to spatial locations. It optimizes the mapping to preserve both gene</span>
<span class="linenos" data-linenos="168 "></span><span class="sd">    expression patterns and spatial structure.</span>
<span class="linenos" data-linenos="169 "></span>
<span class="linenos" data-linenos="170 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="171 "></span><span class="sd">        mode: str, optional (default="clusters")</span>
<span class="linenos" data-linenos="172 "></span><span class="sd">            Mapping mode:</span>
<span class="linenos" data-linenos="173 "></span><span class="sd">            - "clusters": Map cell type proportions (faster)</span>
<span class="linenos" data-linenos="174 "></span><span class="sd">            - "cells": Map individual cells (more detailed)</span>
<span class="linenos" data-linenos="175 "></span><span class="sd">        num_epochs: int, optional (default=500)</span>
<span class="linenos" data-linenos="176 "></span><span class="sd">            Number of training epochs. More epochs may improve results</span>
<span class="linenos" data-linenos="177 "></span><span class="sd">            but increase training time.</span>
<span class="linenos" data-linenos="178 "></span><span class="sd">        device: str, optional (default="cuda:0")</span>
<span class="linenos" data-linenos="179 "></span><span class="sd">            Computing device to use:</span>
<span class="linenos" data-linenos="180 "></span><span class="sd">            - "cuda:0" (or other GPU index) for GPU acceleration</span>
<span class="linenos" data-linenos="181 "></span><span class="sd">            - "cpu" for CPU computation</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">        **kwargs: Any</span>
<span class="linenos" data-linenos="183 "></span><span class="sd">            Additional arguments passed to tangram.map_cells_to_space:</span>
<span class="linenos" data-linenos="184 "></span><span class="sd">            - density_prior: Prior on spatial density</span>
<span class="linenos" data-linenos="185 "></span><span class="sd">            - lambda_d: Density regularization strength</span>
<span class="linenos" data-linenos="186 "></span><span class="sd">            - lambda_g1: Gene-expression regularization</span>
<span class="linenos" data-linenos="187 "></span><span class="sd">            - lambda_g2: Spatial regularization</span>
<span class="linenos" data-linenos="188 "></span><span class="sd">            - lambda_r: Entropy regularization</span>
<span class="linenos" data-linenos="189 "></span>
<span class="linenos" data-linenos="190 "></span><span class="sd">    Notes:</span>
<span class="linenos" data-linenos="191 "></span><span class="sd">        - Automatically stores mapping in self.ad_map</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        - Projects cell type annotations to spatial data</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">        - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        - Progress is shown during training</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">        - GPU acceleration recommended for large datasets</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="197 "></span>    <span class="n">ad_map</span> <span class="o">=</span> <span class="n">tg</span><span class="o">.</span><span class="n">map_cells_to_space</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span>
<span class="linenos" data-linenos="198 "></span>        <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">,</span>
<span class="linenos" data-linenos="199 "></span>        <span class="n">cluster_label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">,</span>
<span class="linenos" data-linenos="200 "></span>        <span class="n">num_epochs</span><span class="o">=</span><span class="n">num_epochs</span><span class="p">,</span>
<span class="linenos" data-linenos="201 "></span>        <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
<span class="linenos" data-linenos="202 "></span>        <span class="o">**</span><span class="n">kwargs</span>
<span class="linenos" data-linenos="203 "></span>    <span class="p">)</span>
<span class="linenos" data-linenos="204 "></span>
<span class="linenos" data-linenos="205 "></span>    <span class="n">tg</span><span class="o">.</span><span class="n">project_cell_annotations</span><span class="p">(</span><span class="n">ad_map</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">)</span>
<span class="linenos" data-linenos="206 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">ad_map</span><span class="o">=</span><span class="n">ad_map</span>
<span class="linenos" data-linenos="207 "></span>
<span class="linenos" data-linenos="208 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'...Model train successfully'</span><span class="p">)</span>
<span class="linenos" data-linenos="209 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="210 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.space._tangram.Tangram.cell2location">
<code class="highlight language-python"><span class="n">cell2location</span><span class="p">(</span><span class="n">annotation_list</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.space._tangram.Tangram.cell2location" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Project cell type annotations to spatial coordinates.</p>
<p>This method creates a visualization-ready AnnData object containing the
predicted cell type proportions for each spatial location.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>annotation_list</code></td>
<td>
</td>
<td><p>list, optional (default=None)
List of cell types to include in the projection.
If None, uses all cell types from training data.</p></td>
<td>
<code>None</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
</td>
<td><p>AnnData
Modified spatial data containing:
- Original spatial data
- Cell type proportions in obsm['tangram_ct_pred']
- Normalized proportions in obs for each cell type</p></td>
</tr>
</tbody>
</table>
<details class="notes">
<summary>Notes</summary>
<ul>
<li>Automatically normalizes cell type proportions</li>
<li>Clips extreme values for better visualization</li>
<li>Adds reference annotation to both AnnData objects</li>
<li>Results can be directly used for spatial plotting</li>
</ul>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="212 "></span><span class="k">def</span> <span class="nf">cell2location</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">annotation_list</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="213 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Project cell type annotations to spatial coordinates.</span>
<span class="linenos" data-linenos="214 "></span>
<span class="linenos" data-linenos="215 "></span><span class="sd">    This method creates a visualization-ready AnnData object containing the</span>
<span class="linenos" data-linenos="216 "></span><span class="sd">    predicted cell type proportions for each spatial location.</span>
<span class="linenos" data-linenos="217 "></span>
<span class="linenos" data-linenos="218 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="219 "></span><span class="sd">        annotation_list: list, optional (default=None)</span>
<span class="linenos" data-linenos="220 "></span><span class="sd">            List of cell types to include in the projection.</span>
<span class="linenos" data-linenos="221 "></span><span class="sd">            If None, uses all cell types from training data.</span>
<span class="linenos" data-linenos="222 "></span>
<span class="linenos" data-linenos="223 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="224 "></span><span class="sd">        AnnData</span>
<span class="linenos" data-linenos="225 "></span><span class="sd">            Modified spatial data containing:</span>
<span class="linenos" data-linenos="226 "></span><span class="sd">            - Original spatial data</span>
<span class="linenos" data-linenos="227 "></span><span class="sd">            - Cell type proportions in obsm['tangram_ct_pred']</span>
<span class="linenos" data-linenos="228 "></span><span class="sd">            - Normalized proportions in obs for each cell type</span>
<span class="linenos" data-linenos="229 "></span>
<span class="linenos" data-linenos="230 "></span><span class="sd">    Notes:</span>
<span class="linenos" data-linenos="231 "></span><span class="sd">        - Automatically normalizes cell type proportions</span>
<span class="linenos" data-linenos="232 "></span><span class="sd">        - Clips extreme values for better visualization</span>
<span class="linenos" data-linenos="233 "></span><span class="sd">        - Adds reference annotation to both AnnData objects</span>
<span class="linenos" data-linenos="234 "></span><span class="sd">        - Results can be directly used for spatial plotting</span>
<span class="linenos" data-linenos="235 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="236 "></span>    <span class="n">adata_plot</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="237 "></span>    <span class="k">if</span> <span class="n">annotation_list</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="238 "></span>        <span class="n">annotation_list</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="p">]))</span>
<span class="linenos" data-linenos="239 "></span>
<span class="linenos" data-linenos="240 "></span>    <span class="n">df</span> <span class="o">=</span> <span class="n">adata_plot</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">"tangram_ct_pred"</span><span class="p">][</span><span class="n">annotation_list</span><span class="p">]</span>
<span class="linenos" data-linenos="241 "></span>    <span class="n">construct_obs_plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">adata_plot</span><span class="p">,</span> <span class="n">perc</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="242 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sp</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="243 "></span>    <span class="n">add_reference</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span><span class="p">,</span><span class="s1">'tangram'</span><span class="p">,</span><span class="s1">'cell type classification with Tangram'</span><span class="p">)</span>
<span class="linenos" data-linenos="244 "></span>    <span class="k">return</span> <span class="n">adata_plot</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.space._tangram.Tangram.impute">
<code class="highlight language-python"><span class="n">impute</span><span class="p">(</span><span class="n">ad_map</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ad_sc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.space._tangram.Tangram.impute" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Impute gene expression in spatial data using trained model.</p>
<p>This method uses the trained mapping to predict the expression of all genes
in the spatial locations, including genes not used in the mapping.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ad_map</code></td>
<td>
<code><span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData, optional (default=None)
Mapping result from train(). If None, uses self.ad_map.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>ad_sc</code></td>
<td>
<code><span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData, optional (default=None)
Single-cell reference data. If None, uses self.adata_sc.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>**kwargs</code></td>
<td>
<code><span title="typing.Any">Any</span></code>
</td>
<td><p>Any
Additional arguments passed to tangram.project_genes:
- scale: Whether to scale imputed values
- filter_genes: Whether to filter genes before imputation
- filter_threshold: Expression threshold for filtering</p></td>
<td>
<code>{}</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<code><span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData
Spatial data with imputed gene expression for all genes
in the single-cell reference data.</p></td>
</tr>
</tbody>
</table>
<details class="notes">
<summary>Notes</summary>
<ul>
<li>Uses mapping weights to transfer expression</li>
<li>Can impute genes not used in original mapping</li>
<li>Useful for analyzing spatial patterns of any gene</li>
<li>Computationally intensive for large gene sets</li>
</ul>
</details>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/space/lib/python3.10/site-packages/omicverse/space/_tangram.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="246 "></span><span class="k">def</span> <span class="nf">impute</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
<span class="linenos" data-linenos="247 "></span>           <span class="n">ad_map</span><span class="p">:</span> <span class="n">AnnData</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="248 "></span>           <span class="n">ad_sc</span><span class="p">:</span> <span class="n">AnnData</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="249 "></span>            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">AnnData</span><span class="p">:</span>
<span class="linenos" data-linenos="250 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""Impute gene expression in spatial data using trained model.</span>
<span class="linenos" data-linenos="251 "></span>
<span class="linenos" data-linenos="252 "></span><span class="sd">    This method uses the trained mapping to predict the expression of all genes</span>
<span class="linenos" data-linenos="253 "></span><span class="sd">    in the spatial locations, including genes not used in the mapping.</span>
<span class="linenos" data-linenos="254 "></span>
<span class="linenos" data-linenos="255 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="256 "></span><span class="sd">        ad_map: AnnData, optional (default=None)</span>
<span class="linenos" data-linenos="257 "></span><span class="sd">            Mapping result from train(). If None, uses self.ad_map.</span>
<span class="linenos" data-linenos="258 "></span><span class="sd">        ad_sc: AnnData, optional (default=None)</span>
<span class="linenos" data-linenos="259 "></span><span class="sd">            Single-cell reference data. If None, uses self.adata_sc.</span>
<span class="linenos" data-linenos="260 "></span><span class="sd">        **kwargs: Any</span>
<span class="linenos" data-linenos="261 "></span><span class="sd">            Additional arguments passed to tangram.project_genes:</span>
<span class="linenos" data-linenos="262 "></span><span class="sd">            - scale: Whether to scale imputed values</span>
<span class="linenos" data-linenos="263 "></span><span class="sd">            - filter_genes: Whether to filter genes before imputation</span>
<span class="linenos" data-linenos="264 "></span><span class="sd">            - filter_threshold: Expression threshold for filtering</span>
<span class="linenos" data-linenos="265 "></span>
<span class="linenos" data-linenos="266 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="267 "></span><span class="sd">        AnnData</span>
<span class="linenos" data-linenos="268 "></span><span class="sd">            Spatial data with imputed gene expression for all genes</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">            in the single-cell reference data.</span>
<span class="linenos" data-linenos="270 "></span>
<span class="linenos" data-linenos="271 "></span><span class="sd">    Notes:</span>
<span class="linenos" data-linenos="272 "></span><span class="sd">        - Uses mapping weights to transfer expression</span>
<span class="linenos" data-linenos="273 "></span><span class="sd">        - Can impute genes not used in original mapping</span>
<span class="linenos" data-linenos="274 "></span><span class="sd">        - Useful for analyzing spatial patterns of any gene</span>
<span class="linenos" data-linenos="275 "></span><span class="sd">        - Computationally intensive for large gene sets</span>
<span class="linenos" data-linenos="276 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="277 "></span>    <span class="k">if</span> <span class="n">ad_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="278 "></span>        <span class="n">ad_map</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ad_map</span>
<span class="linenos" data-linenos="279 "></span>    <span class="k">if</span> <span class="n">ad_sc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="280 "></span>        <span class="n">ad_sc</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_sc</span>
<span class="linenos" data-linenos="281 "></span>    <span class="n">ad_ge</span> <span class="o">=</span> <span class="n">tg</span><span class="o">.</span><span class="n">project_genes</span><span class="p">(</span><span class="n">adata_map</span><span class="o">=</span><span class="n">ad_map</span><span class="p">,</span> 
<span class="linenos" data-linenos="282 "></span>                             <span class="n">adata_sc</span><span class="o">=</span><span class="n">ad_sc</span><span class="p">,</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="linenos" data-linenos="283 "></span>    <span class="k">return</span> <span class="n">ad_ge</span>
</code></pre></div>
</details>
</div>
</div>
</div>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2025, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>