
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api pp - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.pp.preprocess">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api pp
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess/">
        Preprocessing the data of scRNA-seq with omicverse
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_via/">
        Trajectory Inference with VIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_via_velo/">
        Trajectory Inference with VIA and scVelo
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with cellfategenie
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_genesets/">
        Timing-associated geneset analysis with cellfategenie
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis in single cell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.preprocess">
    preprocess()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.highly_variable_features">
    highly_variable_features()
  </a>
<nav aria-label="highly_variable_features()" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.highly_variable_features--examples">
    Examples
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.remove_cc_genes">
    remove_cc_genes()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.identify_robust_genes">
    identify_robust_genes()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.scale">
    scale()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.regress">
    regress()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.regress_and_scale">
    regress_and_scale()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.pp.pca">
    pca()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api pp</h1>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.preprocess">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">'shiftlog|pearson'</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mi">50</span> <span class="o">*</span> <span class="mf">10000.0</span><span class="p">,</span> <span class="n">n_HVGs</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">'human'</span><span class="p">,</span> <span class="n">no_cc</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.preprocess" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Preprocesses the AnnData object adata using either a scanpy or a pearson residuals workflow for size normalization
and highly variable genes (HVGs) selection, and calculates signature scores if necessary. </p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>The data matrix.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>mode</code></td>
<td>
</td>
<td><p>The mode for size normalization and HVGs selection. It can be either 'scanpy' or 'pearson'. If 'scanpy', performs size normalization using scanpy's normalize_total() function and selects HVGs 
using pegasus' highly_variable_features() function with batch correction. If 'pearson', selects HVGs 
using scanpy's experimental.pp.highly_variable_genes() function with pearson residuals method and performs 
size normalization using scanpy's experimental.pp.normalize_pearson_residuals() function. </p></td>
<td>
<code>'shiftlog|pearson'</code>
</td>
</tr>
<tr>
<td><code>target_sum</code></td>
<td>
</td>
<td><p>The target total count after normalization.</p></td>
<td>
<code>50 * 10000.0</code>
</td>
</tr>
<tr>
<td><code>n_HVGs</code></td>
<td>
</td>
<td><p>the number of HVGs to select.</p></td>
<td>
<code>2000</code>
</td>
</tr>
<tr>
<td><code>organism</code></td>
<td>
</td>
<td><p>The organism of the data. It can be either 'human' or 'mouse'. </p></td>
<td>
<code>'human'</code>
</td>
</tr>
<tr>
<td><code>no_cc</code></td>
<td>
</td>
<td><p>Whether to remove cc-correlated genes from HVGs.</p></td>
<td>
<code>False</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>The preprocessed data matrix.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="341 "></span><span class="k">def</span> <span class="nf">preprocess</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">'shiftlog|pearson'</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mi">50</span><span class="o">*</span><span class="mf">1e4</span><span class="p">,</span> <span class="n">n_HVGs</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
<span class="linenos" data-linenos="342 "></span>    <span class="n">organism</span><span class="o">=</span><span class="s1">'human'</span><span class="p">,</span> <span class="n">no_cc</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="linenos" data-linenos="343 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="344 "></span><span class="sd">    Preprocesses the AnnData object adata using either a scanpy or a pearson residuals workflow for size normalization</span>
<span class="linenos" data-linenos="345 "></span><span class="sd">    and highly variable genes (HVGs) selection, and calculates signature scores if necessary. </span>
<span class="linenos" data-linenos="346 "></span>
<span class="linenos" data-linenos="347 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="348 "></span><span class="sd">        adata: The data matrix.</span>
<span class="linenos" data-linenos="349 "></span><span class="sd">        mode: The mode for size normalization and HVGs selection. It can be either 'scanpy' or 'pearson'. If 'scanpy', performs size normalization using scanpy's normalize_total() function and selects HVGs </span>
<span class="linenos" data-linenos="350 "></span><span class="sd">            using pegasus' highly_variable_features() function with batch correction. If 'pearson', selects HVGs </span>
<span class="linenos" data-linenos="351 "></span><span class="sd">            using scanpy's experimental.pp.highly_variable_genes() function with pearson residuals method and performs </span>
<span class="linenos" data-linenos="352 "></span><span class="sd">            size normalization using scanpy's experimental.pp.normalize_pearson_residuals() function. </span>
<span class="linenos" data-linenos="353 "></span><span class="sd">        target_sum: The target total count after normalization.</span>
<span class="linenos" data-linenos="354 "></span><span class="sd">        n_HVGs: the number of HVGs to select.</span>
<span class="linenos" data-linenos="355 "></span><span class="sd">        organism: The organism of the data. It can be either 'human' or 'mouse'. </span>
<span class="linenos" data-linenos="356 "></span><span class="sd">        no_cc: Whether to remove cc-correlated genes from HVGs.</span>
<span class="linenos" data-linenos="357 "></span>
<span class="linenos" data-linenos="358 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="359 "></span><span class="sd">        adata: The preprocessed data matrix. </span>
<span class="linenos" data-linenos="360 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="361 "></span>
<span class="linenos" data-linenos="362 "></span>    <span class="c1"># Log-normalization, HVGs identification</span>
<span class="linenos" data-linenos="363 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'Begin robust gene identification'</span><span class="p">)</span>
<span class="linenos" data-linenos="364 "></span>    <span class="n">identify_robust_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">percent_cells</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<span class="linenos" data-linenos="365 "></span>    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'robust'</span><span class="p">]]</span>
<span class="linenos" data-linenos="366 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'End of robust gene identification.'</span><span class="p">)</span>
<span class="linenos" data-linenos="367 "></span>    <span class="n">method_list</span> <span class="o">=</span> <span class="n">mode</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'|'</span><span class="p">)</span>
<span class="linenos" data-linenos="368 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Begin size normalization: </span><span class="si">{</span><span class="n">method_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1"> and HVGs selection </span><span class="si">{</span><span class="n">method_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="369 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'counts'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="370 "></span>
<span class="linenos" data-linenos="371 "></span>    <span class="k">if</span> <span class="n">method_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">'shiftlog'</span><span class="p">:</span> <span class="c1"># Size normalization + scanpy batch aware HVGs selection</span>
<span class="linenos" data-linenos="372 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span>
<span class="linenos" data-linenos="373 "></span>            <span class="n">adata</span><span class="p">,</span> 
<span class="linenos" data-linenos="374 "></span>            <span class="n">target_sum</span><span class="o">=</span><span class="n">target_sum</span><span class="p">,</span>
<span class="linenos" data-linenos="375 "></span>            <span class="n">exclude_highly_expressed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="376 "></span>            <span class="n">max_fraction</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="linenos" data-linenos="377 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="378 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="linenos" data-linenos="379 "></span>    <span class="k">elif</span> <span class="n">method_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">'pearson'</span><span class="p">:</span>
<span class="linenos" data-linenos="380 "></span>        <span class="c1"># Perason residuals workflow</span>
<span class="linenos" data-linenos="381 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_pearson_residuals</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="linenos" data-linenos="382 "></span>
<span class="linenos" data-linenos="383 "></span>    <span class="k">if</span> <span class="n">method_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">'pearson'</span><span class="p">:</span> <span class="c1"># Size normalization + scanpy batch aware HVGs selection</span>
<span class="linenos" data-linenos="384 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
<span class="linenos" data-linenos="385 "></span>            <span class="n">adata</span><span class="p">,</span> 
<span class="linenos" data-linenos="386 "></span>            <span class="n">flavor</span><span class="o">=</span><span class="s2">"pearson_residuals"</span><span class="p">,</span>
<span class="linenos" data-linenos="387 "></span>            <span class="n">layer</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">,</span>
<span class="linenos" data-linenos="388 "></span>            <span class="n">n_top_genes</span><span class="o">=</span><span class="n">n_HVGs</span><span class="p">,</span>
<span class="linenos" data-linenos="389 "></span>            <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span>
<span class="linenos" data-linenos="390 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="391 "></span>        <span class="k">if</span> <span class="n">no_cc</span><span class="p">:</span>
<span class="linenos" data-linenos="392 "></span>            <span class="n">remove_cc_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span> <span class="n">corr_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="linenos" data-linenos="393 "></span>    <span class="k">elif</span> <span class="n">method_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">'seurat'</span><span class="p">:</span>
<span class="linenos" data-linenos="394 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
<span class="linenos" data-linenos="395 "></span>            <span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="396 "></span>            <span class="n">flavor</span><span class="o">=</span><span class="s2">"seurat_v3"</span><span class="p">,</span>
<span class="linenos" data-linenos="397 "></span>            <span class="n">layer</span><span class="o">=</span><span class="s1">'counts'</span><span class="p">,</span>
<span class="linenos" data-linenos="398 "></span>            <span class="n">n_top_genes</span><span class="o">=</span><span class="n">n_HVGs</span><span class="p">,</span>
<span class="linenos" data-linenos="399 "></span>            <span class="n">batch_key</span><span class="o">=</span><span class="n">batch_key</span><span class="p">,</span>
<span class="linenos" data-linenos="400 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="401 "></span>        <span class="k">if</span> <span class="n">no_cc</span><span class="p">:</span>
<span class="linenos" data-linenos="402 "></span>            <span class="n">remove_cc_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span> <span class="n">corr_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="linenos" data-linenos="403 "></span>
<span class="linenos" data-linenos="404 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">'highly_variable_features'</span><span class="p">])</span>
<span class="linenos" data-linenos="405 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable_features'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable'</span><span class="p">]</span>
<span class="linenos" data-linenos="406 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">'highly_variable'</span><span class="p">])</span>
<span class="linenos" data-linenos="407 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">'means'</span><span class="p">:</span><span class="s1">'mean'</span><span class="p">,</span> <span class="s1">'variances'</span><span class="p">:</span><span class="s1">'var'</span><span class="p">})</span>
<span class="linenos" data-linenos="408 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">'End of size normalization: </span><span class="si">{</span><span class="n">method_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1"> and HVGs selection </span><span class="si">{</span><span class="n">method_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">'</span><span class="p">)</span>
<span class="linenos" data-linenos="409 "></span>
<span class="linenos" data-linenos="410 "></span>    <span class="k">return</span> <span class="n">adata</span> 
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.highly_variable_features">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_features</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">batch</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s1">'pegasus'</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span> <span class="n">span</span><span class="o">=</span><span class="mf">0.02</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">max_disp</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.highly_variable_features" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Highly variable features (HVF) selection. The input data should be logarithmized.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>Annotated data matrix with rows for cells and columns for genes.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>batch</code></td>
<td>
<code>str</code>
</td>
<td><p>A key in data.obs specifying batch information. If <code>batch</code> is not set, do not consider batch effects in selecting highly variable features. Otherwise, if <code>data.obs[batch]</code> is not categorical, <code>data.obs[batch]</code> will be automatically converted into categorical before highly variable feature selection.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>flavor</code></td>
<td>
<code>str</code>
</td>
<td><p>The HVF selection method to use. Available choices are <code>"pegasus"</code> or <code>"Seurat"</code>.</p></td>
<td>
<code>'pegasus'</code>
</td>
</tr>
<tr>
<td><code>n_top</code></td>
<td>
<code>int</code>
</td>
<td><p>Number of genes to be selected as HVF. if <code>None</code>, no gene will be selected.</p></td>
<td>
<code>2000</code>
</td>
</tr>
<tr>
<td><code>span</code></td>
<td>
<code>float</code>
</td>
<td><p>Only applicable when <code>flavor</code> is <code>"pegasus"</code>. The smoothing factor used by <em>scikit-learn loess</em> model in pegasus HVF selection method.</p></td>
<td>
<code>0.02</code>
</td>
</tr>
<tr>
<td><code>min_disp</code></td>
<td>
<code>float</code>
</td>
<td><p>Minimum normalized dispersion.</p></td>
<td>
<code>0.5</code>
</td>
</tr>
<tr>
<td><code>max_disp</code></td>
<td>
<code>float</code>
</td>
<td><p>Maximum normalized dispersion. Set it to <code>np.inf</code> for infinity bound.</p></td>
<td>
<code>np.inf</code>
</td>
</tr>
<tr>
<td><code>min_mean</code></td>
<td>
<code>float</code>
</td>
<td><p>Minimum mean.</p></td>
<td>
<code>0.0125</code>
</td>
</tr>
<tr>
<td><code>max_mean</code></td>
<td>
<code>float</code>
</td>
<td><p>Maximum mean.</p></td>
<td>
<code>7</code>
</td>
</tr>
<tr>
<td><code>n_jobs</code></td>
<td>
<code>int</code>
</td>
<td><p>Number of threads to be used during calculation. If <code>-1</code>, all physical CPU cores will be used.</p></td>
<td>
<code>-1</code>
</td>
</tr>
</tbody>
</table>
<p>Update <code>adata.var</code>:
    * <code>highly_variable_features</code>: replace with Boolean type array indicating the selected highly variable features.</p>
<h4 id="omicverse.pp.highly_variable_features--examples">Examples<a class="headerlink" href="#omicverse.pp.highly_variable_features--examples" title="Permanent link">¶</a></h4>
<blockquote>
<blockquote>
<blockquote>
<p>ov.pp.highly_variable_features(data)
ov.pp.highly_variable_features(data, batch="Channel")</p>
</blockquote>
</blockquote>
</blockquote>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="228 "></span><span class="k">def</span> <span class="nf">highly_variable_features</span><span class="p">(</span>
<span class="linenos" data-linenos="229 "></span>    <span class="n">data</span><span class="p">:</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
<span class="linenos" data-linenos="230 "></span>    <span class="n">batch</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="231 "></span>    <span class="n">flavor</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">"pegasus"</span><span class="p">,</span>
<span class="linenos" data-linenos="232 "></span>    <span class="n">n_top</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span>
<span class="linenos" data-linenos="233 "></span>    <span class="n">span</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.02</span><span class="p">,</span>
<span class="linenos" data-linenos="234 "></span>    <span class="n">min_disp</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
<span class="linenos" data-linenos="235 "></span>    <span class="n">max_disp</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span>
<span class="linenos" data-linenos="236 "></span>    <span class="n">min_mean</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0125</span><span class="p">,</span>
<span class="linenos" data-linenos="237 "></span>    <span class="n">max_mean</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span>
<span class="linenos" data-linenos="238 "></span>    <span class="n">n_jobs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span>
<span class="linenos" data-linenos="239 "></span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="240 "></span><span class="w">    </span><span class="sd">""" Highly variable features (HVF) selection. The input data should be logarithmized.</span>
<span class="linenos" data-linenos="241 "></span>
<span class="linenos" data-linenos="242 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="243 "></span><span class="sd">        data: Annotated data matrix with rows for cells and columns for genes.</span>
<span class="linenos" data-linenos="244 "></span><span class="sd">        batch: A key in data.obs specifying batch information. If `batch` is not set, do not consider batch effects in selecting highly variable features. Otherwise, if `data.obs[batch]` is not categorical, `data.obs[batch]` will be automatically converted into categorical before highly variable feature selection.</span>
<span class="linenos" data-linenos="245 "></span><span class="sd">        flavor: The HVF selection method to use. Available choices are ``"pegasus"`` or ``"Seurat"``.</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">        n_top: Number of genes to be selected as HVF. if ``None``, no gene will be selected.</span>
<span class="linenos" data-linenos="247 "></span><span class="sd">        span: Only applicable when ``flavor`` is ``"pegasus"``. The smoothing factor used by *scikit-learn loess* model in pegasus HVF selection method.</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">        min_disp: Minimum normalized dispersion.</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">        max_disp: Maximum normalized dispersion. Set it to ``np.inf`` for infinity bound.</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">        min_mean: Minimum mean.</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">        max_mean: Maximum mean.</span>
<span class="linenos" data-linenos="252 "></span><span class="sd">        n_jobs: Number of threads to be used during calculation. If ``-1``, all physical CPU cores will be used.</span>
<span class="linenos" data-linenos="253 "></span>
<span class="linenos" data-linenos="254 "></span>
<span class="linenos" data-linenos="255 "></span><span class="sd">    Update ``adata.var``:</span>
<span class="linenos" data-linenos="256 "></span><span class="sd">        * ``highly_variable_features``: replace with Boolean type array indicating the selected highly variable features.</span>
<span class="linenos" data-linenos="257 "></span>
<span class="linenos" data-linenos="258 "></span><span class="sd">    Examples</span>
<span class="linenos" data-linenos="259 "></span><span class="sd">    --------</span>
<span class="linenos" data-linenos="260 "></span><span class="sd">    &gt;&gt;&gt; ov.pp.highly_variable_features(data)</span>
<span class="linenos" data-linenos="261 "></span><span class="sd">    &gt;&gt;&gt; ov.pp.highly_variable_features(data, batch="Channel")</span>
<span class="linenos" data-linenos="262 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="263 "></span>
<span class="linenos" data-linenos="264 "></span>    <span class="k">if</span> <span class="n">flavor</span> <span class="o">==</span> <span class="s2">"pegasus"</span><span class="p">:</span>
<span class="linenos" data-linenos="265 "></span>        <span class="n">select_hvf_pegasus</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">batch</span><span class="p">,</span> <span class="n">n_top</span><span class="o">=</span><span class="n">n_top</span><span class="p">,</span> <span class="n">span</span><span class="o">=</span><span class="n">span</span><span class="p">)</span>
<span class="linenos" data-linenos="266 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="267 "></span>        <span class="k">assert</span> <span class="n">flavor</span> <span class="o">==</span> <span class="s2">"Seurat"</span>
<span class="linenos" data-linenos="268 "></span>        <span class="n">select_hvf_seurat</span><span class="p">(</span>
<span class="linenos" data-linenos="269 "></span>            <span class="n">data</span><span class="p">,</span>
<span class="linenos" data-linenos="270 "></span>            <span class="n">batch</span><span class="p">,</span>
<span class="linenos" data-linenos="271 "></span>            <span class="n">n_top</span><span class="o">=</span><span class="n">n_top</span><span class="p">,</span>
<span class="linenos" data-linenos="272 "></span>            <span class="n">min_disp</span><span class="o">=</span><span class="n">min_disp</span><span class="p">,</span>
<span class="linenos" data-linenos="273 "></span>            <span class="n">max_disp</span><span class="o">=</span><span class="n">max_disp</span><span class="p">,</span>
<span class="linenos" data-linenos="274 "></span>            <span class="n">min_mean</span><span class="o">=</span><span class="n">min_mean</span><span class="p">,</span>
<span class="linenos" data-linenos="275 "></span>            <span class="n">max_mean</span><span class="o">=</span><span class="n">max_mean</span><span class="p">,</span>
<span class="linenos" data-linenos="276 "></span>            <span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">,</span>
<span class="linenos" data-linenos="277 "></span>        <span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>
<span class="linenos" data-linenos="279 "></span>    <span class="n">data</span><span class="o">.</span><span class="n">uns</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">"_tmp_fmat_highly_variable_features"</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span> <span class="c1"># Pop up cached feature matrix</span>
<span class="linenos" data-linenos="280 "></span>
<span class="linenos" data-linenos="281 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable_features'</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2"> highly variable features have been selected."</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.remove_cc_genes">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">remove_cc_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s1">'human'</span><span class="p">,</span> <span class="n">corr_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.remove_cc_genes" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Update adata.var['highly_variable_features'] discarding cc correlated genes. 
Taken from Cospar, Wang et al., 2023.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>Annotated data matrix with rows for cells and columns for genes.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>organism</code></td>
<td>
<code>str</code>
</td>
<td><p>Organism of the dataset. Available choices are <code>"human"</code> or <code>"mouse"</code>.</p></td>
<td>
<code>'human'</code>
</td>
</tr>
<tr>
<td><code>corr_threshold</code></td>
<td>
<code>float</code>
</td>
<td><p>Threshold for correlation with cc genes. Genes having a correlation with cc genes &gt; corr_threshold will be discarded.</p></td>
<td>
<code>0.1</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="307 "></span><span class="k">def</span> <span class="nf">remove_cc_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">organism</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'human'</span><span class="p">,</span> <span class="n">corr_threshold</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.1</span><span class="p">):</span>
<span class="linenos" data-linenos="308 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="309 "></span><span class="sd">    Update adata.var['highly_variable_features'] discarding cc correlated genes. </span>
<span class="linenos" data-linenos="310 "></span><span class="sd">    Taken from Cospar, Wang et al., 2023.</span>
<span class="linenos" data-linenos="311 "></span>
<span class="linenos" data-linenos="312 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="313 "></span><span class="sd">        adata: Annotated data matrix with rows for cells and columns for genes.</span>
<span class="linenos" data-linenos="314 "></span><span class="sd">        organism: Organism of the dataset. Available choices are ``"human"`` or ``"mouse"``.</span>
<span class="linenos" data-linenos="315 "></span><span class="sd">        corr_threshold: Threshold for correlation with cc genes. Genes having a correlation with cc genes &gt; corr_threshold will be discarded.</span>
<span class="linenos" data-linenos="316 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="317 "></span>    <span class="c1"># Get cc genes</span>
<span class="linenos" data-linenos="318 "></span>    <span class="n">cycling_genes</span> <span class="o">=</span> <span class="n">load_signatures_from_file</span><span class="p">(</span><span class="n">predefined_signatures</span><span class="p">[</span><span class="sa">f</span><span class="s1">'cell_cycle_</span><span class="si">{</span><span class="n">organism</span><span class="si">}</span><span class="s1">'</span><span class="p">])</span>
<span class="linenos" data-linenos="319 "></span>    <span class="n">cc_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">cycling_genes</span><span class="p">[</span><span class="s1">'G1/S'</span><span class="p">])</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">cycling_genes</span><span class="p">[</span><span class="s1">'G2/M'</span><span class="p">]))</span>
<span class="linenos" data-linenos="320 "></span>    <span class="n">cc_genes</span> <span class="o">=</span> <span class="p">[</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cc_genes</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="p">]</span>
<span class="linenos" data-linenos="321 "></span>
<span class="linenos" data-linenos="322 "></span>    <span class="c1"># Compute corr</span>
<span class="linenos" data-linenos="323 "></span>    <span class="n">cc_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">cc_genes</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="324 "></span>    <span class="n">HVGs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable_features'</span><span class="p">]]</span>
<span class="linenos" data-linenos="325 "></span>    <span class="n">hvgs_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">HVGs</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">A</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="326 "></span>    <span class="n">cc_corr</span> <span class="o">=</span> <span class="n">corr2_coeff</span><span class="p">(</span><span class="n">hvgs_expression</span><span class="p">,</span> <span class="n">cc_expression</span><span class="p">)</span>
<span class="linenos" data-linenos="327 "></span>
<span class="linenos" data-linenos="328 "></span>    <span class="c1"># Discard genes having the maximum correlation with one of the cc &gt; corr_threshold</span>
<span class="linenos" data-linenos="329 "></span>    <span class="n">max_corr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">cc_corr</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="330 "></span>    <span class="n">HVGs_no_cc</span> <span class="o">=</span> <span class="n">HVGs</span><span class="p">[</span><span class="n">max_corr</span> <span class="o">&lt;</span> <span class="n">corr_threshold</span><span class="p">]</span>
<span class="linenos" data-linenos="331 "></span>    <span class="nb">print</span><span class="p">(</span>
<span class="linenos" data-linenos="332 "></span>        <span class="sa">f</span><span class="s1">'Number of selected non-cycling highly variable genes: </span><span class="si">{</span><span class="n">HVGs_no_cc</span><span class="o">.</span><span class="n">size</span><span class="si">}</span><span class="se">\n</span><span class="s1">'</span>
<span class="linenos" data-linenos="333 "></span>        <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">max_corr</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">corr_threshold</span><span class="p">)</span><span class="si">}</span><span class="s1"> cell cycle correlated genes will be removed...'</span>
<span class="linenos" data-linenos="334 "></span>    <span class="p">)</span>
<span class="linenos" data-linenos="335 "></span>    <span class="c1"># Update </span>
<span class="linenos" data-linenos="336 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'highly_variable_features'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">HVGs_no_cc</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.identify_robust_genes">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">identify_robust_genes</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">percent_cells</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.identify_robust_genes" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Identify robust genes as candidates for HVG selection and remove genes that are not expressed in any cells.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>data</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>Use current selected modality in data, which should contain one RNA expression matrix.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>percent_cells</code></td>
<td>
<code>float</code>
</td>
<td><p>Only assign genes to be <code>robust</code> that are expressed in at least <code>percent_cells</code> % of cells.</p></td>
<td>
<code>0.05</code>
</td>
</tr>
</tbody>
</table>
<p>Update <code>data.var</code>:</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>* ``n_cells``: Total number of cells in which each gene is measured.
<span class="linenos" data-linenos="2 "></span>* ``percent_cells``: Percent of cells in which each gene is measured.
<span class="linenos" data-linenos="3 "></span>* ``robust``: Boolean type indicating if a gene is robust based on the QC metrics.
<span class="linenos" data-linenos="4 "></span>* ``highly_variable_features``: Boolean type indicating if a gene is a highly variable feature. By default, set all robust genes as highly variable features.
</code></pre></div>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="17 "></span><span class="k">def</span> <span class="nf">identify_robust_genes</span><span class="p">(</span><span class="n">data</span><span class="p">:</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">percent_cells</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="18 "></span><span class="w">    </span><span class="sd">""" </span>
<span class="linenos" data-linenos="19 "></span><span class="sd">    Identify robust genes as candidates for HVG selection and remove genes that are not expressed in any cells.</span>
<span class="linenos" data-linenos="20 "></span>
<span class="linenos" data-linenos="21 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="22 "></span><span class="sd">        data: Use current selected modality in data, which should contain one RNA expression matrix.</span>
<span class="linenos" data-linenos="23 "></span><span class="sd">        percent_cells: Only assign genes to be ``robust`` that are expressed in at least ``percent_cells`` % of cells.</span>
<span class="linenos" data-linenos="24 "></span>
<span class="linenos" data-linenos="25 "></span>
<span class="linenos" data-linenos="26 "></span><span class="sd">    Update ``data.var``:</span>
<span class="linenos" data-linenos="27 "></span>
<span class="linenos" data-linenos="28 "></span><span class="sd">        * ``n_cells``: Total number of cells in which each gene is measured.</span>
<span class="linenos" data-linenos="29 "></span><span class="sd">        * ``percent_cells``: Percent of cells in which each gene is measured.</span>
<span class="linenos" data-linenos="30 "></span><span class="sd">        * ``robust``: Boolean type indicating if a gene is robust based on the QC metrics.</span>
<span class="linenos" data-linenos="31 "></span><span class="sd">        * ``highly_variable_features``: Boolean type indicating if a gene is a highly variable feature. By default, set all robust genes as highly variable features.</span>
<span class="linenos" data-linenos="32 "></span>
<span class="linenos" data-linenos="33 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="34 "></span>
<span class="linenos" data-linenos="35 "></span>    <span class="n">prior_n</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="linenos" data-linenos="36 "></span>
<span class="linenos" data-linenos="37 "></span>    <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
<span class="linenos" data-linenos="38 "></span>        <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"n_cells"</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">getnnz</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="39 "></span>        <span class="n">data</span><span class="o">.</span><span class="n">_inplace_subset_var</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"n_cells"</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="40 "></span>        <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"percent_cells"</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"n_cells"</span><span class="p">]</span> <span class="o">/</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">*</span> <span class="mi">100</span>
<span class="linenos" data-linenos="41 "></span>        <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"robust"</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"percent_cells"</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">percent_cells</span>
<span class="linenos" data-linenos="42 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="43 "></span>        <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"robust"</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="linenos" data-linenos="44 "></span>
<span class="linenos" data-linenos="45 "></span>    <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"highly_variable_features"</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">"robust"</span><span class="p">]</span>  <span class="c1"># default all robust genes are "highly" variable</span>
<span class="linenos" data-linenos="46 "></span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"After filtration, </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">prior_n</span><span class="si">}</span><span class="s2"> genes are kept. Among </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> genes, </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">'robust'</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span><span class="si">}</span><span class="s2"> genes are robust."</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.scale">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">layers_add</span><span class="o">=</span><span class="s1">'scaled'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.scale" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Scale the input AnnData object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'scaled' that stores
the expression matrix that has been scaled to unit variance and zero mean.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="420 "></span><span class="k">def</span> <span class="nf">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">layers_add</span><span class="o">=</span><span class="s1">'scaled'</span><span class="p">):</span>
<span class="linenos" data-linenos="421 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="422 "></span><span class="sd">    Scale the input AnnData object.</span>
<span class="linenos" data-linenos="423 "></span>
<span class="linenos" data-linenos="424 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="425 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape.</span>
<span class="linenos" data-linenos="426 "></span>
<span class="linenos" data-linenos="427 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="428 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'scaled' that stores</span>
<span class="linenos" data-linenos="429 "></span><span class="sd">            the expression matrix that has been scaled to unit variance and zero mean.</span>
<span class="linenos" data-linenos="430 "></span>
<span class="linenos" data-linenos="431 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="432 "></span>    <span class="n">adata_mock</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">max_value</span><span class="o">=</span><span class="n">max_value</span><span class="p">)</span>
<span class="linenos" data-linenos="433 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layers_add</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_mock</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="434 "></span>    <span class="k">del</span> <span class="n">adata_mock</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.regress">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.regress" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Regress out covariates from the input AnnData object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape. Should contain columns 'mito_perc' and 'nUMIs'
that represent the percentage of mitochondrial genes and the total number of UMI counts, respectively.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'regressed' that stores
the expression matrix with covariates regressed out.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="436 "></span><span class="k">def</span> <span class="nf">regress</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
<span class="linenos" data-linenos="437 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="438 "></span><span class="sd">    Regress out covariates from the input AnnData object.</span>
<span class="linenos" data-linenos="439 "></span>
<span class="linenos" data-linenos="440 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="441 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape. Should contain columns 'mito_perc' and 'nUMIs'</span>
<span class="linenos" data-linenos="442 "></span><span class="sd">            that represent the percentage of mitochondrial genes and the total number of UMI counts, respectively.</span>
<span class="linenos" data-linenos="443 "></span>
<span class="linenos" data-linenos="444 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="445 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'regressed' that stores</span>
<span class="linenos" data-linenos="446 "></span><span class="sd">            the expression matrix with covariates regressed out.</span>
<span class="linenos" data-linenos="447 "></span>
<span class="linenos" data-linenos="448 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="449 "></span>    <span class="n">adata_mock</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_out</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s1">'mito_perc'</span><span class="p">,</span> <span class="s1">'nUMIs'</span><span class="p">],</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="450 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'regressed'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_mock</span><span class="o">.</span><span class="n">X</span>
<span class="linenos" data-linenos="451 "></span>    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.regress_and_scale">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_and_scale</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.regress_and_scale" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Regress out covariates from the input AnnData object and scale the resulting expression matrix.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape. Should contain a layer called 'regressed'
that stores the expression matrix with covariates regressed out.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'regressed_and_scaled'
that stores the expression matrix with covariates regressed out and then scaled.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="453 "></span><span class="k">def</span> <span class="nf">regress_and_scale</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
<span class="linenos" data-linenos="454 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="455 "></span><span class="sd">    Regress out covariates from the input AnnData object and scale the resulting expression matrix.</span>
<span class="linenos" data-linenos="456 "></span>
<span class="linenos" data-linenos="457 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="458 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape. Should contain a layer called 'regressed'</span>
<span class="linenos" data-linenos="459 "></span><span class="sd">            that stores the expression matrix with covariates regressed out.</span>
<span class="linenos" data-linenos="460 "></span>
<span class="linenos" data-linenos="461 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="462 "></span><span class="sd">        adata : Annotated data matrix with n_obs x n_vars shape. Adds a new layer called 'regressed_and_scaled'</span>
<span class="linenos" data-linenos="463 "></span><span class="sd">            that stores the expression matrix with covariates regressed out and then scaled.</span>
<span class="linenos" data-linenos="464 "></span>
<span class="linenos" data-linenos="465 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="466 "></span>    <span class="k">if</span> <span class="s1">'regressed'</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">:</span>
<span class="linenos" data-linenos="467 "></span>        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="s1">'Regress out covariates first!'</span><span class="p">)</span>
<span class="linenos" data-linenos="468 "></span>    <span class="n">adata_mock</span><span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos="469 "></span>    <span class="n">adata_mock</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata_mock</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'regressed'</span><span class="p">]</span>
<span class="linenos" data-linenos="470 "></span>    <span class="n">adata_mock</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">adata_mock</span><span class="p">)</span>
<span class="linenos" data-linenos="471 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'regressed_and_scaled'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_mock</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'scaled'</span><span class="p">]</span>
<span class="linenos" data-linenos="472 "></span>
<span class="linenos" data-linenos="473 "></span>    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.pp.pca">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">'scaled'</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.pp.pca" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Performs Principal Component Analysis (PCA) on the data stored in a scanpy AnnData object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>Annotated data matrix with rows representing cells and columns representing features.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>n_pcs</code></td>
<td>
</td>
<td><p>Number of principal components to calculate.</p></td>
<td>
<code>50</code>
</td>
</tr>
<tr>
<td><code>layer</code></td>
<td>
</td>
<td><p>The name of the layer in <code>adata</code> where the data to be analyzed is stored. Defaults to the 'scaled' layer,
and falls back to 'lognorm' if that layer does not exist. Raises a KeyError if the specified layer is not present.</p></td>
<td>
<code>'scaled'</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p>The original AnnData object with the calculated PCA embeddings and other information stored in its <code>obsm</code>, <code>varm</code>,
and <code>uns</code> fields.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/pp/_preprocess.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="506 "></span><span class="k">def</span> <span class="nf">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">'scaled'</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="linenos" data-linenos="507 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="508 "></span><span class="sd">    Performs Principal Component Analysis (PCA) on the data stored in a scanpy AnnData object.</span>
<span class="linenos" data-linenos="509 "></span>
<span class="linenos" data-linenos="510 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="511 "></span><span class="sd">        adata : Annotated data matrix with rows representing cells and columns representing features.</span>
<span class="linenos" data-linenos="512 "></span><span class="sd">        n_pcs : Number of principal components to calculate.</span>
<span class="linenos" data-linenos="513 "></span><span class="sd">        layer : The name of the layer in `adata` where the data to be analyzed is stored. Defaults to the 'scaled' layer,</span>
<span class="linenos" data-linenos="514 "></span><span class="sd">            and falls back to 'lognorm' if that layer does not exist. Raises a KeyError if the specified layer is not present.</span>
<span class="linenos" data-linenos="515 "></span>
<span class="linenos" data-linenos="516 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="517 "></span><span class="sd">        adata : The original AnnData object with the calculated PCA embeddings and other information stored in its `obsm`, `varm`,</span>
<span class="linenos" data-linenos="518 "></span><span class="sd">            and `uns` fields.</span>
<span class="linenos" data-linenos="519 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="520 "></span>    <span class="k">if</span> <span class="s1">'lognorm'</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">:</span>
<span class="linenos" data-linenos="521 "></span>        <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">'lognorm'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
<span class="linenos" data-linenos="522 "></span>    <span class="k">if</span> <span class="n">layer</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">:</span> 
<span class="linenos" data-linenos="523 "></span>        <span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">layer</span><span class="p">]</span>
<span class="linenos" data-linenos="524 "></span>        <span class="n">key</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">layer</span><span class="si">}</span><span class="s1">|original'</span>
<span class="linenos" data-linenos="525 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="526 "></span>        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s1">'Selected layer </span><span class="si">{</span><span class="n">layer</span><span class="si">}</span><span class="s1"> is not present. Compute it first!'</span><span class="p">)</span>
<span class="linenos" data-linenos="527 "></span>
<span class="linenos" data-linenos="528 "></span>    <span class="n">model</span> <span class="o">=</span> <span class="n">my_PCA</span><span class="p">()</span>
<span class="linenos" data-linenos="529 "></span>    <span class="n">model</span><span class="o">.</span><span class="n">calculate_PCA</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="n">n_pcs</span><span class="p">)</span>
<span class="linenos" data-linenos="530 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">key</span> <span class="o">+</span> <span class="s1">'|X_pca'</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">embs</span>
<span class="linenos" data-linenos="531 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">key</span> <span class="o">+</span> <span class="s1">'|pca_loadings'</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">loads</span>
<span class="linenos" data-linenos="532 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">key</span> <span class="o">+</span> <span class="s1">'|pca_var_ratios'</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">var_ratios</span>
<span class="linenos" data-linenos="533 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">key</span> <span class="o">+</span> <span class="s1">'|cum_sum_eigenvalues'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">var_ratios</span><span class="p">)</span>
<span class="linenos" data-linenos="534 "></span>    <span class="k">if</span> <span class="n">inplace</span><span class="p">:</span>
<span class="linenos" data-linenos="535 "></span>        <span class="k">return</span> <span class="kc">None</span>
<span class="linenos" data-linenos="536 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="537 "></span>        <span class="k">return</span> <span class="n">adata</span>  
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2023, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>