
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api cpdb - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.single.cpdb_network_cal">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api cpdb
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.cpdb_network_cal">
    cpdb_network_cal()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.cpdb_plot_network">
    cpdb_plot_network()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.cpdb_plot_interaction">
    cpdb_plot_interaction()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.cpdb_interaction_filtered">
    cpdb_interaction_filtered()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.cpdb_submeans_exacted">
    cpdb_submeans_exacted()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api cpdb</h1>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.cpdb_network_cal">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">cpdb_network_cal</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">pvals</span><span class="p">,</span> <span class="n">celltype_key</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.cpdb_network_cal" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Calculate a CPDB (Cell Phone Database) network using gene expression data and return a dictionary of results.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>An AnnData object containing gene expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pvals</code></td>
<td>
<code>list</code>
</td>
<td><p>A list or array of p-values for each gene in the expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>celltype_key</code></td>
<td>
<code>str</code>
</td>
<td><p>The name of the cell type key in adata.obs.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>cpdb_dict</code></td> <td>
<code>dict</code>
</td>
<td><p>A dictionary of results generated by <code>kpy.plot_cpdb_heatmap()</code> from the <code>ktplotspy</code> library.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_cpdb.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="38 "></span><span class="k">def</span> <span class="nf">cpdb_network_cal</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">pvals</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span><span class="n">celltype_key</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span><span class="o">-&gt;</span><span class="nb">dict</span><span class="p">:</span>
<span class="linenos" data-linenos="39 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="40 "></span><span class="sd">    Calculate a CPDB (Cell Phone Database) network using gene expression data and return a dictionary of results.</span>
<span class="linenos" data-linenos="41 "></span>
<span class="linenos" data-linenos="42 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="43 "></span><span class="sd">        adata: An AnnData object containing gene expression data.</span>
<span class="linenos" data-linenos="44 "></span><span class="sd">        pvals: A list or array of p-values for each gene in the expression data.</span>
<span class="linenos" data-linenos="45 "></span><span class="sd">        celltype_key: The name of the cell type key in adata.obs.</span>
<span class="linenos" data-linenos="46 "></span>
<span class="linenos" data-linenos="47 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="48 "></span><span class="sd">        cpdb_dict: A dictionary of results generated by `kpy.plot_cpdb_heatmap()` from the `ktplotspy` library.</span>
<span class="linenos" data-linenos="49 "></span>
<span class="linenos" data-linenos="50 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="51 "></span>    <span class="n">check_kpy</span><span class="p">()</span>
<span class="linenos" data-linenos="52 "></span>    <span class="k">global</span> <span class="n">kpy_install</span>
<span class="linenos" data-linenos="53 "></span>    <span class="k">if</span> <span class="n">kpy_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="54 "></span>        <span class="n">global_imports</span><span class="p">(</span><span class="s2">"ktplotspy"</span><span class="p">,</span><span class="s2">"kpy"</span><span class="p">)</span>
<span class="linenos" data-linenos="55 "></span>
<span class="linenos" data-linenos="56 "></span>    <span class="n">cpdb_dict</span><span class="o">=</span><span class="n">kpy</span><span class="o">.</span><span class="n">plot_cpdb_heatmap</span><span class="p">(</span>
<span class="linenos" data-linenos="57 "></span>        <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="58 "></span>        <span class="n">pvals</span> <span class="o">=</span> <span class="n">pvals</span><span class="p">,</span>
<span class="linenos" data-linenos="59 "></span>        <span class="n">celltype_key</span> <span class="o">=</span> <span class="n">celltype_key</span><span class="p">,</span>
<span class="linenos" data-linenos="60 "></span>        <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>
<span class="linenos" data-linenos="61 "></span>        <span class="n">title</span> <span class="o">=</span> <span class="s2">""</span><span class="p">,</span>
<span class="linenos" data-linenos="62 "></span>        <span class="n">symmetrical</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="linenos" data-linenos="63 "></span>        <span class="n">return_tables</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="64 "></span>    <span class="p">)</span>
<span class="linenos" data-linenos="65 "></span>    <span class="k">return</span> <span class="n">cpdb_dict</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.cpdb_plot_network">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">cpdb_plot_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">interaction_edges</span><span class="p">,</span> <span class="n">celltype_key</span><span class="p">,</span> <span class="n">nodecolor_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">edgeswidth_scale</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">nodesize_scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pos_scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">pos_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">legend_ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">legend_bbox</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="n">legend_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">return_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.cpdb_plot_network" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Plot a network of interactions between cell types using gene expression data.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>An AnnData object containing gene expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>interaction_edges</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>A DataFrame containing the edges of the network, including the source, target, and count of interactions.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>celltype_key</code></td>
<td>
<code>str</code>
</td>
<td><p>The name of the cell type key in adata.obs.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>nodecolor_dict</code></td>
<td>
</td>
<td><p>A dictionary mapping cell type names to node colors. If not provided, uses the colors specified in adata.uns.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>edgeswidth_scale</code></td>
<td>
<code>int</code>
</td>
<td><p>The scaling factor for edge width. Default is 10.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>nodesize_scale</code></td>
<td>
<code>int</code>
</td>
<td><p>The scaling factor for node size. Default is 1.</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>pos_scale</code></td>
<td>
<code>int</code>
</td>
<td><p>The scaling factor for node positions. Default is 1.</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>pos_size</code></td>
<td>
<code>int</code>
</td>
<td><p>The size of the node positions. Default is 10.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The size of the plot. Default is (5,5).</p></td>
<td>
<code>(5, 5)</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
<code>str</code>
</td>
<td><p>The title of the plot. Default is ''.</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>legend_ncol</code></td>
<td>
<code>int</code>
</td>
<td><p>The number of columns in the legend. Default is 3.</p></td>
<td>
<code>3</code>
</td>
</tr>
<tr>
<td><code>legend_bbox</code></td>
<td>
<code>tuple</code>
</td>
<td><p>The location of the legend. Default is (1,0.2).</p></td>
<td>
<code>(1, 0.2)</code>
</td>
</tr>
<tr>
<td><code>legend_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>The font size of the legend. Default is 10.</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>return_graph</code></td>
<td>
<code>bool</code>
</td>
<td><p>If True, returns the network graph. Default is False.</p></td>
<td>
<code>False</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<code><span title="networkx">nx</span>.<span title="networkx.Graph">Graph</span></code>
</td>
<td><p>ax or G: The plot if <code>return_graph</code> is False, otherwise the network graph.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_cpdb.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 68 "></span><span class="k">def</span> <span class="nf">cpdb_plot_network</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">interaction_edges</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
<span class="linenos" data-linenos=" 69 "></span>                      <span class="n">celltype_key</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">nodecolor_dict</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos=" 70 "></span>                      <span class="n">edgeswidth_scale</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">nodesize_scale</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="linenos" data-linenos=" 71 "></span>                      <span class="n">pos_scale</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">pos_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span><span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span>
<span class="linenos" data-linenos=" 72 "></span>                      <span class="n">legend_ncol</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">legend_bbox</span><span class="p">:</span><span class="nb">tuple</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.2</span><span class="p">),</span><span class="n">legend_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
<span class="linenos" data-linenos=" 73 "></span>                     <span class="n">return_graph</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">:</span>
<span class="linenos" data-linenos=" 74 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos=" 75 "></span><span class="sd">    Plot a network of interactions between cell types using gene expression data.</span>
<span class="linenos" data-linenos=" 76 "></span>
<span class="linenos" data-linenos=" 77 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos=" 78 "></span><span class="sd">        adata: An AnnData object containing gene expression data.</span>
<span class="linenos" data-linenos=" 79 "></span><span class="sd">        interaction_edges: A DataFrame containing the edges of the network, including the source, target, and count of interactions.</span>
<span class="linenos" data-linenos=" 80 "></span><span class="sd">        celltype_key: The name of the cell type key in adata.obs.</span>
<span class="linenos" data-linenos=" 81 "></span><span class="sd">        nodecolor_dict: A dictionary mapping cell type names to node colors. If not provided, uses the colors specified in adata.uns.</span>
<span class="linenos" data-linenos=" 82 "></span><span class="sd">        edgeswidth_scale: The scaling factor for edge width. Default is 10.</span>
<span class="linenos" data-linenos=" 83 "></span><span class="sd">        nodesize_scale: The scaling factor for node size. Default is 1.</span>
<span class="linenos" data-linenos=" 84 "></span><span class="sd">        pos_scale: The scaling factor for node positions. Default is 1.</span>
<span class="linenos" data-linenos=" 85 "></span><span class="sd">        pos_size: The size of the node positions. Default is 10.</span>
<span class="linenos" data-linenos=" 86 "></span><span class="sd">        figsize: The size of the plot. Default is (5,5).</span>
<span class="linenos" data-linenos=" 87 "></span><span class="sd">        title: The title of the plot. Default is ''.</span>
<span class="linenos" data-linenos=" 88 "></span><span class="sd">        legend_ncol: The number of columns in the legend. Default is 3.</span>
<span class="linenos" data-linenos=" 89 "></span><span class="sd">        legend_bbox: The location of the legend. Default is (1,0.2).</span>
<span class="linenos" data-linenos=" 90 "></span><span class="sd">        legend_fontsize: The font size of the legend. Default is 10.</span>
<span class="linenos" data-linenos=" 91 "></span><span class="sd">        return_graph: If True, returns the network graph. Default is False.</span>
<span class="linenos" data-linenos=" 92 "></span>
<span class="linenos" data-linenos=" 93 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos=" 94 "></span><span class="sd">        ax or G: The plot if `return_graph` is False, otherwise the network graph.</span>
<span class="linenos" data-linenos=" 95 "></span>
<span class="linenos" data-linenos=" 96 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos=" 97 "></span>    <span class="n">check_kpy</span><span class="p">()</span>
<span class="linenos" data-linenos=" 98 "></span>    <span class="k">global</span> <span class="n">kpy_install</span>
<span class="linenos" data-linenos=" 99 "></span>    <span class="k">if</span> <span class="n">kpy_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="100 "></span>        <span class="n">global_imports</span><span class="p">(</span><span class="s2">"ktplotspy"</span><span class="p">,</span><span class="s2">"kpy"</span><span class="p">)</span>
<span class="linenos" data-linenos="101 "></span>
<span class="linenos" data-linenos="102 "></span>    <span class="c1">#set Digraph of cellphonedb</span>
<span class="linenos" data-linenos="103 "></span>    <span class="n">G</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">DiGraph</span><span class="p">()</span>
<span class="linenos" data-linenos="104 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">interaction_edges</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
<span class="linenos" data-linenos="105 "></span>        <span class="n">G</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span><span class="n">interaction_edges</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'SOURCE'</span><span class="p">],</span>
<span class="linenos" data-linenos="106 "></span>                   <span class="n">interaction_edges</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'TARGET'</span><span class="p">],</span>
<span class="linenos" data-linenos="107 "></span>                   <span class="n">weight</span><span class="o">=</span><span class="n">interaction_edges</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">'COUNT'</span><span class="p">],)</span>
<span class="linenos" data-linenos="108 "></span>
<span class="linenos" data-linenos="109 "></span>    <span class="c1">#set celltypekey's color</span>
<span class="linenos" data-linenos="110 "></span>    <span class="k">if</span> <span class="n">nodecolor_dict</span><span class="o">!=</span><span class="kc">None</span><span class="p">:</span>
<span class="linenos" data-linenos="111 "></span>        <span class="n">type_color_all</span><span class="o">=</span><span class="n">nodecolor_dict</span>
<span class="linenos" data-linenos="112 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="113 "></span>        <span class="k">if</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">_colors'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">celltype_key</span><span class="p">)</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
<span class="linenos" data-linenos="114 "></span>            <span class="n">type_color_all</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'</span><span class="si">{}</span><span class="s1">_colors'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">celltype_key</span><span class="p">)]))</span>
<span class="linenos" data-linenos="115 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="116 "></span>            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">28</span><span class="p">:</span>
<span class="linenos" data-linenos="117 "></span>                <span class="n">type_color_all</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">default_102</span><span class="p">))</span>
<span class="linenos" data-linenos="118 "></span>            <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="119 "></span>                <span class="n">type_color_all</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">,</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">zeileis_28</span><span class="p">))</span>
<span class="linenos" data-linenos="120 "></span>
<span class="linenos" data-linenos="121 "></span>    <span class="c1">#set G_nodes_dict</span>
<span class="linenos" data-linenos="122 "></span>    <span class="n">nodes</span><span class="o">=</span><span class="p">[]</span>
<span class="linenos" data-linenos="123 "></span>    <span class="n">G_degree</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">degree</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()))</span>
<span class="linenos" data-linenos="124 "></span>
<span class="linenos" data-linenos="125 "></span>
<span class="linenos" data-linenos="126 "></span>    <span class="n">G_nodes_dict</span><span class="o">=</span><span class="p">{}</span>
<span class="linenos" data-linenos="127 "></span>    <span class="n">links</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="128 "></span>    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
<span class="linenos" data-linenos="129 "></span>        <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G_nodes_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="linenos" data-linenos="130 "></span>            <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="131 "></span>        <span class="k">if</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">G_nodes_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
<span class="linenos" data-linenos="132 "></span>            <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos="133 "></span>        <span class="n">links</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">"source"</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">"target"</span><span class="p">:</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]})</span>
<span class="linenos" data-linenos="134 "></span>        <span class="n">weight</span><span class="o">=</span><span class="n">G</span><span class="o">.</span><span class="n">get_edge_data</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])[</span><span class="s1">'weight'</span><span class="p">]</span>
<span class="linenos" data-linenos="135 "></span>        <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">+=</span><span class="n">weight</span>
<span class="linenos" data-linenos="136 "></span>        <span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">+=</span><span class="n">weight</span>
<span class="linenos" data-linenos="137 "></span>
<span class="linenos" data-linenos="138 "></span>    <span class="c1">#plot</span>
<span class="linenos" data-linenos="139 "></span>    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span> 
<span class="linenos" data-linenos="140 "></span>    <span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">pos_scale</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="p">(</span><span class="n">pos_size</span><span class="p">)</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">order</span><span class="p">()))</span>
<span class="linenos" data-linenos="141 "></span>    <span class="n">p</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<span class="linenos" data-linenos="142 "></span>
<span class="linenos" data-linenos="143 "></span>    <span class="n">nodesize</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">G_nodes_dict</span><span class="p">[</span><span class="n">u</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()])</span><span class="o">/</span><span class="n">nodesize_scale</span>
<span class="linenos" data-linenos="144 "></span>    <span class="n">nodecolos</span><span class="o">=</span><span class="p">[</span><span class="n">type_color_all</span><span class="p">[</span><span class="n">u</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()]</span>
<span class="linenos" data-linenos="145 "></span>    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_nodes</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">nodelist</span><span class="o">=</span><span class="n">p</span><span class="p">,</span><span class="n">node_size</span><span class="o">=</span><span class="n">nodesize</span><span class="p">,</span><span class="n">node_color</span><span class="o">=</span><span class="n">nodecolos</span><span class="p">)</span>
<span class="linenos" data-linenos="146 "></span>
<span class="linenos" data-linenos="147 "></span>    <span class="n">edgewidth</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">G</span><span class="o">.</span><span class="n">get_edge_data</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">)[</span><span class="s1">'weight'</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">()])</span><span class="o">/</span><span class="n">edgeswidth_scale</span>
<span class="linenos" data-linenos="148 "></span>    <span class="n">nx</span><span class="o">.</span><span class="n">draw_networkx_edges</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="n">edgewidth</span><span class="p">)</span>
<span class="linenos" data-linenos="149 "></span>
<span class="linenos" data-linenos="150 "></span>
<span class="linenos" data-linenos="151 "></span>    <span class="c1">#label_options = {"ec": "white", "fc": "white", "alpha": 0.6}</span>
<span class="linenos" data-linenos="152 "></span>    <span class="c1">#nx.draw_networkx_labels(G, pos, font_size=10,) #bbox=label_options)</span>
<span class="linenos" data-linenos="153 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="154 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">"off"</span><span class="p">)</span>
<span class="linenos" data-linenos="155 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="linenos" data-linenos="156 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="mf">1.5</span><span class="p">)</span>
<span class="linenos" data-linenos="157 "></span>
<span class="linenos" data-linenos="158 "></span>    <span class="n">labels</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
<span class="linenos" data-linenos="159 "></span>    <span class="c1">#用label和color列表生成mpatches.Patch对象，它将作为句柄来生成legend</span>
<span class="linenos" data-linenos="160 "></span>    <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="n">type_color_all</span><span class="p">[</span><span class="n">u</span><span class="p">]</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">]</span>
<span class="linenos" data-linenos="161 "></span>    <span class="n">patches</span> <span class="o">=</span> <span class="p">[</span><span class="n">mpatches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">type_color_all</span><span class="p">[</span><span class="n">u</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">u</span><span class="p">)</span> <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">labels</span> <span class="p">]</span> 
<span class="linenos" data-linenos="162 "></span>
<span class="linenos" data-linenos="163 "></span>    <span class="c1">#plt.xlim(-0.05, 1.05)</span>
<span class="linenos" data-linenos="164 "></span>    <span class="c1">#plt.ylim(-0.05, 1.05)</span>
<span class="linenos" data-linenos="165 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">"off"</span><span class="p">)</span>
<span class="linenos" data-linenos="166 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
<span class="linenos" data-linenos="167 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">patches</span><span class="p">,</span>
<span class="linenos" data-linenos="168 "></span>               <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="n">legend_bbox</span><span class="p">,</span> 
<span class="linenos" data-linenos="169 "></span>               <span class="n">ncol</span><span class="o">=</span><span class="n">legend_ncol</span><span class="p">,</span>
<span class="linenos" data-linenos="170 "></span>               <span class="n">fontsize</span><span class="o">=</span><span class="n">legend_fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="171 "></span>    <span class="k">if</span> <span class="n">return_graph</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="172 "></span>        <span class="k">return</span> <span class="n">G</span>
<span class="linenos" data-linenos="173 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="174 "></span>        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.cpdb_plot_interaction">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">cpdb_plot_interaction</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type1</span><span class="p">,</span> <span class="n">cell_type2</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">pvals</span><span class="p">,</span> <span class="n">celltype_key</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_significant_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">max_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">highlight_size</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">standard_scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap_name</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">ytickslabel_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">xtickslabel_fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">title_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.cpdb_plot_interaction" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Plot a CellPhoneDB interaction.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>AnnData object containing the data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_type1</code></td>
<td>
<code>str</code>
</td>
<td><p>Name of cell type 1.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_type2</code></td>
<td>
<code>str</code>
</td>
<td><p>Name of cell type 2.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>means</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>DataFrame containing the means for each interaction.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pvals</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>DataFrame containing the p-values for each interaction.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>celltype_key</code></td>
<td>
<code>str</code>
</td>
<td><p>Key in adata.obs that contains the cell type information.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>genes</code></td>
<td>
</td>
<td><p>List of genes to include in the plot. If None, all genes are included.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>keep_significant_only</code></td>
<td>
<code>bool</code>
</td>
<td><p>Whether to keep only significant interactions.</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>Figure size.</p></td>
<td>
<code>(4, 8)</code>
</td>
</tr>
<tr>
<td><code>title</code></td>
<td>
<code>str</code>
</td>
<td><p>Title of the plot.</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>max_size</code></td>
<td>
<code>int</code>
</td>
<td><p>Maximum size of the dots.</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>highlight_size</code></td>
<td>
<code>float</code>
</td>
<td><p>Size of the dots for the highlighted interaction.</p></td>
<td>
<code>0.75</code>
</td>
</tr>
<tr>
<td><code>standard_scale</code></td>
<td>
<code>bool</code>
</td>
<td><p>Whether to standard scale the data.</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>cmap_name</code></td>
<td>
<code>str</code>
</td>
<td><p>Name of the colormap to use.</p></td>
<td>
<code>'viridis'</code>
</td>
</tr>
<tr>
<td><code>ytickslabel_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>Fontsize of the yticks labels.</p></td>
<td>
<code>8</code>
</td>
</tr>
<tr>
<td><code>xtickslabel_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>Fontsize of the xticks labels.</p></td>
<td>
<code>8</code>
</td>
</tr>
<tr>
<td><code>title_fontsize</code></td>
<td>
<code>int</code>
</td>
<td><p>Fontsize of the title.</p></td>
<td>
<code>10</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ax</code></td> <td>
<code>matplotlib.<span title="matplotlib.axes">axes</span>.<span title="matplotlib.axes._axes">_axes</span>.<span title="matplotlib.axes._axes.Axes">Axes</span></code>
</td>
<td><p>Axes object containing the plot.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_cpdb.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="177 "></span><span class="k">def</span> <span class="nf">cpdb_plot_interaction</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">cell_type1</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">cell_type2</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
<span class="linenos" data-linenos="178 "></span>                          <span class="n">means</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">pvals</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
<span class="linenos" data-linenos="179 "></span>                          <span class="n">celltype_key</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="180 "></span>                         <span class="n">keep_significant_only</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span><span class="n">title</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span>
<span class="linenos" data-linenos="181 "></span>                         <span class="n">max_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">highlight_size</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span><span class="n">standard_scale</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="linenos" data-linenos="182 "></span>                         <span class="n">cmap_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span>
<span class="linenos" data-linenos="183 "></span>                         <span class="n">ytickslabel_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span><span class="n">xtickslabel_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span><span class="n">title_fontsize</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">_axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">:</span>
<span class="linenos" data-linenos="184 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="185 "></span><span class="sd">    Plot a CellPhoneDB interaction.</span>
<span class="linenos" data-linenos="186 "></span>
<span class="linenos" data-linenos="187 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="188 "></span><span class="sd">        adata: AnnData object containing the data.</span>
<span class="linenos" data-linenos="189 "></span><span class="sd">        cell_type1: Name of cell type 1.</span>
<span class="linenos" data-linenos="190 "></span><span class="sd">        cell_type2: Name of cell type 2.</span>
<span class="linenos" data-linenos="191 "></span><span class="sd">        means: DataFrame containing the means for each interaction.</span>
<span class="linenos" data-linenos="192 "></span><span class="sd">        pvals: DataFrame containing the p-values for each interaction.</span>
<span class="linenos" data-linenos="193 "></span><span class="sd">        celltype_key: Key in adata.obs that contains the cell type information.</span>
<span class="linenos" data-linenos="194 "></span><span class="sd">        genes: List of genes to include in the plot. If None, all genes are included.</span>
<span class="linenos" data-linenos="195 "></span><span class="sd">        keep_significant_only: Whether to keep only significant interactions.</span>
<span class="linenos" data-linenos="196 "></span><span class="sd">        figsize: Figure size.</span>
<span class="linenos" data-linenos="197 "></span><span class="sd">        title: Title of the plot.</span>
<span class="linenos" data-linenos="198 "></span><span class="sd">        max_size: Maximum size of the dots.</span>
<span class="linenos" data-linenos="199 "></span><span class="sd">        highlight_size: Size of the dots for the highlighted interaction.</span>
<span class="linenos" data-linenos="200 "></span><span class="sd">        standard_scale: Whether to standard scale the data.</span>
<span class="linenos" data-linenos="201 "></span><span class="sd">        cmap_name: Name of the colormap to use.</span>
<span class="linenos" data-linenos="202 "></span><span class="sd">        ytickslabel_fontsize: Fontsize of the yticks labels.</span>
<span class="linenos" data-linenos="203 "></span><span class="sd">        xtickslabel_fontsize: Fontsize of the xticks labels.</span>
<span class="linenos" data-linenos="204 "></span><span class="sd">        title_fontsize: Fontsize of the title.</span>
<span class="linenos" data-linenos="205 "></span>
<span class="linenos" data-linenos="206 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="207 "></span><span class="sd">        ax: Axes object containing the plot.</span>
<span class="linenos" data-linenos="208 "></span>
<span class="linenos" data-linenos="209 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="210 "></span>    <span class="n">check_kpy</span><span class="p">()</span>
<span class="linenos" data-linenos="211 "></span>    <span class="k">global</span> <span class="n">kpy_install</span>
<span class="linenos" data-linenos="212 "></span>    <span class="k">if</span> <span class="n">kpy_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="213 "></span>        <span class="n">global_imports</span><span class="p">(</span><span class="s2">"ktplotspy"</span><span class="p">,</span><span class="s2">"kpy"</span><span class="p">)</span>
<span class="linenos" data-linenos="214 "></span>
<span class="linenos" data-linenos="215 "></span>    <span class="n">fig</span><span class="o">=</span><span class="n">kpy</span><span class="o">.</span><span class="n">plot_cpdb</span><span class="p">(</span>
<span class="linenos" data-linenos="216 "></span>        <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="217 "></span>        <span class="n">cell_type1</span> <span class="o">=</span> <span class="n">cell_type1</span><span class="p">,</span>
<span class="linenos" data-linenos="218 "></span>        <span class="n">cell_type2</span> <span class="o">=</span> <span class="n">cell_type2</span><span class="p">,</span> 
<span class="linenos" data-linenos="219 "></span>        <span class="n">means</span> <span class="o">=</span> <span class="n">means</span><span class="p">,</span>
<span class="linenos" data-linenos="220 "></span>        <span class="n">pvals</span> <span class="o">=</span> <span class="n">pvals</span><span class="p">,</span>
<span class="linenos" data-linenos="221 "></span>        <span class="n">celltype_key</span> <span class="o">=</span> <span class="n">celltype_key</span><span class="p">,</span>
<span class="linenos" data-linenos="222 "></span>        <span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span><span class="p">,</span>
<span class="linenos" data-linenos="223 "></span>        <span class="n">keep_significant_only</span><span class="o">=</span><span class="n">keep_significant_only</span><span class="p">,</span>
<span class="linenos" data-linenos="224 "></span>        <span class="n">figsize</span> <span class="o">=</span> <span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="225 "></span>        <span class="n">title</span> <span class="o">=</span> <span class="s2">""</span><span class="p">,</span>
<span class="linenos" data-linenos="226 "></span>        <span class="n">max_size</span> <span class="o">=</span> <span class="n">max_size</span><span class="p">,</span>
<span class="linenos" data-linenos="227 "></span>        <span class="n">highlight_size</span> <span class="o">=</span> <span class="n">highlight_size</span><span class="p">,</span>
<span class="linenos" data-linenos="228 "></span>        <span class="n">standard_scale</span> <span class="o">=</span> <span class="n">standard_scale</span><span class="p">,</span>
<span class="linenos" data-linenos="229 "></span>        <span class="n">cmap_name</span><span class="o">=</span><span class="n">cmap_name</span>
<span class="linenos" data-linenos="230 "></span>    <span class="p">)</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
<span class="linenos" data-linenos="231 "></span>
<span class="linenos" data-linenos="232 "></span>    <span class="c1">#ytickslabels</span>
<span class="linenos" data-linenos="233 "></span>    <span class="n">labels</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">get_ticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos="234 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">ytickslabel_fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="235 "></span>
<span class="linenos" data-linenos="236 "></span>    <span class="c1">#xtickslabels</span>
<span class="linenos" data-linenos="237 "></span>    <span class="n">labels</span><span class="o">=</span><span class="n">fig</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">get_ticklabels</span><span class="p">()</span>
<span class="linenos" data-linenos="238 "></span>    <span class="n">plt</span><span class="o">.</span><span class="n">setp</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="n">xtickslabel_fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="239 "></span>
<span class="linenos" data-linenos="240 "></span>    <span class="n">fig</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="n">title_fontsize</span><span class="p">)</span>
<span class="linenos" data-linenos="241 "></span>
<span class="linenos" data-linenos="242 "></span>    <span class="k">return</span> <span class="n">fig</span><span class="o">.</span><span class="n">get_axes</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.cpdb_interaction_filtered">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">cpdb_interaction_filtered</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type1</span><span class="p">,</span> <span class="n">cell_type2</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">pvals</span><span class="p">,</span> <span class="n">celltype_key</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keep_significant_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="n">max_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">highlight_size</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">standard_scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap_name</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.cpdb_interaction_filtered" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Returns a list of unique interaction groups from a CellPhoneDB analysis filtered by significance.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
<code>anndata.<span title="anndata.AnnData">AnnData</span></code>
</td>
<td><p>Annotated data matrix containing normalized gene expression data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_type1</code></td>
<td>
<code>str</code>
</td>
<td><p>The cell type name for the first interacting partner.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_type2</code></td>
<td>
<code>str</code>
</td>
<td><p>The cell type name for the second interacting partner.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>means</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>Dataframe containing the means of gene expression for each interaction.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>pvals</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>Dataframe containing the p-values for each interaction.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>celltype_key</code></td>
<td>
<code>str</code>
</td>
<td><p>The column name in the <code>adata.obs</code> dataframe containing cell type annotations.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>genes</code></td>
<td>
</td>
<td><p>List of gene names to filter on. If None, no filtering is applied.</p></td>
<td>
<code>None</code>
</td>
</tr>
<tr>
<td><code>keep_significant_only</code></td>
<td>
<code>bool</code>
</td>
<td><p>Whether to filter on significant interactions only (i.e. interactions with FDR &lt; 0.05).</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
<code>tuple</code>
</td>
<td><p>Figure size. Width and height of the figure in inches. </p></td>
<td>
<code>(0, 0)</code>
</td>
</tr>
<tr>
<td><code>max_size</code></td>
<td>
<code>int</code>
</td>
<td><p>Maximum size of the markers in the plot.</p></td>
<td>
<code>1</code>
</td>
</tr>
<tr>
<td><code>highlight_size</code></td>
<td>
<code>float</code>
</td>
<td><p>Size of the markers for the highlighted interaction.</p></td>
<td>
<code>0.75</code>
</td>
</tr>
<tr>
<td><code>standard_scale</code></td>
<td>
<code>bool</code>
</td>
<td><p>Whether to standard scale the data.</p></td>
<td>
<code>True</code>
</td>
</tr>
<tr>
<td><code>cmap_name</code></td>
<td>
<code>str</code>
</td>
<td><p>Name of the colormap to use.</p></td>
<td>
<code>'viridis'</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>genes</code></td> <td>
<code>list</code>
</td>
<td><p>List of unique interaction groups from a CellPhoneDB analysis filtered by significance.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_cpdb.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="265 "></span><span class="k">def</span> <span class="nf">cpdb_interaction_filtered</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span><span class="n">cell_type1</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">cell_type2</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
<span class="linenos" data-linenos="266 "></span>                              <span class="n">means</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">pvals</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">celltype_key</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">genes</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="linenos" data-linenos="267 "></span>                         <span class="n">keep_significant_only</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">figsize</span><span class="p">:</span><span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span>
<span class="linenos" data-linenos="268 "></span>                         <span class="n">max_size</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">highlight_size</span><span class="p">:</span><span class="nb">float</span> <span class="o">=</span> <span class="mf">0.75</span><span class="p">,</span><span class="n">standard_scale</span><span class="p">:</span><span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span><span class="n">cmap_name</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,)</span><span class="o">-&gt;</span><span class="nb">list</span><span class="p">:</span>
<span class="linenos" data-linenos="269 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="270 "></span><span class="sd">    Returns a list of unique interaction groups from a CellPhoneDB analysis filtered by significance.</span>
<span class="linenos" data-linenos="271 "></span>
<span class="linenos" data-linenos="272 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="273 "></span><span class="sd">        adata: Annotated data matrix containing normalized gene expression data.</span>
<span class="linenos" data-linenos="274 "></span><span class="sd">        cell_type1: The cell type name for the first interacting partner.</span>
<span class="linenos" data-linenos="275 "></span><span class="sd">        cell_type2: The cell type name for the second interacting partner.</span>
<span class="linenos" data-linenos="276 "></span><span class="sd">        means: Dataframe containing the means of gene expression for each interaction.</span>
<span class="linenos" data-linenos="277 "></span><span class="sd">        pvals: Dataframe containing the p-values for each interaction.</span>
<span class="linenos" data-linenos="278 "></span><span class="sd">        celltype_key: The column name in the `adata.obs` dataframe containing cell type annotations.</span>
<span class="linenos" data-linenos="279 "></span><span class="sd">        genes: List of gene names to filter on. If None, no filtering is applied.</span>
<span class="linenos" data-linenos="280 "></span><span class="sd">        keep_significant_only: Whether to filter on significant interactions only (i.e. interactions with FDR &lt; 0.05).</span>
<span class="linenos" data-linenos="281 "></span><span class="sd">        figsize: Figure size. Width and height of the figure in inches. </span>
<span class="linenos" data-linenos="282 "></span><span class="sd">        max_size: Maximum size of the markers in the plot.</span>
<span class="linenos" data-linenos="283 "></span><span class="sd">        highlight_size: Size of the markers for the highlighted interaction.</span>
<span class="linenos" data-linenos="284 "></span><span class="sd">        standard_scale: Whether to standard scale the data.</span>
<span class="linenos" data-linenos="285 "></span><span class="sd">        cmap_name: Name of the colormap to use.</span>
<span class="linenos" data-linenos="286 "></span>
<span class="linenos" data-linenos="287 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="288 "></span><span class="sd">        genes: List of unique interaction groups from a CellPhoneDB analysis filtered by significance.</span>
<span class="linenos" data-linenos="289 "></span>
<span class="linenos" data-linenos="290 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="291 "></span>    <span class="n">check_kpy</span><span class="p">()</span>
<span class="linenos" data-linenos="292 "></span>    <span class="k">global</span> <span class="n">kpy_install</span>
<span class="linenos" data-linenos="293 "></span>    <span class="k">if</span> <span class="n">kpy_install</span><span class="o">==</span><span class="kc">True</span><span class="p">:</span>
<span class="linenos" data-linenos="294 "></span>        <span class="n">global_imports</span><span class="p">(</span><span class="s2">"ktplotspy"</span><span class="p">,</span><span class="s2">"kpy"</span><span class="p">)</span>
<span class="linenos" data-linenos="295 "></span>
<span class="linenos" data-linenos="296 "></span>    <span class="n">res</span><span class="o">=</span><span class="n">kpy</span><span class="o">.</span><span class="n">plot_cpdb</span><span class="p">(</span>
<span class="linenos" data-linenos="297 "></span>        <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">,</span>
<span class="linenos" data-linenos="298 "></span>        <span class="n">cell_type1</span> <span class="o">=</span> <span class="n">cell_type1</span><span class="p">,</span>
<span class="linenos" data-linenos="299 "></span>        <span class="n">cell_type2</span> <span class="o">=</span> <span class="n">cell_type2</span><span class="p">,</span> 
<span class="linenos" data-linenos="300 "></span>        <span class="n">means</span> <span class="o">=</span> <span class="n">means</span><span class="p">,</span>
<span class="linenos" data-linenos="301 "></span>        <span class="n">pvals</span> <span class="o">=</span> <span class="n">pvals</span><span class="p">,</span>
<span class="linenos" data-linenos="302 "></span>        <span class="n">celltype_key</span> <span class="o">=</span> <span class="n">celltype_key</span><span class="p">,</span>
<span class="linenos" data-linenos="303 "></span>        <span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span><span class="p">,</span>
<span class="linenos" data-linenos="304 "></span>        <span class="n">keep_significant_only</span><span class="o">=</span><span class="n">keep_significant_only</span><span class="p">,</span>
<span class="linenos" data-linenos="305 "></span>        <span class="n">figsize</span> <span class="o">=</span> <span class="n">figsize</span><span class="p">,</span>
<span class="linenos" data-linenos="306 "></span>        <span class="n">title</span> <span class="o">=</span> <span class="s2">""</span><span class="p">,</span>
<span class="linenos" data-linenos="307 "></span>        <span class="n">max_size</span> <span class="o">=</span> <span class="n">max_size</span><span class="p">,</span>
<span class="linenos" data-linenos="308 "></span>        <span class="n">highlight_size</span> <span class="o">=</span> <span class="n">highlight_size</span><span class="p">,</span>
<span class="linenos" data-linenos="309 "></span>        <span class="n">standard_scale</span> <span class="o">=</span> <span class="n">standard_scale</span><span class="p">,</span>
<span class="linenos" data-linenos="310 "></span>        <span class="n">cmap_name</span><span class="o">=</span><span class="n">cmap_name</span><span class="p">,</span>
<span class="linenos" data-linenos="311 "></span>        <span class="n">return_table</span><span class="o">=</span><span class="kc">True</span>
<span class="linenos" data-linenos="312 "></span>    <span class="p">)</span>
<span class="linenos" data-linenos="313 "></span>
<span class="linenos" data-linenos="314 "></span>    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">'interaction_group'</span><span class="p">]))</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.cpdb_submeans_exacted">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">cpdb_submeans_exacted</span><span class="p">(</span><span class="n">means</span><span class="p">,</span> <span class="n">cell_names</span><span class="p">,</span> <span class="n">cell_type</span><span class="o">=</span><span class="s1">'ligand'</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.cpdb_submeans_exacted" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Returns a subset of the means DataFrame for a given cell type and cell name.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>means</code></td>
<td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>DataFrame containing the means for each interaction.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_names</code></td>
<td>
<code>str</code>
</td>
<td><p>Name of the cell type.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cell_type</code></td>
<td>
<code>str</code>
</td>
<td><p>Whether the cell type is a ligand or a receptor.</p></td>
<td>
<code>'ligand'</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>means</code></td> <td>
<code><span title="pandas">pd</span>.<span title="pandas.DataFrame">DataFrame</span></code>
</td>
<td><p>Subset of the means DataFrame for a given cell type and cell name.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_cpdb.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="244 "></span><span class="k">def</span> <span class="nf">cpdb_submeans_exacted</span><span class="p">(</span><span class="n">means</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span><span class="n">cell_names</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="n">cell_type</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">'ligand'</span><span class="p">)</span><span class="o">-&gt;</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="linenos" data-linenos="245 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="246 "></span><span class="sd">    Returns a subset of the means DataFrame for a given cell type and cell name.</span>
<span class="linenos" data-linenos="247 "></span>
<span class="linenos" data-linenos="248 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="249 "></span><span class="sd">        means: DataFrame containing the means for each interaction.</span>
<span class="linenos" data-linenos="250 "></span><span class="sd">        cell_names: Name of the cell type.</span>
<span class="linenos" data-linenos="251 "></span><span class="sd">        cell_type: Whether the cell type is a ligand or a receptor.</span>
<span class="linenos" data-linenos="252 "></span>
<span class="linenos" data-linenos="253 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="254 "></span><span class="sd">        means: Subset of the means DataFrame for a given cell type and cell name.</span>
<span class="linenos" data-linenos="255 "></span>
<span class="linenos" data-linenos="256 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="257 "></span>    <span class="k">if</span> <span class="n">cell_type</span><span class="o">==</span><span class="s1">'ligand'</span><span class="p">:</span>
<span class="linenos" data-linenos="258 "></span>        <span class="n">means_columns</span><span class="o">=</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="o">+</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">'</span><span class="si">{}</span><span class="s1">\|'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell_names</span><span class="p">))]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="259 "></span>    <span class="k">elif</span> <span class="n">cell_type</span><span class="o">==</span><span class="s1">'receptor'</span><span class="p">:</span>
<span class="linenos" data-linenos="260 "></span>        <span class="n">means_columns</span><span class="o">=</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="o">+</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">means</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">'\|</span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cell_names</span><span class="p">))]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="261 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="262 "></span>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">'cell_type must be ligand or receptor'</span><span class="p">)</span>
<span class="linenos" data-linenos="263 "></span>    <span class="k">return</span> <span class="n">means</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">means_columns</span><span class="p">]</span>
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2024, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>