
<!DOCTYPE html>

<html class="no-js" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<meta content="Zehua Zeng" name="author"/>
<link href="../../../img/favicon.ico" rel="icon"/>
<meta content="mkdocs-1.4.2, mkdocs-material-9.1.2" name="generator"/>
<title>Api scdrug - omicverse Readthedocs</title>
<link href="../../../assets/stylesheets/main.7bf56d0a.min.css" rel="stylesheet"/>
<link href="../../../assets/stylesheets/palette.a0c5b2b5.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<link href="../../../css/custom.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            
                .gscrollbar-fixer { padding-right: 15px; }
                .gdesc-inner { font-size: 0.75rem; }
                body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color);}
                body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color);}
                body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color);}
                </style><script src="../../../assets/javascripts/glightbox.min.js"></script></head>
<body data-md-color-accent="" data-md-color-primary="" data-md-color-scheme="default" dir="ltr">
<script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#omicverse.single.Drug_Response">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<header class="md-header md-header--shadow" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="omicverse Readthedocs" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            omicverse Readthedocs
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Api scdrug
            
          </span>
</div>
</div>
</div>
<form class="md-header__option" data-md-component="palette">
<input aria-label="Switch to dark mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: light)" data-md-color-primary="" data-md-color-scheme="default" id="__palette_1" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_2" hidden="" title="Switch to dark mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"></path></svg>
</label>
<input aria-label="Switch to light mode" class="md-option" data-md-color-accent="" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-primary="" data-md-color-scheme="slate" id="__palette_2" name="__palette" type="radio"/>
<label class="md-header__button md-icon" for="__palette_1" hidden="" title="Switch to light mode">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"></path></svg>
</label>
</form>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="omicverse Readthedocs" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="omicverse Readthedocs">
<img alt="logo" src="../../../img/ico.png"/>
</a>
    omicverse Readthedocs
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/Starlitnightly/omicverse" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 448 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"></path></svg>
</div>
<div class="md-source__repository">
    GitHub
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
        Index
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Installation_guild/">
        Installation
      </a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Tutorial
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_3">
<span class="md-nav__icon md-icon"></span>
          Tutorial
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
          Bulk
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_1_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_1">
<span class="md-nav__icon md-icon"></span>
          Bulk
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_bulk_combat/">
        Batch correction in Bulk RNA-seq or microarray data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deg/">
        Different Expression Analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_deseq2/">
        Different Expression Analysis with DEseq2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_network/">
        Protein-Protein interaction (PPI) analysis by String-db
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_wgcna/">
        WGCNA (Weighted gene co-expression network analysis) analysis
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk/t_tcga/">
        TCGA database preprocess
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
          Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_2">
<span class="md-nav__icon md-icon"></span>
          Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_1" id="__nav_3_2_1_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_1_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_1">
<span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_cpu/">
        Preprocessing the data of scRNA-seq with omicverse[CPU-GPU-mixed]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_preprocess_gpu/">
        Preprocessing the data of scRNA-seq with omicverse[GPU][Bug]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cluster/">
        Clustering space
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_single_batch/">
        Data integration and batch correction
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cnmf/">
        Consensus Non-negative Matrix factorization (cNMF)
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_lazy/">
        Lazy analysis of scRNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_2" id="__nav_3_2_2_label" tabindex="0">
          Annotation
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_2_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_2">
<span class="md-nav__icon md-icon"></span>
          Annotation
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_gptanno/">
        Automatic cell type annotation with GPT/Other
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellmatch/">
        Mapping Cell Names to the Cell Ontology/Taxonomy
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellanno/">
        Celltype auto annotation with SCSA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metatime/">
        Celltype auto annotation with MetaTiME
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_tosica/">
        Celltype annotation migration(mapping) with TOSICA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scmulan/">
        Celltype auto annotation with scMulan
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellvote/">
        Consensus annotation with CellVote
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
          Trajectory
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_3_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_3">
<span class="md-nav__icon md-icon"></span>
          Trajectory
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cytotrace/">
        Prediction of absolute developmental potential using CytoTrace2
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_traj/">
        Basic Trajectory Inference
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_stavia/">
        Trajectory Inference with StaVIA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate_gene/">
        Timing-associated genes analysis with TimeFateKernel
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellfate/">
        Identify the driver regulators of cell fate decisions
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_4" id="__nav_3_2_4_label" tabindex="0">
          Cell Structure
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_4_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_4">
<span class="md-nav__icon md-icon"></span>
          Cell Structure
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_metacells/">
        Inference of MetaCell from Single-Cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_deg_single/">
        Differential expression and celltype analysis [All Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdeg/">
        Differential expression analysis [Meta Cell]
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_aucell/">
        Pathway analysis with AUCell
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_cellphonedb/">
        Cell interaction with CellPhoneDB
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_scdrug/">
        Drug response predict with scDrug
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_simba/">
        Data integration and batch correction with SIMBA
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_2_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_2_5" id="__nav_3_2_5_label" tabindex="0">
          Multi omics
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_2_5_label" class="md-nav" data-md-level="3">
<label class="md-nav__title" for="__nav_3_2_5">
<span class="md-nav__icon md-icon"></span>
          Multi omics
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa/">
        Multi omics analysis by MOFA
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_mofa_glue/">
        Multi omics analysis by MOFA and GLUE
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-single/t_anno_trans/">
        Celltype annotation transfer in multi-omics
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_3" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
          Bulk2Single
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_3_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_3">
<span class="md-nav__icon md-icon"></span>
          Bulk2Single
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulktrajblend/">
        Bulk RNA-seq generate 'interrupted' cells to interpolate scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_bulk2single/">
        Bulk RNA-seq to Single RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-bulk2single/t_single2spatial/">
        Single RNA-seq to Spatial RNA-seq
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_4" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
          Space
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_4_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_4">
<span class="md-nav__icon md-icon"></span>
          Space
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_crop_rotate/">
        Crop and Rotation of spatial transcriptomic data
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_cluster_space/">
        Spatial clustering and denoising expressions
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_spaceflow/">
        Identifying Pseudo-Spatial Map
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_mapping/">
        Mapping single-cell profile onto spatial profile
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_starfysh/">
        Deconvolution spatial transcriptomic without scRNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_stt/">
        Spatial transition tensor of single cells
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_commot_flowsig/">
        Spatial Communication
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_staligner/">
        Spatial integration and clustering
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_gaston/">
        Spatial IsoDepth Calculation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-space/t_slat/">
        Single cell spatial alignment tools
      </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_3_5" type="checkbox"/>
<label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
          Plotting
          <span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_3_5_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_3_5">
<span class="md-nav__icon md-icon"></span>
          Plotting
        </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_single/">
        Visualization of single cell RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_bulk/">
        Visualization of Bulk RNA-seq
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Tutorials-plotting/t_visualize_colorsystem/">
        Color system
      </a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
        API documentation
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Release_notes/">
        Release notes
      </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../Developer_guild/">
        Developer guild
      </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.Drug_Response">
    Drug_Response
  </a>
<nav aria-label="Drug_Response" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._scdrug.Drug_Response.__init__">
    __init__()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._scdrug.Drug_Response.load_model">
    load_model()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._scdrug.Drug_Response.sensitivity_prediction">
    sensitivity_prediction()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._scdrug.Drug_Response.cell_death_proportion">
    cell_death_proportion()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single._scdrug.Drug_Response.draw_plot">
    draw_plot()
  </a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.autoResolution">
    autoResolution()
  </a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#omicverse.single.writeGEP">
    writeGEP()
  </a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<h1>Api scdrug</h1>
<div class="doc doc-object doc-class">
<h2 class="doc doc-heading" id="omicverse.single.Drug_Response">
<code>omicverse.single.Drug_Response</code>
<a class="headerlink" href="#omicverse.single.Drug_Response" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Drug_Response class for drug response prediction.
The raw code could be found at https://github.com/ailabstw/scDrug</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="127 "></span><span class="k">class</span> <span class="nc">Drug_Response</span><span class="p">:</span>
<span class="linenos" data-linenos="128 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="129 "></span><span class="sd">    Drug_Response class for drug response prediction.</span>
<span class="linenos" data-linenos="130 "></span><span class="sd">    The raw code could be found at https://github.com/ailabstw/scDrug</span>
<span class="linenos" data-linenos="131 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="132 "></span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="p">,</span><span class="n">scriptpath</span><span class="p">,</span><span class="n">modelpath</span><span class="p">,</span><span class="n">output</span><span class="o">=</span><span class="s1">'./'</span><span class="p">,</span><span class="n">model</span><span class="o">=</span><span class="s1">'GDSC'</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="s1">'All'</span><span class="p">,</span>
<span class="linenos" data-linenos="133 "></span>                 <span class="n">cell</span><span class="o">=</span><span class="s1">'A549'</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">n_drugs</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="linenos" data-linenos="134 "></span>
<span class="linenos" data-linenos="135 "></span><span class="w">        </span><span class="sd">"""</span>
<span class="linenos" data-linenos="136 "></span><span class="sd">        Initializes the Drug_Response class.</span>
<span class="linenos" data-linenos="137 "></span>
<span class="linenos" data-linenos="138 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="139 "></span><span class="sd">            adata: Annotated data matrix with cells as rows and genes as columns.</span>
<span class="linenos" data-linenos="140 "></span><span class="sd">            scriptpath: Path to the directory containing the CaDRReS scripts for the analysis. You need to download the scripts according to `git clone https://github.com/CSB5/CaDRReS-Sc.git` and set the path to the directory.</span>
<span class="linenos" data-linenos="141 "></span><span class="sd">            modelpath: Path to the directory containing the pre-trained models.</span>
<span class="linenos" data-linenos="142 "></span><span class="sd">                You need to download the models according to `Pyomic.utils.download_GDSC_data()` and `Pyomic.utils.download_CaDRReS_model()` and set the path to the directory.</span>
<span class="linenos" data-linenos="143 "></span><span class="sd">            output: Path to the directory where the output files will be saved (default: './').</span>
<span class="linenos" data-linenos="144 "></span><span class="sd">            model: The name of the pre-trained model to be used for the analysis (default: 'GDSC').</span>
<span class="linenos" data-linenos="145 "></span><span class="sd">            clusters: The cluster labels to be used for the analysis. Default is all cells (default: 'All').</span>
<span class="linenos" data-linenos="146 "></span><span class="sd">            cell: The cell line to be analyzed (default: 'A549').</span>
<span class="linenos" data-linenos="147 "></span><span class="sd">            cpus: The number of CPUs to be used for the analysis (default: 4).</span>
<span class="linenos" data-linenos="148 "></span><span class="sd">            n_drugs: The number of top drugs to be selected based on the predicted sensitivity (default: 10).</span>
<span class="linenos" data-linenos="149 "></span>
<span class="linenos" data-linenos="150 "></span><span class="sd">        Returns:</span>
<span class="linenos" data-linenos="151 "></span><span class="sd">            None</span>
<span class="linenos" data-linenos="152 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="153 "></span>
<span class="linenos" data-linenos="154 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">model</span>
<span class="linenos" data-linenos="155 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span>
<span class="linenos" data-linenos="156 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="157 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="o">=</span><span class="n">output</span>
<span class="linenos" data-linenos="158 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">n_drugs</span><span class="o">=</span><span class="n">n_drugs</span>
<span class="linenos" data-linenos="159 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">=</span><span class="n">modelpath</span>
<span class="linenos" data-linenos="160 "></span>
<span class="linenos" data-linenos="161 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">=</span> <span class="n">scriptpath</span>
<span class="linenos" data-linenos="162 "></span>        <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">scriptpath</span><span class="p">))</span>
<span class="linenos" data-linenos="163 "></span>
<span class="linenos" data-linenos="164 "></span>        <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">load_model</span><span class="p">()</span>
<span class="linenos" data-linenos="167 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">drug_info</span><span class="p">()</span>
<span class="linenos" data-linenos="168 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">bulk_exp</span><span class="p">()</span>
<span class="linenos" data-linenos="169 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">sc_exp</span><span class="p">()</span>
<span class="linenos" data-linenos="170 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">kernel_feature_preparartion</span><span class="p">()</span>
<span class="linenos" data-linenos="171 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">sensitivity_prediction</span><span class="p">()</span>
<span class="linenos" data-linenos="172 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="173 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">+</span><span class="s1">'masked_drugs.csv'</span><span class="p">)[</span><span class="s1">'GDSC'</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'int64'</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'str'</span><span class="p">))</span>
<span class="linenos" data-linenos="174 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">cell_death_proportion</span><span class="p">()</span>
<span class="linenos" data-linenos="175 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="176 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">+</span><span class="s1">'masked_drugs.csv'</span><span class="p">)[</span><span class="s1">'PRISM'</span><span class="p">])</span>
<span class="linenos" data-linenos="177 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">output_result</span><span class="p">()</span>
<span class="linenos" data-linenos="178 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">figure_output</span><span class="p">()</span>
<span class="linenos" data-linenos="179 "></span>
<span class="linenos" data-linenos="180 "></span>    <span class="k">def</span> <span class="nf">load_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="181 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">        load the pre-trained model.</span>
<span class="linenos" data-linenos="183 "></span>
<span class="linenos" data-linenos="184 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="185 "></span>        <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="186 "></span>        <span class="c1">### IC50/AUC prediction</span>
<span class="linenos" data-linenos="187 "></span>        <span class="c1">## Read pre-trained model</span>
<span class="linenos" data-linenos="188 "></span>        <span class="c1">#model_dir = '/Users/fernandozeng/Desktop/analysis/scDrug/CaDRReS-Sc-model/'</span>
<span class="linenos" data-linenos="189 "></span>        <span class="n">model_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span>
<span class="linenos" data-linenos="190 "></span>        <span class="n">obj_function</span> <span class="o">=</span> <span class="n">widgets</span><span class="o">.</span><span class="n">Dropdown</span><span class="p">(</span><span class="n">options</span><span class="o">=</span><span class="p">[</span><span class="s1">'cadrres-wo-sample-bias'</span><span class="p">,</span> <span class="s1">'cadrres-wo-sample-bias-weight'</span><span class="p">],</span> <span class="n">description</span><span class="o">=</span><span class="s1">'Objetice function'</span><span class="p">)</span>
<span class="linenos" data-linenos="191 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span> <span class="o">=</span> <span class="n">obj_function</span><span class="o">.</span><span class="n">value</span>
<span class="linenos" data-linenos="192 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="193 "></span>            <span class="n">model_file</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">+</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">_param_dict_all_genes.pickle'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="194 "></span>        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'PRISM'</span><span class="p">:</span>
<span class="linenos" data-linenos="195 "></span>            <span class="n">model_file</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">+</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">_param_dict_prism.pickle'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="196 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="197 "></span>            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">'Wrong model name.'</span><span class="p">)</span>
<span class="linenos" data-linenos="198 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">model_file</span><span class="p">)</span>
<span class="linenos" data-linenos="199 "></span>
<span class="linenos" data-linenos="200 "></span>    <span class="k">def</span> <span class="nf">drug_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="201 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="202 "></span><span class="sd">        read the drug information.</span>
<span class="linenos" data-linenos="203 "></span>
<span class="linenos" data-linenos="204 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="205 "></span>        <span class="c1">## Read drug information</span>
<span class="linenos" data-linenos="206 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="207 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/preprocessed_data/GDSC/drug_stat.csv'</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="208 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="linenos" data-linenos="209 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="210 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/preprocessed_data/PRISM/PRISM_drug_info.csv'</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="s1">'broad_id'</span><span class="p">)</span>
<span class="linenos" data-linenos="211 "></span>
<span class="linenos" data-linenos="212 "></span>    <span class="k">def</span> <span class="nf">bulk_exp</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="213 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="214 "></span><span class="sd">        extract the bulk gene expression data.</span>
<span class="linenos" data-linenos="215 "></span>
<span class="linenos" data-linenos="216 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="217 "></span>        <span class="c1">## Read test data</span>
<span class="linenos" data-linenos="218 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="219 "></span>            <span class="c1">#GDSC_exp exists in the data folder</span>
<span class="linenos" data-linenos="220 "></span>            <span class="n">files</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/data/GDSC'</span><span class="p">)</span>
<span class="linenos" data-linenos="221 "></span>            <span class="k">if</span> <span class="s1">'GDSC_exp.tsv'</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
<span class="linenos" data-linenos="222 "></span>                <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span> <span class="o">+</span> <span class="s1">'GDSC_exp.tsv.gz'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">'</span><span class="se">\t</span><span class="s1">'</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="223 "></span>                <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="224 "></span>            <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="225 "></span>                <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/data/GDSC/GDSC_exp.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">'</span><span class="se">\t</span><span class="s1">'</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="226 "></span>                <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="linenos" data-linenos="227 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="228 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/data/CCLE/CCLE_expression.csv'</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="229 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="o">=</span><span class="s1">' ('</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="linenos" data-linenos="230 "></span>
<span class="linenos" data-linenos="231 "></span>    <span class="k">def</span> <span class="nf">sc_exp</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="232 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="233 "></span><span class="sd">        Load cluster-specific gene expression profile</span>
<span class="linenos" data-linenos="234 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="235 "></span>        <span class="c1">## Load cluster-specific gene expression profile</span>
<span class="linenos" data-linenos="236 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span> <span class="o">==</span> <span class="s1">'All'</span><span class="p">:</span>
<span class="linenos" data-linenos="237 "></span>            <span class="n">clusters</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
<span class="linenos" data-linenos="238 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="239 "></span>            <span class="n">clusters</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">','</span><span class="p">)]</span>
<span class="linenos" data-linenos="240 "></span>
<span class="linenos" data-linenos="241 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">cluster_norm_exp_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">clusters</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="linenos" data-linenos="242 "></span>        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
<span class="linenos" data-linenos="243 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">cluster_norm_exp_df</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span> <span class="o">=</span>  <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span><span class="o">==</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> \
<span class="linenos" data-linenos="244 "></span>                                                 <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span><span class="o">==</span><span class="n">cluster</span><span class="p">])</span> <span class="k">else</span> <span class="mf">0.0</span>
<span class="linenos" data-linenos="245 "></span>
<span class="linenos" data-linenos="246 "></span>    <span class="k">def</span> <span class="nf">kernel_feature_preparartion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="247 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="248 "></span><span class="sd">        kernel feature preparation</span>
<span class="linenos" data-linenos="249 "></span>
<span class="linenos" data-linenos="250 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="251 "></span>        <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="252 "></span>        <span class="c1">## Read essential genes list</span>
<span class="linenos" data-linenos="253 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="254 "></span>            <span class="n">ess_gene_list</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="255 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="256 "></span>            <span class="n">ess_gene_list</span> <span class="o">=</span> <span class="n">utility</span><span class="o">.</span><span class="n">get_gene_list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">+</span> <span class="s1">'/preprocessed_data/PRISM/feature_genes.txt'</span><span class="p">)</span>
<span class="linenos" data-linenos="257 "></span>
<span class="linenos" data-linenos="258 "></span>        <span class="c1">## Calculate fold-change</span>
<span class="linenos" data-linenos="259 "></span>        <span class="n">cell_line_log2_mean_fc_exp_df</span><span class="p">,</span> <span class="n">cell_line_mean_exp_df</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">gexp</span><span class="o">.</span><span class="n">normalize_log2_mean_fc</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene_exp_df</span><span class="p">)</span>
<span class="linenos" data-linenos="260 "></span>
<span class="linenos" data-linenos="261 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">adata_exp_mean</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="linenos" data-linenos="262 "></span>        <span class="n">cluster_norm_exp_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cluster_norm_exp_df</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata_exp_mean</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="linenos" data-linenos="263 "></span>
<span class="linenos" data-linenos="264 "></span>        <span class="c1">## Calculate kernel feature</span>
<span class="linenos" data-linenos="265 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">test_kernel_df</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">gexp</span><span class="o">.</span><span class="n">calculate_kernel_feature</span><span class="p">(</span><span class="n">cluster_norm_exp_df</span><span class="p">,</span> <span class="n">cell_line_log2_mean_fc_exp_df</span><span class="p">,</span> <span class="n">ess_gene_list</span><span class="p">)</span>
<span class="linenos" data-linenos="266 "></span>
<span class="linenos" data-linenos="267 "></span>    <span class="k">def</span> <span class="nf">sensitivity_prediction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="268 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">        Predict drug sensitivity</span>
<span class="linenos" data-linenos="270 "></span>
<span class="linenos" data-linenos="271 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="272 "></span>        <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="273 "></span>        <span class="c1">## Drug response prediction</span>
<span class="linenos" data-linenos="274 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="275 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'...Predicting drug response for using CaDRReS(GDSC): </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">))</span>
<span class="linenos" data-linenos="276 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="p">,</span> <span class="n">P_test_df</span><span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_from_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_kernel_df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="277 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'...done!'</span><span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="279 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'...Predicting drug response for using CaDRReS(PRISM): </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">))</span>
<span class="linenos" data-linenos="280 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="p">,</span> <span class="n">P_test_df</span><span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_from_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_kernel_df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="281 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s1">'...done!'</span><span class="p">)</span>
<span class="linenos" data-linenos="282 "></span>
<span class="linenos" data-linenos="283 "></span>    <span class="k">def</span> <span class="nf">cell_death_proportion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="284 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="285 "></span><span class="sd">        Predict cell death proportion and cell death percentage at the ref_type dosage</span>
<span class="linenos" data-linenos="286 "></span>
<span class="linenos" data-linenos="287 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="288 "></span>        <span class="c1">### Drug kill prediction</span>
<span class="linenos" data-linenos="289 "></span>        <span class="n">ref_type</span> <span class="o">=</span> <span class="s1">'log2_median_ic50'</span>
<span class="linenos" data-linenos="290 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span><span class="p">]</span>
<span class="linenos" data-linenos="291 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">]</span>
<span class="linenos" data-linenos="292 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">]</span>
<span class="linenos" data-linenos="293 "></span>
<span class="linenos" data-linenos="294 "></span>        <span class="c1">## Predict cell death percentage at the ref_type dosage</span>
<span class="linenos" data-linenos="295 "></span>        <span class="n">pred_delta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">values</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="p">[</span><span class="n">ref_type</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="linenos" data-linenos="296 "></span>        <span class="n">pred_cv_df</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="n">pred_delta_df</span><span class="p">)))</span>
<span class="linenos" data-linenos="297 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">pred_kill_df</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">-</span> <span class="n">pred_cv_df</span>
<span class="linenos" data-linenos="298 "></span>
<span class="linenos" data-linenos="299 "></span>    <span class="k">def</span> <span class="nf">output_result</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="300 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="301 "></span>            <span class="n">drug_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'Drug ID'</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">,</span> 
<span class="linenos" data-linenos="302 "></span>                                    <span class="s1">'Drug Name'</span><span class="p">:</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_id</span><span class="p">][</span><span class="s1">'Drug Name'</span><span class="p">]</span> <span class="k">for</span> <span class="n">drug_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">]})</span>
<span class="linenos" data-linenos="303 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">T</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">/</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="linenos" data-linenos="304 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="305 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_frame</span><span class="p">(</span><span class="n">drug_df</span><span class="p">)</span>
<span class="linenos" data-linenos="306 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'IC50_prediction.csv'</span><span class="p">))</span>
<span class="linenos" data-linenos="307 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_kill_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_frame</span><span class="p">(</span><span class="n">drug_df</span><span class="p">)</span>
<span class="linenos" data-linenos="308 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_kill_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'drug_kill_prediction.csv'</span><span class="p">))</span>
<span class="linenos" data-linenos="309 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="310 "></span>            <span class="n">drug_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="linenos" data-linenos="311 "></span>            <span class="n">drug_list</span>  <span class="o">=</span> <span class="p">[</span><span class="n">d</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">drug_list</span> <span class="k">if</span> <span class="n">d</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span><span class="p">]</span>
<span class="linenos" data-linenos="312 "></span>            <span class="n">drug_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'Drug ID'</span><span class="p">:</span><span class="n">drug_list</span><span class="p">,</span>
<span class="linenos" data-linenos="313 "></span>                                    <span class="s1">'Drug Name'</span><span class="p">:[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">d</span><span class="p">,</span> <span class="s1">'name'</span><span class="p">]</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">drug_list</span><span class="p">]})</span>
<span class="linenos" data-linenos="314 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">drug_list</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="315 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">min</span><span class="p">())</span><span class="o">/</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
<span class="linenos" data-linenos="316 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="317 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_frame</span><span class="p">(</span><span class="n">drug_df</span><span class="p">)</span>
<span class="linenos" data-linenos="318 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'PRISM_prediction.csv'</span><span class="p">))</span>
<span class="linenos" data-linenos="319 "></span>
<span class="linenos" data-linenos="320 "></span>    <span class="k">def</span> <span class="nf">draw_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">()):</span>
<span class="linenos" data-linenos="321 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="322 "></span><span class="sd">        plot heatmap of drug response prediction</span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span><span class="sd">        Arguments:</span>
<span class="linenos" data-linenos="325 "></span><span class="sd">            df : `pandas.DataFrame` drug response prediction dataframe</span>
<span class="linenos" data-linenos="326 "></span><span class="sd">            n_drug : `int` number of drugs to be plotted</span>
<span class="linenos" data-linenos="327 "></span><span class="sd">            name : `str` name of the plot</span>
<span class="linenos" data-linenos="328 "></span><span class="sd">            figsize : `tuple` size of the plot</span>
<span class="linenos" data-linenos="329 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="330 "></span>        <span class="k">def</span> <span class="nf">select_drug</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="p">):</span>
<span class="linenos" data-linenos="331 "></span>            <span class="n">selected_drugs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="332 "></span>            <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">'Drug Name'</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span>
<span class="linenos" data-linenos="333 "></span>            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">columns</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="nb">int</span><span class="p">):</span>
<span class="linenos" data-linenos="334 "></span>                <span class="k">for</span> <span class="n">drug_name</span> <span class="ow">in</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">[:</span><span class="n">n_drug</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
<span class="linenos" data-linenos="335 "></span>                    <span class="k">if</span> <span class="n">drug_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">selected_drugs</span><span class="p">:</span>
<span class="linenos" data-linenos="336 "></span>                        <span class="n">selected_drugs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">drug_name</span><span class="p">)</span>
<span class="linenos" data-linenos="337 "></span>            <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">selected_drugs</span><span class="p">,</span> <span class="p">:]</span>
<span class="linenos" data-linenos="338 "></span>            <span class="k">return</span> <span class="n">df_tmp</span>
<span class="linenos" data-linenos="339 "></span>
<span class="linenos" data-linenos="340 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="341 "></span>            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span> 
<span class="linenos" data-linenos="342 "></span>            <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">n_drug</span><span class="p">,:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Blues'</span><span class="p">,</span> \
<span class="linenos" data-linenos="343 "></span>                        <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linecolor</span><span class="o">=</span><span class="s1">'lightgrey'</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">'shrink'</span><span class="p">:</span> <span class="mf">.2</span><span class="p">,</span> <span class="s1">'label'</span><span class="p">:</span> <span class="s1">'Drug Sensitivity'</span><span class="p">},</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="linenos" data-linenos="344 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Cluster'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="345 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Drug'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="346 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">set_size</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="347 "></span>            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<span class="linenos" data-linenos="348 "></span>                <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="349 "></span>                <span class="n">spine</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">'lightgrey'</span><span class="p">)</span> 
<span class="linenos" data-linenos="350 "></span>            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">.png'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">'tight'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos="351 "></span>            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos="352 "></span>
<span class="linenos" data-linenos="353 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="354 "></span>            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_drug</span><span class="o">*</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">5</span><span class="p">)))</span> 
<span class="linenos" data-linenos="355 "></span>            <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">select_drug</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Reds'</span><span class="p">,</span> \
<span class="linenos" data-linenos="356 "></span>                        <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linecolor</span><span class="o">=</span><span class="s1">'lightgrey'</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">'shrink'</span><span class="p">:</span> <span class="mf">.2</span><span class="p">,</span> <span class="s1">'label'</span><span class="p">:</span> <span class="s1">'Drug Sensitivity'</span><span class="p">},</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="357 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Cluster'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="358 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Drug'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="359 "></span>            <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">set_size</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="360 "></span>            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<span class="linenos" data-linenos="361 "></span>                <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="362 "></span>                <span class="n">spine</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">'lightgrey'</span><span class="p">)</span> 
<span class="linenos" data-linenos="363 "></span>            <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">.png'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">'tight'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos="364 "></span>            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos="365 "></span>
<span class="linenos" data-linenos="366 "></span>    <span class="k">def</span> <span class="nf">figure_output</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="367 "></span><span class="w">        </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="368 "></span><span class="sd">        plot figures</span>
<span class="linenos" data-linenos="369 "></span>
<span class="linenos" data-linenos="370 "></span><span class="sd">        """</span>
<span class="linenos" data-linenos="371 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Ploting figures...'</span><span class="p">)</span>
<span class="linenos" data-linenos="372 "></span>        <span class="c1">## GDSC figures</span>
<span class="linenos" data-linenos="373 "></span>        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="374 "></span>            <span class="n">tmp_pred_ic50_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="375 "></span>            <span class="n">tmp_pred_ic50_df</span> <span class="o">=</span> <span class="n">tmp_pred_ic50_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="nb">sum</span><span class="o">=</span><span class="n">tmp_pred_ic50_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">'sum'</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="376 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">draw_plot</span><span class="p">(</span><span class="n">tmp_pred_ic50_df</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">'GDSC prediction'</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">40</span><span class="p">))</span>
<span class="linenos" data-linenos="377 "></span>            <span class="n">tmp_pred_kill_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_kill_df</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="378 "></span>            <span class="n">tmp_pred_kill_df</span> <span class="o">=</span> <span class="n">tmp_pred_kill_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">tmp_pred_kill_df</span><span class="o">&gt;=</span><span class="mi">50</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
<span class="linenos" data-linenos="379 "></span>            <span class="n">tmp_pred_kill_df</span> <span class="o">=</span> <span class="n">tmp_pred_kill_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="nb">sum</span><span class="o">=</span><span class="n">tmp_pred_kill_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">'sum'</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="380 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">draw_plot</span><span class="p">(</span><span class="n">tmp_pred_kill_df</span><span class="p">,</span> <span class="n">n_drug</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">'predicted cell death'</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
<span class="linenos" data-linenos="381 "></span>
<span class="linenos" data-linenos="382 "></span>        <span class="c1">## PRISM figures</span>
<span class="linenos" data-linenos="383 "></span>        <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="384 "></span>            <span class="n">tmp_pred_auc_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="o">.</span><span class="n">T</span>
<span class="linenos" data-linenos="385 "></span>            <span class="c1">#tmp_pred_auc_df = tmp_pred_auc_df.assign(sum=tmp_pred_auc_df.sum(axis=1)).sort_values(by='sum', ascending=True)</span>
<span class="linenos" data-linenos="386 "></span>            <span class="bp">self</span><span class="o">.</span><span class="n">draw_plot</span><span class="p">(</span><span class="n">tmp_pred_auc_df</span><span class="p">,</span> <span class="n">n_drug</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">n_drugs</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">'PRISM prediction'</span><span class="p">)</span>  
<span class="linenos" data-linenos="387 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'done!'</span><span class="p">)</span>  
</code></pre></div>
</details>
<div class="doc doc-children">
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._scdrug.Drug_Response.__init__">
<code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">scriptpath</span><span class="p">,</span> <span class="n">modelpath</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="s1">'./'</span><span class="p">,</span> <span class="n">model</span><span class="o">=</span><span class="s1">'GDSC'</span><span class="p">,</span> <span class="n">clusters</span><span class="o">=</span><span class="s1">'All'</span><span class="p">,</span> <span class="n">cell</span><span class="o">=</span><span class="s1">'A549'</span><span class="p">,</span> <span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">n_drugs</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single._scdrug.Drug_Response.__init__" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Initializes the Drug_Response class.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p>Annotated data matrix with cells as rows and genes as columns.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>scriptpath</code></td>
<td>
</td>
<td><p>Path to the directory containing the CaDRReS scripts for the analysis. You need to download the scripts according to <code>git clone https://github.com/CSB5/CaDRReS-Sc.git</code> and set the path to the directory.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>modelpath</code></td>
<td>
</td>
<td><p>Path to the directory containing the pre-trained models.
You need to download the models according to <code>Pyomic.utils.download_GDSC_data()</code> and <code>Pyomic.utils.download_CaDRReS_model()</code> and set the path to the directory.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>output</code></td>
<td>
</td>
<td><p>Path to the directory where the output files will be saved (default: './').</p></td>
<td>
<code>'./'</code>
</td>
</tr>
<tr>
<td><code>model</code></td>
<td>
</td>
<td><p>The name of the pre-trained model to be used for the analysis (default: 'GDSC').</p></td>
<td>
<code>'GDSC'</code>
</td>
</tr>
<tr>
<td><code>clusters</code></td>
<td>
</td>
<td><p>The cluster labels to be used for the analysis. Default is all cells (default: 'All').</p></td>
<td>
<code>'All'</code>
</td>
</tr>
<tr>
<td><code>cell</code></td>
<td>
</td>
<td><p>The cell line to be analyzed (default: 'A549').</p></td>
<td>
<code>'A549'</code>
</td>
</tr>
<tr>
<td><code>cpus</code></td>
<td>
</td>
<td><p>The number of CPUs to be used for the analysis (default: 4).</p></td>
<td>
<code>4</code>
</td>
</tr>
<tr>
<td><code>n_drugs</code></td>
<td>
</td>
<td><p>The number of top drugs to be selected based on the predicted sensitivity (default: 10).</p></td>
<td>
<code>10</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>
</td>
<td><p>None</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="132 "></span><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span><span class="n">adata</span><span class="p">,</span><span class="n">scriptpath</span><span class="p">,</span><span class="n">modelpath</span><span class="p">,</span><span class="n">output</span><span class="o">=</span><span class="s1">'./'</span><span class="p">,</span><span class="n">model</span><span class="o">=</span><span class="s1">'GDSC'</span><span class="p">,</span><span class="n">clusters</span><span class="o">=</span><span class="s1">'All'</span><span class="p">,</span>
<span class="linenos" data-linenos="133 "></span>             <span class="n">cell</span><span class="o">=</span><span class="s1">'A549'</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">n_drugs</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="linenos" data-linenos="134 "></span>
<span class="linenos" data-linenos="135 "></span><span class="w">    </span><span class="sd">"""</span>
<span class="linenos" data-linenos="136 "></span><span class="sd">    Initializes the Drug_Response class.</span>
<span class="linenos" data-linenos="137 "></span>
<span class="linenos" data-linenos="138 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="139 "></span><span class="sd">        adata: Annotated data matrix with cells as rows and genes as columns.</span>
<span class="linenos" data-linenos="140 "></span><span class="sd">        scriptpath: Path to the directory containing the CaDRReS scripts for the analysis. You need to download the scripts according to `git clone https://github.com/CSB5/CaDRReS-Sc.git` and set the path to the directory.</span>
<span class="linenos" data-linenos="141 "></span><span class="sd">        modelpath: Path to the directory containing the pre-trained models.</span>
<span class="linenos" data-linenos="142 "></span><span class="sd">            You need to download the models according to `Pyomic.utils.download_GDSC_data()` and `Pyomic.utils.download_CaDRReS_model()` and set the path to the directory.</span>
<span class="linenos" data-linenos="143 "></span><span class="sd">        output: Path to the directory where the output files will be saved (default: './').</span>
<span class="linenos" data-linenos="144 "></span><span class="sd">        model: The name of the pre-trained model to be used for the analysis (default: 'GDSC').</span>
<span class="linenos" data-linenos="145 "></span><span class="sd">        clusters: The cluster labels to be used for the analysis. Default is all cells (default: 'All').</span>
<span class="linenos" data-linenos="146 "></span><span class="sd">        cell: The cell line to be analyzed (default: 'A549').</span>
<span class="linenos" data-linenos="147 "></span><span class="sd">        cpus: The number of CPUs to be used for the analysis (default: 4).</span>
<span class="linenos" data-linenos="148 "></span><span class="sd">        n_drugs: The number of top drugs to be selected based on the predicted sensitivity (default: 10).</span>
<span class="linenos" data-linenos="149 "></span>
<span class="linenos" data-linenos="150 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="151 "></span><span class="sd">        None</span>
<span class="linenos" data-linenos="152 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="153 "></span>
<span class="linenos" data-linenos="154 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">model</span>
<span class="linenos" data-linenos="155 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span>
<span class="linenos" data-linenos="156 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">clusters</span><span class="o">=</span><span class="n">clusters</span>
<span class="linenos" data-linenos="157 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="o">=</span><span class="n">output</span>
<span class="linenos" data-linenos="158 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">n_drugs</span><span class="o">=</span><span class="n">n_drugs</span>
<span class="linenos" data-linenos="159 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">=</span><span class="n">modelpath</span>
<span class="linenos" data-linenos="160 "></span>
<span class="linenos" data-linenos="161 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">scriptpath</span> <span class="o">=</span> <span class="n">scriptpath</span>
<span class="linenos" data-linenos="162 "></span>    <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">scriptpath</span><span class="p">))</span>
<span class="linenos" data-linenos="163 "></span>
<span class="linenos" data-linenos="164 "></span>    <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="165 "></span>
<span class="linenos" data-linenos="166 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">load_model</span><span class="p">()</span>
<span class="linenos" data-linenos="167 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">drug_info</span><span class="p">()</span>
<span class="linenos" data-linenos="168 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">bulk_exp</span><span class="p">()</span>
<span class="linenos" data-linenos="169 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">sc_exp</span><span class="p">()</span>
<span class="linenos" data-linenos="170 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">kernel_feature_preparartion</span><span class="p">()</span>
<span class="linenos" data-linenos="171 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">sensitivity_prediction</span><span class="p">()</span>
<span class="linenos" data-linenos="172 "></span>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="173 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">+</span><span class="s1">'masked_drugs.csv'</span><span class="p">)[</span><span class="s1">'GDSC'</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'int64'</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'str'</span><span class="p">))</span>
<span class="linenos" data-linenos="174 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">cell_death_proportion</span><span class="p">()</span>
<span class="linenos" data-linenos="175 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="176 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span><span class="o">+</span><span class="s1">'masked_drugs.csv'</span><span class="p">)[</span><span class="s1">'PRISM'</span><span class="p">])</span>
<span class="linenos" data-linenos="177 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">output_result</span><span class="p">()</span>
<span class="linenos" data-linenos="178 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">figure_output</span><span class="p">()</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._scdrug.Drug_Response.load_model">
<code class="highlight language-python"><span class="n">load_model</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.single._scdrug.Drug_Response.load_model" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>load the pre-trained model.</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="180 "></span><span class="k">def</span> <span class="nf">load_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="181 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="182 "></span><span class="sd">    load the pre-trained model.</span>
<span class="linenos" data-linenos="183 "></span>
<span class="linenos" data-linenos="184 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="185 "></span>    <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="186 "></span>    <span class="c1">### IC50/AUC prediction</span>
<span class="linenos" data-linenos="187 "></span>    <span class="c1">## Read pre-trained model</span>
<span class="linenos" data-linenos="188 "></span>    <span class="c1">#model_dir = '/Users/fernandozeng/Desktop/analysis/scDrug/CaDRReS-Sc-model/'</span>
<span class="linenos" data-linenos="189 "></span>    <span class="n">model_dir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">modelpath</span>
<span class="linenos" data-linenos="190 "></span>    <span class="n">obj_function</span> <span class="o">=</span> <span class="n">widgets</span><span class="o">.</span><span class="n">Dropdown</span><span class="p">(</span><span class="n">options</span><span class="o">=</span><span class="p">[</span><span class="s1">'cadrres-wo-sample-bias'</span><span class="p">,</span> <span class="s1">'cadrres-wo-sample-bias-weight'</span><span class="p">],</span> <span class="n">description</span><span class="o">=</span><span class="s1">'Objetice function'</span><span class="p">)</span>
<span class="linenos" data-linenos="191 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span> <span class="o">=</span> <span class="n">obj_function</span><span class="o">.</span><span class="n">value</span>
<span class="linenos" data-linenos="192 "></span>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="193 "></span>        <span class="n">model_file</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">+</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">_param_dict_all_genes.pickle'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="194 "></span>    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'PRISM'</span><span class="p">:</span>
<span class="linenos" data-linenos="195 "></span>        <span class="n">model_file</span> <span class="o">=</span> <span class="n">model_dir</span> <span class="o">+</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">_param_dict_prism.pickle'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="196 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="197 "></span>        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">'Wrong model name.'</span><span class="p">)</span>
<span class="linenos" data-linenos="198 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">model_file</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._scdrug.Drug_Response.sensitivity_prediction">
<code class="highlight language-python"><span class="n">sensitivity_prediction</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.single._scdrug.Drug_Response.sensitivity_prediction" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Predict drug sensitivity</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="267 "></span><span class="k">def</span> <span class="nf">sensitivity_prediction</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="268 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="269 "></span><span class="sd">    Predict drug sensitivity</span>
<span class="linenos" data-linenos="270 "></span>
<span class="linenos" data-linenos="271 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="272 "></span>    <span class="kn">from</span> <span class="nn">cadrres_sc</span> <span class="kn">import</span> <span class="n">pp</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">evaluation</span><span class="p">,</span> <span class="n">utility</span>
<span class="linenos" data-linenos="273 "></span>    <span class="c1">## Drug response prediction</span>
<span class="linenos" data-linenos="274 "></span>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="275 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Predicting drug response for using CaDRReS(GDSC): </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">))</span>
<span class="linenos" data-linenos="276 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="p">,</span> <span class="n">P_test_df</span><span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_from_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_kernel_df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="277 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...done!'</span><span class="p">)</span>
<span class="linenos" data-linenos="278 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="279 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...Predicting drug response for using CaDRReS(PRISM): </span><span class="si">{}</span><span class="s1">'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">))</span>
<span class="linenos" data-linenos="280 "></span>        <span class="bp">self</span><span class="o">.</span><span class="n">pred_auc_df</span><span class="p">,</span> <span class="n">P_test_df</span><span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict_from_model</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cadrres_model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">test_kernel_df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">model_spec_name</span><span class="p">)</span>
<span class="linenos" data-linenos="281 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'...done!'</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._scdrug.Drug_Response.cell_death_proportion">
<code class="highlight language-python"><span class="n">cell_death_proportion</span><span class="p">()</span></code>
<a class="headerlink" href="#omicverse.single._scdrug.Drug_Response.cell_death_proportion" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>Predict cell death proportion and cell death percentage at the ref_type dosage</p>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="283 "></span><span class="k">def</span> <span class="nf">cell_death_proportion</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="linenos" data-linenos="284 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="285 "></span><span class="sd">    Predict cell death proportion and cell death percentage at the ref_type dosage</span>
<span class="linenos" data-linenos="286 "></span>
<span class="linenos" data-linenos="287 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="288 "></span>    <span class="c1">### Drug kill prediction</span>
<span class="linenos" data-linenos="289 "></span>    <span class="n">ref_type</span> <span class="o">=</span> <span class="s1">'log2_median_ic50'</span>
<span class="linenos" data-linenos="290 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">masked_drugs</span><span class="p">]</span>
<span class="linenos" data-linenos="291 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">]</span>
<span class="linenos" data-linenos="292 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_list</span><span class="p">]</span>
<span class="linenos" data-linenos="293 "></span>
<span class="linenos" data-linenos="294 "></span>    <span class="c1">## Predict cell death percentage at the ref_type dosage</span>
<span class="linenos" data-linenos="295 "></span>    <span class="n">pred_delta_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">values</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_info_df</span><span class="p">[</span><span class="n">ref_type</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">pred_ic50_df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="linenos" data-linenos="296 "></span>    <span class="n">pred_cv_df</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="n">pred_delta_df</span><span class="p">)))</span>
<span class="linenos" data-linenos="297 "></span>    <span class="bp">self</span><span class="o">.</span><span class="n">pred_kill_df</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">-</span> <span class="n">pred_cv_df</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h3 class="doc doc-heading" id="omicverse.single._scdrug.Drug_Response.draw_plot">
<code class="highlight language-python"><span class="n">draw_plot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">())</span></code>
<a class="headerlink" href="#omicverse.single._scdrug.Drug_Response.draw_plot" title="Permanent link">¶</a></h3>
<div class="doc doc-contents">
<p>plot heatmap of drug response prediction</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>df</code></td>
<td>
</td>
<td><p><code>pandas.DataFrame</code> drug response prediction dataframe</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>n_drug</code></td>
<td>
</td>
<td><p><code>int</code> number of drugs to be plotted</p></td>
<td>
<code>10</code>
</td>
</tr>
<tr>
<td><code>name</code></td>
<td>
</td>
<td><p><code>str</code> name of the plot</p></td>
<td>
<code>''</code>
</td>
</tr>
<tr>
<td><code>figsize</code></td>
<td>
</td>
<td><p><code>tuple</code> size of the plot</p></td>
<td>
<code>()</code>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="320 "></span><span class="k">def</span> <span class="nf">draw_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">''</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">()):</span>
<span class="linenos" data-linenos="321 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="322 "></span><span class="sd">    plot heatmap of drug response prediction</span>
<span class="linenos" data-linenos="323 "></span>
<span class="linenos" data-linenos="324 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="325 "></span><span class="sd">        df : `pandas.DataFrame` drug response prediction dataframe</span>
<span class="linenos" data-linenos="326 "></span><span class="sd">        n_drug : `int` number of drugs to be plotted</span>
<span class="linenos" data-linenos="327 "></span><span class="sd">        name : `str` name of the plot</span>
<span class="linenos" data-linenos="328 "></span><span class="sd">        figsize : `tuple` size of the plot</span>
<span class="linenos" data-linenos="329 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="330 "></span>    <span class="k">def</span> <span class="nf">select_drug</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="p">):</span>
<span class="linenos" data-linenos="331 "></span>        <span class="n">selected_drugs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos="332 "></span>        <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">'Drug Name'</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">:]</span>
<span class="linenos" data-linenos="333 "></span>        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">columns</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="nb">int</span><span class="p">):</span>
<span class="linenos" data-linenos="334 "></span>            <span class="k">for</span> <span class="n">drug_name</span> <span class="ow">in</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">[:</span><span class="n">n_drug</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
<span class="linenos" data-linenos="335 "></span>                <span class="k">if</span> <span class="n">drug_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">selected_drugs</span><span class="p">:</span>
<span class="linenos" data-linenos="336 "></span>                    <span class="n">selected_drugs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">drug_name</span><span class="p">)</span>
<span class="linenos" data-linenos="337 "></span>        <span class="n">df_tmp</span> <span class="o">=</span> <span class="n">df_tmp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">selected_drugs</span><span class="p">,</span> <span class="p">:]</span>
<span class="linenos" data-linenos="338 "></span>        <span class="k">return</span> <span class="n">df_tmp</span>
<span class="linenos" data-linenos="339 "></span>
<span class="linenos" data-linenos="340 "></span>    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">==</span> <span class="s1">'GDSC'</span><span class="p">:</span>
<span class="linenos" data-linenos="341 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span> 
<span class="linenos" data-linenos="342 "></span>        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="n">n_drug</span><span class="p">,:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Blues'</span><span class="p">,</span> \
<span class="linenos" data-linenos="343 "></span>                    <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linecolor</span><span class="o">=</span><span class="s1">'lightgrey'</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">'shrink'</span><span class="p">:</span> <span class="mf">.2</span><span class="p">,</span> <span class="s1">'label'</span><span class="p">:</span> <span class="s1">'Drug Sensitivity'</span><span class="p">},</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="linenos" data-linenos="344 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Cluster'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="345 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Drug'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="346 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">set_size</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="347 "></span>        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<span class="linenos" data-linenos="348 "></span>            <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="349 "></span>            <span class="n">spine</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">'lightgrey'</span><span class="p">)</span> 
<span class="linenos" data-linenos="350 "></span>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">.png'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">'tight'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos="351 "></span>        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos="352 "></span>
<span class="linenos" data-linenos="353 "></span>    <span class="k">else</span><span class="p">:</span>
<span class="linenos" data-linenos="354 "></span>        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_drug</span><span class="o">*</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">5</span><span class="p">)))</span> 
<span class="linenos" data-linenos="355 "></span>        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">select_drug</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_drug</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Reds'</span><span class="p">,</span> \
<span class="linenos" data-linenos="356 "></span>                    <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">linecolor</span><span class="o">=</span><span class="s1">'lightgrey'</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">'shrink'</span><span class="p">:</span> <span class="mf">.2</span><span class="p">,</span> <span class="s1">'label'</span><span class="p">:</span> <span class="s1">'Drug Sensitivity'</span><span class="p">},</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos="357 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Cluster'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="358 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">'Drug'</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="359 "></span>        <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">axes</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">label</span><span class="o">.</span><span class="n">set_size</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="linenos" data-linenos="360 "></span>        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">spine</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
<span class="linenos" data-linenos="361 "></span>            <span class="n">spine</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="362 "></span>            <span class="n">spine</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">'lightgrey'</span><span class="p">)</span> 
<span class="linenos" data-linenos="363 "></span>        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output</span><span class="p">,</span> <span class="s1">'</span><span class="si">{}</span><span class="s1">.png'</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">)),</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">'tight'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="linenos" data-linenos="364 "></span>        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div>
</details>
</div>
</div>
</div>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.autoResolution">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">autoResolution</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.autoResolution" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Automatically determine clustering resolution</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td>
<td>
</td>
<td><p><code>scanpy.AnnData</code> The single cell data.</p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>cpus</code></td>
<td>
</td>
<td><p><code>int</code>, optional (default: 4) The number of cpus used for parallel computing.</p></td>
<td>
<code>4</code>
</td>
</tr>
</tbody>
</table>
<p><strong>Returns:</strong></p>
<table>
<thead>
<tr>
<th>Name</th> <th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata</code></td> <td>
</td>
<td><p><code>scanpy.AnnData</code> The single cell data with the clustering resolution.</p></td>
</tr>
<tr>
<td><code>res</code></td> <td>
</td>
<td><p><code>float</code> The clustering resolution.</p></td>
</tr>
<tr>
<td><code>df_sil</code></td> <td>
</td>
<td><p><code>pandas.DataFrame</code> The silhouette score of each clustering resolution.</p></td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 24 "></span><span class="k">def</span> <span class="nf">autoResolution</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">cpus</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
<span class="linenos" data-linenos=" 25 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos=" 26 "></span><span class="sd">    Automatically determine clustering resolution</span>
<span class="linenos" data-linenos=" 27 "></span>
<span class="linenos" data-linenos=" 28 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos=" 29 "></span><span class="sd">        adata : `scanpy.AnnData` The single cell data.</span>
<span class="linenos" data-linenos=" 30 "></span><span class="sd">        cpus : `int`, optional (default: 4) The number of cpus used for parallel computing.</span>
<span class="linenos" data-linenos=" 31 "></span>
<span class="linenos" data-linenos=" 32 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos=" 33 "></span><span class="sd">        adata : `scanpy.AnnData` The single cell data with the clustering resolution.</span>
<span class="linenos" data-linenos=" 34 "></span><span class="sd">        res : `float` The clustering resolution.</span>
<span class="linenos" data-linenos=" 35 "></span><span class="sd">        df_sil: `pandas.DataFrame` The silhouette score of each clustering resolution.</span>
<span class="linenos" data-linenos=" 36 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos=" 37 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s2">"Automatically determine clustering resolution..."</span><span class="p">)</span>
<span class="linenos" data-linenos=" 38 "></span>    <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="linenos" data-linenos=" 39 "></span>    <span class="k">def</span> <span class="nf">subsample_clustering</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">sample_n</span><span class="p">,</span> <span class="n">subsample_n</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">subsample</span><span class="p">):</span>
<span class="linenos" data-linenos=" 40 "></span>        <span class="n">subadata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">subsample</span><span class="p">]</span>
<span class="linenos" data-linenos=" 41 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">subadata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="n">resolution</span><span class="p">)</span>
<span class="linenos" data-linenos=" 42 "></span>        <span class="n">cluster</span> <span class="o">=</span> <span class="n">subadata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos=" 43 "></span>
<span class="linenos" data-linenos=" 44 "></span>        <span class="n">subsampling_n</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">sample_n</span><span class="p">,</span> <span class="n">sample_n</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
<span class="linenos" data-linenos=" 45 "></span>        <span class="n">coclustering_n</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">sample_n</span><span class="p">,</span> <span class="n">sample_n</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
<span class="linenos" data-linenos=" 46 "></span>
<span class="linenos" data-linenos=" 47 "></span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">subsample_n</span><span class="p">):</span>
<span class="linenos" data-linenos=" 48 "></span>            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">subsample_n</span><span class="p">):</span>
<span class="linenos" data-linenos=" 49 "></span>                <span class="n">x</span> <span class="o">=</span> <span class="n">subsample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="linenos" data-linenos=" 50 "></span>                <span class="n">y</span> <span class="o">=</span> <span class="n">subsample</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
<span class="linenos" data-linenos=" 51 "></span>                <span class="n">subsampling_n</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="linenos" data-linenos=" 52 "></span>                <span class="k">if</span> <span class="n">cluster</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">cluster</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
<span class="linenos" data-linenos=" 53 "></span>                    <span class="n">coclustering_n</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="linenos" data-linenos=" 54 "></span>        <span class="k">return</span> <span class="p">(</span><span class="n">subsampling_n</span><span class="p">,</span> <span class="n">coclustering_n</span><span class="p">)</span>
<span class="linenos" data-linenos=" 55 "></span>    <span class="n">rep_n</span> <span class="o">=</span> <span class="mi">5</span>
<span class="linenos" data-linenos=" 56 "></span>    <span class="n">subset</span> <span class="o">=</span> <span class="mf">0.8</span>
<span class="linenos" data-linenos=" 57 "></span>    <span class="n">sample_n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
<span class="linenos" data-linenos=" 58 "></span>    <span class="n">subsample_n</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sample_n</span> <span class="o">*</span> <span class="n">subset</span><span class="p">)</span>
<span class="linenos" data-linenos=" 59 "></span>    <span class="n">resolutions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">0.4</span><span class="p">,</span> <span class="mf">1.4</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="linenos" data-linenos=" 60 "></span>    <span class="n">silhouette_avg</span> <span class="o">=</span> <span class="p">{}</span>
<span class="linenos" data-linenos=" 61 "></span>    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos=" 62 "></span>    <span class="n">best_resolution</span> <span class="o">=</span> <span class="mi">0</span>
<span class="linenos" data-linenos=" 63 "></span>    <span class="n">highest_sil</span> <span class="o">=</span> <span class="mi">0</span>
<span class="linenos" data-linenos=" 64 "></span>    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">resolutions</span><span class="p">:</span>
<span class="linenos" data-linenos=" 65 "></span>        <span class="n">r</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos=" 66 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s2">"Clustering test: resolution = "</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="linenos" data-linenos=" 67 "></span>        <span class="n">sub_start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="linenos" data-linenos=" 68 "></span>        <span class="n">subsamples</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">sample_n</span><span class="p">,</span> <span class="n">subsample_n</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rep_n</span><span class="p">)]</span>
<span class="linenos" data-linenos=" 69 "></span>        <span class="n">p</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">cpus</span><span class="p">)</span>
<span class="linenos" data-linenos=" 70 "></span>        <span class="n">func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">subsample_clustering</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">sample_n</span><span class="p">,</span> <span class="n">subsample_n</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="linenos" data-linenos=" 71 "></span>        <span class="n">resultList</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">func</span><span class="p">,</span> <span class="n">subsamples</span><span class="p">)</span>
<span class="linenos" data-linenos=" 72 "></span>        <span class="n">p</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos=" 73 "></span>        <span class="n">p</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
<span class="linenos" data-linenos=" 74 "></span>
<span class="linenos" data-linenos=" 75 "></span>        <span class="n">subsampling_n</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">result</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">resultList</span><span class="p">])</span>
<span class="linenos" data-linenos=" 76 "></span>        <span class="n">coclustering_n</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">result</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">resultList</span><span class="p">])</span>
<span class="linenos" data-linenos=" 77 "></span>
<span class="linenos" data-linenos=" 78 "></span>        <span class="n">subsampling_n</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">subsampling_n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)]</span> <span class="o">=</span> <span class="mf">1e6</span>
<span class="linenos" data-linenos=" 79 "></span>        <span class="n">distance</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">coclustering_n</span> <span class="o">/</span> <span class="n">subsampling_n</span>
<span class="linenos" data-linenos=" 80 "></span>        <span class="n">np</span><span class="o">.</span><span class="n">fill_diagonal</span><span class="p">(</span><span class="n">distance</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
<span class="linenos" data-linenos=" 81 "></span>
<span class="linenos" data-linenos=" 82 "></span>        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">louvain</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">'louvain_r'</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>
<span class="linenos" data-linenos=" 83 "></span>        <span class="n">silhouette_avg</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)]</span> <span class="o">=</span> <span class="n">silhouette_score</span><span class="p">(</span><span class="n">distance</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain_r'</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)],</span> <span class="n">metric</span><span class="o">=</span><span class="s2">"precomputed"</span><span class="p">)</span>
<span class="linenos" data-linenos=" 84 "></span>        <span class="k">if</span> <span class="n">silhouette_avg</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)]</span> <span class="o">&gt;</span> <span class="n">highest_sil</span><span class="p">:</span>
<span class="linenos" data-linenos=" 85 "></span>            <span class="n">highest_sil</span> <span class="o">=</span> <span class="n">silhouette_avg</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)]</span>
<span class="linenos" data-linenos=" 86 "></span>            <span class="n">best_resolution</span> <span class="o">=</span> <span class="n">r</span>
<span class="linenos" data-linenos=" 87 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s2">"robustness score = "</span><span class="p">,</span> <span class="n">silhouette_avg</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">r</span><span class="p">)])</span>
<span class="linenos" data-linenos=" 88 "></span>        <span class="n">sub_end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="linenos" data-linenos=" 89 "></span>        <span class="nb">print</span><span class="p">(</span><span class="s1">'time: </span><span class="si">{}</span><span class="s1">'</span><span class="p">,</span> <span class="n">sub_end</span> <span class="o">-</span> <span class="n">sub_start</span><span class="p">)</span>
<span class="linenos" data-linenos=" 90 "></span>        <span class="nb">print</span><span class="p">()</span>
<span class="linenos" data-linenos=" 91 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain_r'</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">best_resolution</span><span class="p">)]</span>
<span class="linenos" data-linenos=" 92 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s2">"resolution with highest score: "</span><span class="p">,</span> <span class="n">best_resolution</span><span class="p">)</span>
<span class="linenos" data-linenos=" 93 "></span>    <span class="n">res</span> <span class="o">=</span> <span class="n">best_resolution</span>
<span class="linenos" data-linenos=" 94 "></span>    <span class="c1"># write silhouette record to uns and remove the clustering results except for the one with the best resolution</span>
<span class="linenos" data-linenos=" 95 "></span>    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">'sihouette score'</span><span class="p">]</span> <span class="o">=</span> <span class="n">silhouette_avg</span>
<span class="linenos" data-linenos=" 96 "></span>    <span class="c1"># draw lineplot</span>
<span class="linenos" data-linenos=" 97 "></span>    <span class="n">df_sil</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">silhouette_avg</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">'silhouette score'</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">silhouette_avg</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
<span class="linenos" data-linenos=" 98 "></span>    <span class="n">df_sil</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">line</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s1">'.-'</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">'green'</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">'Auto Resolution'</span><span class="p">,</span> <span class="n">xticks</span><span class="o">=</span><span class="n">resolutions</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">'resolution'</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">'silhouette score'</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos=" 99 "></span>    <span class="c1">#pp.savefig()</span>
<span class="linenos" data-linenos="100 "></span>    <span class="c1">#plt.close()</span>
<span class="linenos" data-linenos="101 "></span>    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="linenos" data-linenos="102 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'time: </span><span class="si">{}</span><span class="s1">'</span><span class="p">,</span> <span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">)</span>
<span class="linenos" data-linenos="103 "></span>    <span class="k">return</span> <span class="n">adata</span><span class="p">,</span> <span class="n">res</span><span class="p">,</span> <span class="n">df_sil</span>
</code></pre></div>
</details>
</div>
</div>
<div class="doc doc-object doc-function">
<h2 class="doc doc-heading" id="omicverse.single.writeGEP">
<code class="highlight language-python"><span class="n">omicverse</span><span class="o">.</span><span class="n">single</span><span class="o">.</span><span class="n">writeGEP</span><span class="p">(</span><span class="n">adata_GEP</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span></code>
<a class="headerlink" href="#omicverse.single.writeGEP" title="Permanent link">¶</a></h2>
<div class="doc doc-contents first">
<p>Write the gene expression profile to a file</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>adata_GEP</code></td>
<td>
</td>
<td><p><code>scanpy.AnnData</code> The single cell data with gene expression profile. </p></td>
<td>
<em>required</em>
</td>
</tr>
<tr>
<td><code>path</code></td>
<td>
</td>
<td><p><code>str</code> The path to save the gene expression profile.</p></td>
<td>
<em>required</em>
</td>
</tr>
</tbody>
</table>
<details class="quote">
<summary>Source code in <code>/Users/fernandozeng/miniforge3/envs/scbasset/lib/python3.8/site-packages/omicverse/single/_scdrug.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="105 "></span><span class="k">def</span> <span class="nf">writeGEP</span><span class="p">(</span><span class="n">adata_GEP</span><span class="p">,</span><span class="n">path</span><span class="p">):</span>
<span class="linenos" data-linenos="106 "></span><span class="w">    </span><span class="sa">r</span><span class="sd">"""</span>
<span class="linenos" data-linenos="107 "></span><span class="sd">    Write the gene expression profile to a file</span>
<span class="linenos" data-linenos="108 "></span>
<span class="linenos" data-linenos="109 "></span><span class="sd">    Arguments:</span>
<span class="linenos" data-linenos="110 "></span><span class="sd">        adata_GEP : `scanpy.AnnData` The single cell data with gene expression profile. </span>
<span class="linenos" data-linenos="111 "></span><span class="sd">        path : `str` The path to save the gene expression profile.</span>
<span class="linenos" data-linenos="112 "></span>
<span class="linenos" data-linenos="113 "></span><span class="sd">    Returns:</span>
<span class="linenos" data-linenos="114 "></span>
<span class="linenos" data-linenos="115 "></span><span class="sd">    """</span>
<span class="linenos" data-linenos="116 "></span>    <span class="nb">print</span><span class="p">(</span><span class="s1">'Exporting GEP...'</span><span class="p">)</span>
<span class="linenos" data-linenos="117 "></span>    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_GEP</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e6</span><span class="p">)</span>
<span class="linenos" data-linenos="118 "></span>    <span class="n">mat</span> <span class="o">=</span> <span class="n">adata_GEP</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="linenos" data-linenos="119 "></span>    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="linenos" data-linenos="120 "></span>        <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
<span class="linenos" data-linenos="121 "></span>    <span class="n">GEP_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata_GEP</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="linenos" data-linenos="122 "></span>    <span class="n">GEP_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">adata_GEP</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">'louvain'</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos="123 "></span>    <span class="c1"># GEP_df = GEP_df.loc[adata.var.index[adata.var.highly_variable==True]]</span>
<span class="linenos" data-linenos="124 "></span>    <span class="n">GEP_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="125 "></span>    <span class="n">GEP_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">'GEP.txt'</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">'</span><span class="se">\t</span><span class="s1">'</span><span class="p">)</span>
</code></pre></div>
</details>
</div>
</div>
</article>
</div>
</div>
</main>
<footer class="md-footer">
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
<div class="md-copyright__highlight">
      Copyright © 2019-2024, 112 Lab, USTB
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
<script src="../../../assets/javascripts/bundle.fc8c2696.min.js"></script>
<script src="../../../javascripts/config.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
<script>document$.subscribe(() => {const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});})</script></body>
</html>